// @bmc-ux/obsolete/esm2015/lib/list-builder/index.js
import { DragDropModule } from '@angular/cdk/drag-drop';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptListBuilderItemComponent } from './list-builder-item/list-builder-item.component';
import { AdaptListBuilderComponent } from './list-builder.component';
import { AdaptRxLabelModule, AdaptTooltipModule, AdaptCustomActionsModule, AdaptDragonInputHandlerModule, AdaptEmptyStateModule, AdaptIconModule, AdaptPopoverModule, AdaptHighlightModule, AdaptIconConfig } from '@bmc-ux/adapt-angular';
import { AdaptTextFieldModule } from '../textfield/index';
import { AdaptTypeaheadSubModule } from '../typeahead/index';
import { AdaptSelectSearchFilter, AdaptSelectSearchFilterModule } from '../select/index';
import { ai_dots } from '@bmc-ux/dpl-iconfont';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptListBuilderModule {
    constructor(iconService) {
        iconService.registerIcons([ai_dots]);
    }
    static forRoot() {
        return {
            ngModule: AdaptListBuilderModule,
            providers: [AdaptSelectSearchFilter]
        };
    }
}
AdaptListBuilderModule.ɵfac = function AdaptListBuilderModule_Factory(t) { return new (t || AdaptListBuilderModule)(i0.ɵɵinject(i1.AdaptIconConfig)); };
AdaptListBuilderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptListBuilderModule });
AdaptListBuilderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            DragDropModule,
            FormsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptEmptyStateModule,
            AdaptSelectSearchFilterModule,
            AdaptTextFieldModule,
            AdaptTypeaheadSubModule,
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule,
            AdaptCustomActionsModule,
            AdaptHighlightModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptListBuilderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    DragDropModule,
                    FormsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptEmptyStateModule,
                    AdaptSelectSearchFilterModule,
                    AdaptTextFieldModule,
                    AdaptTypeaheadSubModule,
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule,
                    AdaptCustomActionsModule,
                    AdaptHighlightModule,
                ],
                declarations: [AdaptListBuilderComponent, AdaptListBuilderItemComponent],
                exports: [AdaptListBuilderComponent]
            }]
    }], function () { return [{ type: i1.AdaptIconConfig }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptListBuilderModule, { declarations: [AdaptListBuilderComponent, AdaptListBuilderItemComponent], imports: [CommonModule,
        DragDropModule,
        FormsModule,
        AdaptIconModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptEmptyStateModule,
        AdaptSelectSearchFilterModule,
        AdaptTextFieldModule,
        AdaptTypeaheadSubModule,
        AdaptDragonInputHandlerModule,
        AdaptRxLabelModule,
        AdaptCustomActionsModule,
        AdaptHighlightModule], exports: [AdaptListBuilderComponent] }); })();
export * from './list-builder.component';
export * from './list-builder.model';
