// @bmc-ux/adapt-angular/esm2015/collapse/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptCollapseDirective } from './collapse.directive';
import { AdaptExpandoDirective } from './expando.directive';
import * as i0 from "@angular/core";
export class AdaptCollapseModule {
}
AdaptCollapseModule.ɵfac = function AdaptCollapseModule_Factory(t) { return new (t || AdaptCollapseModule)(); };
AdaptCollapseModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCollapseModule });
AdaptCollapseModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCollapseModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptCollapseDirective, AdaptExpandoDirective],
                exports: [AdaptCollapseDirective, AdaptExpandoDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCollapseModule, { declarations: [AdaptCollapseDirective, AdaptExpandoDirective], imports: [CommonModule], exports: [AdaptCollapseDirective, AdaptExpandoDirective] }); })();
export * from './collapse.directive';
export * from './expando.directive';
