// @bmc-ux/adapt-angular/esm2015/rx/form-control/rx-control-label/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ai_info_circle } from '@bmc-ux/dpl-iconfont';
import { AdaptIconModule, AdaptIconConfig } from '../../../icon/index';
import { AdaptPopoverModule } from '../../../popover/index';
import { AdaptTooltipModule } from '../../../tooltip/index';
import { AdaptRxControlLabelComponent } from './adapt-rx-control-label.component';
import { AdaptRxLabelTooltipComponent } from './adapt-rx-label-tooltip.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../icon/index";
export class AdaptRxLabelModule {
    constructor(iconService) {
        iconService.registerIcons([ai_info_circle]);
    }
}
AdaptRxLabelModule.ɵfac = function AdaptRxLabelModule_Factory(t) { return new (t || AdaptRxLabelModule)(i0.ɵɵinject(i1.AdaptIconConfig)); };
AdaptRxLabelModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxLabelModule });
AdaptRxLabelModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptIconModule,
            AdaptPopoverModule,
            AdaptTooltipModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxLabelModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptIconModule,
                    AdaptPopoverModule,
                    AdaptTooltipModule
                ],
                declarations: [
                    AdaptRxControlLabelComponent,
                    AdaptRxLabelTooltipComponent
                ],
                exports: [
                    AdaptRxControlLabelComponent,
                    AdaptRxLabelTooltipComponent
                ]
            }]
    }], function () { return [{ type: i1.AdaptIconConfig }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxLabelModule, { declarations: [AdaptRxControlLabelComponent,
        AdaptRxLabelTooltipComponent], imports: [CommonModule,
        AdaptIconModule,
        AdaptPopoverModule,
        AdaptTooltipModule], exports: [AdaptRxControlLabelComponent,
        AdaptRxLabelTooltipComponent] }); })();
export * from './adapt-rx-label-tooltip.component';
export * from './adapt-rx-control-label-tooltip.model';
export * from './adapt-rx-control-label.component';
