// @bmc-ux/obsolete/esm2015/lib/checkbox2/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptCheckbox2Component } from './checkbox2.component';
import { AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptCheckbox2Module {
}
AdaptCheckbox2Module.ɵfac = function AdaptCheckbox2Module_Factory(t) { return new (t || AdaptCheckbox2Module)(); };
AdaptCheckbox2Module.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCheckbox2Module });
AdaptCheckbox2Module.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[FormsModule, CommonModule, AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule.forRoot()]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCheckbox2Module, [{
        type: NgModule,
        args: [{
                imports: [FormsModule, CommonModule, AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule.forRoot()],
                exports: [AdaptCheckbox2Component],
                declarations: [AdaptCheckbox2Component]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCheckbox2Module, { declarations: [AdaptCheckbox2Component], imports: [FormsModule, CommonModule, AdaptIconModule, AdaptTooltipModule, i1.AdaptPopoverModule], exports: [AdaptCheckbox2Component] }); })();
export * from './checkbox2.component';
