// @bmc-ux/adapt-table/esm2015/lib/table-header-cell/table-header-cell.component.js
import { Component, Input, } from '@angular/core';
import { AdaptTableComponent } from '../adapt-table.component';
import { AdaptTableClasses } from '../adapt-table.models';
import * as i0 from "@angular/core";
import * as i1 from "../adapt-table.component";
import * as i2 from "@angular/common";
import * as i3 from "../adapt-table-filters/adapt-table-filters.component";
import * as i4 from "../primeng-lib/table/sortable-column.directive";
import * as i5 from "./adapt-table-header-cell-menu/adapt-table-header-cell-menu.component";
function AdaptTableHeaderCellComponent_ng_container_6_ng_template_3_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n          ");
} }
function AdaptTableHeaderCellComponent_ng_container_6_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n          ");
    i0.ɵɵtemplate(1, AdaptTableHeaderCellComponent_ng_container_6_ng_template_3_ng_template_1_Template, 1, 0, "ng-template", 7);
    i0.ɵɵtext(2, "\n        ");
} if (rf & 2) {
    const columnConfig_r2 = i0.ɵɵnextContext().$implicit;
    const ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", columnConfig_r2.headerTitleTemplate)("ngTemplateOutletContext", ctx_r4.getColumnHeaderTemplateParams(columnConfig_r2));
} }
function AdaptTableHeaderCellComponent_ng_container_6_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n          ");
    i0.ɵɵelementStart(1, "span", 8);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n        ");
} if (rf & 2) {
    const columnConfig_r2 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(columnConfig_r2.header);
} }
function AdaptTableHeaderCellComponent_ng_container_6_span_9_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 11);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const columnConfig_r2 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r11 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", ctx_r11.getDisplayNumberOfSortingOrder(columnConfig_r2), "\n            ");
} }
const _c0 = function (a0, a1, a2) { return { "d-icon-sorting": a0, "d-icon-triangle_up_adapt": a1, "d-icon-triangle_down_adapt": a2 }; };
function AdaptTableHeaderCellComponent_ng_container_6_span_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 9);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptTableHeaderCellComponent_ng_container_6_span_9_span_2_Template, 2, 1, "span", 10);
    i0.ɵɵtext(3, "\n          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const columnConfig_r2 = i0.ɵɵnextContext().$implicit;
    const ctx_r7 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction3(4, _c0, !ctx_r7.isSortedColumn(columnConfig_r2), ctx_r7.isSortedColumn(columnConfig_r2) && ctx_r7.table.isColumnSortOrderAsc(columnConfig_r2.field), ctx_r7.isSortedColumn(columnConfig_r2) && !ctx_r7.table.isColumnSortOrderAsc(columnConfig_r2.field)))("pSortableColumnDisabled", !ctx_r7.table.isSortableColumn(columnConfig_r2))("pSortableColumn", columnConfig_r2.field);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r7.getDisplayNumberOfSortingOrder(columnConfig_r2));
} }
function AdaptTableHeaderCellComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵtext(2, "\n        ");
    i0.ɵɵtemplate(3, AdaptTableHeaderCellComponent_ng_container_6_ng_template_3_Template, 3, 2, "ng-template", 3);
    i0.ɵɵtext(4, "\n        ");
    i0.ɵɵtemplate(5, AdaptTableHeaderCellComponent_ng_container_6_ng_template_5_Template, 4, 1, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵelementStart(7, "span", 5);
    i0.ɵɵtext(8, "\u00A0");
    i0.ɵɵtemplate(9, AdaptTableHeaderCellComponent_ng_container_6_span_9_Template, 4, 8, "span", 6);
    i0.ɵɵtext(10, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n      ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const columnConfig_r2 = ctx.$implicit;
    const _r5 = i0.ɵɵreference(6);
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", columnConfig_r2.headerTitleTemplate)("ngIfElse", _r5);
    i0.ɵɵadvance(6);
    i0.ɵɵproperty("ngIf", ctx_r0.table.isSortableColumn(columnConfig_r2));
} }
function AdaptTableHeaderCellComponent_adapt_table_header_cell_menu_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "adapt-table-header-cell-menu", 12);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("table", ctx_r1.table)("headerCellMenuTexts", ctx_r1.texts)("columns", ctx_r1.renderingColumn.columnsConfigs);
} }
export class AdaptTableHeaderCellComponent {
    constructor(table) {
        this.table = table;
        this.tableClasses = AdaptTableClasses;
    }
    getColumnHeaderTemplateParams(column) {
        return {
            column: column,
            index: this.index,
        };
    }
    getFilterableColumns(columns) {
        return columns.filter(column => this.table.isFilterableColumn(column));
    }
    getDisplayNumberOfSortingOrder(column) {
        const multiSortMeta = this.table.multiSortMeta || [];
        if (this.table.isMultipleSortMode && multiSortMeta.length >= 2) {
            const sortField = column.field;
            const indexOfSort = multiSortMeta.findIndex(sortMeta => sortMeta.field === sortField);
            return indexOfSort >= 0 ? indexOfSort + 1 : null;
        }
        return null;
    }
    isSortedColumn(column) {
        return this.table.isSorted(column.field);
    }
    getSortedColumnSvgViewBox(column) {
        return this.table.isColumnSortOrderAsc(column.field) ? '0 -6 18 24' : '0 6 18 24';
    }
}
AdaptTableHeaderCellComponent.ɵfac = function AdaptTableHeaderCellComponent_Factory(t) { return new (t || AdaptTableHeaderCellComponent)(i0.ɵɵdirectiveInject(i1.AdaptTableComponent)); };
AdaptTableHeaderCellComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptTableHeaderCellComponent, selectors: [["adapt-table-header-cell"]], inputs: { renderingColumn: "renderingColumn", index: "index", texts: "texts" }, decls: 16, vars: 14, consts: [[4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "table", "headerCellMenuTexts", "columns", 4, "ngIf"], [3, "columns", "index", "texts"], [3, "ngIf", "ngIfElse"], ["headerTitleDefaultTemplate", ""], [1, "c-header__sort-icon-wrapper"], ["class", "c-header__sort-icon", 3, "ngClass", "pSortableColumnDisabled", "pSortableColumn", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "c-header__separator"], [1, "c-header__sort-icon", 3, "ngClass", "pSortableColumnDisabled", "pSortableColumn"], ["class", "at__sorted-column-number", 4, "ngIf"], [1, "at__sorted-column-number"], [3, "table", "headerCellMenuTexts", "columns"]], template: function AdaptTableHeaderCellComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div");
        i0.ɵɵtext(1, "\n  ");
        i0.ɵɵelementStart(2, "div");
        i0.ɵɵtext(3, "\n    ");
        i0.ɵɵelementStart(4, "div");
        i0.ɵɵtext(5, "\n      ");
        i0.ɵɵtemplate(6, AdaptTableHeaderCellComponent_ng_container_6_Template, 12, 3, "ng-container", 0);
        i0.ɵɵtext(7, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(8, "\n\n    ");
        i0.ɵɵtemplate(9, AdaptTableHeaderCellComponent_adapt_table_header_cell_menu_9_Template, 2, 3, "adapt-table-header-cell-menu", 1);
        i0.ɵɵtext(10, "\n\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(11, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(12, "\n\n");
        i0.ɵɵelementStart(13, "adapt-table-filters", 2);
        i0.ɵɵtext(14, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(15, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(2);
        i0.ɵɵclassMap(ctx.tableClasses.HeaderContainer);
        i0.ɵɵadvance(2);
        i0.ɵɵclassMapInterpolate1("", ctx.tableClasses.HeaderTitle, " text-url-break-out");
        i0.ɵɵclassProp("c-header__separator_no-wrap", !ctx.table.wrapCellText);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngForOf", ctx.renderingColumn.columnsConfigs)("ngForTrackBy", ctx.table.trackByForColumns);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.table.isMultipleSortMode && ctx.table.isSortableRenderingColumn(ctx.renderingColumn));
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("columns", ctx.getFilterableColumns(ctx.renderingColumn.columnsConfigs))("index", ctx.index)("texts", ctx.texts);
    } }, directives: [i2.NgForOf, i2.NgIf, i3.AdaptTableFiltersComponent, i2.NgTemplateOutlet, i2.NgClass, i4.SortableColumn, i5.AdaptTableHeaderCellMenuComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTableHeaderCellComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-table-header-cell',
                templateUrl: './table-header-cell.html'
            }]
    }], function () { return [{ type: i1.AdaptTableComponent }]; }, { renderingColumn: [{
            type: Input
        }], index: [{
            type: Input
        }], texts: [{
            type: Input
        }] }); })();
