// @bmc-ux/adapt-angular/esm2015/empty-state/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptEmptyStateComponent } from './empty-state.component';
import * as i0 from "@angular/core";
export class AdaptEmptyStateModule {
}
AdaptEmptyStateModule.ɵfac = function AdaptEmptyStateModule_Factory(t) { return new (t || AdaptEmptyStateModule)(); };
AdaptEmptyStateModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptEmptyStateModule });
AdaptEmptyStateModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptEmptyStateModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [AdaptEmptyStateComponent],
                declarations: [AdaptEmptyStateComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptEmptyStateModule, { declarations: [AdaptEmptyStateComponent], imports: [CommonModule], exports: [AdaptEmptyStateComponent] }); })();
export * from './empty-state.component';
export * from './empty-state.types';
export * from './empty-state.common';
