// @bmc-ux/adapt-angular/esm2015/slider/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptSliderComponent } from './slider.component';
import { AdaptTooltipModule } from '../tooltip/index';
import * as i0 from "@angular/core";
export class AdaptSliderModule {
}
AdaptSliderModule.ɵfac = function AdaptSliderModule_Factory(t) { return new (t || AdaptSliderModule)(); };
AdaptSliderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSliderModule });
AdaptSliderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptTooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSliderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptTooltipModule],
                exports: [AdaptSliderComponent],
                declarations: [AdaptSliderComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSliderModule, { declarations: [AdaptSliderComponent], imports: [CommonModule, AdaptTooltipModule], exports: [AdaptSliderComponent] }); })();
export * from './slider.component';
