// @bmc-ux/adapt-angular/esm2015/slider/slider.component.js
import { ChangeDetectorRef, Component, ElementRef, forwardRef, Input, NgZone, Renderer2, ViewChild, ViewEncapsulation } from '@angular/core';
import { Slider } from 'primeng/slider';
import { DomHandler } from 'primeng/dom';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularService } from '../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "primeng/dom";
import * as i2 from "../common/radar/adapt-radar";
import * as i3 from "@angular/common";
import * as i4 from "../tooltip/tooltip.directive";
const _c0 = ["bar"];
const _c1 = function (a0) { return { "width": a0 }; };
function AdaptSliderComponent_span_3_Template(rf, ctx) { if (rf & 1) {
    const _r15 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "span", 12);
    i0.ɵɵlistener("mousedown", function AdaptSliderComponent_span_3_Template_span_mousedown_0_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r14 = i0.ɵɵnextContext(); return ctx_r14.onMouseDown($event); })("touchstart", function AdaptSliderComponent_span_3_Template_span_touchstart_0_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r16 = i0.ɵɵnextContext(); return ctx_r16.onTouchStart($event); })("touchmove", function AdaptSliderComponent_span_3_Template_span_touchmove_0_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r17 = i0.ɵɵnextContext(); return ctx_r17.onTouchMove($event); })("touchend", function AdaptSliderComponent_span_3_Template_span_touchend_0_listener() { i0.ɵɵrestoreView(_r15); const ctx_r18 = i0.ɵɵnextContext(); return ctx_r18.dragging = false; });
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵstyleProp("cursor", ctx_r1.dragging ? "-webkit-grabbing" : "")("transition", ctx_r1.dragging ? "none" : null);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(5, _c1, ctx_r1.orientation === "horizontal" ? ctx_r1.handleValue + "%" : null));
} }
const _c2 = function (a1, a2) { return { "transition": "none", "left": a1, "width": a2 }; };
function AdaptSliderComponent_span_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13);
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction2(1, _c2, ctx_r2.handleValues[0] + "%", ctx_r2.handleValues[1] - ctx_r2.handleValues[0] + "%"));
} }
const _c3 = function (a0) { return { "height": a0 }; };
function AdaptSliderComponent_span_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 14);
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(1, _c3, ctx_r3.handleValue + "%"));
} }
const _c4 = function (a0, a1) { return { "left": a0, "bottom": a1 }; };
function AdaptSliderComponent_span_9_Template(rf, ctx) { if (rf & 1) {
    const _r21 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "span", 15, 16);
    i0.ɵɵlistener("mousedown", function AdaptSliderComponent_span_9_Template_span_mousedown_0_listener($event) { i0.ɵɵrestoreView(_r21); const ctx_r20 = i0.ɵɵnextContext(); return ctx_r20.onMouseDown($event); })("touchstart", function AdaptSliderComponent_span_9_Template_span_touchstart_0_listener($event) { i0.ɵɵrestoreView(_r21); const ctx_r22 = i0.ɵɵnextContext(); return ctx_r22.onTouchStart($event); })("touchmove", function AdaptSliderComponent_span_9_Template_span_touchmove_0_listener($event) { i0.ɵɵrestoreView(_r21); const ctx_r23 = i0.ɵɵnextContext(); return ctx_r23.onTouchMove($event); })("touchend", function AdaptSliderComponent_span_9_Template_span_touchend_0_listener() { i0.ɵɵrestoreView(_r21); const ctx_r24 = i0.ɵɵnextContext(); return ctx_r24.dragging = false; });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    const _r5 = i0.ɵɵreference(12);
    i0.ɵɵstyleProp("cursor", ctx_r4.dragging ? "-webkit-grabbing" : "")("transition", ctx_r4.dragging ? "none" : null);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction2(7, _c4, ctx_r4.orientation === "horizontal" ? +ctx_r4.handleValue >= 99 ? 98 + "%" : +ctx_r4.handleValue + "%" : null, ctx_r4.orientation === "vertical" ? +ctx_r4.handleValue >= 98 ? 97 + "%" : +ctx_r4.handleValue + "%" : null))("adaptRadarDisableEventSending", true)("adaptTooltip", ctx_r4.dragging ? "" : _r5);
} }
function AdaptSliderComponent_ng_template_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0);
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext();
    i0.ɵɵtextInterpolate(ctx_r6.getTooltipValue(ctx_r6.value));
} }
const _c5 = function (a0) { return { "left": a0 }; };
const _c6 = function (a0) { return { "ui-slider-handle-active": a0 }; };
function AdaptSliderComponent_span_14_Template(rf, ctx) { if (rf & 1) {
    const _r27 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "span", 17, 18);
    i0.ɵɵlistener("mousedown", function AdaptSliderComponent_span_14_Template_span_mousedown_0_listener($event) { i0.ɵɵrestoreView(_r27); const ctx_r26 = i0.ɵɵnextContext(); return ctx_r26.onMouseDown($event, 0); })("touchstart", function AdaptSliderComponent_span_14_Template_span_touchstart_0_listener($event) { i0.ɵɵrestoreView(_r27); const ctx_r28 = i0.ɵɵnextContext(); return ctx_r28.onTouchStart($event, 0); })("touchmove", function AdaptSliderComponent_span_14_Template_span_touchmove_0_listener($event) { i0.ɵɵrestoreView(_r27); const ctx_r29 = i0.ɵɵnextContext(); return ctx_r29.onTouchMove($event, 0); })("touchend", function AdaptSliderComponent_span_14_Template_span_touchend_0_listener() { i0.ɵɵrestoreView(_r27); const ctx_r30 = i0.ɵɵnextContext(); return ctx_r30.dragging = false; });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext();
    const _r8 = i0.ɵɵreference(17);
    i0.ɵɵstyleProp("transition", ctx_r7.dragging ? "none" : null)("cursor", ctx_r7.dragging ? "-webkit-grabbing" : "");
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(8, _c5, ctx_r7.handleValues[0] + "%"))("ngClass", i0.ɵɵpureFunction1(10, _c6, ctx_r7.handleIndex === 0))("adaptRadarDisableEventSending", true)("adaptTooltip", ctx_r7.dragging ? "" : _r8);
} }
function AdaptSliderComponent_ng_template_16_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0);
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext();
    i0.ɵɵtextInterpolate(ctx_r9.getTooltipValue(ctx_r9.values[0]));
} }
function AdaptSliderComponent_ul_19_ng_container_2_li_2_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "span", 25);
} if (rf & 2) {
    const t_r34 = i0.ɵɵnextContext().$implicit;
    const ctx_r37 = i0.ɵɵnextContext(3);
    i0.ɵɵstyleProp("background", ctx_r37.handleValue >= t_r34 ? "#fff" : "#d9d9d9");
} }
function AdaptSliderComponent_ul_19_ng_container_2_li_2_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "span", 25);
} if (rf & 2) {
    i0.ɵɵstyleProp("background", "#d9d9d9");
} }
function AdaptSliderComponent_ul_19_ng_container_2_li_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 22);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵtemplate(2, AdaptSliderComponent_ul_19_ng_container_2_li_2_span_2_Template, 1, 2, "span", 23);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵtemplate(4, AdaptSliderComponent_ul_19_ng_container_2_li_2_span_4_Template, 1, 2, "span", 23);
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵelementStart(6, "span", 24);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const t_r34 = ctx.$implicit;
    const last_r35 = ctx.last;
    const i_r36 = ctx.index;
    const ctx_r33 = i0.ɵɵnextContext(3);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(4, _c5, !last_r35 ? 100 / (ctx_r33.ticks.length - 1) * i_r36 + "%" : 98 + "%"));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r33.rangeHorizontal);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r33.rangeHorizontal);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(t_r34);
} }
function AdaptSliderComponent_ul_19_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵtemplate(2, AdaptSliderComponent_ul_19_ng_container_2_li_2_Template, 9, 6, "li", 21);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r31 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r31.ticks);
} }
function AdaptSliderComponent_ul_19_ng_container_4_li_2_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "span", 25);
} if (rf & 2) {
    const t_r41 = i0.ɵɵnextContext().$implicit;
    const ctx_r44 = i0.ɵɵnextContext(3);
    i0.ɵɵstyleProp("background", ctx_r44.handleValue >= t_r41 ? "#fff" : "#d9d9d9");
} }
function AdaptSliderComponent_ul_19_ng_container_4_li_2_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "span", 25);
} if (rf & 2) {
    i0.ɵɵstyleProp("background", "#d9d9d9");
} }
const _c7 = function (a0) { return { "bottom": a0 }; };
function AdaptSliderComponent_ul_19_ng_container_4_li_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 27);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵtemplate(2, AdaptSliderComponent_ul_19_ng_container_4_li_2_span_2_Template, 1, 2, "span", 23);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵtemplate(4, AdaptSliderComponent_ul_19_ng_container_4_li_2_span_4_Template, 1, 2, "span", 23);
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵelementStart(6, "span", 24);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const t_r41 = ctx.$implicit;
    const last_r42 = ctx.last;
    const i_r43 = ctx.index;
    const ctx_r40 = i0.ɵɵnextContext(3);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(4, _c7, !last_r42 ? 100 / (ctx_r40.ticks.length - 1) * i_r43 + "%" : 98 + "%"));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r40.rangeHorizontal);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r40.rangeHorizontal);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(t_r41);
} }
function AdaptSliderComponent_ul_19_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵtemplate(2, AdaptSliderComponent_ul_19_ng_container_4_li_2_Template, 9, 6, "li", 26);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r32 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r32.ticks);
} }
function AdaptSliderComponent_ul_19_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "ul", 19);
    i0.ɵɵtext(1, "\n\n    ");
    i0.ɵɵtemplate(2, AdaptSliderComponent_ul_19_ng_container_2_Template, 4, 1, "ng-container", 20);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵtemplate(4, AdaptSliderComponent_ul_19_ng_container_4_Template, 4, 1, "ng-container", 20);
    i0.ɵɵtext(5, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("is-vertical", ctx_r10.orientation === "vertical");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r10.orientation !== "vertical");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r10.orientation === "vertical");
} }
function AdaptSliderComponent_span_21_Template(rf, ctx) { if (rf & 1) {
    const _r49 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "span", 17, 28);
    i0.ɵɵlistener("mousedown", function AdaptSliderComponent_span_21_Template_span_mousedown_0_listener($event) { i0.ɵɵrestoreView(_r49); const ctx_r48 = i0.ɵɵnextContext(); return ctx_r48.onMouseDown($event, 1); })("touchstart", function AdaptSliderComponent_span_21_Template_span_touchstart_0_listener($event) { i0.ɵɵrestoreView(_r49); const ctx_r50 = i0.ɵɵnextContext(); return ctx_r50.onTouchStart($event, 1); })("touchmove", function AdaptSliderComponent_span_21_Template_span_touchmove_0_listener($event) { i0.ɵɵrestoreView(_r49); const ctx_r51 = i0.ɵɵnextContext(); return ctx_r51.onTouchMove($event, 1); })("touchend", function AdaptSliderComponent_span_21_Template_span_touchend_0_listener() { i0.ɵɵrestoreView(_r49); const ctx_r52 = i0.ɵɵnextContext(); return ctx_r52.dragging = false; });
    i0.ɵɵtext(2, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext();
    const _r12 = i0.ɵɵreference(24);
    i0.ɵɵstyleProp("cursor", ctx_r11.dragging ? "-webkit-grabbing" : "")("transition", ctx_r11.dragging ? "none" : null);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction1(8, _c5, ctx_r11.handleValues[1] + "%"))("ngClass", i0.ɵɵpureFunction1(10, _c6, ctx_r11.handleIndex === 1))("adaptRadarDisableEventSending", true)("adaptTooltip", ctx_r11.dragging ? "" : _r12);
} }
function AdaptSliderComponent_ng_template_23_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0);
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext();
    i0.ɵɵtextInterpolate(ctx_r13.getTooltipValue(ctx_r13.values[1]));
} }
const _c8 = function (a1, a2, a3, a4) { return { "ui-slider ui-widget ui-widget-content ui-corner-all": true, "ui-state-disabled": a1, "ui-slider-horizontal": a2, "ui-slider-vertical": a3, "ui-slider-animate": a4 }; };
// TODO: Refactor this if possible
/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
export const ADAPT_SLIDER_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => AdaptSliderComponent),
    multi: true
};
export class AdaptSliderComponent extends Slider {
    constructor(el, domHandler, renderer, ngZone, cd, _adaptRadarService) {
        super(el, renderer, ngZone, cd);
        this.el = el;
        this.domHandler = domHandler;
        this.renderer = renderer;
        this._adaptRadarService = _adaptRadarService;
        /**
         * When specified, allows one boundary value to be picked
         * @docs-default false
         */
        this.rangeHorizontal = false;
        /**
         * @docs-default false
         */
        this.tick = false;
        /**
         * When specified, allows move between steps depending on click event. Depending on 'step' attr
         * @docs-default false
         */
        this.customStepChange = false;
        this.adaptRadarDisableEventSending = false;
    }
    ngAfterViewInit() {
        this._elementIntersectObserver = this._adaptRadarService.subscribeOnElementIntersect(this.adaptRadarDisableEventSending, this.el, AdaptRadarAngularCustomEventName.AdaptSlider);
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        if (this._elementIntersectObserver) {
            this._elementIntersectObserver.disconnect();
            this._elementIntersectObserver = null;
        }
    }
    moveBetweenStepRange(value) {
        const CALCULATED_VALUE = Math.round(value / this.step) * this.step;
        this.updateValue(CALCULATED_VALUE);
        this.updateHandleValue();
    }
    handleStepChange(newValue, oldValue) {
        const savedValue = newValue;
        const diff = (newValue - oldValue);
        if (diff < 0 && (diff * -1) >= this.step / 2) {
            newValue = oldValue - this.step;
            this.updateValue(newValue);
            this.updateHandleValue();
            if (this.customStepChange) {
                this.moveBetweenStepRange(savedValue);
            }
        }
        else if (diff > 0 && diff >= this.step / 2) {
            newValue = oldValue + this.step;
            this.updateValue(newValue);
            this.updateHandleValue();
            if (this.customStepChange) {
                this.moveBetweenStepRange(savedValue);
            }
        }
    }
    getTooltipValue(value) {
        return this.formatter ? this.formatter(value) : value;
    }
}
AdaptSliderComponent.ɵfac = function AdaptSliderComponent_Factory(t) { return new (t || AdaptSliderComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.DomHandler), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i2.AdaptRadarAngularService)); };
AdaptSliderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptSliderComponent, selectors: [["adapt-slider"]], viewQuery: function AdaptSliderComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.bar = _t.first);
    } }, hostVars: 2, hostBindings: function AdaptSliderComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("is-vertical", ctx.orientation === "vertical");
    } }, inputs: { rangeHorizontal: "rangeHorizontal", tick: "tick", ticks: "ticks", customStepChange: "customStepChange", formatter: "formatter", adaptRadarDisableEventSending: "adaptRadarDisableEventSending" }, features: [i0.ɵɵProvidersFeature([ADAPT_SLIDER_VALUE_ACCESSOR, DomHandler]), i0.ɵɵInheritDefinitionFeature], decls: 27, vars: 16, consts: [[3, "ngStyle", "ngClass", "click"], ["bar", ""], ["class", "ui-slider-range ui-widget-header ui-corner-all", 3, "cursor", "transition", "ngStyle", "mousedown", "touchstart", "touchmove", "touchend", 4, "ngIf"], ["class", "ui-slider-range ui-widget-header ui-corner-all", 3, "ngStyle", 4, "ngIf"], ["class", "ui-slider-range ui-slider-range-min ui-widget-header ui-corner-all", 3, "ngStyle", 4, "ngIf"], ["class", "ui-slider-handle ui-state-default ui-corner-all", "placement", "top", 3, "cursor", "transition", "ngStyle", "adaptRadarDisableEventSending", "adaptTooltip", "mousedown", "touchstart", "touchmove", "touchend", 4, "ngIf"], ["tipContent", ""], ["class", "ui-slider-handle ui-state-default ui-corner-all", "placement", "top", 3, "transition", "cursor", "ngStyle", "ngClass", "adaptRadarDisableEventSending", "adaptTooltip", "mousedown", "touchstart", "touchmove", "touchend", 4, "ngIf"], ["tipContentRange", ""], ["class", "slider-ticks", 3, "is-vertical", 4, "ngIf"], ["class", "ui-slider-handle ui-state-default ui-corner-all", "placement", "top", 3, "cursor", "transition", "ngStyle", "ngClass", "adaptRadarDisableEventSending", "adaptTooltip", "mousedown", "touchstart", "touchmove", "touchend", 4, "ngIf"], ["tipContentRange2", ""], [1, "ui-slider-range", "ui-widget-header", "ui-corner-all", 3, "ngStyle", "mousedown", "touchstart", "touchmove", "touchend"], [1, "ui-slider-range", "ui-widget-header", "ui-corner-all", 3, "ngStyle"], [1, "ui-slider-range", "ui-slider-range-min", "ui-widget-header", "ui-corner-all", 3, "ngStyle"], ["placement", "top", 1, "ui-slider-handle", "ui-state-default", "ui-corner-all", 3, "ngStyle", "adaptRadarDisableEventSending", "adaptTooltip", "mousedown", "touchstart", "touchmove", "touchend"], ["sliderHandle", ""], ["placement", "top", 1, "ui-slider-handle", "ui-state-default", "ui-corner-all", 3, "ngStyle", "ngClass", "adaptRadarDisableEventSending", "adaptTooltip", "mousedown", "touchstart", "touchmove", "touchend"], ["sliderHandleStart", ""], [1, "slider-ticks"], [4, "ngIf"], ["class", "slider-ticks__tick", 3, "ngStyle", 4, "ngFor", "ngForOf"], [1, "slider-ticks__tick", 3, "ngStyle"], ["class", "slider-ticks__line", 3, "background", 4, "ngIf"], [1, "slider-ticks__value"], [1, "slider-ticks__line"], ["class", "slider-ticks__tick slider-ticks__tick--vertical", 3, "ngStyle", 4, "ngFor", "ngForOf"], [1, "slider-ticks__tick", "slider-ticks__tick--vertical", 3, "ngStyle"], ["sliderHandleEnd", ""]], template: function AdaptSliderComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0, 1);
        i0.ɵɵlistener("click", function AdaptSliderComponent_Template_div_click_0_listener($event) { return ctx.onBarClick($event); });
        i0.ɵɵtext(2, "\n\n  ");
        i0.ɵɵtemplate(3, AdaptSliderComponent_span_3_Template, 2, 7, "span", 2);
        i0.ɵɵtext(4, "\n\n  ");
        i0.ɵɵtemplate(5, AdaptSliderComponent_span_5_Template, 2, 4, "span", 3);
        i0.ɵɵtext(6, "\n\n  ");
        i0.ɵɵtemplate(7, AdaptSliderComponent_span_7_Template, 2, 3, "span", 4);
        i0.ɵɵtext(8, "\n\n  ");
        i0.ɵɵtemplate(9, AdaptSliderComponent_span_9_Template, 2, 10, "span", 5);
        i0.ɵɵtext(10, "\n  ");
        i0.ɵɵtemplate(11, AdaptSliderComponent_ng_template_11_Template, 1, 1, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(13, "\n\n  ");
        i0.ɵɵtemplate(14, AdaptSliderComponent_span_14_Template, 2, 12, "span", 7);
        i0.ɵɵtext(15, "\n  ");
        i0.ɵɵtemplate(16, AdaptSliderComponent_ng_template_16_Template, 1, 1, "ng-template", null, 8, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(18, "\n\n  ");
        i0.ɵɵtemplate(19, AdaptSliderComponent_ul_19_Template, 6, 4, "ul", 9);
        i0.ɵɵtext(20, "\n\n  ");
        i0.ɵɵtemplate(21, AdaptSliderComponent_span_21_Template, 3, 12, "span", 10);
        i0.ɵɵtext(22, "\n  ");
        i0.ɵɵtemplate(23, AdaptSliderComponent_ng_template_23_Template, 1, 1, "ng-template", null, 11, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(25, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(26, "\n");
    } if (rf & 2) {
        i0.ɵɵclassMap(ctx.styleClass);
        i0.ɵɵproperty("ngStyle", ctx.style)("ngClass", i0.ɵɵpureFunction4(11, _c8, ctx.disabled, ctx.orientation === "horizontal", ctx.orientation === "vertical", ctx.animate));
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.rangeHorizontal);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.range);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.orientation === "vertical");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx.range);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", ctx.range);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", ctx.ticks && ctx.ticks.length > 0);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.range);
    } }, directives: [i3.NgStyle, i3.NgClass, i3.NgIf, i4.AdaptTooltipDirective, i3.NgForOf], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSliderComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-slider',
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {
                    '[class.is-vertical]': 'orientation === "vertical"'
                },
                templateUrl: './slider.component.html',
                providers: [ADAPT_SLIDER_VALUE_ACCESSOR, DomHandler],
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.DomHandler }, { type: i0.Renderer2 }, { type: i0.NgZone }, { type: i0.ChangeDetectorRef }, { type: i2.AdaptRadarAngularService }]; }, { rangeHorizontal: [{
            type: Input
        }], tick: [{
            type: Input
        }], ticks: [{
            type: Input
        }], customStepChange: [{
            type: Input
        }], formatter: [{
            type: Input
        }], adaptRadarDisableEventSending: [{
            type: Input
        }], bar: [{
            type: ViewChild,
            args: ['bar', { static: true }]
        }] }); })();
