// @helix/platform/esm2015/ui-kit/json-viewer/json-viewer.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RxJsonViewerComponent } from './json-viewer/json-viewer.component';
import * as i0 from "@angular/core";
export class RxJsonViewerModule {
}
RxJsonViewerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxJsonViewerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxJsonViewerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxJsonViewerModule, declarations: [RxJsonViewerComponent], imports: [CommonModule], exports: [RxJsonViewerComponent] });
RxJsonViewerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxJsonViewerModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxJsonViewerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [RxJsonViewerComponent],
                    exports: [RxJsonViewerComponent]
                }]
        }] });
