// @helix/platform/esm2015/view/components/field-components/selection-field/design/selection-field-design.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptRxRadiobuttonModule, AdaptRxSelectModule } from '@bmc-ux/adapt-angular';
import { SelectionFieldDesignComponent } from './selection-field-design.component';
import { FormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class SelectionFieldDesignModule {
}
SelectionFieldDesignModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectionFieldDesignModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SelectionFieldDesignModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectionFieldDesignModule, declarations: [SelectionFieldDesignComponent], imports: [CommonModule, AdaptRxRadiobuttonModule, FormsModule, AdaptRxSelectModule] });
SelectionFieldDesignModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectionFieldDesignModule, imports: [[CommonModule, AdaptRxRadiobuttonModule, FormsModule, AdaptRxSelectModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectionFieldDesignModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AdaptRxRadiobuttonModule, FormsModule, AdaptRxSelectModule],
                    declarations: [SelectionFieldDesignComponent],
                    entryComponents: [SelectionFieldDesignComponent]
                }]
        }] });
