// @bmc-ux/adapt-angular/esm2015/common/focus-key-manager/index.js
import { NgModule } from '@angular/core';
import { AdaptFocusKeyManagerItemDirective } from './adapt-focus-key-manager-item.directive';
import * as i0 from "@angular/core";
export class AdaptFocusKeyManagerModule {
}
AdaptFocusKeyManagerModule.ɵfac = function AdaptFocusKeyManagerModule_Factory(t) { return new (t || AdaptFocusKeyManagerModule)(); };
AdaptFocusKeyManagerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptFocusKeyManagerModule });
AdaptFocusKeyManagerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFocusKeyManagerModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [
                    AdaptFocusKeyManagerItemDirective
                ],
                exports: [
                    AdaptFocusKeyManagerItemDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptFocusKeyManagerModule, { declarations: [AdaptFocusKeyManagerItemDirective], exports: [AdaptFocusKeyManagerItemDirective] }); })();
export * from './adapt-focus-key-manager-item.directive';
