// @helix/platform/esm2015/view/components/field-components/integer-field/design/integer-field-design.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptRxTextfieldModule } from '@bmc-ux/adapt-angular';
import { IntegerFieldDesignComponent } from './integer-field-design.component';
import * as i0 from "@angular/core";
export class IntegerFieldDesignModule {
}
IntegerFieldDesignModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: IntegerFieldDesignModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
IntegerFieldDesignModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: IntegerFieldDesignModule, declarations: [IntegerFieldDesignComponent], imports: [CommonModule, AdaptRxTextfieldModule, FormsModule] });
IntegerFieldDesignModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: IntegerFieldDesignModule, imports: [[CommonModule, AdaptRxTextfieldModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: IntegerFieldDesignModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AdaptRxTextfieldModule, FormsModule],
                    declarations: [IntegerFieldDesignComponent],
                    entryComponents: [IntegerFieldDesignComponent]
                }]
        }] });
