// @bmc-ux/adapt-angular/esm2015/subnav/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptDropdownModule } from '../dropdown/index';
import { AdaptSubnavComponent } from './subnav.component';
import { AdaptSubnavDropdownComponent } from './subnav-dropdown.component';
import { AdaptSubnavItemComponent } from './subnav-item.component';
import * as i0 from "@angular/core";
export class AdaptSubnavModule {
}
AdaptSubnavModule.ɵfac = function AdaptSubnavModule_Factory(t) { return new (t || AdaptSubnavModule)(); };
AdaptSubnavModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSubnavModule });
AdaptSubnavModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptDropdownModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSubnavModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptDropdownModule
                ],
                declarations: [
                    AdaptSubnavComponent,
                    AdaptSubnavDropdownComponent,
                    AdaptSubnavItemComponent
                ],
                exports: [
                    AdaptSubnavComponent,
                    AdaptSubnavDropdownComponent,
                    AdaptSubnavItemComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSubnavModule, { declarations: [AdaptSubnavComponent,
        AdaptSubnavDropdownComponent,
        AdaptSubnavItemComponent], imports: [CommonModule,
        AdaptDropdownModule], exports: [AdaptSubnavComponent,
        AdaptSubnavDropdownComponent,
        AdaptSubnavItemComponent] }); })();
export * from './subnav-dropdown.component';
export * from './subnav-item.component';
export * from './subnav.component';
