// @bmc-ux/adapt-angular/esm2015/profile/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptDropdownModule } from '../dropdown/index';
import { AdaptListKeyManagerModule } from '../common/list-key-manager/index';
import { AdaptProfileComponent } from './profile.component';
import { AdaptProfileAvatarComponent } from './profile-avatar.component';
import * as i0 from "@angular/core";
export class AdaptProfileModule {
}
AdaptProfileModule.ɵfac = function AdaptProfileModule_Factory(t) { return new (t || AdaptProfileModule)(); };
AdaptProfileModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptProfileModule });
AdaptProfileModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptDropdownModule,
            AdaptListKeyManagerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptProfileModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptDropdownModule,
                    AdaptListKeyManagerModule
                ],
                exports: [AdaptProfileComponent, AdaptProfileAvatarComponent],
                declarations: [AdaptProfileComponent, AdaptProfileAvatarComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptProfileModule, { declarations: [AdaptProfileComponent, AdaptProfileAvatarComponent], imports: [CommonModule,
        AdaptDropdownModule,
        AdaptListKeyManagerModule], exports: [AdaptProfileComponent, AdaptProfileAvatarComponent] }); })();
export * from './profile.component';
export * from './profile-avatar.component';
export * from './profile.model';
