// @bmc-ux/adapt-angular/esm2015/profile/profile.component.js
import { Component, ElementRef, Input, NgZone, ViewChild } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { DOWN_ARROW, TAB } from '@angular/cdk/keycodes';
import { ReplaySubject, Subject } from 'rxjs';
import { take, takeUntil } from 'rxjs/operators';
import { AdaptFocusMonitorService } from '../common/focus-monitor.service';
import { AdaptTranslateService } from '../common/i18n/index';
import { getEventKeyCode } from '../common/key-codes';
import { AdaptListKeyManagerDirective } from '../common/list-key-manager';
import { AdaptDeviceDetectionService } from '../common/device-detection.service';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularCustomEventType, AdaptRadarAngularService, AdaptRadarAngularGenericDirective } from '../common/radar/adapt-radar';
import { AdaptDockedPanelService } from '../docked-panel/index';
import { AdaptDropdownDirective } from '../dropdown/index';
import * as i0 from "@angular/core";
import * as i1 from "../docked-panel/index";
import * as i2 from "../common/i18n/index";
import * as i3 from "@angular/cdk/a11y";
import * as i4 from "../common/radar/adapt-radar";
import * as i5 from "../common/device-detection.service";
import * as i6 from "@angular/common";
import * as i7 from "../dropdown/dropdown.component";
import * as i8 from "./profile-avatar.component";
import * as i9 from "../common/list-key-manager/adapt-list-key-manager.directive";
import * as i10 from "../common/list-key-manager/adapt-list-key-manager-item.directive";
const _c0 = ["dropdownDirective"];
const _c1 = ["adaptListKeyManagerDirective"];
function AdaptProfileComponent_div_1_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelement(2, "adapt-profile-avatar", 8);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("profileImage", ctx_r3.profile.image)("profileUserName", ctx_r3.profile.userName)("formatter", ctx_r3.profile.formatter);
} }
function AdaptProfileComponent_div_1_span_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 9);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "span", 10);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "span", 11);
    i0.ɵɵtext(6);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.profile.desc);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.profile.userName);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_4_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n          ");
    i0.ɵɵelement(2, "adapt-profile-avatar", 18);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("customClass", "a-profile__avatar--big")("profileImage", ctx_r11.profile.image)("profileUserName", ctx_r11.profile.userName)("formatter", ctx_r11.profile.formatter);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 17);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵtemplate(2, AdaptProfileComponent_div_1_ng_template_12_li_4_ng_container_2_Template, 4, 4, "ng-container", 5);
    i0.ɵɵtext(3, "\n\n        ");
    i0.ɵɵelementStart(4, "span", 9);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementStart(6, "span", 10);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n          ");
    i0.ɵɵelementStart(9, "span", 11);
    i0.ɵɵtext(10);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r7.profile.hideAvatar);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r7.profile.desc);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r7.profile.userName);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_8_a_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 23);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(4);
    i0.ɵɵpropertyInterpolate("href", ctx_r12.profile.view.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵproperty("id", ctx_r12.profile.view.id);
    i0.ɵɵattribute("target", ctx_r12.getTarget(ctx_r12.profile.view));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r12.profile.view.title);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_8_button_9_Template(rf, ctx) { if (rf & 1) {
    const _r15 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 24);
    i0.ɵɵlistener("click", function AdaptProfileComponent_div_1_ng_template_12_li_8_button_9_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r14 = i0.ɵɵnextContext(4); return ctx_r14.profile.view.action($event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("id", ctx_r13.profile.view.id);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r13.profile.view.title);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelementStart(2, "span", 19);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n        ");
    i0.ɵɵelementStart(5, "span", 20);
    i0.ɵɵtext(6, "\n\n        ");
    i0.ɵɵtemplate(7, AdaptProfileComponent_div_1_ng_template_12_li_8_a_7_Template, 2, 4, "a", 21);
    i0.ɵɵtext(8, "\n\n        ");
    i0.ɵɵtemplate(9, AdaptProfileComponent_div_1_ng_template_12_li_8_button_9_Template, 2, 2, "button", 22);
    i0.ɵɵtext(10, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r8.profile.header);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r8.profile.view.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r8.profile.view.action);
} }
const _c2 = function (a0) { return { "active": a0 }; };
function AdaptProfileComponent_div_1_ng_template_12_li_10_a_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 30);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r16 = i0.ɵɵnextContext().$implicit;
    const ctx_r18 = i0.ɵɵnextContext(3);
    i0.ɵɵclassMapInterpolate1("a-dropdown__link dropdown-item ", pItem_r16.className, "");
    i0.ɵɵpropertyInterpolate("href", pItem_r16.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵproperty("item", pItem_r16)("id", pItem_r16.id)("ngClass", i0.ɵɵpureFunction1(9, _c2, pItem_r16.active));
    i0.ɵɵattribute("target", ctx_r18.getTarget(pItem_r16));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(pItem_r16.name);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_10_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r25 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 31);
    i0.ɵɵlistener("click", function AdaptProfileComponent_div_1_ng_template_12_li_10_button_4_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r25); const pItem_r16 = i0.ɵɵnextContext().$implicit; return pItem_r16.action($event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r16 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-dropdown__link dropdown-item ", pItem_r16.className, "");
    i0.ɵɵproperty("item", pItem_r16)("id", pItem_r16.id)("ngClass", i0.ɵɵpureFunction1(7, _c2, pItem_r16.active));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", pItem_r16.name, "\n        ");
} }
function AdaptProfileComponent_div_1_ng_template_12_li_10_div_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r16 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-dropdown__link-disabled dropdown-item disabled ", pItem_r16.className, "");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", pItem_r16.name, "\n        ");
} }
function AdaptProfileComponent_div_1_ng_template_12_li_10_div_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "div", 32);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 25);
    i0.ɵɵtext(1, "\n\n        ");
    i0.ɵɵtemplate(2, AdaptProfileComponent_div_1_ng_template_12_li_10_a_2_Template, 2, 11, "a", 26);
    i0.ɵɵtext(3, "\n\n        ");
    i0.ɵɵtemplate(4, AdaptProfileComponent_div_1_ng_template_12_li_10_button_4_Template, 2, 9, "button", 27);
    i0.ɵɵtext(5, "\n\n        ");
    i0.ɵɵtemplate(6, AdaptProfileComponent_div_1_ng_template_12_li_10_div_6_Template, 2, 4, "div", 28);
    i0.ɵɵtext(7, "\n\n        ");
    i0.ɵɵtemplate(8, AdaptProfileComponent_div_1_ng_template_12_li_10_div_8_Template, 1, 0, "div", 29);
    i0.ɵɵtext(9, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r16 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !pItem_r16.disabled && !pItem_r16.action && !pItem_r16.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !pItem_r16.disabled && pItem_r16.action && !pItem_r16.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", pItem_r16.disabled && !pItem_r16.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", pItem_r16.separated && !pItem_r16.hide);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_12_a_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 35);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r28 = i0.ɵɵnextContext(4);
    i0.ɵɵpropertyInterpolate("href", ctx_r28.profile.signOut.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵproperty("id", ctx_r28.profile.signOut.id);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r28.profile.signOut.title);
} }
function AdaptProfileComponent_div_1_ng_template_12_li_12_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r31 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 36);
    i0.ɵɵlistener("click", function AdaptProfileComponent_div_1_ng_template_12_li_12_button_4_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r31); const ctx_r30 = i0.ɵɵnextContext(4); return ctx_r30.profile.signOut.action($event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r29 = i0.ɵɵnextContext(4);
    i0.ɵɵproperty("id", ctx_r29.profile.signOut.id);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", ctx_r29.profile.signOut.title, "\n        ");
} }
function AdaptProfileComponent_div_1_ng_template_12_li_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 25);
    i0.ɵɵtext(1, "\n\n        ");
    i0.ɵɵtemplate(2, AdaptProfileComponent_div_1_ng_template_12_li_12_a_2_Template, 2, 3, "a", 33);
    i0.ɵɵtext(3, "\n\n        ");
    i0.ɵɵtemplate(4, AdaptProfileComponent_div_1_ng_template_12_li_12_button_4_Template, 2, 2, "button", 34);
    i0.ɵɵtext(5, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r10.profile.signOut.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r10.profile.signOut.action);
} }
function AdaptProfileComponent_div_1_ng_template_12_Template(rf, ctx) { if (rf & 1) {
    const _r33 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n    ");
    i0.ɵɵelementStart(1, "ul", 12, 13);
    i0.ɵɵlistener("keyManagerStateChange", function AdaptProfileComponent_div_1_ng_template_12_Template_ul_keyManagerStateChange_1_listener($event) { i0.ɵɵrestoreView(_r33); const ctx_r32 = i0.ɵɵnextContext(2); return ctx_r32.keyManagerStateChange($event); })("keydown", function AdaptProfileComponent_div_1_ng_template_12_Template_ul_keydown_1_listener($event) { i0.ɵɵrestoreView(_r33); const ctx_r34 = i0.ɵɵnextContext(2); return ctx_r34.dropdownMenuKeydown($event); });
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵtemplate(4, AdaptProfileComponent_div_1_ng_template_12_li_4_Template, 13, 3, "li", 14);
    i0.ɵɵtext(5, "\n\n      ");
    i0.ɵɵprojection(6, 1);
    i0.ɵɵtext(7, "\n\n      ");
    i0.ɵɵtemplate(8, AdaptProfileComponent_div_1_ng_template_12_li_8_Template, 12, 3, "li", 5);
    i0.ɵɵtext(9, "\n\n      ");
    i0.ɵɵtemplate(10, AdaptProfileComponent_div_1_ng_template_12_li_10_Template, 10, 4, "li", 15);
    i0.ɵɵtext(11, "\n\n      ");
    i0.ɵɵtemplate(12, AdaptProfileComponent_div_1_ng_template_12_li_12_Template, 6, 2, "li", 16);
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n  ");
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("a-dropdown-menu__mobile", !ctx_r5.isDesktop);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !ctx_r5.profile.typeLong);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r5.profile.view);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r5.profile.menu);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r5.profile.signOut);
} }
const _c3 = function (a0, a1, a2) { return { "a-profile--dark": a0, "a-profile--long": a1, "a-profile--small": a2 }; };
const _c4 = function (a0) { return { "a-profile__menu--small": a0 }; };
function AdaptProfileComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r36 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 2, 3);
    i0.ɵɵlistener("onOpen", function AdaptProfileComponent_div_1_Template_div_onOpen_0_listener() { i0.ɵɵrestoreView(_r36); const ctx_r35 = i0.ɵɵnextContext(); return ctx_r35.handleProfileDropdownOpen(); })("onClose", function AdaptProfileComponent_div_1_Template_div_onClose_0_listener() { i0.ɵɵrestoreView(_r36); const ctx_r37 = i0.ɵɵnextContext(); return ctx_r37.handleProfileDropdownClose(); });
    i0.ɵɵtext(2, "\n\n  ");
    i0.ɵɵelementStart(3, "button", 4);
    i0.ɵɵlistener("keydown", function AdaptProfileComponent_div_1_Template_button_keydown_3_listener($event) { i0.ɵɵrestoreView(_r36); const ctx_r38 = i0.ɵɵnextContext(); return ctx_r38.dropdownMenuToggleKeydown($event); });
    i0.ɵɵtext(4, "\n\n    ");
    i0.ɵɵtemplate(5, AdaptProfileComponent_div_1_ng_container_5_Template, 4, 3, "ng-container", 5);
    i0.ɵɵtext(6, "\n\n    ");
    i0.ɵɵtemplate(7, AdaptProfileComponent_div_1_span_7_Template, 8, 2, "span", 6);
    i0.ɵɵtext(8, "\n\n    ");
    i0.ɵɵprojection(9);
    i0.ɵɵtext(10, "\n\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n\n  ");
    i0.ɵɵtemplate(12, AdaptProfileComponent_div_1_ng_template_12_Template, 15, 6, "ng-template", 7);
    i0.ɵɵtext(13, "\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("holdFocusInMenu", true)("autoFocusFirst", false)("restoreFocusAfterClose", true)("focusNextElementAfterClose", true)("placement", "bottom-right")("ngClass", i0.ɵɵpureFunction3(10, _c3, !ctx_r0.profile.inverted, ctx_r0.profile.typeLong, !ctx_r0.profile.typeLong));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(14, _c4, !ctx_r0.profile.typeLong));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r0.profile.hideAvatar);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0.profile.typeLong);
} }
function AdaptProfileComponent_div_3_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelement(2, "adapt-profile-avatar", 18);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r39 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("customClass", "a-profile__avatar--big")("profileImage", ctx_r39.profile.image)("profileUserName", ctx_r39.profile.userName)("formatter", ctx_r39.profile.formatter);
} }
function AdaptProfileComponent_div_3_a_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 44);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r40 = i0.ɵɵnextContext(2);
    i0.ɵɵpropertyInterpolate("href", ctx_r40.profile.signOut.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵproperty("id", ctx_r40.profile.signOut.id);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("\n         ", ctx_r40.profile.signOut.title, "");
} }
function AdaptProfileComponent_div_3_button_16_Template(rf, ctx) { if (rf & 1) {
    const _r45 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 45);
    i0.ɵɵlistener("click", function AdaptProfileComponent_div_3_button_16_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r45); const ctx_r44 = i0.ɵɵnextContext(2); ctx_r44.profile.signOut.action($event); return ctx_r44.closeHamburger(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r41 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("id", ctx_r41.profile.signOut.id);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("\n              ", ctx_r41.profile.signOut.title, "\n      ");
} }
function AdaptProfileComponent_div_3_li_26_a_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 51);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r46 = i0.ɵɵnextContext(3);
    i0.ɵɵpropertyInterpolate("href", ctx_r46.profile.view.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵproperty("id", ctx_r46.profile.view.id);
    i0.ɵɵattribute("target", ctx_r46.getTarget(ctx_r46.profile.view));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r46.profile.view.title);
} }
function AdaptProfileComponent_div_3_li_26_button_9_Template(rf, ctx) { if (rf & 1) {
    const _r49 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 52);
    i0.ɵɵlistener("click", function AdaptProfileComponent_div_3_li_26_button_9_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r49); const ctx_r48 = i0.ɵɵnextContext(3); ctx_r48.profile.view.action($event); return ctx_r48.closeHamburger(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r47 = i0.ɵɵnextContext(3);
    i0.ɵɵproperty("id", ctx_r47.profile.view.id);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r47.profile.view.title);
} }
function AdaptProfileComponent_div_3_li_26_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 46);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "span", 47);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "span", 48);
    i0.ɵɵtext(6, "\n\n        ");
    i0.ɵɵtemplate(7, AdaptProfileComponent_div_3_li_26_a_7_Template, 2, 4, "a", 49);
    i0.ɵɵtext(8, "\n\n        ");
    i0.ɵɵtemplate(9, AdaptProfileComponent_div_3_li_26_button_9_Template, 2, 2, "button", 50);
    i0.ɵɵtext(10, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r42 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r42.profile.header);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r42.profile.view.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r42.profile.view.action);
} }
function AdaptProfileComponent_div_3_li_28_a_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 56);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r50 = i0.ɵɵnextContext().$implicit;
    const ctx_r52 = i0.ɵɵnextContext(2);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link ", pItem_r50.className, "");
    i0.ɵɵpropertyInterpolate("href", pItem_r50.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵproperty("id", pItem_r50.id)("ngClass", i0.ɵɵpureFunction1(8, _c2, pItem_r50.active));
    i0.ɵɵattribute("target", ctx_r52.getTarget(pItem_r50));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(pItem_r50.name);
} }
function AdaptProfileComponent_div_3_li_28_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r58 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 57);
    i0.ɵɵlistener("click", function AdaptProfileComponent_div_3_li_28_button_4_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r58); const pItem_r50 = i0.ɵɵnextContext().$implicit; const ctx_r56 = i0.ɵɵnextContext(2); pItem_r50.action($event); return ctx_r56.closeHamburger(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r50 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link ", pItem_r50.className, "");
    i0.ɵɵproperty("id", pItem_r50.id)("ngClass", i0.ɵɵpureFunction1(6, _c2, pItem_r50.active));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", pItem_r50.name, "\n      ");
} }
function AdaptProfileComponent_div_3_li_28_div_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r50 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link disabled ", pItem_r50.className, "");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", pItem_r50.name, "\n      ");
} }
function AdaptProfileComponent_div_3_li_28_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 53);
    i0.ɵɵtext(1, "\n\n      ");
    i0.ɵɵtemplate(2, AdaptProfileComponent_div_3_li_28_a_2_Template, 2, 10, "a", 54);
    i0.ɵɵtext(3, "\n\n      ");
    i0.ɵɵtemplate(4, AdaptProfileComponent_div_3_li_28_button_4_Template, 2, 8, "button", 55);
    i0.ɵɵtext(5, "\n\n      ");
    i0.ɵɵtemplate(6, AdaptProfileComponent_div_3_li_28_div_6_Template, 2, 4, "div", 28);
    i0.ɵɵtext(7, "\n\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pItem_r50 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !pItem_r50.disabled && !pItem_r50.action && !pItem_r50.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !pItem_r50.disabled && pItem_r50.action && !pItem_r50.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", pItem_r50.disabled && !pItem_r50.hide);
} }
const _c5 = function (a0) { return { "a-profile--light": a0 }; };
function AdaptProfileComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 37);
    i0.ɵɵtext(1, "\n\n  ");
    i0.ɵɵelementStart(2, "div", 38);
    i0.ɵɵtext(3, "\n\n    ");
    i0.ɵɵtemplate(4, AdaptProfileComponent_div_3_ng_container_4_Template, 4, 4, "ng-container", 5);
    i0.ɵɵtext(5, "\n\n    ");
    i0.ɵɵelementStart(6, "span", 9);
    i0.ɵɵtext(7, "\n\n      ");
    i0.ɵɵelementStart(8, "span", 10);
    i0.ɵɵtext(9);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n      ");
    i0.ɵɵelementStart(11, "span", 11);
    i0.ɵɵtext(12);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n\n      ");
    i0.ɵɵtemplate(14, AdaptProfileComponent_div_3_a_14_Template, 2, 3, "a", 39);
    i0.ɵɵtext(15, "\n\n      ");
    i0.ɵɵtemplate(16, AdaptProfileComponent_div_3_button_16_Template, 2, 2, "button", 40);
    i0.ɵɵtext(17, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(19, "\n\n  ");
    i0.ɵɵprojection(20, 2);
    i0.ɵɵtext(21, "\n  ");
    i0.ɵɵprojection(22, 3);
    i0.ɵɵtext(23, "\n\n  ");
    i0.ɵɵelementStart(24, "ul", 41);
    i0.ɵɵtext(25, "\n    ");
    i0.ɵɵtemplate(26, AdaptProfileComponent_div_3_li_26_Template, 12, 3, "li", 42);
    i0.ɵɵtext(27, "\n\n    ");
    i0.ɵɵtemplate(28, AdaptProfileComponent_div_3_li_28_Template, 8, 3, "li", 43);
    i0.ɵɵtext(29, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(30, "\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c5, ctx_r1.profile.inverted));
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", !ctx_r1.profile.hideAvatar);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r1.profile.desc);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.profile.userName);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.profile.signOut && ctx_r1.profile.signOut.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.profile.signOut && ctx_r1.profile.signOut.action);
    i0.ɵɵadvance(10);
    i0.ɵɵproperty("ngIf", ctx_r1.profile.view);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r1.profile.menu);
} }
const _c6 = [[["", "profile-btn-content", ""]], [["", "profile-menu-content", ""]], [["", "profile-btn-content-mobile", ""]], [["", "profile-menu-content-mobile", ""]]];
const _c7 = ["[profile-btn-content]", "[profile-menu-content]", "[profile-btn-content-mobile]", "[profile-menu-content-mobile]"];
let PROFILE_ITEM_INDEX = 0;
export class AdaptProfileComponent extends AdaptRadarAngularGenericDirective {
    constructor(dps, ts, _el, _focusMonitor, _ngZone, _adaptRadarService, _deviceDetector, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptProfile);
        this._el = _el;
        this._focusMonitor = _focusMonitor;
        this._ngZone = _ngZone;
        this._adaptRadarService = _adaptRadarService;
        this._deviceDetector = _deviceDetector;
        this.prefixId = 'adapt-profile-menu-item-';
        this.defaultTexts = {};
        this._openFromKeyboard = false;
        this._listKeyManagerDestroy = new Subject();
        this._destroyed$ = new ReplaySubject(1);
        this.hamburgerService = dps;
        this.setDefaultTexts(ts.getCurrentLanguage());
        this.isDesktop = this._deviceDetector.deviceDesktop();
    }
    /**
     * Returns clientWidth from the nativeElement, is used by navigation
     * @docs-default
     */
    getClientWidth() {
        return this._el.nativeElement.clientWidth;
    }
    /**
     * Closes related DockedPanel
     * @docs-default
     */
    closeHamburger() {
        this.hamburgerService.closeAllDockedPanels();
    }
    ngOnInit() {
        // init from input or default values
        this.profile.desc = this.profile.desc || this.defaultTexts.description;
        this.profile.header = this.profile.header || this.defaultTexts.header;
        if (this.profile.view) {
            this.profile.view.title = this.profile.view.title || this.defaultTexts.view;
            this.profile.view.id = this.profile.view.id || this._getPrefixId();
        }
        this.profile.menu.forEach(item => {
            item.id = item.id || this._getPrefixId();
        });
        if (this.profile.signOut) {
            this.profile.signOut.title = this.profile.signOut.title || this.defaultTexts.signOut;
            this.profile.signOut.id = this.profile.signOut.id || this._getPrefixId();
        }
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.clientWidth = this._el.nativeElement.clientWidth;
    }
    ngOnDestroy() {
        this._listKeyManagerDestroy.next();
        this._destroyed$.next(true);
        this._destroyed$.complete();
    }
    setDefaultTexts(translation) {
        this.defaultTexts.description = translation['adapt.profile.description'];
        this.defaultTexts.header = translation['adapt.profile.header'];
        this.defaultTexts.view = translation['adapt.profile.view'];
        this.defaultTexts.signOut = translation['adapt.profile.signOut'];
    }
    dropdownMenuToggleKeydown(event) {
        this._openFromKeyboard = true;
        const eventKeyCode = getEventKeyCode(event);
        if (this.popup.isOpen()) {
            if (eventKeyCode === TAB) {
                this.popup.close();
            }
            else {
                this.dropdownMenuKeydown(event);
            }
        }
        else {
            if (eventKeyCode === DOWN_ARROW) {
                event.preventDefault();
                this.popup.open(event);
            }
        }
    }
    dropdownMenuKeydown(event) {
        if (getEventKeyCode(event) === TAB) {
            event.preventDefault();
        }
        this._listKeyManager.onKeydown(event);
        this._focusMonitor.focusVia(this._listKeyManager.activeItem.getNativeElement(), 'keyboard');
    }
    getTarget(item) {
        return item.target ? item.target : '_self';
    }
    keyManagerStateChange(listKeyManager) {
        this._listKeyManager = listKeyManager;
        if (this._listKeyManager) {
            this._ngZone.onStable
                .pipe(takeUntil(this._destroyed$), take(1))
                .subscribe(() => {
                const activeIndex = this.adaptListKeyManagerDirective.items.toArray()
                    .findIndex(itemDirective => { var _a; return (_a = itemDirective.item) === null || _a === void 0 ? void 0 : _a.active; });
                if (activeIndex !== -1) {
                    this._listKeyManager.setActiveItem(activeIndex);
                }
                else {
                    this._listKeyManager.setFirstItemActive();
                }
                if (this._openFromKeyboard) {
                    this._focusMonitor.focusVia(this._listKeyManager.activeItem.getNativeElement(), 'keyboard');
                }
            });
            if (this.popup) {
                this._listKeyManager.tabOut
                    .pipe(takeUntil(this._listKeyManagerDestroy))
                    .subscribe(() => {
                    this.popup.close();
                });
            }
        }
        else {
            this._listKeyManagerDestroy.next();
        }
    }
    handleProfileDropdownOpen() {
        var _a;
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptProfile, {
            EventType: AdaptRadarAngularCustomEventType.Click,
            Type: ((_a = this.profile) === null || _a === void 0 ? void 0 : _a.typeLong) ? 'long' : 'short'
        });
    }
    handleProfileDropdownClose() {
        this._openFromKeyboard = false;
    }
    _getPrefixId() {
        return (this.profile.prefixId || this.prefixId) + PROFILE_ITEM_INDEX++;
    }
}
AdaptProfileComponent.ɵfac = function AdaptProfileComponent_Factory(t) { return new (t || AdaptProfileComponent)(i0.ɵɵdirectiveInject(i1.AdaptDockedPanelService), i0.ɵɵdirectiveInject(i2.AdaptTranslateService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i3.FocusMonitor), i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i4.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i5.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptProfileComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptProfileComponent, selectors: [["adapt-profile"]], viewQuery: function AdaptProfileComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
        i0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.popup = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.adaptListKeyManagerDirective = _t.first);
    } }, inputs: { profile: "profile", type: "type", getClientWidth: "getClientWidth", closeHamburger: "closeHamburger" }, features: [i0.ɵɵProvidersFeature([AdaptFocusMonitorService]), i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c7, decls: 5, vars: 2, consts: [["class", "a-profile dropdown", "adaptDropdown", "", 3, "adaptRadarDisableEventSending", "holdFocusInMenu", "autoFocusFirst", "restoreFocusAfterClose", "focusNextElementAfterClose", "placement", "ngClass", "onOpen", "onClose", 4, "ngIf"], ["class", "a-profile a-profile__hamburger", 3, "ngClass", 4, "ngIf"], ["adaptDropdown", "", 1, "a-profile", "dropdown", 3, "adaptRadarDisableEventSending", "holdFocusInMenu", "autoFocusFirst", "restoreFocusAfterClose", "focusNextElementAfterClose", "placement", "ngClass", "onOpen", "onClose"], ["dropdownDirective", ""], ["type", "button", "role", "button", "adaptDropdownToggle", "", "aria-label", "Profile menu", 1, "a-profile__menu", "dropdown-toggle", 3, "ngClass", "keydown"], [4, "ngIf"], ["class", "menu-profile__name", 4, "ngIf"], ["adaptDropdownMenuTemplate", ""], [3, "profileImage", "profileUserName", "formatter"], [1, "menu-profile__name"], [1, "menu-profile__signed"], [1, "menu-profile__username"], ["aria-labelledby", "profile", "adaptListKeyManager", "", 1, "a-profile-dropdown-menu", "dropdown-menu", "a-dropdown-menu", "dropdown-menu--right", "a-dropdown-menu__light", 3, "keyManagerStateChange", "keydown"], ["adaptListKeyManagerDirective", "adaptListKeyManager"], ["class", "dropdown-profile__container a-profile__menu", 4, "ngIf"], ["class", "a-dropdown__item", 4, "ngFor", "ngForOf"], ["class", "a-dropdown__item", 4, "ngIf"], [1, "dropdown-profile__container", "a-profile__menu"], [3, "customClass", "profileImage", "profileUserName", "formatter"], [1, "dropdown-profile__title"], [1, "dropdown-profile__edit"], ["adaptListKeyManagerItem", "", 3, "id", "href", 4, "ngIf"], ["adaptListKeyManagerItem", "", "type", "button", "class", "btn btn-link", 3, "id", "click", 4, "ngIf"], ["adaptListKeyManagerItem", "", 3, "id", "href"], ["adaptListKeyManagerItem", "", "type", "button", 1, "btn", "btn-link", 3, "id", "click"], [1, "a-dropdown__item"], ["adaptListKeyManagerItem", "", 3, "item", "id", "class", "ngClass", "href", 4, "ngIf"], ["adaptListKeyManagerItem", "", "type", "button", 3, "item", "id", "class", "ngClass", "click", 4, "ngIf"], [3, "class", 4, "ngIf"], ["class", "dropdown-divider", 4, "ngIf"], ["adaptListKeyManagerItem", "", 3, "item", "id", "ngClass", "href"], ["adaptListKeyManagerItem", "", "type", "button", 3, "item", "id", "ngClass", "click"], [1, "dropdown-divider"], ["adaptListKeyManagerItem", "", "class", "a-dropdown__link dropdown-item dropdown-profile__signout", 3, "href", "id", 4, "ngIf"], ["adaptListKeyManagerItem", "", "type", "button", "class", "a-dropdown__link dropdown-item dropdown-profile__signout", 3, "id", "click", 4, "ngIf"], ["adaptListKeyManagerItem", "", 1, "a-dropdown__link", "dropdown-item", "dropdown-profile__signout", 3, "href", "id"], ["adaptListKeyManagerItem", "", "type", "button", 1, "a-dropdown__link", "dropdown-item", "dropdown-profile__signout", 3, "id", "click"], [1, "a-profile", "a-profile__hamburger", 3, "ngClass"], [1, "a-profile__menu", "a-profile__menu--long"], ["class", "a-dropdown__link dropdown-profile__signout", 3, "href", "id", 4, "ngIf"], ["type", "button", "class", "a-dropdown__link dropdown-profile__signout", 3, "id", "click", 4, "ngIf"], [1, "a-hamburger__menu"], ["class", "menu-profile", 4, "ngIf"], ["class", "a-hamburger__menu-item", 4, "ngFor", "ngForOf"], [1, "a-dropdown__link", "dropdown-profile__signout", 3, "href", "id"], ["type", "button", 1, "a-dropdown__link", "dropdown-profile__signout", 3, "id", "click"], [1, "menu-profile"], [1, "menu-profile__title"], [1, "menu-profile__edit"], [3, "id", "href", 4, "ngIf"], ["type", "button", "class", "btn btn-link", 3, "id", "click", 4, "ngIf"], [3, "id", "href"], ["type", "button", 1, "btn", "btn-link", 3, "id", "click"], [1, "a-hamburger__menu-item"], [3, "id", "class", "ngClass", "href", 4, "ngIf"], ["type", "button", 3, "id", "class", "ngClass", "click", 4, "ngIf"], [3, "id", "ngClass", "href"], ["type", "button", 3, "id", "ngClass", "click"]], template: function AdaptProfileComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c6);
        i0.ɵɵtext(0, "\n");
        i0.ɵɵtemplate(1, AdaptProfileComponent_div_1_Template, 14, 16, "div", 0);
        i0.ɵɵtext(2, "\n\n");
        i0.ɵɵtemplate(3, AdaptProfileComponent_div_3_Template, 31, 10, "div", 1);
        i0.ɵɵtext(4, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.type !== "hamburger");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.type === "hamburger");
    } }, directives: [i6.NgIf, i7.AdaptDropdownDirective, i6.NgClass, i7.AdaptDropdownToggleDirective, i7.AdaptDropdownMenuTemplateDirective, i8.AdaptProfileAvatarComponent, i9.AdaptListKeyManagerDirective, i6.NgForOf, i10.AdaptListKeyManagerItemDirective], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptProfileComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-profile',
                templateUrl: './profile.component.html',
                providers: [AdaptFocusMonitorService]
            }]
    }], function () { return [{ type: i1.AdaptDockedPanelService }, { type: i2.AdaptTranslateService }, { type: i0.ElementRef }, { type: i3.FocusMonitor }, { type: i0.NgZone }, { type: i4.AdaptRadarAngularService }, { type: i5.AdaptDeviceDetectionService }, { type: i0.ElementRef }]; }, { profile: [{
            type: Input
        }], type: [{
            type: Input
        }], popup: [{
            type: ViewChild,
            args: ['dropdownDirective']
        }], adaptListKeyManagerDirective: [{
            type: ViewChild,
            args: ['adaptListKeyManagerDirective', { static: false }]
        }], getClientWidth: [{
            type: Input
        }], closeHamburger: [{
            type: Input
        }] }); })();
