// @bmc-ux/adapt-angular/esm2015/advanced-filtering/advanced-filter/advanced-filter.models.js
// TODO: maybe capitalize names
export var AdvancedFilterOptionDataType;
(function (AdvancedFilterOptionDataType) {
    AdvancedFilterOptionDataType["boolean"] = "boolean";
    AdvancedFilterOptionDataType["time"] = "time";
    AdvancedFilterOptionDataType["date"] = "date";
    AdvancedFilterOptionDataType["datetime"] = "datetime";
    AdvancedFilterOptionDataType["string"] = "string";
    AdvancedFilterOptionDataType["selection"] = "selection";
    AdvancedFilterOptionDataType["number"] = "number";
    AdvancedFilterOptionDataType["custom"] = "custom";
})(AdvancedFilterOptionDataType || (AdvancedFilterOptionDataType = {}));
export const booleanDataTypeBlankValue = 'blank';
export var ActiveSavedFilterSelectionCause;
(function (ActiveSavedFilterSelectionCause) {
    ActiveSavedFilterSelectionCause[ActiveSavedFilterSelectionCause["SelectedByUser"] = 1] = "SelectedByUser";
    ActiveSavedFilterSelectionCause[ActiveSavedFilterSelectionCause["OnFilterChange"] = 2] = "OnFilterChange";
    ActiveSavedFilterSelectionCause[ActiveSavedFilterSelectionCause["OnFiltersClear"] = 3] = "OnFiltersClear";
    ActiveSavedFilterSelectionCause[ActiveSavedFilterSelectionCause["OnSelectedFiltersSetter"] = 4] = "OnSelectedFiltersSetter";
})(ActiveSavedFilterSelectionCause || (ActiveSavedFilterSelectionCause = {}));
