// @helix/platform/esm2015/ui-kit/blade/blade.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RxBladeComponent } from './blade.component';
import * as i0 from "@angular/core";
export class RxBladeModule {
}
RxBladeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBladeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxBladeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBladeModule, declarations: [RxBladeComponent], imports: [CommonModule], exports: [RxBladeComponent] });
RxBladeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBladeModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBladeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [RxBladeComponent],
                    exports: [RxBladeComponent]
                }]
        }] });
