// @bmc-ux/adapt-table/esm2015/lib/primeng-lib/table/editable-row.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import * as i0 from "@angular/core";
// eslint-disable-next-line @angular-eslint/directive-class-suffix
export class EditableRow {
    constructor(el) {
        this.el = el;
    }
    isEnabled() {
        return this.pEditableRowDisabled !== true;
    }
}
EditableRow.ɵfac = function EditableRow_Factory(t) { return new (t || EditableRow)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
EditableRow.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: EditableRow, selectors: [["", "pEditableRow", ""]], inputs: { data: ["pEditableRow", "data"], pEditableRowDisabled: "pEditableRowDisabled" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(EditableRow, [{
        type: Directive,
        args: [{
                // eslint-disable-next-line @angular-eslint/directive-selector
                selector: '[pEditableRow]'
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { data: [{
            type: Input,
            args: ['pEditableRow']
        }], pEditableRowDisabled: [{
            type: Input
        }] }); })();
