// @bmc-ux/adapt-table/esm2015/lib/primeng-lib/table/row-toggler.directive.js
import { Directive, HostListener, Input } from '@angular/core';
import { AdaptTableComponent } from '../../adapt-table.component';
import * as i0 from "@angular/core";
import * as i1 from "../../adapt-table.component";
// eslint-disable-next-line @angular-eslint/directive-class-suffix
export class RowToggler {
    constructor(dt) {
        this.dt = dt;
    }
    onClick(event) {
        if (this.isEnabled()) {
            this.dt.toggleRow(this.data, event);
            event.preventDefault();
        }
    }
    isEnabled() {
        return this.pRowTogglerDisabled !== true;
    }
}
RowToggler.ɵfac = function RowToggler_Factory(t) { return new (t || RowToggler)(i0.ɵɵdirectiveInject(i1.AdaptTableComponent)); };
RowToggler.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: RowToggler, selectors: [["", "pRowToggler", ""]], hostBindings: function RowToggler_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function RowToggler_click_HostBindingHandler($event) { return ctx.onClick($event); });
    } }, inputs: { data: ["pRowToggler", "data"], pRowTogglerDisabled: "pRowTogglerDisabled" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(RowToggler, [{
        type: Directive,
        args: [{
                // eslint-disable-next-line @angular-eslint/directive-selector
                selector: '[pRowToggler]'
            }]
    }], function () { return [{ type: i1.AdaptTableComponent }]; }, { data: [{
            type: Input,
            args: ['pRowToggler']
        }], pRowTogglerDisabled: [{
            type: Input
        }], onClick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
