// @bmc-ux/adapt-table/esm2015/lib/primeng-lib/table/table.service.js
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class TableService {
    constructor() {
        this.sortSource = new Subject();
        this.selectionSource = new Subject();
        this.valueSource = new Subject();
        this.totalRecordsSource = new Subject();
        this.columnsSource = new Subject();
        this.viewRelatedDataSource = new Subject();
        this.sortSource$ = this.sortSource.asObservable();
        this.selectionSource$ = this.selectionSource.asObservable();
        this.valueSource$ = this.valueSource.asObservable();
        this.totalRecordsSource$ = this.totalRecordsSource.asObservable();
        this.columnsSource$ = this.columnsSource.asObservable();
        // The idea under the viewRelatedDataSource is to be triggered each time when the table view related data has changed
        this.viewRelatedDataSource$ = this.viewRelatedDataSource.asObservable();
    }
    onSort(sortMeta) {
        this.sortSource.next(sortMeta);
        this.onViewRelatedDataChange();
    }
    onSelectionChange(selection) {
        this.selectionSource.next(selection);
    }
    onValueChange(value) {
        this.valueSource.next(value);
        this.onViewRelatedDataChange();
    }
    onTotalRecordsChange(value) {
        this.totalRecordsSource.next(value);
    }
    onColumnsChange(columns) {
        this.columnsSource.next(columns);
        this.onViewRelatedDataChange();
    }
    onViewRelatedDataChange() {
        this.viewRelatedDataSource.next();
    }
}
TableService.ɵfac = function TableService_Factory(t) { return new (t || TableService)(); };
TableService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: TableService, factory: TableService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TableService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
