// @bmc-ux/adapt-table/esm2015/lib/primeng-lib/table/table.module.js
/*
* This module was copied from https://github.com/primefaces/primeng/releases/tag/9.0.5
* Reason: maintain all needed features and bug fixes on ADAPT side
* */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EditableRow } from './editable-row.directive';
import { ReorderableColumn } from './reordarable-column.directive';
import { ResizableColumn } from './resizable-column.directive';
import { RowToggler } from './row-toggler.directive';
import { ScrollableView } from './scrollable-view.component';
import { SortIcon } from './sort-icon.component';
import { SortableColumn } from './sortable-column.directive';
import { Table } from './table';
import { TableBody } from './table-body.component';
import { PaginatorModule } from 'primeng/paginator';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { AdaptTableInfiniteScrollingLoaderComponent } from '../../table-infinite-scrolling-loader/table-infinite-scrolling-loader.component';
import { AdaptBusyModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@angular/cdk/scrolling";
export class TableModule {
}
TableModule.ɵfac = function TableModule_Factory(t) { return new (t || TableModule)(); };
TableModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: TableModule });
TableModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            PaginatorModule,
            ScrollingModule,
            AdaptBusyModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(TableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    PaginatorModule,
                    ScrollingModule,
                    AdaptBusyModule
                ],
                exports: [
                    Table,
                    SortableColumn,
                    RowToggler,
                    ResizableColumn,
                    ReorderableColumn,
                    SortIcon,
                    EditableRow,
                    ScrollableView,
                    TableBody,
                    AdaptTableInfiniteScrollingLoaderComponent
                ],
                declarations: [
                    Table,
                    SortableColumn,
                    RowToggler,
                    ResizableColumn,
                    ReorderableColumn,
                    TableBody,
                    ScrollableView,
                    SortIcon,
                    EditableRow,
                    AdaptTableInfiniteScrollingLoaderComponent,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(TableModule, { declarations: [Table,
        SortableColumn,
        RowToggler,
        ResizableColumn,
        ReorderableColumn,
        TableBody,
        ScrollableView,
        SortIcon,
        EditableRow,
        AdaptTableInfiniteScrollingLoaderComponent], imports: [CommonModule,
        PaginatorModule,
        ScrollingModule,
        AdaptBusyModule], exports: [Table,
        SortableColumn,
        RowToggler,
        ResizableColumn,
        ReorderableColumn,
        SortIcon,
        EditableRow,
        ScrollableView,
        TableBody,
        AdaptTableInfiniteScrollingLoaderComponent] }); })();
i0.ɵɵsetComponentScope(ScrollableView, [i1.NgClass, i1.NgStyle, i1.NgTemplateOutlet, i1.NgForOf, i1.NgIf, TableBody,
    AdaptTableInfiniteScrollingLoaderComponent, i2.CdkVirtualScrollViewport, i2.CdkFixedSizeVirtualScroll], []);
