// @bmc-ux/adapt-angular/esm2015/steps/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptTooltipModule } from '../tooltip/index';
import { AdaptStepsComponent } from './steps.component';
import * as i0 from "@angular/core";
export class AdaptStepsModule {
}
AdaptStepsModule.ɵfac = function AdaptStepsModule_Factory(t) { return new (t || AdaptStepsModule)(); };
AdaptStepsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptStepsModule });
AdaptStepsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptTooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptStepsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptTooltipModule],
                exports: [AdaptStepsComponent],
                declarations: [AdaptStepsComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptStepsModule, { declarations: [AdaptStepsComponent], imports: [CommonModule, AdaptTooltipModule], exports: [AdaptStepsComponent] }); })();
export * from './steps.component';
