// @bmc-ux/adapt-table/esm2015/lib/shared/adapt-table-storage.service.js
import { Injectable } from '@angular/core';
import { StorageType } from '../adapt-table.models';
import { isEmpty } from 'lodash-es';
import * as i0 from "@angular/core";
export class AdaptTableStorageService {
    getData(stateKey, storageType) {
        if (storageType === StorageType.Local) {
            try {
                return JSON.parse(localStorage.getItem(stateKey));
            }
            catch (error) {
                return null;
            }
        }
        if (storageType === StorageType.Session) {
            try {
                return JSON.parse(sessionStorage.getItem(stateKey));
            }
            catch (error) {
                return null;
            }
        }
    }
    updateData(stateKey, dataToUpdate, storageType) {
        const restoredState = Object.assign(Object.assign({}, this.getData(stateKey, storageType)), dataToUpdate);
        if (restoredState) {
            // Delete empty data from storage
            Object.keys(restoredState).forEach(key => isEmpty(restoredState[key]) && delete restoredState[key]);
            if (storageType === StorageType.Local) {
                localStorage.setItem(stateKey, JSON.stringify(restoredState));
            }
            if (storageType === StorageType.Session) {
                sessionStorage.setItem(stateKey, JSON.stringify(restoredState));
            }
        }
    }
    // TODO: For future, think how to restore table state correctly
    clearData(stateKey) {
        localStorage.removeItem(stateKey);
        sessionStorage.removeItem(stateKey);
    }
}
AdaptTableStorageService.ɵfac = function AdaptTableStorageService_Factory(t) { return new (t || AdaptTableStorageService)(); };
AdaptTableStorageService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptTableStorageService, factory: AdaptTableStorageService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTableStorageService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
