// @bmc-ux/adapt-angular/esm2015/code-viewer/index.js
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { AdaptRxSwitchModule } from '../rx/rx-switch/index';
import { AdaptCodeViewerComponent } from './code-viewer.component';
import * as i0 from "@angular/core";
export class AdaptCodeViewerModule {
}
AdaptCodeViewerModule.ɵfac = function AdaptCodeViewerModule_Factory(t) { return new (t || AdaptCodeViewerModule)(); };
AdaptCodeViewerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCodeViewerModule });
AdaptCodeViewerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptRxSwitchModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCodeViewerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptRxSwitchModule, FormsModule],
                declarations: [AdaptCodeViewerComponent],
                exports: [AdaptCodeViewerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCodeViewerModule, { declarations: [AdaptCodeViewerComponent], imports: [CommonModule, AdaptRxSwitchModule, FormsModule], exports: [AdaptCodeViewerComponent] }); })();
export * from './code-viewer.component';
export * from './code-viewer.service';
export * from './code-viewer.model';
