// @helix/platform/esm2015/view/components/field-components/decimal-field/design/decimal-field-design.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptRxTextfieldModule } from '@bmc-ux/adapt-angular';
import { DecimalFieldDesignComponent } from './decimal-field-design.component';
import * as i0 from "@angular/core";
export class DecimalFieldDesignModule {
}
DecimalFieldDesignModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: DecimalFieldDesignModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DecimalFieldDesignModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: DecimalFieldDesignModule, declarations: [DecimalFieldDesignComponent], imports: [CommonModule, AdaptRxTextfieldModule, FormsModule] });
DecimalFieldDesignModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: DecimalFieldDesignModule, imports: [[CommonModule, AdaptRxTextfieldModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: DecimalFieldDesignModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AdaptRxTextfieldModule, FormsModule],
                    declarations: [DecimalFieldDesignComponent],
                    entryComponents: [DecimalFieldDesignComponent]
                }]
        }] });
