// @bmc-ux/adapt-angular/esm2015/rx/rx-typeahead/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptDropdownModule } from '../../dropdown/index';
import { AdaptRxTypeaheadComponent } from './adapt-rx-typeahead.component';
import { AdaptHighlightModule } from './../../highlight/index';
import * as i0 from "@angular/core";
export class AdaptRxTypeaheadModule {
}
AdaptRxTypeaheadModule.ɵfac = function AdaptRxTypeaheadModule_Factory(t) { return new (t || AdaptRxTypeaheadModule)(); };
AdaptRxTypeaheadModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxTypeaheadModule });
AdaptRxTypeaheadModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            AdaptRxFormControlModule,
            AdaptDropdownModule,
            CommonModule,
            ScrollingModule,
            AdaptHighlightModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxTypeaheadModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptRxFormControlModule,
                    AdaptDropdownModule,
                    CommonModule,
                    ScrollingModule,
                    AdaptHighlightModule
                ],
                declarations: [
                    AdaptRxTypeaheadComponent
                ],
                exports: [
                    AdaptRxTypeaheadComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxTypeaheadModule, { declarations: [AdaptRxTypeaheadComponent], imports: [AdaptRxFormControlModule,
        AdaptDropdownModule,
        CommonModule,
        ScrollingModule,
        AdaptHighlightModule], exports: [AdaptRxTypeaheadComponent] }); })();
export * from './adapt-rx-typeahead.component';
export * from './adapt-rx-typeahead-config.service';
export * from './adapt-rx-typeahead.types';
