// @bmc-ux/obsolete/esm2015/lib/uploader/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptUploadItemModule, AdaptTruncateFileNameModule, AdaptTooltipModule, AdaptAlertModule, AdaptBusyModule, AdaptButtonModule, AdaptCollapseModule, AdaptFileSizePipeModule, AdaptFileTypesPipeModule, AdaptEmptyStateModule, AdaptIconModule, AdaptPopoverModule, SendFileService, AdaptRxLabelModule } from '@bmc-ux/adapt-angular';
import { AdaptUploaderComponent } from './uploader.component';
import { AdaptSingleUploaderComponent } from './single-uploader/single-uploader.component';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptUploadModule {
    static forRoot(config) {
        return {
            ngModule: AdaptUploadModule,
            providers: [
                (config && config.uploader) ? config.uploader : { provide: 'UploaderService', useClass: SendFileService },
                { provide: SendFileService, useClass: SendFileService }
            ]
        };
    }
}
AdaptUploadModule.ɵfac = function AdaptUploadModule_Factory(t) { return new (t || AdaptUploadModule)(); };
AdaptUploadModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptUploadModule });
AdaptUploadModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptButtonModule,
            AdaptAlertModule.forRoot(),
            AdaptCollapseModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptIconModule,
            AdaptBusyModule,
            AdaptTruncateFileNameModule,
            AdaptUploadItemModule,
            AdaptFileSizePipeModule,
            AdaptFileTypesPipeModule,
            AdaptEmptyStateModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptUploadModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptButtonModule,
                    AdaptAlertModule.forRoot(),
                    AdaptCollapseModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptIconModule,
                    AdaptBusyModule,
                    AdaptTruncateFileNameModule,
                    AdaptUploadItemModule,
                    AdaptFileSizePipeModule,
                    AdaptFileTypesPipeModule,
                    AdaptEmptyStateModule,
                    AdaptRxLabelModule
                ],
                declarations: [
                    AdaptUploaderComponent,
                    AdaptSingleUploaderComponent
                ],
                exports: [AdaptUploaderComponent, AdaptSingleUploaderComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptUploadModule, { declarations: [AdaptUploaderComponent,
        AdaptSingleUploaderComponent], imports: [CommonModule,
        FormsModule,
        AdaptButtonModule, i1.AdaptAlertModule, AdaptCollapseModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptIconModule,
        AdaptBusyModule,
        AdaptTruncateFileNameModule,
        AdaptUploadItemModule,
        AdaptFileSizePipeModule,
        AdaptFileTypesPipeModule,
        AdaptEmptyStateModule,
        AdaptRxLabelModule], exports: [AdaptUploaderComponent, AdaptSingleUploaderComponent] }); })();
export * from './uploader.component';
export * from './single-uploader/single-uploader.component';
