// @bmc-ux/adapt-angular/esm2015/split/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptSplitComponent } from './split.component';
import { AdaptSplitGutterDirective } from './split-gutter.directive';
import { AdaptSplitAreaDirective } from './split-area.directive';
import * as i0 from "@angular/core";
export class AdaptSplitModule {
}
AdaptSplitModule.ɵfac = function AdaptSplitModule_Factory(t) { return new (t || AdaptSplitModule)(); };
AdaptSplitModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSplitModule });
AdaptSplitModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSplitModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptSplitComponent,
                    AdaptSplitAreaDirective,
                    AdaptSplitGutterDirective
                ],
                exports: [
                    AdaptSplitComponent,
                    AdaptSplitAreaDirective,
                    AdaptSplitGutterDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSplitModule, { declarations: [AdaptSplitComponent,
        AdaptSplitAreaDirective,
        AdaptSplitGutterDirective], imports: [CommonModule], exports: [AdaptSplitComponent,
        AdaptSplitAreaDirective,
        AdaptSplitGutterDirective] }); })();
export * from './split.component';
export * from './split-area.directive';
export * from './split-gutter.directive';
