// @bmc-ux/adapt-angular/esm2015/rx/form-control/rx-feedback/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFeedbackComponent } from './adapt-rx-feedback.component';
import * as i0 from "@angular/core";
export class AdaptRxFeedbackModule {
}
AdaptRxFeedbackModule.ɵfac = function AdaptRxFeedbackModule_Factory(t) { return new (t || AdaptRxFeedbackModule)(); };
AdaptRxFeedbackModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxFeedbackModule });
AdaptRxFeedbackModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxFeedbackModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptRxFeedbackComponent
                ],
                exports: [
                    AdaptRxFeedbackComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxFeedbackModule, { declarations: [AdaptRxFeedbackComponent], imports: [CommonModule], exports: [AdaptRxFeedbackComponent] }); })();
export * from './adapt-rx-feedback.component';
export * from './adapt-rx-feedback.service';
export * from './adapt-rx-feedback.model';
