// @bmc-ux/adapt-angular/esm2015/agreement/agreement-admin-modal/agreement-admin-modal.component.js
import { Component, EventEmitter, Input, Output, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { AdaptModalService } from '../../modal/modal.service';
import { AdaptMessageService } from '../../toast/toast-message.service';
import * as i0 from "@angular/core";
import * as i1 from "../../modal/modal.service";
import * as i2 from "../../toast/toast-message.service";
import * as i3 from "../../rx/rx-switch/adapt-rx-switch.component";
import * as i4 from "@angular/forms";
import * as i5 from "@angular/common";
import * as i6 from "../../button/button.component";
import * as i7 from "../../collapse/collapse.directive";
import * as i8 from "../../rx/rx-checkbox/adapt-rx-checkbox.component";
import * as i9 from "../../alert/alert.component";
const _c0 = ["mTemplate"];
function AdaptAgreementAdminModalComponent_ng_template_0_adapt_rx_checkbox_17_Template(rf, ctx) { if (rf & 1) {
    const _r10 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "adapt-rx-checkbox", 15);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_adapt_rx_checkbox_17_Template_adapt_rx_checkbox_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r10); const ctx_r9 = i0.ɵɵnextContext(2); return ctx_r9.state.accountPerformance = $event; })("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_adapt_rx_checkbox_17_Template_adapt_rx_checkbox_ngModelChange_0_listener() { i0.ɵɵrestoreView(_r10); const ctx_r11 = i0.ɵɵnextContext(2); return ctx_r11.handleStateChange(); });
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r3.state.accountPerformance)("label", "Turn on data collection for my account");
} }
function AdaptAgreementAdminModalComponent_ng_template_0_span_21_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1, "this product.");
    i0.ɵɵelementEnd();
} }
function AdaptAgreementAdminModalComponent_ng_template_0_span_23_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1, "BMC\u2019s products and services.");
    i0.ɵɵelementEnd();
} }
const _c1 = "Turning this feature on will enable data collection for all users of this product. Users may change these options at any time in their settings.";
const _c2 = function () { return { content: _c1, type: "inline", variant: "info" }; };
function AdaptAgreementAdminModalComponent_ng_template_0_adapt_alert_26_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "adapt-alert", 16);
} if (rf & 2) {
    i0.ɵɵproperty("config", i0.ɵɵpureFunction0(1, _c2));
} }
const _c3 = "Turning this feature on will enable data collection for all users of BMC\u2019s products and services. Users may change these options at any time in their settings.";
const _c4 = function () { return { content: _c3, type: "inline", variant: "info" }; };
function AdaptAgreementAdminModalComponent_ng_template_0_adapt_alert_28_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "adapt-alert", 16);
} if (rf & 2) {
    i0.ɵɵproperty("config", i0.ɵɵpureFunction0(1, _c4));
} }
function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_adapt_rx_checkbox_6_Template(rf, ctx) { if (rf & 1) {
    const _r14 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "adapt-rx-checkbox", 15);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_adapt_rx_checkbox_6_Template_adapt_rx_checkbox_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r14); const ctx_r13 = i0.ɵɵnextContext(3); return ctx_r13.state.accountMarketing = $event; })("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_adapt_rx_checkbox_6_Template_adapt_rx_checkbox_ngModelChange_0_listener() { i0.ɵɵrestoreView(_r14); const ctx_r15 = i0.ɵɵnextContext(3); return ctx_r15.handleStateChange(); });
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(3);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r12.state.accountMarketing)("label", "Turn on data collection for my account");
} }
const _c5 = "Turning this feature on will enable data collection for all users in your company. Users will be required to opt-in before any data collection is done, and may change these options at any time in their settings.";
const _c6 = function () { return { content: _c5, type: "inline", variant: "info" }; };
const _c7 = function (a0, a1) { return { "d-icon-angle_down": a0, "d-icon-angle_up": a1 }; };
const _c8 = function (a0, a1) { return { "mb-3": a0, "agreement__text-hidden": a1 }; };
function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_Template(rf, ctx) { if (rf & 1) {
    const _r17 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "adapt-rx-switch", 4);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_Template_adapt_rx_switch_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r17); const ctx_r16 = i0.ɵɵnextContext(2); return ctx_r16.state.organizationMarketing = $event; })("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_Template_adapt_rx_switch_ngModelChange_2_listener() { i0.ɵɵrestoreView(_r17); const ctx_r18 = i0.ɵɵnextContext(2); return ctx_r18.handleMarketingStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n\n      ");
    i0.ɵɵelementStart(4, "div", 5);
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵtemplate(6, AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_adapt_rx_checkbox_6_Template, 2, 3, "adapt-rx-checkbox", 6);
    i0.ɵɵtext(7, "\n\n        ");
    i0.ɵɵelementStart(8, "p", 7);
    i0.ɵɵtext(9, "Enabling this feature allows BMC to use the data we anonymously collect to better\n          inform your company of relevant products and services, as well as to reach out to individuals at your organization\n          for feedback and other requests. Enabling this option will improve the relevancy of BMC\u2019s communications, as well\n          as provide your company with more targeted offerings.");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n\n        ");
    i0.ɵɵelement(11, "adapt-alert", 17);
    i0.ɵɵtext(12, "\n\n        ");
    i0.ɵɵelementStart(13, "button", 10);
    i0.ɵɵlistener("click", function AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_Template_button_click_13_listener() { i0.ɵɵrestoreView(_r17); const ctx_r19 = i0.ɵɵnextContext(2); return ctx_r19.isCollapsedMarketingContent = !ctx_r19.isCollapsedMarketingContent; });
    i0.ɵɵtext(14, "\n          ");
    i0.ɵɵelementStart(15, "span");
    i0.ɵɵtext(16, "Click for examples of ways data may be used");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n          ");
    i0.ɵɵelement(18, "span", 11);
    i0.ɵɵtext(19, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(20, "\n\n        ");
    i0.ɵɵelementStart(21, "ul", 12);
    i0.ɵɵtext(22, "\n          ");
    i0.ɵɵelementStart(23, "li");
    i0.ɵɵtext(24, "To better inform sales teams about the way your company leverages BMC products");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n          ");
    i0.ɵɵelementStart(26, "li");
    i0.ɵɵtext(27, "To understand industry trends to better support your company");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(28, "\n          ");
    i0.ɵɵelementStart(29, "li");
    i0.ɵɵtext(30, "To identify communications topics relating to the industry");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(31, "\n          ");
    i0.ɵɵelementStart(32, "li");
    i0.ɵɵtext(33, "To better target messaging around new products and features");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(34, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(35, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(36, "\n\n      ");
    i0.ɵɵelement(37, "hr", 3);
    i0.ɵɵtext(38, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r8.state.organizationMarketing)("label", "Enable data collection for marketing and outreach");
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r8.state.organizationMarketing);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("config", i0.ɵɵpureFunction0(11, _c6));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(12, _c7, ctx_r8.isCollapsedMarketingContent, !ctx_r8.isCollapsedMarketingContent));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("collapse", ctx_r8.isCollapsedMarketingContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(15, _c8, !ctx_r8.isCollapsedMarketingContent, ctx_r8.isCollapsedMarketingContent));
} }
function AdaptAgreementAdminModalComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    const _r21 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n  ");
    i0.ɵɵelementStart(1, "div", 1);
    i0.ɵɵtext(2, "\n    ");
    i0.ɵɵelementStart(3, "h2", 2);
    i0.ɵɵtext(4, "Allow data collection for product improvement");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n\n    ");
    i0.ɵɵelementStart(6, "p");
    i0.ɵɵtext(7, "Help BMC improve your experience with our products.");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n\n    ");
    i0.ɵɵelement(9, "hr", 3);
    i0.ɵɵtext(10, "\n\n    ");
    i0.ɵɵelementContainerStart(11);
    i0.ɵɵtext(12, "\n      ");
    i0.ɵɵelementStart(13, "adapt-rx-switch", 4);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_Template_adapt_rx_switch_ngModelChange_13_listener($event) { i0.ɵɵrestoreView(_r21); const ctx_r20 = i0.ɵɵnextContext(); return ctx_r20.state.organizationPerformance = $event; })("ngModelChange", function AdaptAgreementAdminModalComponent_ng_template_0_Template_adapt_rx_switch_ngModelChange_13_listener() { i0.ɵɵrestoreView(_r21); const ctx_r22 = i0.ɵɵnextContext(); return ctx_r22.handlePerformanceStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n\n      ");
    i0.ɵɵelementStart(15, "div", 5);
    i0.ɵɵtext(16, "\n        ");
    i0.ɵɵtemplate(17, AdaptAgreementAdminModalComponent_ng_template_0_adapt_rx_checkbox_17_Template, 2, 3, "adapt-rx-checkbox", 6);
    i0.ɵɵtext(18, "\n\n        ");
    i0.ɵɵelementStart(19, "p", 7);
    i0.ɵɵtext(20, "Enabling this feature allows BMC to collect anonymous information about\n          individual usage of\n          ");
    i0.ɵɵtemplate(21, AdaptAgreementAdminModalComponent_ng_template_0_span_21_Template, 2, 0, "span", 8);
    i0.ɵɵtext(22, "\n          ");
    i0.ɵɵtemplate(23, AdaptAgreementAdminModalComponent_ng_template_0_span_23_Template, 2, 0, "span", 8);
    i0.ɵɵtext(24, "\n          The data collected will be used to identify defects, optimize product performance, and aid in the\n          development of new product features. Enabling this option will ensure that your user\u2019s needs are taken\n          into account when planning for new product improvements.");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n\n        ");
    i0.ɵɵtemplate(26, AdaptAgreementAdminModalComponent_ng_template_0_adapt_alert_26_Template, 1, 2, "adapt-alert", 9);
    i0.ɵɵtext(27, "\n\n          ");
    i0.ɵɵtemplate(28, AdaptAgreementAdminModalComponent_ng_template_0_adapt_alert_28_Template, 1, 2, "adapt-alert", 9);
    i0.ɵɵtext(29, "\n\n        ");
    i0.ɵɵelementStart(30, "button", 10);
    i0.ɵɵlistener("click", function AdaptAgreementAdminModalComponent_ng_template_0_Template_button_click_30_listener() { i0.ɵɵrestoreView(_r21); const ctx_r23 = i0.ɵɵnextContext(); return ctx_r23.isCollapsedPerformanceContent = !ctx_r23.isCollapsedPerformanceContent; });
    i0.ɵɵtext(31, "\n          ");
    i0.ɵɵelementStart(32, "span");
    i0.ɵɵtext(33, "Click for examples of the types of data we collect");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(34, "\n          ");
    i0.ɵɵelement(35, "span", 11);
    i0.ɵɵtext(36, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(37, "\n\n        ");
    i0.ɵɵelementStart(38, "ul", 12);
    i0.ɵɵtext(39, "\n          ");
    i0.ɵɵelementStart(40, "li");
    i0.ɵɵtext(41, "Time to complete certain tasks");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(42, "\n          ");
    i0.ɵɵelementStart(43, "li");
    i0.ɵɵtext(44, "Counts of clicks or taps on certain elements, like buttons");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(45, "\n          ");
    i0.ɵɵelementStart(46, "li");
    i0.ɵɵtext(47, "The number of errors experienced while using a BMC product, and their type");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(48, "\n          ");
    i0.ɵɵelementStart(49, "li");
    i0.ɵɵtext(50, "Time spent not actively using a BMC product");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(51, "\n          ");
    i0.ɵɵelementStart(52, "li");
    i0.ɵɵtext(53, "Responses to ratings and surveys provided by BMC products");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(54, "\n          ");
    i0.ɵɵelementStart(55, "li");
    i0.ɵɵtext(56, "How often a product is signed in to, or signed out of");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(57, "\n          ");
    i0.ɵɵelementStart(58, "li");
    i0.ɵɵtext(59, "Any movement from one BMC product to another");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(60, "\n          ");
    i0.ɵɵelementStart(61, "li");
    i0.ɵɵtext(62, "Information that allows us to identify a product, or instance of a product");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(63, "\n          ");
    i0.ɵɵelementStart(64, "li");
    i0.ɵɵtext(65, "Diagnostic information needed to technical troubleshooting");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(66, "\n          ");
    i0.ɵɵelementStart(67, "li");
    i0.ɵɵtext(68, "General information about your computer or web browser that could be used to identify technical problems\n          ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(69, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(70, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(71, "\n\n      ");
    i0.ɵɵelement(72, "hr", 3);
    i0.ɵɵtext(73, "\n    ");
    i0.ɵɵelementContainerEnd();
    i0.ɵɵtext(74, "\n\n    ");
    i0.ɵɵtemplate(75, AdaptAgreementAdminModalComponent_ng_template_0_ng_container_75_Template, 39, 18, "ng-container", 8);
    i0.ɵɵtext(76, "\n\n    ");
    i0.ɵɵelementStart(77, "p");
    i0.ɵɵtext(78, "By enabling this feature you consent to BMC\u2019s collection, processing, and use of user analytics data\n      described, but not limited to, the examples outlined above. You can change your selection in your application\u2019s\n      settings at any time.");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(79, "\n\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(80, "\n  ");
    i0.ɵɵelementStart(81, "div", 13);
    i0.ɵɵtext(82, "\n    ");
    i0.ɵɵelementStart(83, "button", 14);
    i0.ɵɵlistener("click", function AdaptAgreementAdminModalComponent_ng_template_0_Template_button_click_83_listener() { const restoredCtx = i0.ɵɵrestoreView(_r21); const close_r2 = restoredCtx.close; const ctx_r24 = i0.ɵɵnextContext(); return ctx_r24.handleSubmit(close_r2); });
    i0.ɵɵtext(84);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(85, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(86, "\n");
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(13);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r1.state.organizationPerformance)("label", "Enable data collection for performance and product improvement");
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r1.state.organizationPerformance);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", !ctx_r1.multiProductUsage);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.multiProductUsage);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !ctx_r1.multiProductUsage);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.multiProductUsage);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(15, _c7, ctx_r1.isCollapsedPerformanceContent, !ctx_r1.isCollapsedPerformanceContent));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("collapse", ctx_r1.isCollapsedPerformanceContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(18, _c8, !ctx_r1.isCollapsedPerformanceContent, ctx_r1.isCollapsedPerformanceContent));
    i0.ɵɵadvance(37);
    i0.ɵɵproperty("ngIf", ctx_r1.state.organizationPerformance);
    i0.ɵɵadvance(9);
    i0.ɵɵtextInterpolate(ctx_r1.getSubmitButtonText());
} }
export class AdaptAgreementAdminModalComponent {
    constructor(_modalService, _messageService) {
        this._modalService = _modalService;
        this._messageService = _messageService;
        this.isCollapsedPerformanceContent = true;
        this.isCollapsedMarketingContent = true;
        this.multiProductUsage = false;
        this.state = {
            accountPerformance: false,
            accountMarketing: false,
            organizationPerformance: false,
            organizationMarketing: false
        };
        this.openModal = false;
        this.showNotificationAfterSubmit = true;
        this.stateChange = new EventEmitter();
        this.submitButtonClick = new EventEmitter();
        this.modalOpen = new EventEmitter();
    }
    ngOnChanges(changes) {
        var _a;
        if ((_a = changes['openModal']) === null || _a === void 0 ? void 0 : _a.currentValue) {
            this.open();
        }
    }
    open() {
        const promise = this._modalService
            .open({
            content: this.mTemplate,
            testID: this.testID,
            blockKeyboard: true,
            beforeDismiss: () => false
        });
        this.modalOpen.emit({ promise });
    }
    handlePerformanceStateChange() {
        if (!this.state.organizationPerformance) {
            this.state.accountPerformance = false;
            this.state.organizationMarketing = false;
            this.state.accountMarketing = false;
        }
        this.handleStateChange();
    }
    handleMarketingStateChange() {
        if (!this.state.organizationMarketing) {
            this.state.accountMarketing = false;
        }
        this.handleStateChange();
    }
    handleStateChange() {
        this.stateChange.emit(this.state);
    }
    handleSubmit(close) {
        this.submitButtonClick.emit(this.state);
        close();
        if (this.showNotificationAfterSubmit) {
            this._messageService.add({
                severity: 'info',
                summary: 'Analytics preferences have been updated',
                detail: 'Your analytics choices have been submitted. If you would like to change these settings you may do so in your application’s settings.',
                link: 'Go to settings',
                linkHandler: this.notificationLinkHandler,
                life: 10000
            });
        }
    }
    getSubmitButtonText() {
        if (this.state.organizationPerformance && this.state.organizationMarketing) {
            return 'Agree and submit';
        }
        if (!this.state.organizationPerformance) {
            return 'Opt-out of data collection';
        }
        if (!this.state.organizationMarketing) {
            return 'Agree to selected and submit';
        }
    }
}
AdaptAgreementAdminModalComponent.ɵfac = function AdaptAgreementAdminModalComponent_Factory(t) { return new (t || AdaptAgreementAdminModalComponent)(i0.ɵɵdirectiveInject(i1.AdaptModalService), i0.ɵɵdirectiveInject(i2.AdaptMessageService)); };
AdaptAgreementAdminModalComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptAgreementAdminModalComponent, selectors: [["adapt-agreement-modal"]], viewQuery: function AdaptAgreementAdminModalComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.mTemplate = _t.first);
    } }, inputs: { multiProductUsage: "multiProductUsage", state: "state", openModal: "openModal", testID: "testID", showNotificationAfterSubmit: "showNotificationAfterSubmit", notificationLinkHandler: "notificationLinkHandler" }, outputs: { stateChange: "stateChange", submitButtonClick: "submitButtonClick", modalOpen: "modalOpen" }, features: [i0.ɵɵNgOnChangesFeature], decls: 3, vars: 0, consts: [["mTemplate", ""], [1, "modal-body"], [1, "agreement__main-title"], [1, "agreement__separator"], [1, "agreement__switcher", 3, "adaptRadarDisableEventSending", "ngModel", "label", "ngModelChange"], [1, "agreement__content"], [3, "adaptRadarDisableEventSending", "ngModel", "label", "ngModelChange", 4, "ngIf"], [1, "agreement__content-text"], [4, "ngIf"], [3, "config", 4, "ngIf"], ["type", "button", "adapt-button", "", "btn-type", "tertiary", 1, "p-0", "mt-3", "mb-3", "agreement__collapse-btn", 3, "adaptRadarDisableEventSending", "click"], [1, "p-start-1", 3, "ngClass"], [1, "m-0", 3, "collapse", "adaptRadarDisableEventSending", "ngClass"], [1, "modal-footer"], ["type", "button", 1, "btn", "btn-primary", 3, "click"], [3, "adaptRadarDisableEventSending", "ngModel", "label", "ngModelChange"], [3, "config"], [3, "adaptRadarDisableEventSending", "config"]], template: function AdaptAgreementAdminModalComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, AdaptAgreementAdminModalComponent_ng_template_0_Template, 87, 21, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(2, "\n");
    } }, directives: [i3.AdaptRxSwitchComponent, i4.NgControlStatus, i4.NgModel, i5.NgIf, i6.AdaptButtonComponent, i5.NgClass, i7.AdaptCollapseDirective, i8.AdaptRxCheckboxComponent, i9.AdaptAlertComponent], styles: [".adapt-agreement.adapt-agreement_no-shadow{box-shadow:none}.agreement__main-title{margin:0 0 1rem}.agreement__separator{margin:1rem 0}.agreement__switcher{font-size:.9375rem}.agreement__content{margin-left:4.125em}.adapt-agreement__content_empty-state{font-size:.9375rem;text-align:center;width:100%}.agreement__content-text{margin:1rem 0}.agreement__badge .badge{min-width:70px}.agreement__collapse-btn{color:var(--text-primary)}adapt-alert .text-info{display:flex}.agreement__text-hidden .collapse-block{height:0;overflow:hidden;visibility:hidden}"], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAgreementAdminModalComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-agreement-modal',
                styleUrls: ['../agreement.scss'],
                templateUrl: './agreement-admin-modal.component.html',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.AdaptModalService }, { type: i2.AdaptMessageService }]; }, { multiProductUsage: [{
            type: Input
        }], state: [{
            type: Input
        }], openModal: [{
            type: Input
        }], testID: [{
            type: Input
        }], showNotificationAfterSubmit: [{
            type: Input
        }], notificationLinkHandler: [{
            type: Input
        }], stateChange: [{
            type: Output
        }], submitButtonClick: [{
            type: Output
        }], modalOpen: [{
            type: Output
        }], mTemplate: [{
            type: ViewChild,
            args: ['mTemplate']
        }] }); })();
