// @bmc-ux/obsolete/esm2015/lib/select-inline/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule, AdaptEmptyStateModule, AdaptFocusMonitorDirectiveModule, AdaptDragonInputHandlerModule, AdaptRxLabelModule, AdaptCustomActionsModule } from '@bmc-ux/adapt-angular';
import { AdaptSelectSearchFilterModule } from '../select/index';
import { AdaptSelectInlineComponent } from './select-inline.component';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptSelectInlineModule {
}
AdaptSelectInlineModule.ɵfac = function AdaptSelectInlineModule_Factory(t) { return new (t || AdaptSelectInlineModule)(); };
AdaptSelectInlineModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSelectInlineModule });
AdaptSelectInlineModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            FormsModule,
            ReactiveFormsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptEmptyStateModule,
            AdaptSelectSearchFilterModule,
            AdaptFocusMonitorDirectiveModule,
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule,
            AdaptCustomActionsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSelectInlineModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    FormsModule,
                    ReactiveFormsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptEmptyStateModule,
                    AdaptSelectSearchFilterModule,
                    AdaptFocusMonitorDirectiveModule,
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule,
                    AdaptCustomActionsModule
                ],
                declarations: [AdaptSelectInlineComponent],
                exports: [AdaptSelectInlineComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSelectInlineModule, { declarations: [AdaptSelectInlineComponent], imports: [CommonModule,
        FormsModule,
        FormsModule,
        ReactiveFormsModule,
        AdaptIconModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptEmptyStateModule,
        AdaptSelectSearchFilterModule,
        AdaptFocusMonitorDirectiveModule,
        AdaptDragonInputHandlerModule,
        AdaptRxLabelModule,
        AdaptCustomActionsModule], exports: [AdaptSelectInlineComponent] }); })();
export * from './select-inline.component';
