// @bmc-ux/adapt-angular/esm2015/button-group/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptButtonGroupComponent } from './button-group.component';
import { AdaptButtonModule } from '../button/index';
import * as i0 from "@angular/core";
export class AdaptButtonGroupModule {
}
AdaptButtonGroupModule.ɵfac = function AdaptButtonGroupModule_Factory(t) { return new (t || AdaptButtonGroupModule)(); };
AdaptButtonGroupModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptButtonGroupModule });
AdaptButtonGroupModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptButtonGroupModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptButtonModule],
                exports: [AdaptButtonGroupComponent],
                declarations: [AdaptButtonGroupComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptButtonGroupModule, { declarations: [AdaptButtonGroupComponent], imports: [CommonModule, AdaptButtonModule], exports: [AdaptButtonGroupComponent] }); })();
export * from './button-group.component';
export * from './button-group.models';
