// @bmc-ux/adapt-angular/esm2015/beard/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxLabelModule } from '../rx/form-control/index';
import { AdaptBeardComponent } from './beard.component';
import * as i0 from "@angular/core";
export class AdaptBeardModule {
}
AdaptBeardModule.ɵfac = function AdaptBeardModule_Factory(t) { return new (t || AdaptBeardModule)(); };
AdaptBeardModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptBeardModule });
AdaptBeardModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptRxLabelModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptBeardModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptRxLabelModule],
                declarations: [AdaptBeardComponent],
                exports: [AdaptBeardComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptBeardModule, { declarations: [AdaptBeardComponent], imports: [CommonModule, AdaptRxLabelModule], exports: [AdaptBeardComponent] }); })();
export * from './beard.component';
