// @bmc-ux/adapt-angular/esm2015/common/list-key-manager/index.js
import { NgModule } from '@angular/core';
import { AdaptListKeyManagerItemDirective } from './adapt-list-key-manager-item.directive';
import { AdaptListKeyManagerDirective } from './adapt-list-key-manager.directive';
import * as i0 from "@angular/core";
export class AdaptListKeyManagerModule {
}
AdaptListKeyManagerModule.ɵfac = function AdaptListKeyManagerModule_Factory(t) { return new (t || AdaptListKeyManagerModule)(); };
AdaptListKeyManagerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptListKeyManagerModule });
AdaptListKeyManagerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptListKeyManagerModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [
                    AdaptListKeyManagerItemDirective,
                    AdaptListKeyManagerDirective
                ],
                exports: [
                    AdaptListKeyManagerItemDirective,
                    AdaptListKeyManagerDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptListKeyManagerModule, { declarations: [AdaptListKeyManagerItemDirective,
        AdaptListKeyManagerDirective], exports: [AdaptListKeyManagerItemDirective,
        AdaptListKeyManagerDirective] }); })();
export * from './adapt-list-key-manager.directive';
export * from './adapt-list-key-manager-item.directive';
