// @helix/platform/esm2015/ui-kit/line-loader/line-loader.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RxLineLoaderComponent } from './line-loader.component';
import * as i0 from "@angular/core";
export class RxLineLoaderModule {
}
RxLineLoaderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLineLoaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxLineLoaderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLineLoaderModule, declarations: [RxLineLoaderComponent], imports: [CommonModule], exports: [RxLineLoaderComponent] });
RxLineLoaderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLineLoaderModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLineLoaderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxLineLoaderComponent],
                    imports: [CommonModule],
                    exports: [RxLineLoaderComponent]
                }]
        }] });
