// @helix/platform/esm2015/shared/components/designer/inspector-components/select-expression-dropdown/select-expression-dropdown.module.js
import { NgModule } from '@angular/core';
import { RxSelectExpressionDropdownComponent } from './select-expression-dropdown.component';
import * as i0 from "@angular/core";
export class RxSelectExpressionDropdownModule {
}
RxSelectExpressionDropdownModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxSelectExpressionDropdownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxSelectExpressionDropdownModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxSelectExpressionDropdownModule, declarations: [RxSelectExpressionDropdownComponent], exports: [RxSelectExpressionDropdownComponent] });
RxSelectExpressionDropdownModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxSelectExpressionDropdownModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxSelectExpressionDropdownModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxSelectExpressionDropdownComponent],
                    imports: [],
                    exports: [RxSelectExpressionDropdownComponent]
                }]
        }] });
