// @bmc-ux/adapt-angular/esm2015/code-highlighter/code-highlighter.directive.js
import { Directive, ElementRef } from '@angular/core';
import 'prismjs/components/prism-bash.js';
import 'prismjs/plugins/toolbar/prism-toolbar.js';
import 'prismjs/plugins/command-line/prism-command-line.js';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "../common/radar/adapt-radar";
export class AdaptCodeHighlighterDirective extends AdaptRadarAngularGenericDirective {
    constructor(el, _adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptCodeHighlighter);
        this.el = el;
    }
    ngOnInit() {
        Prism.highlightElement(this.el.nativeElement);
    }
}
AdaptCodeHighlighterDirective.ɵfac = function AdaptCodeHighlighterDirective_Factory(t) { return new (t || AdaptCodeHighlighterDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptCodeHighlighterDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptCodeHighlighterDirective, selectors: [["", "adaptCode", ""], ["", "adapt-code", ""]], features: [i0.ɵɵInheritDefinitionFeature] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCodeHighlighterDirective, [{
        type: Directive,
        args: [{
                selector: '[adaptCode], [adapt-code]'
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, null); })();
