// @bmc-ux/adapt-angular/esm2015/code-highlighter/index.js
/**
 *  Work in progress, this component currently is not in AdaptModule package.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptCodeHighlighterDirective } from './code-highlighter.directive';
import * as i0 from "@angular/core";
export class AdaptCodeHighlighterModule {
}
AdaptCodeHighlighterModule.ɵfac = function AdaptCodeHighlighterModule_Factory(t) { return new (t || AdaptCodeHighlighterModule)(); };
AdaptCodeHighlighterModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCodeHighlighterModule });
AdaptCodeHighlighterModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCodeHighlighterModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptCodeHighlighterDirective],
                exports: [AdaptCodeHighlighterDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCodeHighlighterModule, { declarations: [AdaptCodeHighlighterDirective], imports: [CommonModule], exports: [AdaptCodeHighlighterDirective] }); })();
export * from './code-highlighter.directive';
