// @bmc-ux/adapt-table/esm2015/lib/adapt-table.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { AdaptTooltipModule, AdaptRxSearchModule, AdaptRxSelectModule, AdaptDropdownModule, AdaptPopoverModule, AdaptEmptyStateModule, AdaptBusyModule, AdaptFocusHolderModule, AdaptAccordionModule, AdaptTagModule, AdaptRxCheckboxModule, AdaptResizeModule, AdaptListKeyManagerModule, AdaptButtonModule, AdaptRxTextfieldModule, AdaptRxRadiobuttonModule } from '@bmc-ux/adapt-angular';
import { AdaptTableComponent } from './adapt-table.component';
import { AdaptTableFiltersComponent } from './adapt-table-filters/adapt-table-filters.component';
import { AdaptTableMergedFiltersComponent } from './adapt-table-filters/adapt-table-merged-filters/adapt-table-merged-filters.component';
import { FormsModule } from '@angular/forms';
import { AdaptTableToolbarComponent } from './adapt-table-toolbar/adapt-table-toolbar.component';
import { AdaptTableActionsComponent } from './adapt-table-actions/adapt-table-actions.component';
import { AdaptHeaderSelectionCheckboxComponent } from './header-selection-checkbox/header-selection-checkbox.component';
import { TableModule } from './primeng-lib/table/table.module';
import { AdaptDataDirective } from './shared/adapt-data-directive';
import { SortByPipe } from './shared/sortby.pipe';
import { AdaptTableStateDirective } from './shared/adapt-table-state.directive';
import { AdaptTableHeaderCellComponent } from './table-header-cell/table-header-cell.component';
import { AdaptTableHeaderCellMenuComponent } from './table-header-cell/adapt-table-header-cell-menu/adapt-table-header-cell-menu.component';
import { AdaptPaginatorComponent } from './adapt-paginator/adapt-paginator.component';
import { AdaptTableToolbarItemDirective } from './adapt-table-toolbar/adapt-table-toolbar-item.directive';
import { AdaptRowSelectionCheckboxComponent } from './row-selection-checkbox/row-selection-checkbox.component';
import { AdaptSelectMigrationComponent } from './adapt-table-filters/select-migration.component';
import { AdaptFrozenColumnDirective } from './shared/adapt-frozen-column-directive';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
import * as i2 from "@angular/common";
import * as i3 from "@angular/cdk/drag-drop";
import * as i4 from "./primeng-lib/table/table-body.component";
import * as i5 from "./primeng-lib/table/scrollable-view.component";
import * as i6 from "./primeng-lib/table/resizable-column.directive";
import * as i7 from "./primeng-lib/table/reordarable-column.directive";
import * as i8 from "@angular/forms";
import * as i9 from "./primeng-lib/table/editable-row.directive";
import * as i10 from "./primeng-lib/table/row-toggler.directive";
// TODO: double check forRoot
export class AdaptTableModule {
    static forRoot() {
        return {
            ngModule: AdaptTableModule
        };
    }
}
AdaptTableModule.ɵfac = function AdaptTableModule_Factory(t) { return new (t || AdaptTableModule)(); };
AdaptTableModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTableModule });
AdaptTableModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            DragDropModule,
            TableModule,
            AdaptTooltipModule,
            AdaptRxSearchModule,
            AdaptRxSelectModule,
            FormsModule,
            AdaptDropdownModule.forRoot(),
            AdaptRxTextfieldModule,
            AdaptRxRadiobuttonModule,
            AdaptPopoverModule.forRoot(),
            AdaptEmptyStateModule,
            AdaptBusyModule,
            AdaptFocusHolderModule,
            AdaptAccordionModule,
            AdaptTagModule,
            AdaptRxCheckboxModule,
            AdaptResizeModule,
            AdaptListKeyManagerModule,
            AdaptButtonModule
        ], TableModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTableModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    AdaptTableComponent,
                    AdaptTableFiltersComponent,
                    AdaptTableMergedFiltersComponent,
                    AdaptTableToolbarComponent,
                    AdaptTableActionsComponent,
                    AdaptDataDirective,
                    AdaptFrozenColumnDirective,
                    AdaptTableStateDirective,
                    AdaptTableHeaderCellComponent,
                    AdaptPaginatorComponent,
                    AdaptTableToolbarItemDirective,
                    AdaptHeaderSelectionCheckboxComponent,
                    AdaptRowSelectionCheckboxComponent,
                    AdaptTableHeaderCellMenuComponent,
                    SortByPipe,
                    AdaptSelectMigrationComponent
                ],
                imports: [
                    CommonModule,
                    DragDropModule,
                    TableModule,
                    AdaptTooltipModule,
                    AdaptRxSearchModule,
                    AdaptRxSelectModule,
                    FormsModule,
                    AdaptDropdownModule.forRoot(),
                    AdaptRxTextfieldModule,
                    AdaptRxRadiobuttonModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptEmptyStateModule,
                    AdaptBusyModule,
                    AdaptFocusHolderModule,
                    AdaptAccordionModule,
                    AdaptTagModule,
                    AdaptRxCheckboxModule,
                    AdaptResizeModule,
                    AdaptListKeyManagerModule,
                    AdaptButtonModule
                ],
                exports: [
                    AdaptTableComponent,
                    AdaptTableStateDirective,
                    TableModule
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTableModule, { declarations: [AdaptTableComponent,
        AdaptTableFiltersComponent,
        AdaptTableMergedFiltersComponent,
        AdaptTableToolbarComponent,
        AdaptTableActionsComponent,
        AdaptDataDirective,
        AdaptFrozenColumnDirective,
        AdaptTableStateDirective,
        AdaptTableHeaderCellComponent,
        AdaptPaginatorComponent,
        AdaptTableToolbarItemDirective,
        AdaptHeaderSelectionCheckboxComponent,
        AdaptRowSelectionCheckboxComponent,
        AdaptTableHeaderCellMenuComponent,
        SortByPipe,
        AdaptSelectMigrationComponent], imports: [CommonModule,
        DragDropModule,
        TableModule,
        AdaptTooltipModule,
        AdaptRxSearchModule,
        AdaptRxSelectModule,
        FormsModule, i1.AdaptDropdownModule, AdaptRxTextfieldModule,
        AdaptRxRadiobuttonModule, i1.AdaptPopoverModule, AdaptEmptyStateModule,
        AdaptBusyModule,
        AdaptFocusHolderModule,
        AdaptAccordionModule,
        AdaptTagModule,
        AdaptRxCheckboxModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptButtonModule], exports: [AdaptTableComponent,
        AdaptTableStateDirective,
        TableModule] }); })();
i0.ɵɵsetComponentScope(AdaptTableComponent, [i2.NgClass, i3.CdkDropList, i2.NgStyle, i2.NgIf, AdaptTableToolbarComponent, i2.NgTemplateOutlet, i4.TableBody, i1.AdaptResizeDirective, i5.ScrollableView, AdaptPaginatorComponent, i1.AdaptBusyDirective, i2.NgForOf, i6.ResizableColumn, i7.ReorderableColumn, AdaptFrozenColumnDirective,
    AdaptDataDirective,
    AdaptTableHeaderCellComponent,
    AdaptHeaderSelectionCheckboxComponent, i1.AdaptTooltipDirective, i3.CdkDrag, i3.CdkDragHandle, i1.AdaptRxCheckboxComponent, i8.NgControlStatus, i8.NgModel, i9.EditableRow, i1.AdaptRxRadiobuttonComponent, i10.RowToggler, AdaptRowSelectionCheckboxComponent, i1.AdaptEmptyStateComponent, AdaptTableActionsComponent, i1.AdaptRxTextfieldComponent, i1.AdaptButtonComponent], []);
