// @bmc-ux/adapt-angular/esm2015/pagination/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptPaginationComponent } from './pagination.component';
import * as i0 from "@angular/core";
export class AdaptPaginationModule {
}
AdaptPaginationModule.ɵfac = function AdaptPaginationModule_Factory(t) { return new (t || AdaptPaginationModule)(); };
AdaptPaginationModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptPaginationModule });
AdaptPaginationModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptPaginationModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [AdaptPaginationComponent],
                exports: [AdaptPaginationComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptPaginationModule, { declarations: [AdaptPaginationComponent], imports: [CommonModule], exports: [AdaptPaginationComponent] }); })();
export * from './pagination.component';
export * from './pagination.model';
