// @bmc-ux/adapt-angular/esm2015/pagination/pagination.component.js
import { ChangeDetectorRef, Component, ElementRef, Input } from '@angular/core';
import { Paginator } from 'primeng/paginator';
import { AdaptTranslateService } from '../common/i18n/index';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularService } from '../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "../common/i18n/index";
import * as i2 from "../common/radar/adapt-radar";
import * as i3 from "@angular/common";
const _c0 = function (a0) { return { "active": a0 }; };
function AdaptPaginationComponent_ul_2_li_11_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "li", 8);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "button", 9);
    i0.ɵɵlistener("click", function AdaptPaginationComponent_ul_2_li_11_Template_button_click_2_listener() { const restoredCtx = i0.ɵɵrestoreView(_r5); const pageLink_r3 = restoredCtx.$implicit; const ctx_r4 = i0.ɵɵnextContext(2); return ctx_r4.changePage(pageLink_r3 - 1); });
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const pageLink_r3 = ctx.$implicit;
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(3, _c0, pageLink_r3 - 1 === ctx_r2.getPage()));
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("data-testid", ctx_r2.testID + "_item_" + pageLink_r3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(pageLink_r3);
} }
const _c1 = function (a0) { return { "disabled": a0 }; };
function AdaptPaginationComponent_ul_2_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "ul", 2);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "li", 3);
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵelementStart(4, "button", 4);
    i0.ɵɵlistener("click", function AdaptPaginationComponent_ul_2_Template_button_click_4_listener($event) { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return ctx_r6.changePageToPrev($event); });
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵelementStart(6, "span", 5);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n    ");
    i0.ɵɵtemplate(11, AdaptPaginationComponent_ul_2_li_11_Template, 5, 5, "li", 6);
    i0.ɵɵtext(12, "\n    ");
    i0.ɵɵelementStart(13, "li", 7);
    i0.ɵɵtext(14, "\n      ");
    i0.ɵɵelementStart(15, "button", 4);
    i0.ɵɵlistener("click", function AdaptPaginationComponent_ul_2_Template_button_click_15_listener($event) { i0.ɵɵrestoreView(_r7); const ctx_r8 = i0.ɵɵnextContext(); return ctx_r8.changePageToNext($event); });
    i0.ɵɵtext(16, "\n        ");
    i0.ɵɵelementStart(17, "span", 5);
    i0.ɵɵtext(18);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(19, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(20, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(21, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r0.styleClass);
    i0.ɵɵproperty("ngStyle", ctx_r0.style)("ngClass", "pagination");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(15, _c1, ctx_r0.isFirstPage()));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tabindex", ctx_r0.isFirstPage() ? -1 : null);
    i0.ɵɵattribute("aria-label", ctx_r0.left)("data-testid", ctx_r0.testID + "_left");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r0.left);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngForOf", ctx_r0.pageLinks);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(17, _c1, ctx_r0.isLastPage()));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tabindex", ctx_r0.isLastPage() ? -1 : null);
    i0.ɵɵattribute("aria-label", ctx_r0.right)("data-testid", ctx_r0.testID + "_right");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r0.right);
} }
function AdaptPaginationComponent_ul_4_Template(rf, ctx) { if (rf & 1) {
    const _r10 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "ul", 2);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "li", 3);
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵelementStart(4, "button", 4);
    i0.ɵɵlistener("click", function AdaptPaginationComponent_ul_4_Template_button_click_4_listener($event) { i0.ɵɵrestoreView(_r10); const ctx_r9 = i0.ɵɵnextContext(); return ctx_r9.changePageToPrev($event); });
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵelementStart(6, "span", 10);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n    ");
    i0.ɵɵelementStart(11, "li", 11);
    i0.ɵɵtext(12, "\n      ");
    i0.ɵɵelementStart(13, "button", 4);
    i0.ɵɵlistener("click", function AdaptPaginationComponent_ul_4_Template_button_click_13_listener($event) { i0.ɵɵrestoreView(_r10); const ctx_r11 = i0.ɵɵnextContext(); return ctx_r11.changePageToNext($event); });
    i0.ɵɵtext(14, "\n        ");
    i0.ɵɵelementStart(15, "span", 12);
    i0.ɵɵtext(16);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(19, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r1.styleClass);
    i0.ɵɵproperty("ngStyle", ctx_r1.style)("ngClass", "pagination pager d-flex justify-content-start");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(14, _c1, ctx_r1.isFirstPage()));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tabindex", ctx_r1.isFirstPage() ? -1 : null);
    i0.ɵɵattribute("aria-label", ctx_r1.left)("data-testid", ctx_r1.testID + "_left");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.left);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(16, _c1, ctx_r1.isLastPage()));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("tabindex", ctx_r1.isLastPage() ? -1 : null);
    i0.ɵɵattribute("aria-label", ctx_r1.right)("data-testid", ctx_r1.testID + "_right");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.right);
} }
let nextUniqueId = 0;
export class AdaptPaginationComponent extends Paginator {
    constructor(ts, cd, _adaptRadarService, _elem) {
        super(cd);
        this.ts = ts;
        this._adaptRadarService = _adaptRadarService;
        this._elem = _elem;
        this.uniqueId = `adapt-pagination-${++nextUniqueId}`;
        /**
         * Type of pagination - 'pager' | 'prevnext'
         * @docs-default pager
         */
        this.mode = 'pager';
        /**
         * String for test id data attribute. For test purposes
         * @docs-default adapt-pagination-${++nextUniqueId}
         */
        this.testID = this.uniqueId;
        this.adaptRadarDisableEventSending = false;
        const translate = ts.getCurrentLanguage();
        this.left = translate['adapt.common.previous'];
        this.right = translate['adapt.common.next'];
    }
    ngAfterViewInit() {
        this._elementIntersectObserver = this._adaptRadarService.subscribeOnElementIntersect(this.adaptRadarDisableEventSending, this._elem, AdaptRadarAngularCustomEventName.AdaptPagination);
    }
    ngOnDestroy() {
        if (this._elementIntersectObserver) {
            this._elementIntersectObserver.disconnect();
            this._elementIntersectObserver = null;
        }
    }
}
AdaptPaginationComponent.ɵfac = function AdaptPaginationComponent_Factory(t) { return new (t || AdaptPaginationComponent)(i0.ɵɵdirectiveInject(i1.AdaptTranslateService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i2.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptPaginationComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptPaginationComponent, selectors: [["adapt-pagination"]], inputs: { mode: "mode", left: "left", right: "right", testID: "testID", adaptRadarDisableEventSending: "adaptRadarDisableEventSending" }, features: [i0.ɵɵInheritDefinitionFeature], decls: 7, vars: 2, consts: [["aria-label", "Page navigation"], [3, "class", "ngStyle", "ngClass", 4, "ngIf"], [3, "ngStyle", "ngClass"], [1, "page-item", "page-prev", 3, "ngClass"], ["type", "button", 1, "btn", "page-link", 3, "tabindex", "click"], [1, "sr-only"], ["class", "page-item", 3, "ngClass", 4, "ngFor", "ngForOf"], [1, "page-item", "page-next", 3, "ngClass"], [1, "page-item", 3, "ngClass"], ["type", "button", 1, "btn", "page-link", 3, "click"], [1, "ml-1"], [1, "page-item", "page-next", "ml-auto", 3, "ngClass"], [1, "mr-1"]], template: function AdaptPaginationComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "nav", 0);
        i0.ɵɵtext(1, "\n  ");
        i0.ɵɵtemplate(2, AdaptPaginationComponent_ul_2_Template, 22, 19, "ul", 1);
        i0.ɵɵtext(3, "\n  ");
        i0.ɵɵtemplate(4, AdaptPaginationComponent_ul_4_Template, 20, 18, "ul", 1);
        i0.ɵɵtext(5, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(6, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.mode === "pager");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.mode === "prevnext");
    } }, directives: [i3.NgIf, i3.NgStyle, i3.NgClass, i3.NgForOf], styles: [".pager[_ngcontent-%COMP%]   .page-link[_ngcontent-%COMP%] {  width: 100%;  }"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptPaginationComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-pagination',
                templateUrl: './pagination.component.html',
                styles: ['.pager .page-link {  width: 100%;  }']
            }]
    }], function () { return [{ type: i1.AdaptTranslateService }, { type: i0.ChangeDetectorRef }, { type: i2.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { mode: [{
            type: Input
        }], left: [{
            type: Input
        }], right: [{
            type: Input
        }], testID: [{
            type: Input
        }], adaptRadarDisableEventSending: [{
            type: Input
        }] }); })();
