// @bmc-ux/adapt-angular/esm2015/close/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptCloseComponent } from './close.component';
import * as i0 from "@angular/core";
export class AdaptCloseModule {
}
AdaptCloseModule.ɵfac = function AdaptCloseModule_Factory(t) { return new (t || AdaptCloseModule)(); };
AdaptCloseModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCloseModule });
AdaptCloseModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCloseModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [AdaptCloseComponent],
                declarations: [AdaptCloseComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCloseModule, { declarations: [AdaptCloseComponent], imports: [CommonModule], exports: [AdaptCloseComponent] }); })();
export * from './close.component';
