// @bmc-ux/obsolete/esm2015/lib/rating/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptIconModule, AdaptTooltipModule } from '@bmc-ux/adapt-angular';
import { AdaptRatingComponent } from './rating.component';
import { AdaptRatingVotingComponent } from './rating-voting.component';
import { AdaptRatingLikeComponent } from './rating-like.component';
import { AdaptRatingBinaryComponent } from './rating-binary.component';
import * as i0 from "@angular/core";
export class AdaptRatingModule {
}
AdaptRatingModule.ɵfac = function AdaptRatingModule_Factory(t) { return new (t || AdaptRatingModule)(); };
AdaptRatingModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRatingModule });
AdaptRatingModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptTooltipModule, AdaptIconModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRatingModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptTooltipModule, AdaptIconModule, FormsModule],
                exports: [AdaptRatingComponent, AdaptRatingVotingComponent, AdaptRatingLikeComponent, AdaptRatingBinaryComponent],
                declarations: [AdaptRatingComponent, AdaptRatingVotingComponent, AdaptRatingLikeComponent, AdaptRatingBinaryComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRatingModule, { declarations: [AdaptRatingComponent, AdaptRatingVotingComponent, AdaptRatingLikeComponent, AdaptRatingBinaryComponent], imports: [CommonModule, AdaptTooltipModule, AdaptIconModule, FormsModule], exports: [AdaptRatingComponent, AdaptRatingVotingComponent, AdaptRatingLikeComponent, AdaptRatingBinaryComponent] }); })();
export * from './rating.component';
export * from './rating.model';
export * from './rating-like.component';
export * from './rating-like.enum';
export * from './rating-voting.component';
export * from './rating-binary.component';
export * from './rating.model';
