// @bmc-ux/obsolete/esm2015/lib/rating/rating-voting.component.js
import { Component, ElementRef, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { AdaptRadarAngularCustomEventName, AdaptTranslateService, TooltipPlacement } from '@bmc-ux/adapt-angular';
import { AdaptRadarGenericDirective, AdaptRadarService } from '@bmc-ux/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
import * as i2 from "@bmc-ux/adapt-radar";
import * as i3 from "@angular/common";
function AdaptRatingVotingComponent_button_0_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r2.thanksText);
} }
const _c0 = function (a0, a1, a2, a3, a4, a5) { return { "d-icon-star": a0, "d-icon-star_o": a1, "d-icon-thumb_up_adapt": a2, "d-icon-thumb_up_o_adapt": a3, "d-icon-flag": a4, "d-icon-flag_o": a5 }; };
function AdaptRatingVotingComponent_button_0_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 4);
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵpropertyInterpolate("adaptTooltip", ctx_r3.tooltip !== "" ? ctx_r3.tooltip : "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction6(5, _c0, ctx_r3.type === "star" && ctx_r3.isActive, ctx_r3.type === "star" && !ctx_r3.isActive, ctx_r3.type === "like" && ctx_r3.isActive, ctx_r3.type === "like" && !ctx_r3.isActive, ctx_r3.type === "flag" && ctx_r3.isActive, ctx_r3.type === "flag" && !ctx_r3.isActive))("placement", ctx_r3.tooltipPlacement)("adaptRadarDisableEventSending", true);
    i0.ɵɵattribute("aria-label", ctx_r3.tooltip || "");
} }
const _c1 = function (a0, a1, a2, a3, a4) { return { "a-rating-voting__btn": a0, "btn btn-secondary btn-sm": a1, "is-star-icon": a2, "is-active": a3, "active": a4 }; };
function AdaptRatingVotingComponent_button_0_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 2);
    i0.ɵɵlistener("click", function AdaptRatingVotingComponent_button_0_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(); return ctx_r4.handleClick(); });
    i0.ɵɵtext(1, "\n\n  ");
    i0.ɵɵtemplate(2, AdaptRatingVotingComponent_button_0_span_2_Template, 2, 1, "span", 1);
    i0.ɵɵtext(3, "\n\n  ");
    i0.ɵɵtemplate(4, AdaptRatingVotingComponent_button_0_span_4_Template, 2, 12, "span", 3);
    i0.ɵɵtext(5, "\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction5(3, _c1, ctx_r0.type !== "thanks", ctx_r0.type === "thanks", ctx_r0.type === "star", ctx_r0.isActive, ctx_r0.type === "thanks" && ctx_r0.isActive));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0.type === "thanks");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0.type !== "thanks");
} }
function AdaptRatingVotingComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementStart(2, "span", 5);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", ctx_r1.thanksAbbrMessage, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1("\"", ctx_r1.thanksText, "\"");
} }
const RATING_VOTING_TYPES = [
    'star',
    'flag',
    'like',
    'thanks',
];
export class RatingVotingTypeError extends Error {
    constructor(type) {
        super(`Rating voting type "${type}" isn't supported (allowed: star, like, flag, thanks).`);
    }
}
export class AdaptRatingVotingComponent extends AdaptRadarGenericDirective {
    constructor(ts, _adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptRatingVoting);
        this.ts = ts;
        this.isActive = false;
        this.wasClick = true;
        this.type = 'star';
        // eslint-disable-next-line @angular-eslint/no-output-on-prefix, @angular-eslint/no-output-rename
        this.onVotingChanged = new EventEmitter();
        this._tooltip = '';
        this._tooltipPlacement = 'top';
        this.setDefaultTexts(ts.getCurrentLanguage());
    }
    get thanksText() {
        return this._thanksText;
    }
    set thanksText(text) {
        this._thanksText = text;
    }
    get thanksAbbrMessage() {
        return this._thanksAbbrMessage;
    }
    set thanksAbbrMessage(text) {
        this._thanksAbbrMessage = text;
    }
    get tooltip() {
        return this._tooltip;
    }
    set tooltip(v) {
        this._tooltip = v;
    }
    get tooltipPlacement() {
        return this._tooltipPlacement;
    }
    set tooltipPlacement(placement) {
        this._tooltipPlacement = placement;
    }
    setDefaultTexts(translation) {
        this.thanksText = translation['adapt.rating.vote.thank'];
        this.thanksAbbrMessage = translation['adapt.rating.vote.confirm'];
    }
    handleClick() {
        this.isActive = !this.isActive;
        this.onVotingChanged.emit({ type: this.type, isActive: this.isActive });
        if (this.type === 'thanks') {
            this.wasClick = false;
        }
    }
    ngOnChanges() {
        if (this.type && RATING_VOTING_TYPES.indexOf(this.type) === -1) {
            throw new RatingVotingTypeError(this.type);
        }
    }
}
AdaptRatingVotingComponent.ɵfac = function AdaptRatingVotingComponent_Factory(t) { return new (t || AdaptRatingVotingComponent)(i0.ɵɵdirectiveInject(i1.AdaptTranslateService), i0.ɵɵdirectiveInject(i2.AdaptRadarService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptRatingVotingComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptRatingVotingComponent, selectors: [["adapt-rating-voting"]], hostVars: 2, hostBindings: function AdaptRatingVotingComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("a-rating-voting", true);
    } }, inputs: { type: "type", thanksText: "thanksText", thanksAbbrMessage: "thanksAbbrMessage", tooltip: "tooltip", tooltipPlacement: "tooltipPlacement" }, outputs: { onVotingChanged: "voting-change" }, features: [i0.ɵɵInheritDefinitionFeature, i0.ɵɵNgOnChangesFeature], decls: 4, vars: 2, consts: [["type", "button", 3, "ngClass", "click", 4, "ngIf"], [4, "ngIf"], ["type", "button", 3, "ngClass", "click"], ["class", "a-rating-voting__icon", 3, "ngClass", "placement", "adaptRadarDisableEventSending", "adaptTooltip", 4, "ngIf"], [1, "a-rating-voting__icon", 3, "ngClass", "placement", "adaptRadarDisableEventSending", "adaptTooltip"], [1, "font-weight-semibold"]], template: function AdaptRatingVotingComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, AdaptRatingVotingComponent_button_0_Template, 6, 9, "button", 0);
        i0.ɵɵtext(1, "\n\n");
        i0.ɵɵtemplate(2, AdaptRatingVotingComponent_span_2_Template, 4, 2, "span", 1);
        i0.ɵɵtext(3, "\n\n\n\n\n");
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.wasClick);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx.wasClick);
    } }, directives: [i3.NgIf, i3.NgClass, i1.AdaptTooltipDirective], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRatingVotingComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-rating-voting',
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {
                    '[class.a-rating-voting]': 'true'
                },
                templateUrl: './rating-voting.component.html',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.AdaptTranslateService }, { type: i2.AdaptRadarService }, { type: i0.ElementRef }]; }, { type: [{
            type: Input
        }], onVotingChanged: [{
            type: Output,
            args: ['voting-change']
        }], thanksText: [{
            type: Input
        }], thanksAbbrMessage: [{
            type: Input
        }], tooltip: [{
            type: Input
        }], tooltipPlacement: [{
            type: Input
        }] }); })();
