// @bmc-ux/obsolete/esm2015/lib/rating/rating-binary.component.js
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewEncapsulation } from '@angular/core';
import { AdaptRadarAngularCustomEventName, AdaptTranslateService } from '@bmc-ux/adapt-angular';
import { AdaptRadarGenericDirective, AdaptRadarService } from '@bmc-ux/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
import * as i2 from "@bmc-ux/adapt-radar";
import * as i3 from "@angular/common";
const _c0 = function (a0) { return { "d-icon-thumb_up_adapt": a0 }; };
const _c1 = function (a0) { return { "d-icon-thumb_down_adapt": a0 }; };
function AdaptRatingBinaryComponent_div_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r5 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "button", 4);
    i0.ɵɵlistener("click", function AdaptRatingBinaryComponent_div_1_ng_container_2_Template_button_click_2_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r4 = i0.ɵɵnextContext(2); return ctx_r4.handleClick($event, "like"); });
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵelementStart(4, "span");
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n    ");
    i0.ɵɵelementStart(8, "button", 4);
    i0.ɵɵlistener("click", function AdaptRatingBinaryComponent_div_1_ng_container_2_Template_button_click_8_listener($event) { i0.ɵɵrestoreView(_r5); const ctx_r6 = i0.ɵɵnextContext(2); return ctx_r6.handleClick($event, "dislike"); });
    i0.ɵɵtext(9, "\n      ");
    i0.ɵɵelementStart(10, "span");
    i0.ɵɵtext(11);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c0, ctx_r2.type === "icon"));
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("sr-only", ctx_r2.type === "icon");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r2.btnTextLeft);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(10, _c1, ctx_r2.type === "icon"));
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("sr-only", ctx_r2.type === "icon");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r2.btnTextRight);
} }
function AdaptRatingBinaryComponent_div_1_div_4_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r7.showMessage);
} }
function AdaptRatingBinaryComponent_div_1_div_4_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementStart(2, "span", 6);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵprojection(5);
    i0.ɵɵtext(6, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", ctx_r8.textMessage, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate3("", ctx_r8.quotesLeft, "", ctx_r8.showBtnValue, "", ctx_r8.quotesRight, ". ");
} }
function AdaptRatingBinaryComponent_div_1_div_4_a_6_Template(rf, ctx) { if (rf & 1) {
    const _r11 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 7);
    i0.ɵɵlistener("click", function AdaptRatingBinaryComponent_div_1_div_4_a_6_Template_a_click_0_listener($event) { i0.ɵɵrestoreView(_r11); const ctx_r10 = i0.ɵɵnextContext(3); return ctx_r10.handleClick($event, ""); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r9.undoButton);
} }
function AdaptRatingBinaryComponent_div_1_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵtemplate(2, AdaptRatingBinaryComponent_div_1_div_4_span_2_Template, 2, 1, "span", 3);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵtemplate(4, AdaptRatingBinaryComponent_div_1_div_4_span_4_Template, 7, 4, "span", 3);
    i0.ɵɵtext(5, "\n    ");
    i0.ɵɵtemplate(6, AdaptRatingBinaryComponent_div_1_div_4_a_6_Template, 2, 1, "a", 5);
    i0.ɵɵtext(7, "\n  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r3.type === "icon");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r3.type === "text");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r3.undoButton !== "");
} }
function AdaptRatingBinaryComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 2);
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵtemplate(2, AdaptRatingBinaryComponent_div_1_ng_container_2_Template, 14, 12, "ng-container", 3);
    i0.ɵɵtext(3, "\n\n  ");
    i0.ɵɵtemplate(4, AdaptRatingBinaryComponent_div_1_div_4_Template, 8, 3, "div", 3);
    i0.ɵɵtext(5, "\n\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0.toggle);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r0.toggle);
} }
function AdaptRatingBinaryComponent_div_3_button_11_Template(rf, ctx) { if (rf & 1) {
    const _r15 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 14);
    i0.ɵɵlistener("click", function AdaptRatingBinaryComponent_div_3_button_11_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r15); const ctx_r14 = i0.ɵɵnextContext(2); return ctx_r14.changeValue("decrease"); });
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "span", 10);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r12.btnTextRight);
} }
function AdaptRatingBinaryComponent_div_3_div_13_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r18 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 7);
    i0.ɵɵlistener("click", function AdaptRatingBinaryComponent_div_3_div_13_a_2_Template_a_click_0_listener($event) { i0.ɵɵrestoreView(_r18); const ctx_r17 = i0.ɵɵnextContext(3); return ctx_r17.cancelValue($event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r16.undoButton);
} }
function AdaptRatingBinaryComponent_div_3_div_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 15);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵtemplate(2, AdaptRatingBinaryComponent_div_3_div_13_a_2_Template, 2, 1, "a", 5);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r13.undoButton !== "");
} }
function AdaptRatingBinaryComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r20 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 8);
    i0.ɵɵtext(1, "\n\n    ");
    i0.ɵɵelementStart(2, "button", 9);
    i0.ɵɵlistener("click", function AdaptRatingBinaryComponent_div_3_Template_button_click_2_listener() { i0.ɵɵrestoreView(_r20); const ctx_r19 = i0.ɵɵnextContext(); return ctx_r19.changeValue("increase"); });
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵelementStart(4, "span", 10);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n    ");
    i0.ɵɵelementStart(8, "div", 11);
    i0.ɵɵtext(9);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n    ");
    i0.ɵɵtemplate(11, AdaptRatingBinaryComponent_div_3_button_11_Template, 5, 1, "button", 12);
    i0.ɵɵtext(12, "\n\n    ");
    i0.ɵɵtemplate(13, AdaptRatingBinaryComponent_div_3_div_13_Template, 4, 1, "div", 13);
    i0.ɵɵtext(14, "\n\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("a-rating_btn-hidden", !ctx_r1.toggle);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.btnTextLeft);
    i0.ɵɵadvance(3);
    i0.ɵɵclassProp("font-weight-semibold", !ctx_r1.toggle);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r1.value);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.toggle);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r1.toggle);
} }
const _c2 = ["*"];
const RATING_BINARY_TYPE = ['icon', 'text', 'count'];
export class RatingBinaryTypeError extends Error {
    constructor(type) {
        super(`Rating binary type "${type}" isn't supported (allowed: 'icon', 'text', 'count').`);
    }
}
export class AdaptRatingBinaryComponent extends AdaptRadarGenericDirective {
    constructor(_element, _renderer, ts, _adaptRadarService) {
        super(_adaptRadarService, _element, AdaptRadarAngularCustomEventName.AdaptRatingBinary);
        this._element = _element;
        this._renderer = _renderer;
        this.ts = ts;
        this.toggle = true;
        // eslint-disable-next-line @angular-eslint/no-output-on-prefix, @angular-eslint/no-output-rename
        this.onChangeValue = new EventEmitter();
        this._type = 'text';
        this._value = 12;
        this.setDefaultTexts(ts.getCurrentLanguage());
    }
    get type() {
        return this._type;
    }
    set type(t) {
        this._type = t;
    }
    get likeMessage() {
        return this._likeMessage;
    }
    set likeMessage(message) {
        this._likeMessage = message;
    }
    get dislikeMessage() {
        return this._dislikeMessage;
    }
    set dislikeMessage(message) {
        this._dislikeMessage = message;
    }
    get textMessage() {
        return this._textMessage;
    }
    set textMessage(v) {
        this._textMessage = v;
    }
    get value() {
        return this._value;
    }
    set value(v) {
        this._value = v;
    }
    setDefaultTexts(translation) {
        this.btnTextLeft = translation['adapt.rating.binary.left'];
        this.btnTextRight = translation['adapt.rating.binary.right'];
        this.undoButton = translation['adapt.common.undo'];
        this.quotesLeft = translation['adapt.common.doubleQuotesLeft'];
        this.quotesRight = translation['adapt.common.doubleQuotesRight'];
        this._likeMessage = translation['adapt.rating.binary.like'];
        this._dislikeMessage = translation['adapt.rating.binary.dislike'];
        this._textMessage = translation['adapt.rating.binary.confirm'];
    }
    handleClick(e, type = '') {
        e.preventDefault();
        this._setHostStyles();
        this._toggleState();
        switch (type) {
            case 'like':
                this.showMessage = this.likeMessage;
                this.showBtnValue = this.btnTextLeft;
                this.onChangeValue.emit({ value: this.btnTextLeft });
                break;
            case 'dislike':
                this.showMessage = this.dislikeMessage;
                this.showBtnValue = this.btnTextRight;
                this.onChangeValue.emit({ value: this.btnTextRight });
                break;
            default:
                this.onChangeValue.emit({ value: this.undoButton });
                break;
        }
    }
    changeValue(action) {
        this._setHostStyles();
        this._toggleState();
        switch (action) {
            case 'increase':
                this.savedValue = this.value;
                this.value++;
                this.onChangeValue.emit({ value: this.value });
                break;
            case 'decrease':
                this.savedValue = this.value;
                this.value--;
                this.onChangeValue.emit({ value: this.value });
                break;
        }
    }
    cancelValue(e) {
        e.preventDefault();
        this.reset();
    }
    reset() {
        this.value = this.savedValue;
        this.onChangeValue.emit({ value: this.undoButton });
        this._toggleState();
    }
    ngOnChanges() {
        if (this.type && RATING_BINARY_TYPE.indexOf(this.type) === -1) {
            throw new RatingBinaryTypeError(this.type);
        }
    }
    _setHostStyles() {
        if (this._element) {
            const { nativeElement } = this._element;
            const { offsetHeight } = nativeElement;
            this._renderer.setStyle(nativeElement, 'minHeight', `${offsetHeight}px`);
            this._renderer.addClass(nativeElement, 'a-rating-center');
        }
    }
    _toggleState() {
        this.toggle = !this.toggle;
    }
}
AdaptRatingBinaryComponent.ɵfac = function AdaptRatingBinaryComponent_Factory(t) { return new (t || AdaptRatingBinaryComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i1.AdaptTranslateService), i0.ɵɵdirectiveInject(i2.AdaptRadarService)); };
AdaptRatingBinaryComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptRatingBinaryComponent, selectors: [["adapt-rating-binary"]], hostVars: 2, hostBindings: function AdaptRatingBinaryComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("a-rating-binary", true);
    } }, inputs: { btnTextLeft: "btnTextLeft", btnTextRight: "btnTextRight", undoButton: "undoButton", type: "type", likeMessage: "likeMessage", dislikeMessage: "dislikeMessage", textMessage: "textMessage", value: "value" }, outputs: { onChangeValue: "change-value" }, features: [i0.ɵɵInheritDefinitionFeature, i0.ɵɵNgOnChangesFeature], ngContentSelectors: _c2, decls: 5, vars: 2, consts: [["class", "btn-group", 4, "ngIf"], ["class", "btn-group a-rating-group", 4, "ngIf"], [1, "btn-group"], [4, "ngIf"], ["type", "button", 1, "btn", "btn-secondary", 3, "ngClass", "click"], ["href", "#", 3, "click", 4, "ngIf"], [1, "font-weight-semibold"], ["href", "#", 3, "click"], [1, "btn-group", "a-rating-group"], ["type", "button", 1, "btn", "btn-secondary", "d-icon-triangle_up", 3, "click"], [1, "sr-only"], [1, "a-rating-group__value"], ["type", "button", "class", "btn btn-secondary d-icon-triangle_down", 3, "click", 4, "ngIf"], ["class", "a-rating__undo-container", 4, "ngIf"], ["type", "button", 1, "btn", "btn-secondary", "d-icon-triangle_down", 3, "click"], [1, "a-rating__undo-container"]], template: function AdaptRatingBinaryComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtext(0, "\n");
        i0.ɵɵtemplate(1, AdaptRatingBinaryComponent_div_1_Template, 6, 2, "div", 0);
        i0.ɵɵtext(2, "\n\n");
        i0.ɵɵtemplate(3, AdaptRatingBinaryComponent_div_3_Template, 15, 8, "div", 1);
        i0.ɵɵtext(4, "\n\n\n\n\n");
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.type !== "count");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.type === "count");
    } }, directives: [i3.NgIf, i3.NgClass], styles: ["\n    /* Hot fix for IE 11 */\n    .a-rating-center::after {\n      content: '';\n      display: inline-block;\n      min-height: inherit;\n      font-size: 0;\n    }\n  "], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRatingBinaryComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-rating-binary',
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {
                    '[class.a-rating-binary]': 'true'
                },
                templateUrl: './rating-binary.component.html',
                encapsulation: ViewEncapsulation.None,
                styles: [`
    /* Hot fix for IE 11 */
    .a-rating-center::after {
      content: '';
      display: inline-block;
      min-height: inherit;
      font-size: 0;
    }
  `]
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.AdaptTranslateService }, { type: i2.AdaptRadarService }]; }, { btnTextLeft: [{
            type: Input
        }], btnTextRight: [{
            type: Input
        }], undoButton: [{
            type: Input
        }], onChangeValue: [{
            type: Output,
            args: ['change-value']
        }], type: [{
            type: Input
        }], likeMessage: [{
            type: Input
        }], dislikeMessage: [{
            type: Input
        }], textMessage: [{
            type: Input
        }], value: [{
            type: Input
        }] }); })();
