// ng-hcaptcha/esm2020/lib/ng-hcaptcha.module.mjs
import { NgModule } from '@angular/core';
import { NgHcaptchaComponent } from './ng-hcaptcha.component';
import { CAPTCHA_CONFIG } from './ng-hcaptcha-config';
import { NgHcaptchaInvisibleButtonDirective } from './ng-hcaptcha-invisible-button.directive';
import { NgHcaptchaService } from './ng-hcaptcha.service';
import * as i0 from "@angular/core";
export class NgHcaptchaModule {
    static forRoot(config) {
        return {
            ngModule: NgHcaptchaModule,
            providers: [
                NgHcaptchaService,
                {
                    provide: CAPTCHA_CONFIG,
                    useValue: config || []
                },
            ]
        };
    }
}
NgHcaptchaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgHcaptchaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaModule, declarations: [NgHcaptchaComponent, NgHcaptchaInvisibleButtonDirective], exports: [NgHcaptchaComponent, NgHcaptchaInvisibleButtonDirective] });
NgHcaptchaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [NgHcaptchaComponent, NgHcaptchaInvisibleButtonDirective],
                    exports: [NgHcaptchaComponent, NgHcaptchaInvisibleButtonDirective],
                }]
        }] });
