// ng-hcaptcha/esm2020/lib/ng-hcaptcha.service.mjs
import { Inject, Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { loadHCaptcha } from "./hcaptcha-utils";
import { CAPTCHA_CONFIG } from "./ng-hcaptcha-config";
import * as i0 from "@angular/core";
export class NgHcaptchaService {
    constructor(captchaConfig) {
        this.captchaConfig = captchaConfig;
    }
    verify() {
        return new Observable((subscriber) => {
            loadHCaptcha(this.captchaConfig.languageCode).subscribe(() => {
                setTimeout((context) => {
                    // Create hCaptcha element
                    if (!this.hCaptchaElement) {
                        this.hCaptchaElement = document.createElement('div');
                        document.body.appendChild(this.hCaptchaElement);
                    }
                    // Render hCaptcha using the defined options
                    if (!this.hCaptchaWidgetId) {
                        // Configure hCaptcha
                        const options = {
                            sitekey: this.captchaConfig.siteKey,
                            size: 'invisible',
                            callback: (res) => {
                                subscriber.next(res);
                                subscriber.complete();
                                this.resetHcaptcha();
                            },
                            'expired-callback': (res) => {
                                subscriber.error(res);
                                this.resetHcaptcha();
                            },
                            'error-callback': (err) => {
                                subscriber.error(err);
                                this.resetHcaptcha();
                            },
                        };
                        this.hCaptchaWidgetId = window.hcaptcha.render(this.hCaptchaElement, options);
                    }
                    // Immediately execute hCaptcha
                    window.hcaptcha.execute(this.hCaptchaWidgetId);
                }, 50, this);
            });
        });
    }
    resetHcaptcha() {
        window.hcaptcha.remove(this.hCaptchaWidgetId);
        this.hCaptchaElement = null;
        this.hCaptchaWidgetId = null;
    }
}
NgHcaptchaService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaService, deps: [{ token: CAPTCHA_CONFIG }], target: i0.ɵɵFactoryTarget.Injectable });
NgHcaptchaService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.5", ngImport: i0, type: NgHcaptchaService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [CAPTCHA_CONFIG]
                }] }]; } });
