// @helix/platform/esm2015/shared/components/dynamic-component-renderer/dynamic-component-renderer.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RxDynamicComponentRendererComponent } from './dynamic-component-renderer.component';
import * as i0 from "@angular/core";
export class RxDynamicComponentRendererModule {
}
RxDynamicComponentRendererModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxDynamicComponentRendererModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxDynamicComponentRendererModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxDynamicComponentRendererModule, declarations: [RxDynamicComponentRendererComponent], imports: [CommonModule], exports: [RxDynamicComponentRendererComponent] });
RxDynamicComponentRendererModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxDynamicComponentRendererModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxDynamicComponentRendererModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [RxDynamicComponentRendererComponent],
                    exports: [RxDynamicComponentRendererComponent],
                    entryComponents: [RxDynamicComponentRendererComponent]
                }]
        }] });
