// @bmc-ux/adapt-angular/esm2015/truncate-file-name/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptTruncateFileNameDirective } from './truncate-file-name.directive';
import * as i0 from "@angular/core";
export class AdaptTruncateFileNameModule {
}
AdaptTruncateFileNameModule.ɵfac = function AdaptTruncateFileNameModule_Factory(t) { return new (t || AdaptTruncateFileNameModule)(); };
AdaptTruncateFileNameModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTruncateFileNameModule });
AdaptTruncateFileNameModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTruncateFileNameModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptTruncateFileNameDirective],
                exports: [AdaptTruncateFileNameDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTruncateFileNameModule, { declarations: [AdaptTruncateFileNameDirective], imports: [CommonModule], exports: [AdaptTruncateFileNameDirective] }); })();
export * from './truncate-file-name.directive';
