// ngx-kjua/esm2020/app/modules/ngx-kjua/ngx-kjua.component.mjs
import { ChangeDetectionStrategy, Component, Inject, Input, Output, PLATFORM_ID, ViewChild, EventEmitter } from "@angular/core";
import * as kjua from "kjua-svg";
import * as i0 from "@angular/core";
export class NgxKjuaComponent {
    constructor(platformId) {
        this.platformId = platformId;
        /**
         * render method
         */
        this.render = "svg";
        /**
         * render pixel-perfect lines
         */
        this.crisp = true;
        /**
         * minimum version= 1..40
         */
        this.minVersion = 1;
        /**
         * error correction level
         */
        this.ecLevel = "L";
        /**
         * size in pixel
         */
        this.size = 200;
        /**
         * pixel-ratio; undefined for devicePixelRatio
         */
        this.ratio = undefined;
        /**
         * code color
         */
        this.fill = "#333";
        /**
         * background color
         */
        this.back = "#fff";
        /**
         * content
         */
        this.text = "";
        /**
         * roundend corners in pc= 0..100
         */
        this.rounded = 0;
        /**
         * quiet zone in modules
         */
        this.quiet = 0;
        /**
         * modes
         */
        this.mode = "plain";
        /**
         * label/image size and pos in pc= 0..100
         */
        this.mSize = 30;
        this.mPosX = 50;
        this.mPosY = 50;
        this.image = undefined;
        this.imageAsCode = false;
        /**
         * label
         */
        this.label = "";
        this.fontname = "sans-serif";
        this.fontcolor = "#333";
        this.fontoutline = true;
        /**
         * If true, rendering is done inside "requestAnimationFrame"-call.
         * Use this if you want to generate more than one code (e.g. batch)
         */
        this.renderAsync = false;
        this.codeFinished = new EventEmitter();
        this.viewInitialized = false;
    }
    ngAfterViewInit() {
        this.viewInitialized = true;
        this.updateView();
    }
    ngOnChanges(changes) {
        if (this.viewInitialized) {
            this.updateView();
        }
    }
    get template() {
        const settings = {
            render: this.render,
            crisp: this.crisp,
            minVersion: this.minVersion,
            ecLevel: this.ecLevel,
            size: this.size,
            ratio: this.ratio,
            fill: this.fill,
            back: this.back,
            text: this.text,
            rounded: this.rounded,
            quiet: this.quiet,
            mode: this.mode,
            mSize: this.mSize,
            mPosX: this.mPosX,
            mPosY: this.mPosY,
            label: this.label,
            fontname: this.fontname,
            fontcolor: this.fontcolor,
            image: this.image,
            fontoutline: this.fontoutline,
            imageAsCode: this.imageAsCode,
            elementId: this.elementId
        };
        return kjua(settings);
    }
    renderCode() {
        this.div.nativeElement.innerHTML = "";
        const node = this.template;
        this.div.nativeElement.appendChild(node);
        this.codeFinished.next(node);
    }
    updateView() {
        this.div.nativeElement.style.width = +this.size;
        this.div.nativeElement.style.height = +this.size;
        if (typeof this.image === "string") {
            const img = new Image();
            img.crossOrigin = "anonymous";
            img.onload = () => {
                if (this.renderAsync) {
                    requestAnimationFrame(() => this.renderCode());
                }
                else {
                    this.renderCode();
                }
            };
            img.src = 'data:image/png;base64,' + this.image;
        }
        else {
            if (this.renderAsync) {
                requestAnimationFrame(() => this.renderCode());
            }
            else {
                this.renderCode();
            }
        }
    }
}
NgxKjuaComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NgxKjuaComponent, deps: [{ token: PLATFORM_ID }], target: i0.ɵɵFactoryTarget.Component });
NgxKjuaComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.6", type: NgxKjuaComponent, selector: "ngx-kjua", inputs: { render: "render", crisp: "crisp", minVersion: "minVersion", ecLevel: "ecLevel", size: "size", ratio: "ratio", fill: "fill", back: "back", text: "text", rounded: "rounded", quiet: "quiet", mode: "mode", mSize: "mSize", mPosX: "mPosX", mPosY: "mPosY", image: "image", imageAsCode: "imageAsCode", label: "label", fontname: "fontname", fontcolor: "fontcolor", fontoutline: "fontoutline", renderAsync: "renderAsync", cssClass: "cssClass", elementId: "elementId" }, outputs: { codeFinished: "codeFinished" }, viewQueries: [{ propertyName: "div", first: true, predicate: ["elem"], descendants: true }], usesOnChanges: true, ngImport: i0, template: `
    <div [class]="cssClass" #elem></div>`, isInline: true, styles: [":host{display:block}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NgxKjuaComponent, decorators: [{
            type: Component,
            args: [{
                    selector: "ngx-kjua",
                    template: `
    <div [class]="cssClass" #elem></div>`,
                    styles: [`
    :host {
      display: block;
    }
  `],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return [{ type: Object, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }]; }, propDecorators: { render: [{
                type: Input
            }], crisp: [{
                type: Input
            }], minVersion: [{
                type: Input
            }], ecLevel: [{
                type: Input
            }], size: [{
                type: Input
            }], ratio: [{
                type: Input
            }], fill: [{
                type: Input
            }], back: [{
                type: Input
            }], text: [{
                type: Input
            }], rounded: [{
                type: Input
            }], quiet: [{
                type: Input
            }], mode: [{
                type: Input
            }], mSize: [{
                type: Input
            }], mPosX: [{
                type: Input
            }], mPosY: [{
                type: Input
            }], image: [{
                type: Input
            }], imageAsCode: [{
                type: Input
            }], label: [{
                type: Input
            }], fontname: [{
                type: Input
            }], fontcolor: [{
                type: Input
            }], fontoutline: [{
                type: Input
            }], renderAsync: [{
                type: Input
            }], cssClass: [{
                type: Input
            }], elementId: [{
                type: Input
            }], codeFinished: [{
                type: Output
            }], div: [{
                type: ViewChild,
                args: ["elem"]
            }] } });
