// ngx-kjua/esm2020/app/modules/ngx-kjua/ngx-kjua.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { NgxKjuaComponent } from "./ngx-kjua.component";
import * as i0 from "@angular/core";
export class NgxKjuaModule {
}
NgxKjuaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NgxKjuaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxKjuaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NgxKjuaModule, declarations: [NgxKjuaComponent], imports: [CommonModule], exports: [NgxKjuaComponent] });
NgxKjuaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NgxKjuaModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: NgxKjuaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        NgxKjuaComponent
                    ],
                    exports: [
                        NgxKjuaComponent
                    ]
                }]
        }] });
