// @bmc-ux/adapt-angular/esm2015/agreement/agreement-content/agreement-content.component.js
import { Component, ElementRef, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../../common/radar/adapt-radar';
import { AdaptTranslateService } from '../../common/i18n';
import { formatString } from '../../common/utilities';
import * as i0 from "@angular/core";
import * as i1 from "../../common/i18n";
import * as i2 from "../../common/radar/adapt-radar";
import * as i3 from "@angular/common";
import * as i4 from "../../button/button.component";
import * as i5 from "../../collapse/collapse.directive";
import * as i6 from "../../rx/rx-switch/adapt-rx-switch.component";
import * as i7 from "@angular/forms";
function AdaptAgreementContentComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    const _r4 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "h4");
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementStart(5, "div");
    i0.ɵɵtext(6, "\n      ");
    i0.ɵɵelementStart(7, "adapt-rx-switch", 8);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_7_listener($event) { i0.ɵɵrestoreView(_r4); const ctx_r3 = i0.ɵɵnextContext(); return ctx_r3.state.organizationPerformance = $event; })("ngModelChange", function AdaptAgreementContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_7_listener() { i0.ɵɵrestoreView(_r4); const ctx_r5 = i0.ɵɵnextContext(); return ctx_r5.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n    ");
    i0.ɵɵelementStart(10, "div");
    i0.ɵɵtext(11, "\n      ");
    i0.ɵɵelementStart(12, "adapt-rx-switch", 9);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_12_listener($event) { i0.ɵɵrestoreView(_r4); const ctx_r6 = i0.ɵɵnextContext(); return ctx_r6.state.accountPerformance = $event; })("ngModelChange", function AdaptAgreementContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_12_listener() { i0.ɵɵrestoreView(_r4); const ctx_r7 = i0.ɵɵnextContext(); return ctx_r7.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r0.texts.switchersMainTitle);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r0.state.organizationPerformance)("label", ctx_r0.texts.organizationSwitcherLabel);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("disabled", !ctx_r0.state.organizationPerformance)("ngModel", ctx_r0.state.accountPerformance)("label", ctx_r0.texts.accountSwitcherLabel);
} }
function AdaptAgreementContentComponent_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    const _r9 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "adapt-rx-switch", 8);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementContentComponent_ng_container_9_Template_adapt_rx_switch_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r9); const ctx_r8 = i0.ɵɵnextContext(); return ctx_r8.state.accountPerformance = $event; })("ngModelChange", function AdaptAgreementContentComponent_ng_container_9_Template_adapt_rx_switch_ngModelChange_2_listener() { i0.ɵɵrestoreView(_r9); const ctx_r10 = i0.ɵɵnextContext(); return ctx_r10.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r1.state.accountPerformance)("label", ctx_r1.texts.performanceSwitcherLabel);
} }
function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r15 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "h4");
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "div");
    i0.ɵɵtext(6, "\n        ");
    i0.ɵɵelementStart(7, "adapt-rx-switch", 8);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_2_Template_adapt_rx_switch_ngModelChange_7_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r14 = i0.ɵɵnextContext(3); return ctx_r14.state.organizationMarketing = $event; })("ngModelChange", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_2_Template_adapt_rx_switch_ngModelChange_7_listener() { i0.ɵɵrestoreView(_r15); const ctx_r16 = i0.ɵɵnextContext(3); return ctx_r16.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n      ");
    i0.ɵɵelementStart(10, "div");
    i0.ɵɵtext(11, "\n        ");
    i0.ɵɵelementStart(12, "adapt-rx-switch", 9);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_2_Template_adapt_rx_switch_ngModelChange_12_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r17 = i0.ɵɵnextContext(3); return ctx_r17.state.accountMarketing = $event; })("ngModelChange", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_2_Template_adapt_rx_switch_ngModelChange_12_listener() { i0.ɵɵrestoreView(_r15); const ctx_r18 = i0.ɵɵnextContext(3); return ctx_r18.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r12.texts.marketingAndOutreachTitle);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r12.state.organizationMarketing)("label", ctx_r12.texts.organizationSwitcherLabel);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("disabled", !ctx_r12.state.organizationMarketing)("ngModel", ctx_r12.state.accountMarketing)("label", ctx_r12.texts.accountSwitcherLabel);
} }
function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    const _r20 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "adapt-rx-switch", 8);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_4_Template_adapt_rx_switch_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r20); const ctx_r19 = i0.ɵɵnextContext(3); return ctx_r19.state.accountMarketing = $event; })("ngModelChange", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_4_Template_adapt_rx_switch_ngModelChange_2_listener() { i0.ɵɵrestoreView(_r20); const ctx_r21 = i0.ɵɵnextContext(3); return ctx_r21.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r13.state.accountMarketing)("label", ctx_r13.texts.marketingAndOutreachSwitcherLabel);
} }
const _c0 = function (a0, a1) { return { "d-icon-angle_down": a0, "d-icon-angle_up": a1 }; };
const _c1 = function (a0, a1) { return { "mb-5": a0, "agreement__text-hidden": a1 }; };
function AdaptAgreementContentComponent_ng_container_64_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r23 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵtemplate(2, AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_2_Template, 15, 8, "ng-container", 1);
    i0.ɵɵtext(3, "\n\n    ");
    i0.ɵɵtemplate(4, AdaptAgreementContentComponent_ng_container_64_ng_container_2_ng_container_4_Template, 4, 3, "ng-container", 1);
    i0.ɵɵtext(5, "\n\n    ");
    i0.ɵɵelementStart(6, "div", 2);
    i0.ɵɵtext(7, "\n\n      ");
    i0.ɵɵelementStart(8, "button", 3);
    i0.ɵɵlistener("click", function AdaptAgreementContentComponent_ng_container_64_ng_container_2_Template_button_click_8_listener() { i0.ɵɵrestoreView(_r23); const ctx_r22 = i0.ɵɵnextContext(2); return ctx_r22.isCollapsedMarketingContent = !ctx_r22.isCollapsedMarketingContent; });
    i0.ɵɵtext(9, "\n        ");
    i0.ɵɵelementStart(10, "span");
    i0.ɵɵtext(11);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n        ");
    i0.ɵɵelement(13, "span", 4);
    i0.ɵɵtext(14, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(15, "\n\n      ");
    i0.ɵɵelementStart(16, "div", 5);
    i0.ɵɵtext(17, "\n        ");
    i0.ɵɵelementStart(18, "p", 6);
    i0.ɵɵtext(19);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(20, "\n\n        ");
    i0.ɵɵelementStart(21, "ul", 7);
    i0.ɵɵtext(22, "\n          ");
    i0.ɵɵelementStart(23, "li");
    i0.ɵɵtext(24);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n          ");
    i0.ɵɵelementStart(26, "li");
    i0.ɵɵtext(27);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(28, "\n          ");
    i0.ɵɵelementStart(29, "li");
    i0.ɵɵtext(30);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(31, "\n          ");
    i0.ɵɵelementStart(32, "li");
    i0.ɵɵtext(33);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(34, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(35, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(36, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(37, "\n\n    ");
    i0.ɵɵelement(38, "hr", 0);
    i0.ɵɵtext(39, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r11.showOrganizationSettings);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r11.showOrganizationSettings);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r11.texts.learnMoreLabel);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(13, _c0, ctx_r11.isCollapsedMarketingContent, !ctx_r11.isCollapsedMarketingContent));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("collapse", ctx_r11.isCollapsedMarketingContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(16, _c1, !ctx_r11.isCollapsedMarketingContent, ctx_r11.isCollapsedMarketingContent));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r11.texts.marketingAndOutreachNote);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r11.texts.learnMoreListItem11);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r11.texts.learnMoreListItem12);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r11.texts.learnMoreListItem13);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r11.texts.learnMoreListItem14);
} }
function AdaptAgreementContentComponent_ng_container_64_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n  ");
    i0.ɵɵtemplate(2, AdaptAgreementContentComponent_ng_container_64_ng_container_2_Template, 40, 19, "ng-container", 1);
    i0.ɵɵtext(3, "\n");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r2.state.organizationPerformance || ctx_r2.state.accountPerformance);
} }
export class AdaptAgreementContentComponent extends AdaptRadarAngularGenericDirective {
    constructor(_ts, _adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptAgreementContent);
        this._ts = _ts;
        this.isCollapsedPerformanceContent = true;
        this.isCollapsedMarketingContent = true;
        this.texts = {};
        this.multiProductUsage = false;
        this.stateChange = new EventEmitter();
        this._currentLanguage = _ts.getCurrentLanguage();
        this._setDefaultTexts(this._currentLanguage);
    }
    handleStateChange() {
        if (this.showOrganizationSettings) {
            if (!this.state.organizationPerformance) {
                this.state.organizationMarketing = false;
                this.state.accountPerformance = false;
            }
            if (!this.state.organizationMarketing) {
                this.state.accountMarketing = false;
            }
        }
        else {
            if (!this.state.accountPerformance) {
                this.state.accountMarketing = false;
            }
        }
        this.stateChange.emit(this.state);
    }
    get learnMoreNoteText() {
        return formatString(this._currentLanguage['adapt.agreement.common.learnMoreNote'], this.multiProductUsage ? this._currentLanguage['adapt.agreement.common.learnMoreNoteMultiProductUsage'] : this._currentLanguage['adapt.agreement.common.learnMoreNoteNotMultiProductUsage']);
    }
    _setDefaultTexts(translation) {
        this.texts.settingsNote = translation['adapt.agreement.content.settingsNote'];
        this.texts.switchersMainTitle = translation['adapt.agreement.content.switchersMainTitle'];
        this.texts.organizationSwitcherLabel = translation['adapt.agreement.content.organizationSwitcherLabel'];
        this.texts.accountSwitcherLabel = translation['adapt.agreement.content.accountSwitcherLabel'];
        this.texts.performanceSwitcherLabel = translation['adapt.agreement.content.performanceSwitcherLabel'];
        this.texts.learnMoreLabel = translation['adapt.agreement.content.learnMoreLabel'];
        this.texts.learnMoreNoteNotMultiProductUsage = translation['adapt.agreement.common.learnMoreNoteNotMultiProductUsage'];
        this.texts.learnMoreNoteMultiProductUsage = translation['adapt.agreement.common.learnMoreNoteMultiProductUsage'];
        this.texts.learnMoreListItem1 = translation['adapt.agreement.common.learnMoreListItem1'];
        this.texts.learnMoreListItem2 = translation['adapt.agreement.common.learnMoreListItem2'];
        this.texts.learnMoreListItem3 = translation['adapt.agreement.common.learnMoreListItem3'];
        this.texts.learnMoreListItem4 = translation['adapt.agreement.common.learnMoreListItem4'];
        this.texts.learnMoreListItem5 = translation['adapt.agreement.common.learnMoreListItem5'];
        this.texts.learnMoreListItem6 = translation['adapt.agreement.common.learnMoreListItem6'];
        this.texts.learnMoreListItem7 = translation['adapt.agreement.common.learnMoreListItem7'];
        this.texts.learnMoreListItem8 = translation['adapt.agreement.common.learnMoreListItem8'];
        this.texts.learnMoreListItem9 = translation['adapt.agreement.common.learnMoreListItem9'];
        this.texts.learnMoreListItem10 = translation['adapt.agreement.common.learnMoreListItem10'];
        this.texts.learnMoreListItem11 = translation['adapt.agreement.common.learnMoreListItem11'];
        this.texts.learnMoreListItem12 = translation['adapt.agreement.common.learnMoreListItem12'];
        this.texts.learnMoreListItem13 = translation['adapt.agreement.common.learnMoreListItem13'];
        this.texts.learnMoreListItem14 = translation['adapt.agreement.common.learnMoreListItem14'];
        this.texts.marketingAndOutreachTitle = translation['adapt.agreement.content.marketingAndOutreachTitle'];
        this.texts.marketingAndOutreachSwitcherLabel = translation['adapt.agreement.content.marketingAndOutreachSwitcherLabel'];
        this.texts.marketingAndOutreachNote = translation['adapt.agreement.content.marketingAndOutreachNote'];
        this.texts.agreementNote = translation['adapt.agreement.common.agreementNote'];
    }
}
AdaptAgreementContentComponent.ɵfac = function AdaptAgreementContentComponent_Factory(t) { return new (t || AdaptAgreementContentComponent)(i0.ɵɵdirectiveInject(i1.AdaptTranslateService), i0.ɵɵdirectiveInject(i2.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptAgreementContentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptAgreementContentComponent, selectors: [["adapt-agreement-content"]], inputs: { multiProductUsage: "multiProductUsage", state: "state", showOrganizationSettings: "showOrganizationSettings", showMarketingSection: "showMarketingSection" }, outputs: { stateChange: "stateChange" }, features: [i0.ɵɵInheritDefinitionFeature], decls: 69, vars: 28, consts: [[1, "agreement__separator"], [4, "ngIf"], [1, "agreement__content"], ["type", "button", "adapt-button", "", "btn-type", "tertiary", 1, "p-0", "mt-3", "mb-3", "agreement__collapse-btn", 3, "adaptRadarDisableEventSending", "click"], [1, "p-start-1", 3, "ngClass"], [3, "collapse", "adaptRadarDisableEventSending", "ngClass"], [1, "agreement__content-text"], [1, "m-0"], [1, "agreement__switcher", 3, "adaptRadarDisableEventSending", "ngModel", "label", "ngModelChange"], [1, "agreement__switcher", 3, "adaptRadarDisableEventSending", "disabled", "ngModel", "label", "ngModelChange"]], template: function AdaptAgreementContentComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "p");
        i0.ɵɵtext(1);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(2, "\n\n");
        i0.ɵɵelement(3, "hr", 0);
        i0.ɵɵtext(4, "\n");
        i0.ɵɵelementContainerStart(5);
        i0.ɵɵtext(6, "\n  ");
        i0.ɵɵtemplate(7, AdaptAgreementContentComponent_ng_container_7_Template, 15, 8, "ng-container", 1);
        i0.ɵɵtext(8, "\n\n  ");
        i0.ɵɵtemplate(9, AdaptAgreementContentComponent_ng_container_9_Template, 4, 3, "ng-container", 1);
        i0.ɵɵtext(10, "\n\n  ");
        i0.ɵɵelementStart(11, "div", 2);
        i0.ɵɵtext(12, "\n\n    ");
        i0.ɵɵelementStart(13, "button", 3);
        i0.ɵɵlistener("click", function AdaptAgreementContentComponent_Template_button_click_13_listener() { return ctx.isCollapsedPerformanceContent = !ctx.isCollapsedPerformanceContent; });
        i0.ɵɵtext(14, "\n      ");
        i0.ɵɵelementStart(15, "span");
        i0.ɵɵtext(16);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(17, "\n      ");
        i0.ɵɵelement(18, "span", 4);
        i0.ɵɵtext(19, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(20, "\n\n    ");
        i0.ɵɵelementStart(21, "div", 5);
        i0.ɵɵtext(22, "\n\n      ");
        i0.ɵɵelementStart(23, "p", 6);
        i0.ɵɵtext(24);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(25, "\n\n      ");
        i0.ɵɵelementStart(26, "ul", 7);
        i0.ɵɵtext(27, "\n        ");
        i0.ɵɵelementStart(28, "li");
        i0.ɵɵtext(29);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(30, "\n        ");
        i0.ɵɵelementStart(31, "li");
        i0.ɵɵtext(32);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(33, "\n        ");
        i0.ɵɵelementStart(34, "li");
        i0.ɵɵtext(35);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(36, "\n        ");
        i0.ɵɵelementStart(37, "li");
        i0.ɵɵtext(38);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(39, "\n        ");
        i0.ɵɵelementStart(40, "li");
        i0.ɵɵtext(41);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(42, "\n        ");
        i0.ɵɵelementStart(43, "li");
        i0.ɵɵtext(44);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(45, "\n        ");
        i0.ɵɵelementStart(46, "li");
        i0.ɵɵtext(47);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(48, "\n        ");
        i0.ɵɵelementStart(49, "li");
        i0.ɵɵtext(50);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(51, "\n        ");
        i0.ɵɵelementStart(52, "li");
        i0.ɵɵtext(53);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(54, "\n        ");
        i0.ɵɵelementStart(55, "li");
        i0.ɵɵtext(56);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(57, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(58, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(59, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(60, "\n\n  ");
        i0.ɵɵelement(61, "hr", 0);
        i0.ɵɵtext(62, "\n");
        i0.ɵɵelementContainerEnd();
        i0.ɵɵtext(63, "\n");
        i0.ɵɵtemplate(64, AdaptAgreementContentComponent_ng_container_64_Template, 4, 1, "ng-container", 1);
        i0.ɵɵtext(65, "\n");
        i0.ɵɵelementStart(66, "p");
        i0.ɵɵtext(67);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(68, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx.texts.settingsNote);
        i0.ɵɵadvance(6);
        i0.ɵɵproperty("ngIf", ctx.showOrganizationSettings);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx.showOrganizationSettings);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("adaptRadarDisableEventSending", true);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreLabel);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(22, _c0, ctx.isCollapsedPerformanceContent, !ctx.isCollapsedPerformanceContent));
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("collapse", ctx.isCollapsedPerformanceContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(25, _c1, !ctx.isCollapsedPerformanceContent, ctx.isCollapsedPerformanceContent));
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.learnMoreNoteText);
        i0.ɵɵadvance(5);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem1);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem2);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem3);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem4);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem5);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem6);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem7);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem8);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem9);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.learnMoreListItem10);
        i0.ɵɵadvance(8);
        i0.ɵɵproperty("ngIf", ctx.showMarketingSection);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.texts.agreementNote);
    } }, directives: [i3.NgIf, i4.AdaptButtonComponent, i3.NgClass, i5.AdaptCollapseDirective, i6.AdaptRxSwitchComponent, i7.NgControlStatus, i7.NgModel], styles: [".adapt-agreement.adapt-agreement_no-shadow{box-shadow:none}.agreement__main-title{margin:0 0 1rem}.agreement__separator{margin:1rem 0}.agreement__switcher{font-size:.9375rem}.agreement__content{margin-left:4.125em}.adapt-agreement__content_empty-state{font-size:.9375rem;text-align:center;width:100%}.agreement__content-text{margin:1rem 0}.agreement__badge .badge{min-width:70px}.agreement__collapse-btn{color:var(--text-primary)}adapt-alert .text-info{display:flex}.agreement__text-hidden .collapse-block{height:0;overflow:hidden;visibility:hidden}"], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAgreementContentComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-agreement-content',
                styleUrls: ['../agreement.scss'],
                templateUrl: './agreement-content.component.html',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.AdaptTranslateService }, { type: i2.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { multiProductUsage: [{
            type: Input
        }], state: [{
            type: Input
        }], showOrganizationSettings: [{
            type: Input
        }], showMarketingSection: [{
            type: Input
        }], stateChange: [{
            type: Output
        }] }); })();
