// @bmc-ux/adapt-angular/esm2015/menu/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptMenuComponent } from './menu.component';
import { AdaptIconModule } from '../icon/index';
import { AdaptButtonModule } from '../button/index';
import { RouterModule } from '@angular/router';
import { AdaptMenuAnchorDirective, AdaptMenuDirective, AdaptSubmenuComponent } from './submenu/submenu.component';
import { AdaptFocusMonitorDirectiveModule } from '../common/focus-monitor.directive';
import { AdaptMenuMobileViewComponent } from './menu-mobile-view/menu-mobile-view.component';
import { AdaptFocusHolderModule } from '../focus-holder/index';
import { FormsModule } from '@angular/forms';
import { MenuGlobalService } from './menu-global.service';
import { AdaptDragonInputHandlerModule } from '../dragon-input-handler/index';
import { AdaptRxTextfieldModule } from '../rx/rx-textfield/index';
import * as i0 from "@angular/core";
export class AdaptMenuModule {
}
AdaptMenuModule.ɵfac = function AdaptMenuModule_Factory(t) { return new (t || AdaptMenuModule)(); };
AdaptMenuModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptMenuModule });
AdaptMenuModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [
        MenuGlobalService
    ], imports: [[
            CommonModule,
            FormsModule,
            AdaptIconModule,
            AdaptButtonModule,
            RouterModule,
            AdaptFocusHolderModule,
            AdaptFocusMonitorDirectiveModule,
            AdaptRxTextfieldModule,
            AdaptDragonInputHandlerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMenuModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptIconModule,
                    AdaptButtonModule,
                    RouterModule,
                    AdaptFocusHolderModule,
                    AdaptFocusMonitorDirectiveModule,
                    AdaptRxTextfieldModule,
                    AdaptDragonInputHandlerModule
                ],
                exports: [AdaptMenuComponent],
                declarations: [
                    AdaptMenuComponent,
                    AdaptSubmenuComponent,
                    AdaptMenuAnchorDirective,
                    AdaptMenuDirective,
                    AdaptMenuMobileViewComponent
                ],
                entryComponents: [AdaptSubmenuComponent],
                providers: [
                    MenuGlobalService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptMenuModule, { declarations: [AdaptMenuComponent,
        AdaptSubmenuComponent,
        AdaptMenuAnchorDirective,
        AdaptMenuDirective,
        AdaptMenuMobileViewComponent], imports: [CommonModule,
        FormsModule,
        AdaptIconModule,
        AdaptButtonModule,
        RouterModule,
        AdaptFocusHolderModule,
        AdaptFocusMonitorDirectiveModule,
        AdaptRxTextfieldModule,
        AdaptDragonInputHandlerModule], exports: [AdaptMenuComponent] }); })();
export * from './menu.component';
export * from './submenu/submenu.component';
export * from './models/menu-config';
export * from './models/menu-item';
export * from './menu-mobile-view/menu-mobile-view.component';
