// @bmc-ux/adapt-angular/esm2015/alert/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptAlertComponent } from './alert.component';
import * as i0 from "@angular/core";
export class AdaptAlertModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptAlertModule
        };
    }
}
AdaptAlertModule.ɵfac = function AdaptAlertModule_Factory(t) { return new (t || AdaptAlertModule)(); };
AdaptAlertModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAlertModule });
AdaptAlertModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAlertModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptAlertComponent],
                exports: [AdaptAlertComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAlertModule, { declarations: [AdaptAlertComponent], imports: [CommonModule], exports: [AdaptAlertComponent] }); })();
export * from './alert.component';
export * from './alert.config';
