// @bmc-ux/adapt-angular/esm2015/card-carousel/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptCardCarouselComponent, } from './card-carousel.component';
import * as i0 from "@angular/core";
export class AdaptCardCarouselModule {
}
AdaptCardCarouselModule.ɵfac = function AdaptCardCarouselModule_Factory(t) { return new (t || AdaptCardCarouselModule)(); };
AdaptCardCarouselModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCardCarouselModule });
AdaptCardCarouselModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCardCarouselModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                exports: [AdaptCardCarouselComponent],
                declarations: [AdaptCardCarouselComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCardCarouselModule, { declarations: [AdaptCardCarouselComponent], imports: [CommonModule], exports: [AdaptCardCarouselComponent] }); })();
export * from './card-carousel.component';
