// @bmc-ux/adapt-charts/esm2015/adapt-charts.module.js
import { NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptChartComponent } from './common/chart.component';
import { AdaptDropdownModule, AdaptEmptyStateModule, AdaptIconModule, AdaptListKeyManagerModule, AdaptSliderModule, AdaptTooltipModule } from '@bmc-ux/adapt-angular';
import { AdaptHeatmapComponent } from './heatmap/heatmap.component';
import { AdaptLineGraphComponent } from './line-graph/line-graph.component';
import { AdaptScatterPlotComponent } from './scatter-plot/scatter-plot.component';
import { AdaptTreemapComponent } from './treemap/treemap.component';
import { ChartColorService } from './common/services/chart-color.service';
import { ChartLegendComponent } from './common/legend/chart-legend.component';
import { ChartGradientLegendComponent } from './common/legend/chart-gradient-legend.component';
import { CutNumberPipe } from './common/pipes/cut-number.pipe';
import { AdaptPieChartComponent } from './pie-chart/pie-chart.component';
import { AdaptStackedChartComponent } from './stacked-chart/stacked-chart.component';
import { AdaptChartSliderComponent } from './common/chart-slider/chart-slider.component';
import { AdaptAreaGraphComponent } from './area-graph/area-graph.component';
import { AdaptAreaGraphSliderComponent } from './area-graph/area-graph-slider.component';
import { NgxGraphModule } from '@swimlane/ngx-graph';
import { AdaptFlowChartComponent } from './flow-chart/flow-chart.component';
import { AdaptTextEllipsisDirective } from './common/utils/text-ellipsis.directive';
import { FormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class AdaptChartsModule {
}
AdaptChartsModule.ɵfac = function AdaptChartsModule_Factory(t) { return new (t || AdaptChartsModule)(); };
AdaptChartsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptChartsModule });
AdaptChartsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [
        ChartColorService
    ], imports: [[
            AdaptTooltipModule,
            NgxGraphModule,
            FormsModule,
            AdaptSliderModule,
            AdaptEmptyStateModule,
            AdaptDropdownModule,
            CommonModule,
            AdaptListKeyManagerModule,
            AdaptIconModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptChartsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptTooltipModule,
                    NgxGraphModule,
                    FormsModule,
                    AdaptSliderModule,
                    AdaptEmptyStateModule,
                    AdaptDropdownModule,
                    CommonModule,
                    AdaptListKeyManagerModule,
                    AdaptIconModule
                ],
                declarations: [
                    AdaptChartComponent,
                    AdaptHeatmapComponent,
                    AdaptLineGraphComponent,
                    AdaptScatterPlotComponent,
                    AdaptTreemapComponent,
                    ChartLegendComponent,
                    ChartGradientLegendComponent,
                    CutNumberPipe,
                    AdaptPieChartComponent,
                    AdaptStackedChartComponent,
                    AdaptChartSliderComponent,
                    AdaptAreaGraphSliderComponent,
                    AdaptAreaGraphComponent,
                    AdaptFlowChartComponent,
                    AdaptTextEllipsisDirective,
                ],
                providers: [
                    ChartColorService
                ],
                exports: [
                    AdaptChartComponent,
                    AdaptHeatmapComponent,
                    AdaptLineGraphComponent,
                    AdaptScatterPlotComponent,
                    AdaptTreemapComponent,
                    ChartLegendComponent,
                    ChartGradientLegendComponent,
                    AdaptPieChartComponent,
                    AdaptStackedChartComponent,
                    AdaptAreaGraphComponent,
                    AdaptFlowChartComponent,
                    AdaptTextEllipsisDirective
                ],
                schemas: [NO_ERRORS_SCHEMA]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptChartsModule, { declarations: [AdaptChartComponent,
        AdaptHeatmapComponent,
        AdaptLineGraphComponent,
        AdaptScatterPlotComponent,
        AdaptTreemapComponent,
        ChartLegendComponent,
        ChartGradientLegendComponent,
        CutNumberPipe,
        AdaptPieChartComponent,
        AdaptStackedChartComponent,
        AdaptChartSliderComponent,
        AdaptAreaGraphSliderComponent,
        AdaptAreaGraphComponent,
        AdaptFlowChartComponent,
        AdaptTextEllipsisDirective], imports: [AdaptTooltipModule,
        NgxGraphModule,
        FormsModule,
        AdaptSliderModule,
        AdaptEmptyStateModule,
        AdaptDropdownModule,
        CommonModule,
        AdaptListKeyManagerModule,
        AdaptIconModule], exports: [AdaptChartComponent,
        AdaptHeatmapComponent,
        AdaptLineGraphComponent,
        AdaptScatterPlotComponent,
        AdaptTreemapComponent,
        ChartLegendComponent,
        ChartGradientLegendComponent,
        AdaptPieChartComponent,
        AdaptStackedChartComponent,
        AdaptAreaGraphComponent,
        AdaptFlowChartComponent,
        AdaptTextEllipsisDirective] }); })();
i0.ɵɵsetComponentScope(AdaptChartSliderComponent, [i1.NgClass, i1.NgIf, i1.NgSwitch, i1.NgSwitchCase, AdaptStackedChartComponent], []);
i0.ɵɵsetComponentScope(AdaptAreaGraphSliderComponent, [AdaptAreaGraphComponent, i1.NgIf, i1.NgClass], []);
