// @bmc-ux/adapt-angular/esm2015/rx/rx-list-item/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ai_angle_right, ai_dots } from '@bmc-ux/dpl-iconfont';
import { AdaptIconModule, AdaptIconConfig } from '../../icon/index';
import { AdaptRxTextfieldModule } from '../rx-textfield/index';
import { AdaptTooltipModule } from '../../tooltip/index';
import { AdaptRxListItemComponent } from './adapt-rx-list-item.component';
import { AdaptCustomActionsModule } from '../../custom-actions/index';
import { AdaptHighlightModule } from './../../highlight/index';
import * as i0 from "@angular/core";
import * as i1 from "../../icon/index";
export class AdaptRxListItemModule {
    constructor(iconService) {
        iconService.registerIcons([ai_angle_right, ai_dots]);
    }
}
AdaptRxListItemModule.ɵfac = function AdaptRxListItemModule_Factory(t) { return new (t || AdaptRxListItemModule)(i0.ɵɵinject(i1.AdaptIconConfig)); };
AdaptRxListItemModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxListItemModule });
AdaptRxListItemModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptRxTextfieldModule,
            AdaptCustomActionsModule,
            AdaptHighlightModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxListItemModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptRxTextfieldModule,
                    AdaptCustomActionsModule,
                    AdaptHighlightModule
                ],
                declarations: [AdaptRxListItemComponent],
                exports: [AdaptRxListItemComponent]
            }]
    }], function () { return [{ type: i1.AdaptIconConfig }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxListItemModule, { declarations: [AdaptRxListItemComponent], imports: [CommonModule,
        FormsModule,
        AdaptIconModule,
        AdaptTooltipModule,
        AdaptRxTextfieldModule,
        AdaptCustomActionsModule,
        AdaptHighlightModule], exports: [AdaptRxListItemComponent] }); })();
export * from './adapt-rx-list-item.component';
export * from './adapt-rx-list-item.types';
