// @bmc-ux/adapt-angular/esm2015/crop/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptCropComponent } from './crop.component';
import * as i0 from "@angular/core";
export class AdaptCropModule {
}
AdaptCropModule.ɵfac = function AdaptCropModule_Factory(t) { return new (t || AdaptCropModule)(); };
AdaptCropModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCropModule });
AdaptCropModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCropModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptCropComponent
                ],
                exports: [
                    AdaptCropComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCropModule, { declarations: [AdaptCropComponent], imports: [CommonModule], exports: [AdaptCropComponent] }); })();
export * from './crop.component';
export * from './crop.model';
