// @bmc-ux/obsolete/esm2015/lib/textfield-formbuilder/index.js
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { AdaptTextFieldFormBuilderComponent } from './textfield-formbuilder.component';
import { AdaptTextFieldIconsModule } from '../textfield/textfield.icons';
import { AdaptDragonInputHandlerModule, AdaptIconModule, AdaptPopoverModule, AdaptRxLabelModule, AdaptTooltipModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptTextFieldFormBuilderModule {
}
AdaptTextFieldFormBuilderModule.ɵfac = function AdaptTextFieldFormBuilderModule_Factory(t) { return new (t || AdaptTextFieldFormBuilderModule)(); };
AdaptTextFieldFormBuilderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTextFieldFormBuilderModule });
AdaptTextFieldFormBuilderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule,
            ReactiveFormsModule,
            AdaptTextFieldIconsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTextFieldFormBuilderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    FormsModule,
                    CommonModule,
                    ReactiveFormsModule,
                    AdaptTextFieldIconsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule
                ],
                exports: [AdaptTextFieldFormBuilderComponent],
                declarations: [AdaptTextFieldFormBuilderComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTextFieldFormBuilderModule, { declarations: [AdaptTextFieldFormBuilderComponent], imports: [FormsModule,
        CommonModule,
        ReactiveFormsModule,
        AdaptTextFieldIconsModule,
        AdaptIconModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptDragonInputHandlerModule,
        AdaptRxLabelModule], exports: [AdaptTextFieldFormBuilderComponent] }); })();
export * from './textfield-formbuilder.component';
