// @bmc-ux/adapt-angular/esm2015/rx/rx-datetime/adapt-rx-datetime-range.component.js
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Inject, Input, LOCALE_ID, Optional, Output, Renderer2, Self, ViewChild } from '@angular/core';
import { NgControl } from '@angular/forms';
import { trigger } from '@angular/animations';
import { ENTER, SPACE, TAB } from '@angular/cdk/keycodes';
import { Directionality } from '@angular/cdk/bidi';
import { asapScheduler, timer } from 'rxjs';
import { take, takeUntil } from 'rxjs/operators';
import { isArray, isFunction, isNil, pick } from 'lodash-es';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularCustomEventType, AdaptRadarAngularService } from '../../common/radar/adapt-radar';
import { AdaptInteractionSourceService, getInteractionSource } from '../../common/interaction-source';
import { getDateParts } from './common/partition.utils';
import { AdaptRxFeedbackComponent, AdaptRxFormControlDirective } from '../form-control/index';
import { AdaptDeviceDetectionService } from '../../common/device-detection.service';
import { AdaptTranslateService } from '../../common/i18n/index';
import { AdaptDropdownCloseReason, AdaptDropdownDirective } from '../../dropdown/index';
import { SafeCdr } from '../../common/safe-cdr';
import { getEventKeyCode } from '../../common/key-codes';
import { getFocusableElements } from '../../common/focus';
import { SLIDE_ANIMATION, checkIfPropsDefined, getDatetimeModelPropsByMode, checkIsAfter, getStructWithDefaults } from './common/common.utils';
import { RxDatetimePickerMode } from './adapt-rx-datetime.types';
import { AdaptRxDatetimeConfigService } from './adapt-rx-datetime-config.service';
import { AdaptRxDatetimeAdapter } from './adapt-rx-datetime-adapter.service';
import { AdaptRxDatetimeI18nService } from './adapt-rx-datetime-i18n.service';
import { AdaptRxDatetimeParserFormatter } from './adapt-rx-datetime-parser-formatter.service';
import { AdaptRxDatetimePickerComponent } from './datetime-picker/adapt-rx-datetime-picker.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
import * as i2 from "./adapt-rx-datetime-config.service";
import * as i3 from "../../common/device-detection.service";
import * as i4 from "./adapt-rx-datetime-i18n.service";
import * as i5 from "./adapt-rx-datetime-adapter.service";
import * as i6 from "./adapt-rx-datetime-parser-formatter.service";
import * as i7 from "../../common/i18n/index";
import * as i8 from "../../common/radar/adapt-radar";
import * as i9 from "../../common/interaction-source";
import * as i10 from "@angular/cdk/bidi";
import * as i11 from "../form-control/rx-control-label/adapt-rx-control-label.component";
import * as i12 from "@angular/common";
import * as i13 from "../form-control/rx-feedback/adapt-rx-feedback.component";
import * as i14 from "./datetime-picker/adapt-rx-datetime-picker.component";
import * as i15 from "../../dropdown/dropdown.component";
const _c0 = ["inputControlRef"];
const _c1 = ["openPopupRef"];
const _c2 = ["adaptRxFeedbackRef"];
const _c3 = ["startRangeTab"];
const _c4 = ["endRangeTab"];
const _c5 = ["adaptDropdownDirective"];
const _c6 = ["startDatetimePickerComponentRef"];
const _c7 = ["endDatetimePickerComponentRef"];
const _c8 = ["datetimePickerContainer"];
function AdaptRxDatetimeRangeComponent_ng_template_5_span_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r8.texts.emptyStartRange);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_span_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r9.startValueSummary.dateTime);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_9_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 19);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r17 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r17.startValueSummary.date);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_9_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 20);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r18 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r18.startValueSummary.time);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_9_span_1_Template, 2, 1, "span", 17);
    i0.ɵɵtemplate(2, AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_9_div_2_Template, 2, 1, "div", 18);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r10.startValueSummary == null ? null : ctx_r10.startValueSummary.date);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r10.startValueSummary == null ? null : ctx_r10.startValueSummary.time);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_span_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r12.texts.emptyEndRange);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_span_15_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r13.endValueSummary.dateTime);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_16_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 19);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r19 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r19.endValueSummary.date);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_16_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 20);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r20 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r20.endValueSummary.time);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_16_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_16_span_1_Template, 2, 1, "span", 17);
    i0.ɵɵtemplate(2, AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_16_div_2_Template, 2, 1, "div", 18);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r14 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r14.endValueSummary == null ? null : ctx_r14.endValueSummary.date);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r14.endValueSummary == null ? null : ctx_r14.endValueSummary.time);
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_div_18_Template(rf, ctx) { if (rf & 1) {
    const _r23 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 21);
    i0.ɵɵelementStart(1, "adapt-rx-datetime-picker", 22, 23);
    i0.ɵɵlistener("pickerValueChange", function AdaptRxDatetimeRangeComponent_ng_template_5_div_18_Template_adapt_rx_datetime_picker_pickerValueChange_1_listener($event) { i0.ɵɵrestoreView(_r23); const ctx_r22 = i0.ɵɵnextContext(2); return ctx_r22.onStartPickerValueChange($event); })("pickerStateChange", function AdaptRxDatetimeRangeComponent_ng_template_5_div_18_Template_adapt_rx_datetime_picker_pickerStateChange_1_listener() { i0.ɵɵrestoreView(_r23); const ctx_r24 = i0.ɵɵnextContext(2); return ctx_r24.onPickerStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("id", ctx_r15.id + "_startRangePanel");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("mode", ctx_r15.mode)("expandedControlsStyle", false)("hasSeconds", ctx_r15.hasSeconds)("use12HoursTime", ctx_r15.use12HoursTime)("firstDayOfWeek", ctx_r15.firstDayOfWeek)("initialDatetime", ctx_r15.initialDatetime || ctx_r15.defaultInitialDatetime)("activeDatetime", ctx_r15.startPickerActiveDatetime)("dayFilter", ctx_r15.rangeFromDayFilter)("rangeFrom", ctx_r15.rangeFrom)("rangeTo", ctx_r15.rangeTo)("disabled", ctx_r15.disabled)("readonly", ctx_r15.readonly)("texts", ctx_r15.texts)("testID", ctx_r15.testID);
    i0.ɵɵattribute("data-testid", ctx_r15.testID + "_picker");
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_div_19_Template(rf, ctx) { if (rf & 1) {
    const _r27 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 21);
    i0.ɵɵelementStart(1, "adapt-rx-datetime-picker", 22, 24);
    i0.ɵɵlistener("pickerValueChange", function AdaptRxDatetimeRangeComponent_ng_template_5_div_19_Template_adapt_rx_datetime_picker_pickerValueChange_1_listener($event) { i0.ɵɵrestoreView(_r27); const ctx_r26 = i0.ɵɵnextContext(2); return ctx_r26.onEndPickerValueChange($event); })("pickerStateChange", function AdaptRxDatetimeRangeComponent_ng_template_5_div_19_Template_adapt_rx_datetime_picker_pickerStateChange_1_listener() { i0.ɵɵrestoreView(_r27); const ctx_r28 = i0.ɵɵnextContext(2); return ctx_r28.onPickerStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("id", ctx_r16.id + "_endRangePanel");
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("mode", ctx_r16.mode)("expandedControlsStyle", false)("hasSeconds", ctx_r16.hasSeconds)("use12HoursTime", ctx_r16.use12HoursTime)("firstDayOfWeek", ctx_r16.firstDayOfWeek)("initialDatetime", ctx_r16.initialDatetime || ctx_r16.defaultInitialDatetime)("activeDatetime", ctx_r16.endPickerActiveDatetime)("dayFilter", ctx_r16.rangeToDayFilter)("rangeFrom", ctx_r16.rangeFrom)("rangeTo", ctx_r16.rangeTo)("disabled", ctx_r16.disabled)("readonly", ctx_r16.readonly)("texts", ctx_r16.texts)("testID", ctx_r16.testID);
    i0.ɵɵattribute("data-testid", ctx_r16.testID + "_picker");
} }
function AdaptRxDatetimeRangeComponent_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    const _r30 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 7);
    i0.ɵɵelement(1, "adapt-rx-feedback", 8);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(2, "div", 9);
    i0.ɵɵelementStart(3, "button", 10, 11);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_5_Template_button_click_3_listener() { i0.ɵɵrestoreView(_r30); const ctx_r29 = i0.ɵɵnextContext(); return ctx_r29.switchToStartDT(); })("keydown.arrowRight", function AdaptRxDatetimeRangeComponent_ng_template_5_Template_button_keydown_arrowRight_3_listener($event) { i0.ɵɵrestoreView(_r30); const _r11 = i0.ɵɵreference(11); $event.preventDefault(); return _r11.focus(); });
    i0.ɵɵtext(5);
    i0.ɵɵelementStart(6, "div", 12);
    i0.ɵɵtemplate(7, AdaptRxDatetimeRangeComponent_ng_template_5_span_7_Template, 2, 1, "span", 13);
    i0.ɵɵtemplate(8, AdaptRxDatetimeRangeComponent_ng_template_5_span_8_Template, 2, 1, "span", 13);
    i0.ɵɵtemplate(9, AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_9_Template, 3, 2, "ng-container", 13);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(10, "button", 14, 15);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_5_Template_button_click_10_listener() { i0.ɵɵrestoreView(_r30); const ctx_r32 = i0.ɵɵnextContext(); return ctx_r32.switchToEndDT(); })("keydown.arrowLeft", function AdaptRxDatetimeRangeComponent_ng_template_5_Template_button_keydown_arrowLeft_10_listener($event) { i0.ɵɵrestoreView(_r30); const _r7 = i0.ɵɵreference(4); $event.preventDefault(); return _r7.focus(); });
    i0.ɵɵtext(12);
    i0.ɵɵelementStart(13, "div", 12);
    i0.ɵɵtemplate(14, AdaptRxDatetimeRangeComponent_ng_template_5_span_14_Template, 2, 1, "span", 13);
    i0.ɵɵtemplate(15, AdaptRxDatetimeRangeComponent_ng_template_5_span_15_Template, 2, 1, "span", 13);
    i0.ɵɵtemplate(16, AdaptRxDatetimeRangeComponent_ng_template_5_ng_container_16_Template, 3, 2, "ng-container", 13);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(17, "div");
    i0.ɵɵtemplate(18, AdaptRxDatetimeRangeComponent_ng_template_5_div_18_Template, 3, 16, "div", 16);
    i0.ɵɵtemplate(19, AdaptRxDatetimeRangeComponent_ng_template_5_div_19_Template, 3, 16, "div", 16);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("errors", ctx_r2.ngControl.errors)("warningMessage", ctx_r2.warningMessage)("controlTouched", ctx_r2.ngControl.touched)("alertFeedbackStyle", true)("alertFeedbackTruncation", !ctx_r2.isMobile);
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("end-active", !ctx_r2.startRangeActive);
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("font-weight-semibold", ctx_r2.startRangeActive)("has-seconds", ctx_r2.isSecondsWrap);
    i0.ɵɵattribute("aria-controls", ctx_r2.id + "_startRangePanel")("aria-selected", ctx_r2.startRangeActive)("data-testid", ctx_r2.testID + "_start");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r2.texts.startRangeLabel, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r2.startValueSummary);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r2.isSecondsWrap && ctx_r2.startValueSummary);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.isSecondsWrap && ctx_r2.startValueSummary);
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("font-weight-semibold", !ctx_r2.startRangeActive)("has-seconds", ctx_r2.isSecondsWrap);
    i0.ɵɵattribute("aria-controls", ctx_r2.id + "_endRangePanel")("aria-selected", !ctx_r2.startRangeActive)("data-testid", ctx_r2.testID + "_end");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r2.texts.endRangeLabel, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r2.endValueSummary);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r2.isSecondsWrap && ctx_r2.endValueSummary);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.isSecondsWrap && ctx_r2.endValueSummary);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("@slideAnimation", ctx_r2.startRangeActiveAnimation);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.startRangeActive);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r2.startRangeActive);
} }
function AdaptRxDatetimeRangeComponent_ng_template_7_ng_template_1_Template(rf, ctx) { }
function AdaptRxDatetimeRangeComponent_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 25);
    i0.ɵɵtemplate(1, AdaptRxDatetimeRangeComponent_ng_template_7_ng_template_1_Template, 0, 0, "ng-template", 26);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    const _r1 = i0.ɵɵreference(6);
    i0.ɵɵclassProp("a3t-datetime-inline-light", ctx_r4.inlineLight)("a3t-datetime-inline-small", ctx_r4.inlineCompact);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r1);
} }
function AdaptRxDatetimeRangeComponent_ng_template_9_button_5_Template(rf, ctx) { if (rf & 1) {
    const _r42 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 34, 35);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_9_button_5_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r42); const ctx_r41 = i0.ɵɵnextContext(2); return ctx_r41.datetimeOpenHandler($event); })("keydown", function AdaptRxDatetimeRangeComponent_ng_template_9_button_5_Template_button_keydown_0_listener($event) { i0.ɵɵrestoreView(_r42); const ctx_r43 = i0.ɵɵnextContext(2); return ctx_r43.datetimeOpenHandler($event); });
    i0.ɵɵelement(2, "i", 36);
    i0.ɵɵelementStart(3, "span", 37);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const hasErrors_r35 = i0.ɵɵnextContext().$implicit;
    const ctx_r38 = i0.ɵɵnextContext();
    i0.ɵɵclassProp("a3t-datetime--calendar-control_invalid", hasErrors_r35)("disabled-look", ctx_r38.readonly && ctx_r38.disabledStyleForReadonlyState);
    i0.ɵɵproperty("disabled", ctx_r38.disabled);
    i0.ɵɵattribute("data-testid", ctx_r38.testID + "_calendar-control")("aria-invalid", ctx_r38.ngControl.invalid)("aria-readonly", ctx_r38.readonly)("aria-disabled", ctx_r38.disabled);
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("d-icon-calendar", ctx_r38.mode === ctx_r38.PICKER_MODE.Date || ctx_r38.mode === ctx_r38.PICKER_MODE.Year || ctx_r38.mode === ctx_r38.PICKER_MODE.YearMonth || ctx_r38.mode === ctx_r38.PICKER_MODE.Month)("d-icon-calendar_clock_o", ctx_r38.mode === ctx_r38.PICKER_MODE.DateTime)("d-icon-clock_o", ctx_r38.mode === ctx_r38.PICKER_MODE.Time);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r38.mode === ctx_r38.PICKER_MODE.Time ? ctx_r38.texts.clockBtn : ctx_r38.texts.calendarBtn);
} }
function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_ng_template_2_Template(rf, ctx) { }
function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_span_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r50 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r50.texts.applyBtn);
} }
function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_span_15_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r51 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r51.texts.nextBtn);
} }
const _c9 = function (a0, a1) { return [a0, a1]; };
function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    const _r53 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 38, 39);
    i0.ɵɵlistener("keydown", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_div_keydown_0_listener($event) { i0.ɵɵrestoreView(_r53); const ctx_r52 = i0.ɵɵnextContext(2); return ctx_r52.datetimePickerContainerKeydown($event); });
    i0.ɵɵtemplate(2, AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_ng_template_2_Template, 0, 0, "ng-template", 26);
    i0.ɵɵelementStart(3, "div", 40);
    i0.ɵɵelementStart(4, "div");
    i0.ɵɵelementStart(5, "button", 41, 42);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_click_5_listener() { i0.ɵɵrestoreView(_r53); const ctx_r54 = i0.ɵɵnextContext(2); return ctx_r54.pickerCloseAndCancel(); })("keydown.arrowRight", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_keydown_arrowRight_5_listener($event) { i0.ɵɵrestoreView(_r53); const _r48 = i0.ɵɵreference(10); $event.preventDefault(); return _r48.focus(); });
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(8, "div", 43);
    i0.ɵɵelementStart(9, "button", 44, 45);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_click_9_listener() { i0.ɵɵrestoreView(_r53); const ctx_r56 = i0.ɵɵnextContext(2); return ctx_r56.pickerClearAll(); })("keydown.arrowLeft", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_keydown_arrowLeft_9_listener($event) { i0.ɵɵrestoreView(_r53); const _r47 = i0.ɵɵreference(6); $event.preventDefault(); return _r47.focus(); })("keydown.arrowRight", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_keydown_arrowRight_9_listener($event) { i0.ɵɵrestoreView(_r53); const _r49 = i0.ɵɵreference(13); $event.preventDefault(); return _r49.focus(); });
    i0.ɵɵtext(11);
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(12, "button", 46, 47);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_click_12_listener() { i0.ɵɵrestoreView(_r53); const ctx_r59 = i0.ɵɵnextContext(2); return ctx_r59.goToNextStep(); })("keydown.arrowLeft", function AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template_button_keydown_arrowLeft_12_listener($event) { i0.ɵɵrestoreView(_r53); const _r48 = i0.ɵɵreference(10); $event.preventDefault(); return _r48.focus(); });
    i0.ɵɵtemplate(14, AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_span_14_Template, 2, 1, "span", 13);
    i0.ɵɵtemplate(15, AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_span_15_Template, 2, 1, "span", 13);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r39 = i0.ɵɵnextContext(2);
    const _r1 = i0.ɵɵreference(6);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c9, ctx_r39.isMobile ? "a3t-datetime--calendar-width" : "a3t-datetime--calendar-min-width", ctx_r39.popupClass || ""));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", _r1);
    i0.ɵɵadvance(3);
    i0.ɵɵattribute("data-testid", ctx_r39.testID + "_cancel");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r39.texts.cancelBtn, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("data-testid", ctx_r39.testID + "_clear");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1(" ", ctx_r39.texts.clearBtn, " ");
    i0.ɵɵadvance(1);
    i0.ɵɵattribute("data-testid", ctx_r39.testID + "_apply");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r39.isLastWizardStep);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r39.isLastWizardStep);
} }
function AdaptRxDatetimeRangeComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    const _r62 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 27, 28);
    i0.ɵɵlistener("onOpen", function AdaptRxDatetimeRangeComponent_ng_template_9_Template_div_onOpen_0_listener() { i0.ɵɵrestoreView(_r62); const ctx_r61 = i0.ɵɵnextContext(); return ctx_r61.popupOpenChangeHandler(true); })("onClose", function AdaptRxDatetimeRangeComponent_ng_template_9_Template_div_onClose_0_listener($event) { i0.ɵɵrestoreView(_r62); const ctx_r63 = i0.ɵɵnextContext(); return ctx_r63.popupOpenChangeHandler(false, $event); });
    i0.ɵɵelementStart(2, "div", 29);
    i0.ɵɵelementStart(3, "input", 30, 31);
    i0.ɵɵlistener("click", function AdaptRxDatetimeRangeComponent_ng_template_9_Template_input_click_3_listener($event) { i0.ɵɵrestoreView(_r62); const ctx_r64 = i0.ɵɵnextContext(); return ctx_r64.datetimeOpenHandler($event); })("input", function AdaptRxDatetimeRangeComponent_ng_template_9_Template_input_input_3_listener() { i0.ɵɵrestoreView(_r62); const _r37 = i0.ɵɵreference(4); const ctx_r65 = i0.ɵɵnextContext(); return ctx_r65.onInput(_r37.value); })("focus", function AdaptRxDatetimeRangeComponent_ng_template_9_Template_input_focus_3_listener($event) { i0.ɵɵrestoreView(_r62); const ctx_r66 = i0.ɵɵnextContext(); return ctx_r66.focusHandler($event); })("blur", function AdaptRxDatetimeRangeComponent_ng_template_9_Template_input_blur_3_listener($event) { i0.ɵɵrestoreView(_r62); const ctx_r67 = i0.ɵɵnextContext(); return ctx_r67.blurHandler($event); });
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(5, AdaptRxDatetimeRangeComponent_ng_template_9_button_5_Template, 5, 16, "button", 32);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(6, AdaptRxDatetimeRangeComponent_ng_template_9_ng_template_6_Template, 16, 12, "ng-template", 33);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const hasErrors_r35 = ctx.$implicit;
    const _r36 = i0.ɵɵreference(1);
    const ctx_r6 = i0.ɵɵnextContext();
    const _r0 = i0.ɵɵreference(4);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("placement", ctx_r6.placement)("holdFocusInMenu", false)("autoFocusFirst", false)("restoreFocusAfterClose", true)("appendToBody", ctx_r6.appendToBody)("autoClose", "outside")("mobileView", ctx_r6.mobileViewConfig);
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("a3t-datetime--controls-wrapper-active", _r36.isOpen());
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("a3t-datetime--input-control_invalid", hasErrors_r35)("a3t-datetime--input-control_mobile", ctx_r6.isMobile && !ctx_r6.readonly)("a3t-datetime--input-control_readonly", ctx_r6.readonly && !ctx_r6.disabledStyleForReadonlyState)("form-control-disabled-look", ctx_r6.readonly && ctx_r6.disabledStyleForReadonlyState)("form-control-danger", _r0.hasUIErrorState);
    i0.ɵɵproperty("id", ctx_r6.id)("value", ctx_r6.inputValue)("required", ctx_r6.required)("disabled", ctx_r6.disabled)("readOnly", ctx_r6.readonly || ctx_r6.isMobile && !ctx_r6.disabled)("autofocus", ctx_r6.autofocus);
    i0.ɵɵattribute("autocomplete", "off")("tabindex", ctx_r6.tabIndex)("placeholder", ctx_r6.placeholder)("name", ctx_r6.name)("aria-label", ctx_r6.ariaLabel)("aria-labelledby", ctx_r6.ariaLabelledby)("aria-describedby", ctx_r6.ariaDescribedBy)("aria-required", ctx_r6.required)("aria-invalid", ctx_r6.ngControl.invalid)("aria-readonly", ctx_r6.readonly)("aria-disabled", ctx_r6.disabled)("data-testid", ctx_r6.testID);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !(ctx_r6.readonly && !ctx_r6.disabledStyleForReadonlyState));
} }
const _c10 = function (a0) { return { $implicit: a0 }; };
export class AdaptRxDatetimeRangeComponent extends AdaptRxFormControlDirective {
    constructor(ngControl, _datetimeConfigService, _deviceDetectionService, _datetimeI18nService, 
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    _datetimeAdapter, _datetimeParserFormatter, _changeDetectorRef, _renderer, ts, localeID, _adaptRadarService, _interactionSourceOriginService, _dir, _elem) {
        super(ngControl, ts, _adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptRxDatetimeRange, _changeDetectorRef);
        this._datetimeI18nService = _datetimeI18nService;
        this._datetimeAdapter = _datetimeAdapter;
        this._datetimeParserFormatter = _datetimeParserFormatter;
        this._renderer = _renderer;
        this.ts = ts;
        this.localeID = localeID;
        this._interactionSourceOriginService = _interactionSourceOriginService;
        this._dir = _dir;
        this.controlName = 'datetime-range';
        /**
         * Inline mode
         * @docs-default false
         */
        this.inline = false;
        /**
         * The preferred placement of the typeahead
         * @docs-default auto
         */
        this.placement = 'auto';
        /**
         * Attach popup to the body
         * @docs-default false
         */
        this.appendToBody = false;
        /**
         * Light theme for inline mode
         * @docs-default false
         */
        this.inlineLight = true;
        /**
         * Small size for inline mode
         * @docs-default false
         */
        this.inlineCompact = false;
        /**
         * Trigger for disable picker's wizard
         * @docs-default false
         */
        this.disableWizard = false;
        /**
         * Picker's mode
         * @docs-default RxDatetimePickerMode.Date
         */
        this.mode = RxDatetimePickerMode.Date;
        /**
         * Enables seconds for time
         * @docs-default false
         */
        this.hasSeconds = false;
        /**
         * Use disable styles for readonly state
         * @docs-default false
         * @since v10.5.0
         */
        this.disabledStyleForReadonlyState = false;
        /* eslint-disable @angular-eslint/no-output-on-prefix */
        /**
         *  An event fired when the popup is opened or closed
         */
        this.onPopupOpenChange = new EventEmitter();
        /**
         * An event fired when model change
         */
        this.onDatetimeChange = new EventEmitter();
        this.PICKER_MODE = RxDatetimePickerMode;
        this.inputValue = '';
        this.mobileViewConfig = {
            isPicker: true
        };
        this.isMobile = false;
        this.inputRangeSeparator = ' - ';
        this.isLastWizardStep = false;
        this.startRangeActive = true;
        this.startRangeActiveAnimation = 1;
        this.rangeFrom = null;
        this.rangeTo = null;
        this.rangeFromDayFilter = this._rangeFromDayFilter.bind(this);
        this.rangeToDayFilter = this._rangeToDayFilter.bind(this);
        this.startValueSummary = { date: '', time: '', dateTime: '' };
        this.endValueSummary = { date: '', time: '', dateTime: '' };
        this.startPickerActiveDatetime = null;
        this.endPickerActiveDatetime = null;
        this.isRtl = false;
        // TODO: Refactor this if possible
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        this._controlStartValue = null;
        // TODO: Refactor this if possible
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        this._controlEndValue = null;
        // TODO: Refactor this if possible
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        this._backupControlStartValue = null;
        // TODO: Refactor this if possible
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        this._backupControlEndValue = null;
        this._openWithKeyboard = false;
        this.isMobile = _deviceDetectionService.isMobile();
        this._texts = this._defaultTexts = _datetimeConfigService.texts;
        this.isRtl = this._dir.value === 'rtl';
    }
    /**
     * Enables/disables 12-hours time format.
     * By default time format calculates automatically
     * @docs-default
     */
    set use12HoursTime(value) {
        this._use12HoursTime = value;
    }
    get use12HoursTime() {
        return !isNil(this._use12HoursTime) ? this._use12HoursTime : this._datetimeI18nService.getTimeFormat(this.localeID) === 12;
    }
    /**
     * Calendar first day of week.
     * By default calculates automatically
     * @docs-default
     */
    set firstDayOfWeek(value) {
        this._firstDayOfWeek = value;
    }
    get firstDayOfWeek() {
        return !isNil(this._firstDayOfWeek) ? this._firstDayOfWeek : this._datetimeI18nService.getFirstDayOfWeek(this.localeID);
    }
    /**
     * Datetime texts
     * @docs-default
     */
    get texts() {
        return this._texts;
    }
    set texts(value) {
        this._texts = Object.assign(Object.assign({}, this._defaultTexts), value);
    }
    get controlStartValue() {
        return this._datetimeAdapter.fromModel(this._controlStartValue);
    }
    set controlStartValue(value) {
        const nextValue = isFunction(this.defaultDatetime) ?
            getStructWithDefaults(value, this.defaultDatetime(this.mode, this.hasSeconds)) : value;
        this._controlStartValue = this._datetimeAdapter.toModel(nextValue);
        this.startPickerActiveDatetime = this.controlStartValue;
    }
    get controlEndValue() {
        return this._datetimeAdapter.fromModel(this._controlEndValue);
    }
    set controlEndValue(value) {
        const nextValue = isFunction(this.defaultDatetime) ?
            getStructWithDefaults(value, this.defaultDatetime(this.mode, this.hasSeconds)) : value;
        this._controlEndValue = this._datetimeAdapter.toModel(nextValue);
        this.endPickerActiveDatetime = this.controlEndValue;
    }
    get isSecondsWrap() {
        return this.mode === RxDatetimePickerMode.DateTime && !this.inline && this.hasSeconds;
    }
    ngOnInit() {
        super.ngOnInit();
        if (this.inline) {
            this._updateRange();
            this._checkWizardLastStep();
        }
        this._updateInputValue();
    }
    switchToStartDT() {
        this.startRangeActive = true;
        this.startRangeActiveAnimation--;
        this._checkWizardLastStep();
    }
    switchToEndDT() {
        this.startRangeActive = false;
        this.startRangeActiveAnimation++;
        this._checkWizardLastStep();
    }
    datetimeOpenHandler(event) {
        if (!this.nonInteractive && !this.popup.isOpen()) {
            const interactionSource = getInteractionSource(event);
            const keyCode = event ? getEventKeyCode(event) : null;
            this._targetOpen = event ? event.target : null;
            if (this.isMobile && this._targetOpen === this.inputControlRef.nativeElement) {
                this._onTouched();
            }
            if (interactionSource === 'keyboard' && keyCode !== ENTER && keyCode !== SPACE) {
                return;
            }
            this._openWithKeyboard = interactionSource === 'keyboard';
            this.popup.open(event);
        }
    }
    onInput(value) {
        var _a, _b;
        this.inputValue = value;
        const propsToPick = getDatetimeModelPropsByMode(this.mode, this.hasSeconds);
        const splitString = value.split(this.inputRangeSeparator.trim());
        // Start value
        const parsedStartDatetime = this._datetimeParserFormatter.parse((_a = splitString[0]) === null || _a === void 0 ? void 0 : _a.trim(), this.mode, this.hasSeconds, this.use12HoursTime, this.localeID, this.inputFormat);
        if (!parsedStartDatetime) {
            this.controlStartValue = null;
        }
        else if (parsedStartDatetime.parseError) {
            this.controlStartValue = parsedStartDatetime;
        }
        else {
            this.controlStartValue = Object.assign({
                year: null,
                month: null,
                date: null,
                hours: null,
                minutes: null,
                seconds: null
            }, pick(parsedStartDatetime, propsToPick));
        }
        // End value
        const parsedEndDatetime = this._datetimeParserFormatter.parse((_b = splitString[1]) === null || _b === void 0 ? void 0 : _b.trim(), this.mode, this.hasSeconds, this.use12HoursTime, this.localeID, this.inputFormat);
        if (!parsedEndDatetime) {
            this.controlEndValue = null;
        }
        else if (parsedEndDatetime.parseError) {
            this.controlEndValue = parsedEndDatetime;
        }
        else {
            this.controlEndValue = Object.assign({
                year: null,
                month: null,
                date: null,
                hours: null,
                minutes: null,
                seconds: null
            }, pick(parsedEndDatetime, propsToPick));
        }
        this._updateRange();
        if (this.popup.isOpen()) {
            this._checkWizardLastStep();
        }
        this._propagateChange();
    }
    blurHandler(event) {
        super.blurHandler(event);
        if (!this.popup.isOpen()) {
            this._updateInputValue();
            this._onTouched();
        }
    }
    /**
     * Popup toggle handler
     */
    popupOpenChangeHandler(isOpen, closeReason) {
        this.onPopupOpenChange.emit(isOpen);
        if (isOpen) {
            this.defaultInitialDatetime = getDateParts(new Date());
            this.popup.popupAnimationDone
                .pipe(takeUntil(this.destroy$), take(1))
                .subscribe(() => {
                this._popupFocusableElements = getFocusableElements(this.datetimePickerContainer.nativeElement);
                if (this._openWithKeyboard) {
                    this._popupFocusableElements[0].focus();
                }
            });
            this._backupControlStartValue = this._controlStartValue;
            this._backupControlEndValue = this._controlEndValue;
            this.switchToStartDT();
            this._updateRange();
            this._checkWizardLastStep();
        }
        else {
            if (document.activeElement !== this.inputControlRef.nativeElement) {
                this._updateInputValue();
            }
            this._onTouched();
            if (closeReason === AdaptDropdownCloseReason.Esc && this._targetOpen) {
                this._targetOpen.focus();
            }
        }
    }
    datetimePickerContainerKeydown(event) {
        if (getEventKeyCode(event) === TAB) {
            if (event.shiftKey) {
                if (event.target === this._popupFocusableElements[0]) {
                    event.preventDefault();
                    this.popup.close();
                }
            }
            else {
                if (event.target === this._popupFocusableElements[this._popupFocusableElements.length - 1]) {
                    event.preventDefault();
                    this.popup.close();
                }
            }
        }
    }
    pickerCloseAndCancel() {
        this._controlStartValue = this._backupControlStartValue;
        this._controlEndValue = this._backupControlEndValue;
        this.startPickerActiveDatetime = this.controlStartValue;
        this.endPickerActiveDatetime = this.controlEndValue;
        this._propagateChange();
        this._updateInputValue();
        this.popup.close();
        this.inputControlRef.nativeElement.focus();
    }
    pickerClearAll() {
        this._controlStartValue = null;
        this._controlEndValue = null;
        this.startPickerActiveDatetime = null;
        this.endPickerActiveDatetime = null;
        this._updateRange();
        this._propagateChange();
        this._updateInputValue();
        this._checkWizardLastStep();
    }
    goToNextStep() {
        if (this.isLastWizardStep) {
            this.popup.close();
            this.inputControlRef.nativeElement.focus();
        }
        else {
            if (this.startRangeActive) {
                if (this.startDatetimePickerComponent.publicApi.isLastStep()) {
                    this.switchToEndDT();
                    this.endRangeTab.nativeElement.focus();
                }
                else {
                    this.startDatetimePickerComponent.publicApi.nextStep();
                    this._checkWizardLastStep();
                }
            }
            else {
                if (this.endDatetimePickerComponent.publicApi.isLastStep()) {
                    this.switchToStartDT();
                    this.startRangeTab.nativeElement.focus();
                }
                else {
                    this.endDatetimePickerComponent.publicApi.nextStep();
                    this._checkWizardLastStep();
                }
            }
        }
    }
    onStartPickerValueChange(newValue) {
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptRxDatetime, {
            GroupName: this.adaptRadarGroupName,
            EventType: AdaptRadarAngularCustomEventType.ValueChanged,
            Label: this.label,
            Required: this.required,
            ReadOnly: this.readonly,
            Disabled: this.disabled,
            Type: this.mode,
            Range: true,
            Inline: this.inline,
            // InputMethod: getInteractionSourceOrigin(event, this.isMobile)
            /*DynamicEntry: boolean;*/
        });
        this.controlStartValue = newValue;
        this._updateRange();
        this.onDatetimeChange.emit([this._controlStartValue, this._controlEndValue]);
        this._propagateChange();
        this._updateInputValue();
        if (this.inline) {
            this._onTouched();
        }
        else {
            this._checkWizardLastStep();
        }
    }
    onEndPickerValueChange(newValue) {
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptRxDatetime, {
            GroupName: this.adaptRadarGroupName,
            EventType: AdaptRadarAngularCustomEventType.ValueChanged,
            Label: this.label,
            Required: this.required,
            ReadOnly: this.readonly,
            Disabled: this.disabled,
            Type: this.mode,
            Range: true,
            Inline: this.inline,
            /*InputMethod: getInteractionSourceOrigin(event, this.isMobile)*/
            /*DynamicEntry: boolean;*/
        });
        this.controlEndValue = newValue;
        this._updateRange();
        this.onDatetimeChange.emit([this._controlStartValue, this._controlEndValue]);
        this._propagateChange();
        this._updateInputValue();
        if (this.inline) {
            this._onTouched();
        }
        else {
            this._checkWizardLastStep();
        }
    }
    onPickerStateChange() {
        if (!this.inline) {
            // To be sure the picker view already have painted
            timer(0, asapScheduler)
                .pipe(takeUntil(this.destroy$))
                .subscribe(() => {
                this._checkWizardLastStep();
            });
        }
    }
    focusHandler(event) {
        super.focusHandler(event);
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptRxDatetime, {
            GroupName: this.adaptRadarGroupName,
            EventType: AdaptRadarAngularCustomEventType.Focus,
            Label: this.label,
            Required: this.required,
            ReadOnly: this.readonly,
            Disabled: this.disabled,
            Type: this.mode,
            Range: true,
            Inline: this.inline,
            InputMethod: this._interactionSourceOriginService.getLastInteractionSourceOrigin()
            /*DynamicEntry: boolean;*/
        });
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    writeValue(value) {
        var _a;
        if (isArray(value)) {
            this.controlStartValue = this._datetimeAdapter.fromModel(value[0]);
            this.controlEndValue = this._datetimeAdapter.fromModel(value[1]);
        }
        else {
            this.controlStartValue = null;
            this.controlEndValue = null;
        }
        this._updateRange();
        this._updateInputValue();
        if ((_a = this.popup) === null || _a === void 0 ? void 0 : _a.isOpen()) {
            this._checkWizardLastStep();
        }
        SafeCdr.detectChanges(this._changeDetectorRef);
    }
    _focusNativeElement() {
        if (!this.inline) {
            this.inputControlRef.nativeElement.focus();
        }
    }
    _isModelFull(value) {
        return checkIfPropsDefined(value, getDatetimeModelPropsByMode(this.mode, this.hasSeconds));
    }
    _updateInputValue() {
        this.startValueSummary = this._getDateTimeSummary(this.controlStartValue);
        this.endValueSummary = this._getDateTimeSummary(this.controlEndValue);
        const rangeFrom = this.startValueSummary ? this.startValueSummary.dateTime : '';
        const rangeTo = this.endValueSummary ? this.endValueSummary.dateTime : '';
        this.inputValue = rangeFrom + (rangeFrom || rangeTo ? this.inputRangeSeparator : '') + rangeTo;
    }
    _getDateTimeSummary(control) {
        if (!this._isModelFull(control)) {
            return null;
        }
        return {
            date: this._datetimeParserFormatter.format(control, RxDatetimePickerMode.Date, false, false, this.localeID, this.inputFormat),
            time: this._datetimeParserFormatter.format(control, RxDatetimePickerMode.Time, this.hasSeconds, this.use12HoursTime, this.localeID, this.inputFormat),
            dateTime: this._datetimeParserFormatter.format(control, this.mode, this.hasSeconds, this.use12HoursTime, this.localeID, this.inputFormat)
        };
    }
    _updateRange() {
        this.rangeFrom = this.controlStartValue;
        this.rangeTo = this.controlEndValue;
    }
    _checkWizardLastStep() {
        var _a;
        // In PROD mode this.endDatetimePickerComponent is undefined
        // Add extra change detection to prevent fix it
        // https://github.bmc.com/bmc-ux/adapt-angular/issues/6721
        SafeCdr.detectChanges(this._changeDetectorRef);
        this.isLastWizardStep = this.disableWizard ||
            (this._isModelFull(this.controlStartValue) &&
                this._isModelFull(this.controlEndValue) &&
                !this.startRangeActive &&
                ((_a = this.endDatetimePickerComponent) === null || _a === void 0 ? void 0 : _a.publicApi.isLastStep()));
    }
    _rangeToDayFilter(date) {
        return (this.rangeFrom ? checkIsAfter(date, this.rangeFrom, true) : true) &&
            (isFunction(this.dayFilter) ? this.dayFilter(date) : true);
    }
    _rangeFromDayFilter(date) {
        return (this.rangeTo ? checkIsAfter(this.rangeTo, date, true) : true) &&
            (isFunction(this.dayFilter) ? this.dayFilter(date) : true);
    }
    _propagateChange() {
        this._onChange(this._controlStartValue || this._controlEndValue ? [this._controlStartValue, this._controlEndValue] : null);
    }
}
AdaptRxDatetimeRangeComponent.ɵfac = function AdaptRxDatetimeRangeComponent_Factory(t) { return new (t || AdaptRxDatetimeRangeComponent)(i0.ɵɵdirectiveInject(i1.NgControl, 10), i0.ɵɵdirectiveInject(i2.AdaptRxDatetimeConfigService), i0.ɵɵdirectiveInject(i3.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i4.AdaptRxDatetimeI18nService), i0.ɵɵdirectiveInject(i5.AdaptRxDatetimeAdapter), i0.ɵɵdirectiveInject(i6.AdaptRxDatetimeParserFormatter), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i7.AdaptTranslateService), i0.ɵɵdirectiveInject(LOCALE_ID), i0.ɵɵdirectiveInject(i8.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i9.AdaptInteractionSourceService), i0.ɵɵdirectiveInject(i10.Directionality), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptRxDatetimeRangeComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptRxDatetimeRangeComponent, selectors: [["adapt-rx-datetime-range"]], viewQuery: function AdaptRxDatetimeRangeComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
        i0.ɵɵviewQuery(_c1, 5);
        i0.ɵɵviewQuery(_c2, 7);
        i0.ɵɵviewQuery(_c3, 5);
        i0.ɵɵviewQuery(_c4, 5);
        i0.ɵɵviewQuery(_c5, 5);
        i0.ɵɵviewQuery(_c6, 5);
        i0.ɵɵviewQuery(_c7, 5);
        i0.ɵɵviewQuery(_c8, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.inputControlRef = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.openPopupRef = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.adaptRxFeedbackRef = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.startRangeTab = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.endRangeTab = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.popup = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.startDatetimePickerComponent = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.endDatetimePickerComponent = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.datetimePickerContainer = _t.first);
    } }, inputs: { placeholder: "placeholder", inline: "inline", placement: "placement", appendToBody: "appendToBody", inlineLight: "inlineLight", inlineCompact: "inlineCompact", dayFilter: "dayFilter", disableWizard: "disableWizard", mode: "mode", hasSeconds: "hasSeconds", use12HoursTime: "use12HoursTime", firstDayOfWeek: "firstDayOfWeek", initialDatetime: "initialDatetime", defaultDatetime: "defaultDatetime", disabledStyleForReadonlyState: "disabledStyleForReadonlyState", popupClass: "popupClass", texts: "texts", inputFormat: "inputFormat" }, outputs: { onPopupOpenChange: "onPopupOpenChange", onDatetimeChange: "onDatetimeChange" }, features: [i0.ɵɵInheritDefinitionFeature], decls: 11, vars: 18, consts: [[3, "for", "label", "subLabel", "requiredLabel", "showRequiredLabel", "tooltip", "testID"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [3, "errors", "warningMessage", "controlTouched"], ["adaptRxFeedbackRef", ""], ["dateTimePickerTemplate", ""], ["dateTimeInlineTemplate", ""], ["dateTimeDropdownTemplate", ""], [1, "bg-background-inverse", "px-3"], [3, "errors", "warningMessage", "controlTouched", "alertFeedbackStyle", "alertFeedbackTruncation"], ["role", "tablist", 1, "a3t-datetime-range--header"], ["type", "button", "role", "tab", 3, "click", "keydown.arrowRight"], ["startRangeTab", ""], [1, "a3t-datetime-range--header__summary"], [4, "ngIf"], ["type", "button", "role", "tab", 3, "click", "keydown.arrowLeft"], ["endRangeTab", ""], [3, "id", 4, "ngIf"], ["class", "a3t-datetime-range--header__summary-date", 4, "ngIf"], ["class", "a3t-datetime-range--header__summary-time", 4, "ngIf"], [1, "a3t-datetime-range--header__summary-date"], [1, "a3t-datetime-range--header__summary-time"], [3, "id"], [3, "mode", "expandedControlsStyle", "hasSeconds", "use12HoursTime", "firstDayOfWeek", "initialDatetime", "activeDatetime", "dayFilter", "rangeFrom", "rangeTo", "disabled", "readonly", "texts", "testID", "pickerValueChange", "pickerStateChange"], ["startDatetimePickerComponentRef", ""], ["endDatetimePickerComponentRef", ""], [1, "a3t-datetime--calendar-min-width"], [3, "ngTemplateOutlet"], ["adaptDropdown", "", 3, "adaptRadarDisableEventSending", "placement", "holdFocusInMenu", "autoFocusFirst", "restoreFocusAfterClose", "appendToBody", "autoClose", "mobileView", "onOpen", "onClose"], ["adaptDropdownDirective", "adaptDropdown"], ["adaptDropdownAnchor", "", 1, "a3t-datetime--controls-wrapper", "input-group"], ["type", "text", 1, "form-control", "rx-form-control", "a3t-datetime--input-control", 3, "id", "value", "required", "disabled", "readOnly", "autofocus", "click", "input", "focus", "blur"], ["inputControlRef", ""], ["class", "a3t-datetime--calendar-control input-group-append", "type", "button", "role", "button", 3, "disabled", "a3t-datetime--calendar-control_invalid", "disabled-look", "click", "keydown", 4, "ngIf"], ["adaptDropdownMenuTemplate", ""], ["type", "button", "role", "button", 1, "a3t-datetime--calendar-control", "input-group-append", 3, "disabled", "click", "keydown"], ["openPopupRef", ""], [1, "a3t-datetime--calendar-control__icon"], [1, "sr-only"], [1, "dropdown-menu", "p-0", 3, "ngClass", "keydown"], ["datetimePickerContainer", ""], [1, "a3t-datetime--footer"], ["type", "button", 1, "btn", "btn-sm", "btn-link", 3, "click", "keydown.arrowRight"], ["closeAndCancelButton", ""], [1, "d-flex", "justify-content-center"], ["type", "button", 1, "btn", "btn-sm", "btn-secondary", "m-end-1", 3, "click", "keydown.arrowLeft", "keydown.arrowRight"], ["clearButton", ""], ["type", "button", 1, "btn", "btn-sm", "btn-primary", 3, "click", "keydown.arrowLeft"], ["wizardNextButton", ""]], template: function AdaptRxDatetimeRangeComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div");
        i0.ɵɵelement(1, "adapt-rx-control-label", 0);
        i0.ɵɵelementContainer(2, 1);
        i0.ɵɵelement(3, "adapt-rx-feedback", 2, 3);
        i0.ɵɵelementEnd();
        i0.ɵɵtemplate(5, AdaptRxDatetimeRangeComponent_ng_template_5_Template, 20, 32, "ng-template", null, 4, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(7, AdaptRxDatetimeRangeComponent_ng_template_7_Template, 2, 5, "ng-template", null, 5, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtemplate(9, AdaptRxDatetimeRangeComponent_ng_template_9_Template, 7, 39, "ng-template", null, 6, i0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r0 = i0.ɵɵreference(4);
        const _r3 = i0.ɵɵreference(8);
        const _r5 = i0.ɵɵreference(10);
        i0.ɵɵclassProp("has-danger", _r0.hasUIErrorState)("has-warning", ctx.warningMessage);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("for", ctx.id)("label", ctx.label)("subLabel", ctx.subLabel)("requiredLabel", ctx.requiredLabel)("showRequiredLabel", ctx.required)("tooltip", ctx.tooltip)("testID", ctx.testID);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngTemplateOutlet", ctx.inline ? _r3 : _r5)("ngTemplateOutletContext", i0.ɵɵpureFunction1(16, _c10, _r0.hasUIErrorState));
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("errors", ctx.ngControl.errors)("warningMessage", ctx.warningMessage)("controlTouched", ctx.ngControl.touched);
    } }, directives: [i11.AdaptRxControlLabelComponent, i12.NgTemplateOutlet, i13.AdaptRxFeedbackComponent, i12.NgIf, i14.AdaptRxDatetimePickerComponent, i15.AdaptDropdownDirective, i15.AdaptDropdownAnchorDirective, i15.AdaptDropdownMenuTemplateDirective, i12.NgClass], encapsulation: 2, data: { animation: [
            trigger('slideAnimation', SLIDE_ANIMATION)
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxDatetimeRangeComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-rx-datetime-range',
                templateUrl: './adapt-rx-datetime-range.component.html',
                preserveWhitespaces: false,
                animations: [
                    trigger('slideAnimation', SLIDE_ANIMATION)
                ]
            }]
    }], function () { return [{ type: i1.NgControl, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }, { type: i2.AdaptRxDatetimeConfigService }, { type: i3.AdaptDeviceDetectionService }, { type: i4.AdaptRxDatetimeI18nService }, { type: i5.AdaptRxDatetimeAdapter }, { type: i6.AdaptRxDatetimeParserFormatter }, { type: i0.ChangeDetectorRef }, { type: i0.Renderer2 }, { type: i7.AdaptTranslateService }, { type: undefined, decorators: [{
                type: Inject,
                args: [LOCALE_ID]
            }] }, { type: i8.AdaptRadarAngularService }, { type: i9.AdaptInteractionSourceService }, { type: i10.Directionality }, { type: i0.ElementRef }]; }, { placeholder: [{
            type: Input
        }], inline: [{
            type: Input
        }], placement: [{
            type: Input
        }], appendToBody: [{
            type: Input
        }], inlineLight: [{
            type: Input
        }], inlineCompact: [{
            type: Input
        }], dayFilter: [{
            type: Input
        }], disableWizard: [{
            type: Input
        }], mode: [{
            type: Input
        }], hasSeconds: [{
            type: Input
        }], use12HoursTime: [{
            type: Input
        }], firstDayOfWeek: [{
            type: Input
        }], initialDatetime: [{
            type: Input
        }], defaultDatetime: [{
            type: Input
        }], disabledStyleForReadonlyState: [{
            type: Input
        }], popupClass: [{
            type: Input
        }], texts: [{
            type: Input
        }], inputFormat: [{
            type: Input
        }], onPopupOpenChange: [{
            type: Output
        }], onDatetimeChange: [{
            type: Output
        }], inputControlRef: [{
            type: ViewChild,
            args: ['inputControlRef', { static: false }]
        }], openPopupRef: [{
            type: ViewChild,
            args: ['openPopupRef', { static: false }]
        }], adaptRxFeedbackRef: [{
            type: ViewChild,
            args: ['adaptRxFeedbackRef', { static: true }]
        }], startRangeTab: [{
            type: ViewChild,
            args: ['startRangeTab', { static: false }]
        }], endRangeTab: [{
            type: ViewChild,
            args: ['endRangeTab', { static: false }]
        }], popup: [{
            type: ViewChild,
            args: ['adaptDropdownDirective', { static: false }]
        }], startDatetimePickerComponent: [{
            type: ViewChild,
            args: ['startDatetimePickerComponentRef', { static: false }]
        }], endDatetimePickerComponent: [{
            type: ViewChild,
            args: ['endDatetimePickerComponentRef', { static: false }]
        }], datetimePickerContainer: [{
            type: ViewChild,
            args: ['datetimePickerContainer', { static: false }]
        }] }); })();
