// @bmc-ux/adapt-angular/esm2015/rx/rx-datetime/adapt-rx-datetime-adapter.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export function ADAPT_RX_DATETIME_ADAPTER_FACTORY() {
    return new AdaptRxDatetimeStructAdapter();
}
export class AdaptRxDatetimeAdapter {
}
AdaptRxDatetimeAdapter.ɵfac = function AdaptRxDatetimeAdapter_Factory(t) { return new (t || AdaptRxDatetimeAdapter)(); };
AdaptRxDatetimeAdapter.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptRxDatetimeAdapter, factory: function () { return ADAPT_RX_DATETIME_ADAPTER_FACTORY(); }, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxDatetimeAdapter, [{
        type: Injectable,
        args: [{ providedIn: 'root', useFactory: ADAPT_RX_DATETIME_ADAPTER_FACTORY }]
    }], null, null); })();
export class AdaptRxDatetimeStructAdapter extends AdaptRxDatetimeAdapter {
    /**
     * Converts RxDatetimeStruct => RxDatetimeStruct
     */
    fromModel(date) {
        return date;
    }
    /**
     * Converts RxDatetimeStruct => RxDatetimeStruct
     */
    toModel(date) {
        return date;
    }
}
AdaptRxDatetimeStructAdapter.ɵfac = /*@__PURE__*/ function () { let ɵAdaptRxDatetimeStructAdapter_BaseFactory; return function AdaptRxDatetimeStructAdapter_Factory(t) { return (ɵAdaptRxDatetimeStructAdapter_BaseFactory || (ɵAdaptRxDatetimeStructAdapter_BaseFactory = i0.ɵɵgetInheritedFactory(AdaptRxDatetimeStructAdapter)))(t || AdaptRxDatetimeStructAdapter); }; }();
AdaptRxDatetimeStructAdapter.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptRxDatetimeStructAdapter, factory: AdaptRxDatetimeStructAdapter.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxDatetimeStructAdapter, [{
        type: Injectable
    }], null, null); })();
