// @bmc-ux/adapt-angular/esm2015/rx/rx-datetime/adapt-rx-datetime-config.service.js
import { Injectable } from '@angular/core';
import { AdaptTranslateService } from '../../common/i18n/index';
import * as i0 from "@angular/core";
import * as i1 from "../../common/i18n/index";
export class AdaptRxDatetimeConfigService {
    constructor(_ts) {
        this._ts = _ts;
    }
    get texts() {
        const translation = this._ts.getCurrentLanguage();
        return {
            clock: {
                hhBtn: translation['adapt.datetime.hh'],
                mmBtn: translation['adapt.datetime.mm'],
                ssBtn: translation['adapt.datetime.ss'],
                amBtn: translation['adapt.datetime.am'],
                pmBtn: translation['adapt.datetime.pm'],
                hoursLabel: translation['adapt.datetime.hours'],
                minutesLabel: translation['adapt.datetime.minutes'],
                secondsLabel: translation['adapt.datetime.seconds'],
                meridiemLabel: translation['adapt.datetime.meridiem']
            },
            emptyDate: translation['adapt.datetime.emptyDate'],
            emptyTime: translation['adapt.datetime.emptyTime'],
            emptyYear: translation['adapt.datetime.emptyYear'],
            emptyMonth: translation['adapt.datetime.emptyMonth'],
            cancelBtn: translation['adapt.common.cancel'],
            clearBtn: translation['adapt.common.clearAll'],
            nextBtn: translation['adapt.common.next'],
            applyBtn: translation['adapt.common.ok'],
            calendarBtn: translation['adapt.datetime.calendar'],
            clockBtn: translation['adapt.datetime.clock'],
            startRangeLabel: translation['adapt.datetimeRange.startDate'],
            endRangeLabel: translation['adapt.datetimeRange.endDate'],
            emptyStartRange: translation['adapt.datetimeRange.emptyStart'],
            emptyEndRange: translation['adapt.datetimeRange.emptyEnd']
        };
    }
}
AdaptRxDatetimeConfigService.ɵfac = function AdaptRxDatetimeConfigService_Factory(t) { return new (t || AdaptRxDatetimeConfigService)(i0.ɵɵinject(i1.AdaptTranslateService)); };
AdaptRxDatetimeConfigService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptRxDatetimeConfigService, factory: AdaptRxDatetimeConfigService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxDatetimeConfigService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.AdaptTranslateService }]; }, null); })();
