// @bmc-ux/adapt-angular/esm2015/rx/rx-datetime/adapt-rx-datetime.types.js
import { DateTimeKey } from '../../common/common.enums';
// Calendar
export class RxCalendarDay {
    constructor(index) {
        this.index = index;
    }
}
export class RxCalendar {
}
// Year
export class RxYear {
}
export class RxYearConfig {
}
// Month
export class RxMonth {
}
export class RxMonthConfig {
}
// Clock
export var RxClockMode;
(function (RxClockMode) {
    RxClockMode[RxClockMode["Hours"] = 0] = "Hours";
    RxClockMode[RxClockMode["Minutes"] = 1] = "Minutes";
    RxClockMode[RxClockMode["Seconds"] = 2] = "Seconds";
})(RxClockMode || (RxClockMode = {}));
export var RxClockFace;
(function (RxClockFace) {
    RxClockFace[RxClockFace["Hours24"] = 0] = "Hours24";
    RxClockFace[RxClockFace["Hours12"] = 1] = "Hours12";
    RxClockFace[RxClockFace["Minutes"] = 2] = "Minutes";
    RxClockFace[RxClockFace["Seconds"] = 3] = "Seconds";
})(RxClockFace || (RxClockFace = {}));
// DateTimePicker
export var RxDatetimePickerMode;
(function (RxDatetimePickerMode) {
    RxDatetimePickerMode[RxDatetimePickerMode["DateTime"] = 0] = "DateTime";
    RxDatetimePickerMode[RxDatetimePickerMode["Date"] = 1] = "Date";
    RxDatetimePickerMode[RxDatetimePickerMode["Time"] = 2] = "Time";
    RxDatetimePickerMode[RxDatetimePickerMode["YearMonth"] = 3] = "YearMonth";
    RxDatetimePickerMode[RxDatetimePickerMode["Year"] = 4] = "Year";
    RxDatetimePickerMode[RxDatetimePickerMode["Month"] = 5] = "Month";
})(RxDatetimePickerMode || (RxDatetimePickerMode = {}));
export var RxDatetimePickerState;
(function (RxDatetimePickerState) {
    RxDatetimePickerState[RxDatetimePickerState["Date"] = 0] = "Date";
    RxDatetimePickerState[RxDatetimePickerState["TimeHours"] = 1] = "TimeHours";
    RxDatetimePickerState[RxDatetimePickerState["TimeMinutes"] = 2] = "TimeMinutes";
    RxDatetimePickerState[RxDatetimePickerState["TimeSeconds"] = 3] = "TimeSeconds";
    RxDatetimePickerState[RxDatetimePickerState["Year"] = 4] = "Year";
    RxDatetimePickerState[RxDatetimePickerState["Month"] = 5] = "Month";
})(RxDatetimePickerState || (RxDatetimePickerState = {}));
