// @bmc-ux/adapt-angular/esm2015/rx/rx-datetime/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ai_angle_down, ai_angle_left, ai_angle_right, ai_calendar, ai_clock_o } from '@bmc-ux/dpl-iconfont';
import { AdaptDropdownModule } from '../../dropdown/index';
import { AdaptIconModule, AdaptIconConfig } from '../../icon/index';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxCalendarComponent } from './calendar/adapt-rx-calendar.component';
import { AdaptRxClockComponent } from './clock/adapt-rx-clock.component';
import { AdaptRxDatetimeComponent } from './adapt-rx-datetime.component';
import { AdaptRxDatetimePickerComponent } from './datetime-picker/adapt-rx-datetime-picker.component';
import { AdaptRxDatetimeRangeComponent } from './adapt-rx-datetime-range.component';
import { AdaptRxMonthComponent } from './month/adapt-rx-month.component';
import { AdaptRxYearComponent } from './year/adapt-rx-year.component';
import * as i0 from "@angular/core";
import * as i1 from "../../icon/index";
export class AdaptRxDatetimeModule {
    constructor(iconService) {
        iconService.registerIcons([
            ai_angle_down,
            ai_angle_right,
            ai_angle_left,
            ai_calendar,
            ai_clock_o
        ]);
    }
}
AdaptRxDatetimeModule.ɵfac = function AdaptRxDatetimeModule_Factory(t) { return new (t || AdaptRxDatetimeModule)(i0.ɵɵinject(i1.AdaptIconConfig)); };
AdaptRxDatetimeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxDatetimeModule });
AdaptRxDatetimeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptDropdownModule,
            AdaptIconModule,
            AdaptRxFormControlModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxDatetimeModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptDropdownModule,
                    AdaptIconModule,
                    AdaptRxFormControlModule
                ],
                declarations: [
                    AdaptRxCalendarComponent,
                    AdaptRxClockComponent,
                    AdaptRxDatetimeComponent,
                    AdaptRxDatetimePickerComponent,
                    AdaptRxDatetimeRangeComponent,
                    AdaptRxMonthComponent,
                    AdaptRxYearComponent,
                ],
                exports: [
                    AdaptRxCalendarComponent,
                    AdaptRxClockComponent,
                    AdaptRxDatetimeComponent,
                    AdaptRxDatetimePickerComponent,
                    AdaptRxDatetimeRangeComponent,
                    AdaptRxMonthComponent,
                    AdaptRxYearComponent,
                ]
            }]
    }], function () { return [{ type: i1.AdaptIconConfig }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxDatetimeModule, { declarations: [AdaptRxCalendarComponent,
        AdaptRxClockComponent,
        AdaptRxDatetimeComponent,
        AdaptRxDatetimePickerComponent,
        AdaptRxDatetimeRangeComponent,
        AdaptRxMonthComponent,
        AdaptRxYearComponent], imports: [CommonModule,
        AdaptDropdownModule,
        AdaptIconModule,
        AdaptRxFormControlModule], exports: [AdaptRxCalendarComponent,
        AdaptRxClockComponent,
        AdaptRxDatetimeComponent,
        AdaptRxDatetimePickerComponent,
        AdaptRxDatetimeRangeComponent,
        AdaptRxMonthComponent,
        AdaptRxYearComponent] }); })();
export * from './calendar/adapt-rx-calendar.component';
export * from './clock/adapt-rx-clock.component';
export * from './datetime-picker/adapt-rx-datetime-picker.component';
export * from './month/adapt-rx-month.component';
export * from './year/adapt-rx-year.component';
export * from './common/common.utils';
export * from './common/parse.utils';
export * from './common/partition.utils';
export * from './adapt-rx-datetime-adapter.service';
export * from './adapt-rx-datetime-config.service';
export * from './adapt-rx-datetime-i18n.service';
export * from './adapt-rx-datetime-parser-formatter.service';
export * from './adapt-rx-datetime-range.component';
export * from './adapt-rx-datetime.component';
export * from './adapt-rx-datetime.types';
