// @bmc-ux/adapt-angular/esm2015/toast/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ToastModule } from 'primeng/toast';
import { AdaptToastComponent } from './toast.component';
import { AdaptToastItemComponent } from './toast-item.component';
import * as i0 from "@angular/core";
export class AdaptToastModule {
}
AdaptToastModule.ɵfac = function AdaptToastModule_Factory(t) { return new (t || AdaptToastModule)(); };
AdaptToastModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptToastModule });
AdaptToastModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, ToastModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptToastModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, ToastModule],
                declarations: [AdaptToastComponent, AdaptToastItemComponent],
                exports: [AdaptToastComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptToastModule, { declarations: [AdaptToastComponent, AdaptToastItemComponent], imports: [CommonModule, ToastModule], exports: [AdaptToastComponent] }); })();
export * from './toast-item.component';
export * from './toast-message.service';
export * from './toast.component';
export * from './toast.model';
