// @bmc-ux/adapt-angular/esm2015/toast/toast-item.component.js
import { Component, Input, Output, EventEmitter, ViewChild, ElementRef, ViewEncapsulation, NgZone, ChangeDetectorRef } from '@angular/core';
import { trigger, animate, transition, style, keyframes } from '@angular/animations';
import { isFunction } from 'lodash-es';
import { ANIMATION_TIMING_FUNCTION, ANIMATION_DURATION } from '../common/animation';
import { SafeCdr } from '../common/safe-cdr';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
const _c0 = ["nonExpandedDetails"];
const _c1 = ["expandedDetails"];
function AdaptToastItemComponent_ng_container_3_div_12_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13, 14);
    i0.ɵɵtext(2);
    i0.ɵɵpipe(3, "slice");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1("", i0.ɵɵpipeBind3(3, 1, ctx_r10.message.detail, 0, 160), "...");
} }
function AdaptToastItemComponent_ng_container_3_div_12_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13, 15);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r11.message.detail);
} }
function AdaptToastItemComponent_ng_container_3_div_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptToastItemComponent_ng_container_3_div_12_span_2_Template, 4, 5, "span", 9);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptToastItemComponent_ng_container_3_div_12_span_4_Template, 3, 1, "span", 9);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext(2);
    i0.ɵɵstyleProp("transition", "height " + ctx_r7.animationDuration)("height", ctx_r7.expanded ? ctx_r7.totalHeight : ctx_r7.nonExpandedTextHeight, "px");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r7.expanded);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r7.expanded);
} }
function AdaptToastItemComponent_ng_container_3_div_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 13);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r8.message.detail);
} }
function AdaptToastItemComponent_ng_container_3_ng_template_23_Template(rf, ctx) { }
const _c2 = function (a0, a1, a2, a3) { return { "a-toast--info": a0, "a-toast--warning": a1, "a-toast--danger": a2, "a-toast--success": a3 }; };
const _c3 = function (a0, a1) { return { "zIndex": a0, "marginBottom": a1 }; };
function AdaptToastItemComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    const _r15 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "div", 3, 4);
    i0.ɵɵlistener("@toastAnim.done", function AdaptToastItemComponent_ng_container_3_Template_div_animation_toastAnim_done_2_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r14 = i0.ɵɵnextContext(); return ctx_r14.animationDone($event); })("mouseleave", function AdaptToastItemComponent_ng_container_3_Template_div_mouseleave_2_listener() { i0.ɵɵrestoreView(_r15); const ctx_r16 = i0.ɵɵnextContext(); ctx_r16.hideFullMessage(); return ctx_r16.onMouseLeave(); });
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "div", 5);
    i0.ɵɵlistener("mouseenter", function AdaptToastItemComponent_ng_container_3_Template_div_mouseenter_5_listener() { i0.ɵɵrestoreView(_r15); const ctx_r17 = i0.ɵɵnextContext(); ctx_r17.showFullMessage(); return ctx_r17.onMouseEnter(); });
    i0.ɵɵtext(6, "\n        ");
    i0.ɵɵelementStart(7, "div", 6);
    i0.ɵɵtext(8);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n        ");
    i0.ɵɵelementStart(10, "div", 7);
    i0.ɵɵtext(11, "\n          ");
    i0.ɵɵtemplate(12, AdaptToastItemComponent_ng_container_3_div_12_Template, 6, 6, "div", 8);
    i0.ɵɵtext(13, "\n\n          ");
    i0.ɵɵtemplate(14, AdaptToastItemComponent_ng_container_3_div_14_Template, 2, 1, "div", 9);
    i0.ɵɵtext(15, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(16, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n      ");
    i0.ɵɵelementStart(18, "div", 10);
    i0.ɵɵtext(19, "\n        ");
    i0.ɵɵelementStart(20, "button", 11);
    i0.ɵɵlistener("click", function AdaptToastItemComponent_ng_container_3_Template_button_click_20_listener($event) { i0.ɵɵrestoreView(_r15); const ctx_r18 = i0.ɵɵnextContext(); return ctx_r18.onCloseIconClick($event); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(21, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(22, "\n      ");
    i0.ɵɵtemplate(23, AdaptToastItemComponent_ng_container_3_ng_template_23_Template, 0, 0, "ng-template", 12);
    i0.ɵɵtext(24, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    const _r4 = i0.ɵɵreference(10);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("@toastAnim", undefined)("ngClass", i0.ɵɵpureFunction4(8, _c2, ctx_r1.message.severity === "info", ctx_r1.message.severity === "warn", ctx_r1.message.severity === "error", ctx_r1.message.severity === "success"));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction2(13, _c3, ctx_r1.zIndex - ctx_r1.index, ctx_r1.expanded && ctx_r1.message.detail.length >= 160 ? "0" : "10px"));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.message.summary);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r1.message.detail.length >= 160);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.message.detail.length < 160);
    i0.ɵɵadvance(6);
    i0.ɵɵattribute("aria-label", ctx_r1.ariaLabel);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", _r4);
} }
function AdaptToastItemComponent_ng_container_5_div_12_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13, 14);
    i0.ɵɵtext(2);
    i0.ɵɵpipe(3, "slice");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r23 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1("", i0.ɵɵpipeBind3(3, 1, ctx_r23.message.detail, 0, 160), "...");
} }
function AdaptToastItemComponent_ng_container_5_div_12_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13, 15);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r24 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r24.message.detail);
} }
function AdaptToastItemComponent_ng_container_5_div_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptToastItemComponent_ng_container_5_div_12_span_2_Template, 4, 5, "span", 9);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptToastItemComponent_ng_container_5_div_12_span_4_Template, 3, 1, "span", 9);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r20 = i0.ɵɵnextContext(2);
    i0.ɵɵstyleProp("transition", "height " + ctx_r20.animationDuration)("height", ctx_r20.expanded ? ctx_r20.totalHeight : ctx_r20.nonExpandedTextHeight, "px");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r20.expanded);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r20.expanded);
} }
function AdaptToastItemComponent_ng_container_5_div_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 13);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r21 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r21.message.detail);
} }
function AdaptToastItemComponent_ng_container_5_ng_template_23_Template(rf, ctx) { }
function AdaptToastItemComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    const _r28 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "div", 3, 4);
    i0.ɵɵlistener("@toastAnimLeft.done", function AdaptToastItemComponent_ng_container_5_Template_div_animation_toastAnimLeft_done_2_listener($event) { i0.ɵɵrestoreView(_r28); const ctx_r27 = i0.ɵɵnextContext(); return ctx_r27.animationDone($event); })("mouseleave", function AdaptToastItemComponent_ng_container_5_Template_div_mouseleave_2_listener() { i0.ɵɵrestoreView(_r28); const ctx_r29 = i0.ɵɵnextContext(); ctx_r29.hideFullMessage(); return ctx_r29.onMouseLeave(); });
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "div", 5);
    i0.ɵɵlistener("mouseenter", function AdaptToastItemComponent_ng_container_5_Template_div_mouseenter_5_listener() { i0.ɵɵrestoreView(_r28); const ctx_r30 = i0.ɵɵnextContext(); ctx_r30.showFullMessage(); return ctx_r30.onMouseEnter(); });
    i0.ɵɵtext(6, "\n        ");
    i0.ɵɵelementStart(7, "div", 6);
    i0.ɵɵtext(8);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n        ");
    i0.ɵɵelementStart(10, "div", 7);
    i0.ɵɵtext(11, "\n          ");
    i0.ɵɵtemplate(12, AdaptToastItemComponent_ng_container_5_div_12_Template, 6, 6, "div", 8);
    i0.ɵɵtext(13, "\n\n          ");
    i0.ɵɵtemplate(14, AdaptToastItemComponent_ng_container_5_div_14_Template, 2, 1, "div", 9);
    i0.ɵɵtext(15, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(16, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n      ");
    i0.ɵɵelementStart(18, "div", 10);
    i0.ɵɵtext(19, "\n        ");
    i0.ɵɵelementStart(20, "button", 11);
    i0.ɵɵlistener("click", function AdaptToastItemComponent_ng_container_5_Template_button_click_20_listener($event) { i0.ɵɵrestoreView(_r28); const ctx_r31 = i0.ɵɵnextContext(); return ctx_r31.onCloseIconClick($event); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(21, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(22, "\n      ");
    i0.ɵɵtemplate(23, AdaptToastItemComponent_ng_container_5_ng_template_23_Template, 0, 0, "ng-template", 12);
    i0.ɵɵtext(24, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    const _r4 = i0.ɵɵreference(10);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("@toastAnimLeft", undefined)("ngClass", i0.ɵɵpureFunction4(8, _c2, ctx_r2.message.severity === "info", ctx_r2.message.severity === "warn", ctx_r2.message.severity === "error", ctx_r2.message.severity === "success"));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction2(13, _c3, ctx_r2.zIndex - ctx_r2.index, ctx_r2.expanded && ctx_r2.message.detail.length >= 160 ? "0" : "10px"));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r2.message.summary);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r2.message.detail.length >= 160);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r2.message.detail.length < 160);
    i0.ɵɵadvance(6);
    i0.ɵɵattribute("aria-label", ctx_r2.ariaLabel);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", _r4);
} }
function AdaptToastItemComponent_ng_container_7_div_12_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13, 14);
    i0.ɵɵtext(2);
    i0.ɵɵpipe(3, "slice");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r36 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate1("", i0.ɵɵpipeBind3(3, 1, ctx_r36.message.detail, 0, 160), "...");
} }
function AdaptToastItemComponent_ng_container_7_div_12_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 13, 15);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r37 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r37.message.detail);
} }
function AdaptToastItemComponent_ng_container_7_div_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptToastItemComponent_ng_container_7_div_12_span_2_Template, 4, 5, "span", 9);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptToastItemComponent_ng_container_7_div_12_span_4_Template, 3, 1, "span", 9);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r33 = i0.ɵɵnextContext(2);
    i0.ɵɵstyleProp("transition", "height " + ctx_r33.animationDuration)("height", ctx_r33.expanded ? ctx_r33.totalHeight : ctx_r33.nonExpandedTextHeight, "px");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r33.expanded);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r33.expanded);
} }
function AdaptToastItemComponent_ng_container_7_div_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 13);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r34 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r34.message.detail);
} }
function AdaptToastItemComponent_ng_container_7_ng_template_23_Template(rf, ctx) { }
function AdaptToastItemComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    const _r41 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "div", 3, 4);
    i0.ɵɵlistener("@toastAnimTopCenter.done", function AdaptToastItemComponent_ng_container_7_Template_div_animation_toastAnimTopCenter_done_2_listener($event) { i0.ɵɵrestoreView(_r41); const ctx_r40 = i0.ɵɵnextContext(); return ctx_r40.animationDone($event); })("mouseleave", function AdaptToastItemComponent_ng_container_7_Template_div_mouseleave_2_listener() { i0.ɵɵrestoreView(_r41); const ctx_r42 = i0.ɵɵnextContext(); ctx_r42.hideFullMessage(); return ctx_r42.onMouseLeave(); });
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "div", 5);
    i0.ɵɵlistener("mouseenter", function AdaptToastItemComponent_ng_container_7_Template_div_mouseenter_5_listener() { i0.ɵɵrestoreView(_r41); const ctx_r43 = i0.ɵɵnextContext(); ctx_r43.showFullMessage(); return ctx_r43.onMouseEnter(); });
    i0.ɵɵtext(6, "\n        ");
    i0.ɵɵelementStart(7, "div", 6);
    i0.ɵɵtext(8);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n        ");
    i0.ɵɵelementStart(10, "div", 7);
    i0.ɵɵtext(11, "\n          ");
    i0.ɵɵtemplate(12, AdaptToastItemComponent_ng_container_7_div_12_Template, 6, 6, "div", 8);
    i0.ɵɵtext(13, "\n\n          ");
    i0.ɵɵtemplate(14, AdaptToastItemComponent_ng_container_7_div_14_Template, 2, 1, "div", 9);
    i0.ɵɵtext(15, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(16, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n      ");
    i0.ɵɵelementStart(18, "div", 10);
    i0.ɵɵtext(19, "\n        ");
    i0.ɵɵelementStart(20, "button", 11);
    i0.ɵɵlistener("click", function AdaptToastItemComponent_ng_container_7_Template_button_click_20_listener($event) { i0.ɵɵrestoreView(_r41); const ctx_r44 = i0.ɵɵnextContext(); return ctx_r44.onCloseIconClick($event); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(21, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(22, "\n      ");
    i0.ɵɵtemplate(23, AdaptToastItemComponent_ng_container_7_ng_template_23_Template, 0, 0, "ng-template", 12);
    i0.ɵɵtext(24, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext();
    const _r4 = i0.ɵɵreference(10);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("@toastAnimTopCenter", undefined)("ngClass", i0.ɵɵpureFunction4(8, _c2, ctx_r3.message.severity === "info", ctx_r3.message.severity === "warn", ctx_r3.message.severity === "error", ctx_r3.message.severity === "success"));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngStyle", i0.ɵɵpureFunction2(13, _c3, ctx_r3.zIndex - ctx_r3.index, ctx_r3.expanded && ctx_r3.message.detail.length >= 160 ? "0" : "10px"));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r3.message.summary);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r3.message.detail.length >= 160);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r3.message.detail.length < 160);
    i0.ɵɵadvance(6);
    i0.ɵɵattribute("aria-label", ctx_r3.ariaLabel);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", _r4);
} }
function AdaptToastItemComponent_ng_template_9_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r47 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 17);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "button", 18);
    i0.ɵɵlistener("click", function AdaptToastItemComponent_ng_template_9_div_1_Template_button_click_2_listener() { i0.ɵɵrestoreView(_r47); const ctx_r46 = i0.ɵɵnextContext(2); return ctx_r46.handleLinkClick(); });
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r45 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r45.message.link || ctx_r45.link);
} }
function AdaptToastItemComponent_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n    ");
    i0.ɵɵtemplate(1, AdaptToastItemComponent_ng_template_9_div_1_Template, 5, 1, "div", 16);
    i0.ɵɵtext(2, "\n  ");
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r5.message.link || ctx_r5.link);
} }
export class AdaptToastItemComponent {
    constructor(_ngZone, _cdr) {
        this._ngZone = _ngZone;
        this._cdr = _cdr;
        this.expanded = false;
        this.animationDuration = ANIMATION_DURATION.medium;
        /* eslint-disable-next-line @angular-eslint/no-output-on-prefix */
        this.onClose = new EventEmitter();
        /* eslint-disable-next-line @angular-eslint/no-output-on-prefix */
        this.onAnimationEnd = new EventEmitter(); // changed from 'EventEmitter<any>' since 10.14.0
        this.linkClick = new EventEmitter(); // changed from 'EventEmitter<any>' since 10.14.0
        this._placement = 'top-right';
    }
    get placement() {
        return this._placement;
    }
    set placement(value) {
        this._placement = value;
    }
    ngAfterViewInit() {
        this.initTimeout();
    }
    initTimeout() {
        if (!this.message.sticky) {
            this._ngZone.runOutsideAngular(() => {
                this.timeout = window.setTimeout(() => {
                    this.onClose.emit({
                        index: this.index,
                        message: this.message
                    });
                    SafeCdr.detectChanges(this._cdr);
                }, this.message.life || 3000);
            });
        }
    }
    clearTimeout() {
        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }
    }
    onMouseEnter() {
        this.clearTimeout();
    }
    onMouseLeave() {
        this.initTimeout();
    }
    onCloseIconClick(event) {
        this.clearTimeout();
        this.onClose.emit({
            index: this.index,
            message: this.message
        });
        event.preventDefault();
    }
    /**
     * Show full toast message
     */
    showFullMessage() {
        this.expanded = true;
        this.totalHeight = this.nonExpandedTextHeight;
        setTimeout(() => {
            if (this.expandedDetails) {
                this.expandedTextHeight = this.expandedDetails.nativeElement.getBoundingClientRect().height;
                this.totalHeight = this.expandedTextHeight;
            }
        });
    }
    /**
     * Hide full toast message
     */
    hideFullMessage() {
        this.expanded = false;
    }
    /**
     * Emit empty value to handle link click
     */
    handleLinkClick() {
        this.linkClick.next(this.message);
        if (isFunction(this.message.linkHandler)) {
            this.message.linkHandler(this.message);
        }
    }
    animationDone(event) {
        if (event.toState === 'void') {
            this.onAnimationEnd.next(this.message);
        }
        setTimeout(() => {
            if (this.nonExpandedDetails) {
                this.nonExpandedTextHeight = this.nonExpandedDetails.nativeElement.getBoundingClientRect().height;
            }
        });
    }
    ngOnDestroy() {
        this.clearTimeout();
    }
}
AdaptToastItemComponent.ɵfac = function AdaptToastItemComponent_Factory(t) { return new (t || AdaptToastItemComponent)(i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
AdaptToastItemComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptToastItemComponent, selectors: [["adapt-toast-item"]], viewQuery: function AdaptToastItemComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
        i0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.nonExpandedDetails = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.expandedDetails = _t.first);
    } }, inputs: { message: "message", placement: "placement", index: "index", link: "link", ariaLabel: "ariaLabel", zIndex: "zIndex" }, outputs: { onClose: "onClose", onAnimationEnd: "onAnimationEnd", linkClick: "linkClick" }, decls: 13, vars: 3, consts: [["container", ""], [4, "ngIf"], ["linkSection", ""], ["aria-live", "polite", 1, "a-toast", 3, "ngClass", "mouseleave"], ["messageel", ""], [3, "ngStyle", "mouseenter"], [1, "a-toast__summary"], [1, "a-toast__details"], [3, "transition", "height", 4, "ngIf"], ["class", "a-toast__detail", 4, "ngIf"], [1, "a-toast__close"], ["type", "button", 1, "a-toast__close-button", "d-icon-cross_adapt", 3, "click"], [3, "ngTemplateOutlet"], [1, "a-toast__detail"], ["nonExpandedDetails", ""], ["expandedDetails", ""], ["class", "mt-2 ellipsis", 4, "ngIf"], [1, "mt-2", "ellipsis"], [1, "btn", "btn-link", "p-0", 3, "click"]], template: function AdaptToastItemComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", null, 0);
        i0.ɵɵtext(2, "\n  ");
        i0.ɵɵtemplate(3, AdaptToastItemComponent_ng_container_3_Template, 26, 16, "ng-container", 1);
        i0.ɵɵtext(4, "\n\n  ");
        i0.ɵɵtemplate(5, AdaptToastItemComponent_ng_container_5_Template, 26, 16, "ng-container", 1);
        i0.ɵɵtext(6, "\n\n  ");
        i0.ɵɵtemplate(7, AdaptToastItemComponent_ng_container_7_Template, 26, 16, "ng-container", 1);
        i0.ɵɵtext(8, "\n\n  ");
        i0.ɵɵtemplate(9, AdaptToastItemComponent_ng_template_9_Template, 3, 1, "ng-template", null, 2, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(11, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(12, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.placement === "top-right");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.placement === "top-left");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.placement === "top-center");
    } }, directives: [i1.NgIf, i1.NgClass, i1.NgStyle, i1.NgTemplateOutlet], pipes: [i1.SlicePipe], encapsulation: 2, data: { animation: [
            trigger('toastAnim', [
                transition('void => *', [
                    style({ transform: 'translate3d(20px, 0px, 0px)', opacity: '0' }),
                    animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn, style({ transform: 'translate3d(0px, 0px, 0px)', opacity: '1' }))
                ]),
                transition('* => void', [
                    animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut, keyframes([
                        style({
                            transform: 'scale(1)',
                            opacity: '1',
                            paddingBottom: '*',
                            paddingTop: '*',
                            marginBottom: '*',
                            height: '*',
                            offset: 0
                        }),
                        style({
                            transform: 'scale(.8)',
                            opacity: '0',
                            paddingBottom: '*',
                            paddingTop: '*',
                            marginBottom: '*',
                            height: '*',
                            offset: .7
                        }),
                        style({ paddingBottom: '0px', paddingTop: '0px', marginBottom: '0px', height: '0px', offset: 1 })
                    ]))
                ])
            ]),
            trigger('toastAnimLeft', [
                transition('void => *', [
                    style({ transform: 'translate3d(-20px, 0, 0)', opacity: '0' }),
                    animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn, style({ transform: 'translate3d(0, 0, 0)', opacity: '1' }))
                ]),
                transition('* => void', [
                    animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut, keyframes([
                        style({
                            transform: 'scale(1)', opacity: '1', paddingBottom: '*',
                            paddingTop: '*', marginBottom: '*', height: '*', offset: 0
                        }),
                        style({
                            transform: 'scale(.8)', opacity: '0',
                            paddingBottom: '*', paddingTop: '*',
                            marginBottom: '*', height: '*', offset: .7
                        }),
                        style({ paddingBottom: '0px', paddingTop: '0px', marginBottom: '0px', height: '0px', offset: 1 })
                    ]))
                ])
            ]),
            trigger('toastAnimTopCenter', [
                transition('void => *', [
                    style({ transform: 'translate3d(0, -20px, 0)', opacity: '0' }),
                    animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn, style({ transform: 'translate3d(0, 0, 0)', opacity: '1' }))
                ]),
                transition('* => void', [
                    animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut, keyframes([
                        style({
                            transform: 'scale(1)',
                            opacity: '1',
                            paddingBottom: '*',
                            paddingTop: '*',
                            marginBottom: '*',
                            height: '*',
                            offset: 0
                        }),
                        style({
                            transform: 'scale(.8)',
                            opacity: '0',
                            paddingBottom: '*',
                            paddingTop: '*',
                            marginBottom: '*',
                            height: '*',
                            offset: .7
                        }),
                        style({ paddingBottom: '0px', paddingTop: '0px', marginBottom: '0px', height: '0px', offset: 1 })
                    ]))
                ])
            ])
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptToastItemComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-toast-item',
                templateUrl: './toast-item.component.html',
                animations: [
                    trigger('toastAnim', [
                        transition('void => *', [
                            style({ transform: 'translate3d(20px, 0px, 0px)', opacity: '0' }),
                            animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn, style({ transform: 'translate3d(0px, 0px, 0px)', opacity: '1' }))
                        ]),
                        transition('* => void', [
                            animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut, keyframes([
                                style({
                                    transform: 'scale(1)',
                                    opacity: '1',
                                    paddingBottom: '*',
                                    paddingTop: '*',
                                    marginBottom: '*',
                                    height: '*',
                                    offset: 0
                                }),
                                style({
                                    transform: 'scale(.8)',
                                    opacity: '0',
                                    paddingBottom: '*',
                                    paddingTop: '*',
                                    marginBottom: '*',
                                    height: '*',
                                    offset: .7
                                }),
                                style({ paddingBottom: '0px', paddingTop: '0px', marginBottom: '0px', height: '0px', offset: 1 })
                            ]))
                        ])
                    ]),
                    trigger('toastAnimLeft', [
                        transition('void => *', [
                            style({ transform: 'translate3d(-20px, 0, 0)', opacity: '0' }),
                            animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn, style({ transform: 'translate3d(0, 0, 0)', opacity: '1' }))
                        ]),
                        transition('* => void', [
                            animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut, keyframes([
                                style({
                                    transform: 'scale(1)', opacity: '1', paddingBottom: '*',
                                    paddingTop: '*', marginBottom: '*', height: '*', offset: 0
                                }),
                                style({
                                    transform: 'scale(.8)', opacity: '0',
                                    paddingBottom: '*', paddingTop: '*',
                                    marginBottom: '*', height: '*', offset: .7
                                }),
                                style({ paddingBottom: '0px', paddingTop: '0px', marginBottom: '0px', height: '0px', offset: 1 })
                            ]))
                        ])
                    ]),
                    trigger('toastAnimTopCenter', [
                        transition('void => *', [
                            style({ transform: 'translate3d(0, -20px, 0)', opacity: '0' }),
                            animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn, style({ transform: 'translate3d(0, 0, 0)', opacity: '1' }))
                        ]),
                        transition('* => void', [
                            animate(ANIMATION_DURATION.medium + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut, keyframes([
                                style({
                                    transform: 'scale(1)',
                                    opacity: '1',
                                    paddingBottom: '*',
                                    paddingTop: '*',
                                    marginBottom: '*',
                                    height: '*',
                                    offset: 0
                                }),
                                style({
                                    transform: 'scale(.8)',
                                    opacity: '0',
                                    paddingBottom: '*',
                                    paddingTop: '*',
                                    marginBottom: '*',
                                    height: '*',
                                    offset: .7
                                }),
                                style({ paddingBottom: '0px', paddingTop: '0px', marginBottom: '0px', height: '0px', offset: 1 })
                            ]))
                        ])
                    ])
                ],
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i0.NgZone }, { type: i0.ChangeDetectorRef }]; }, { message: [{
            type: Input
        }], placement: [{
            type: Input
        }], index: [{
            type: Input
        }], link: [{
            type: Input
        }], ariaLabel: [{
            type: Input
        }], zIndex: [{
            type: Input
        }], onClose: [{
            type: Output
        }], onAnimationEnd: [{
            type: Output
        }], linkClick: [{
            type: Output
        }], nonExpandedDetails: [{
            type: ViewChild,
            args: ['nonExpandedDetails', { static: false }]
        }], expandedDetails: [{
            type: ViewChild,
            args: ['expandedDetails', { static: false }]
        }] }); })();
