// @bmc-ux/adapt-angular/esm2015/toast/toast-message.service.js
import { Injectable } from '@angular/core';
import { MessageService } from 'primeng/api';
import * as i0 from "@angular/core";
export class AdaptMessageService extends MessageService {
    constructor() {
        super();
    }
    /**
     * Extension of Primeng method with AdaptMessage type
     * @since v10.14.0
     */
    add(message) {
        super.add(message);
    }
    /**
     * Extension of Primeng method with AdaptMessage[] type
     * @since v10.14.0
     */
    addAll(messages) {
        super.addAll(messages);
    }
}
AdaptMessageService.ɵfac = function AdaptMessageService_Factory(t) { return new (t || AdaptMessageService)(); };
AdaptMessageService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptMessageService, factory: AdaptMessageService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMessageService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
