// @bmc-ux/adapt-angular/esm2015/menu/submenu/submenu.component.js
import { ApplicationRef, Component, ComponentFactoryResolver, ContentChild, ChangeDetectorRef, Directive, ElementRef, HostListener, Inject, Injector, Input, NgZone, QueryList, Renderer2, ViewChild, ViewChildren, ViewContainerRef, forwardRef, HostBinding } from '@angular/core';
import { Directionality } from '@angular/cdk/bidi';
import { ESCAPE, LEFT_ARROW, RIGHT_ARROW } from '@angular/cdk/keycodes';
import { transition, trigger, useAnimation } from '@angular/animations';
import { ANIMATION_NAME } from '../../common/animation';
import { AdaptDeviceDetectionService } from '../../common/device-detection.service';
import { AdaptFocusMonitorDirective } from '../../common/focus-monitor.directive';
import { AdaptFocusMonitorService } from '../../common/focus-monitor.service';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularCustomEventType, AdaptRadarAngularService } from '../../common/radar/adapt-radar';
import { AdaptTreeStructureFilteringService } from '../../common/tree-structure-filter';
import { LayerService } from '../../common/layer-service';
import { PopupService } from '../../common/popup';
import { convertPlacement, positionElements } from '../../common/positioning';
import { debounce, isEventFrom } from '../../common/utilities';
import { getInteractionSource } from '../../common/interaction-source';
import { getEventKeyCode } from '../../common/key-codes';
import { MenuGlobalService } from '../menu-global.service';
import { MenuInnerService } from '../menu.service';
import { SafeCdr } from '../../common/safe-cdr';
import * as i0 from "@angular/core";
import * as i1 from "../menu.service";
import * as i2 from "../menu-global.service";
import * as i3 from "../../common/device-detection.service";
import * as i4 from "../../common/focus-monitor.service";
import * as i5 from "@angular/cdk/bidi";
import * as i6 from "../../common/tree-structure-filter";
import * as i7 from "../../common/radar/adapt-radar";
import * as i8 from "@angular/common";
import * as i9 from "../../focus-holder/focus-holder.directive";
import * as i10 from "../../dragon-input-handler/dragon-input-handler.directive";
import * as i11 from "@angular/forms";
import * as i12 from "../../common/focus-monitor.directive";
import * as i13 from "@angular/router";
import * as i14 from "../../common/layer-service";
const _c0 = ["filterResults"];
const _c1 = function (a0) { return { "adapt-search-clear-visible": a0 }; };
function AdaptSubmenuComponent_ul_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "li", 4);
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵelementStart(4, "div", 5);
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵelementStart(6, "input", 6);
    i0.ɵɵlistener("ngModelChange", function AdaptSubmenuComponent_ul_0_ng_container_2_Template_input_ngModelChange_6_listener($event) { i0.ɵɵrestoreView(_r6); const ctx_r5 = i0.ɵɵnextContext(2); return ctx_r5.filterQuery = $event; })("ngModelChange", function AdaptSubmenuComponent_ul_0_ng_container_2_Template_input_ngModelChange_6_listener() { i0.ɵɵrestoreView(_r6); const ctx_r7 = i0.ɵɵnextContext(2); return ctx_r7.onFilteringDebounce(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n        ");
    i0.ɵɵelementStart(8, "button", 7);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_2_Template_button_click_8_listener() { i0.ɵɵrestoreView(_r6); const ctx_r8 = i0.ɵɵnextContext(2); return ctx_r8.onClearFilter(); });
    i0.ɵɵtext(9, "\n          ");
    i0.ɵɵelementStart(10, "span", 8);
    i0.ɵɵtext(11);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(15, "\n    ");
    i0.ɵɵelementStart(16, "li", 9);
    i0.ɵɵtext(17, "\n      ");
    i0.ɵɵelement(18, "div", 10);
    i0.ɵɵtext(19, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(20, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(6);
    i0.ɵɵproperty("ngModel", ctx_r2.filterQuery)("placeholder", ctx_r2.filteringConfig.placeholder);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(4, _c1, !!ctx_r2.filterQuery.length));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r2.filteringConfig.btnClearText);
} }
const _c2 = function (a0) { return { "m-start-2": a0 }; };
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                    ");
    i0.ɵɵelementStart(2, "button", 22);
    i0.ɵɵtext(3, "\n                    ");
    i0.ɵɵelementStart(4, "span", 23);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n                    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n                  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r23 = i0.ɵɵnextContext();
    const item_r18 = ctx_r23.$implicit;
    const i_r19 = ctx_r23.index;
    const ctx_r20 = i0.ɵɵnextContext(5);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link disabled-item disabled ", item_r18.className, "");
    i0.ɵɵproperty("adaptFocusMonitor", item_r18.label + "_" + item_r18.id || item_r18.label + "_" + i_r19);
    i0.ɵɵattribute("data-testid", "filter_" + ctx_r20.generateSubmenuItemTestID(item_r18));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(7, _c2, item_r18.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r18.label);
} }
const _c3 = function (a0, a1) { return { "active": a0, "selected": a1 }; };
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r31 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 27);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_a_2_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r31); const item_r18 = i0.ɵɵnextContext(2).$implicit; const ctx_r29 = i0.ɵɵnextContext(5); return ctx_r29.handleSubMenuLinkClick(item_r18); });
    i0.ɵɵtext(1, "\n                    ");
    i0.ɵɵelementStart(2, "span", 23);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n                    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r32 = i0.ɵɵnextContext(2);
    const item_r18 = ctx_r32.$implicit;
    const i_r19 = ctx_r32.index;
    const ctx_r24 = i0.ɵɵnextContext(5);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link adapt-submenu__link_filter ", item_r18.className, "");
    i0.ɵɵproperty("adaptFocusMonitor", item_r18.label + "_" + item_r18.id || item_r18.label + "_" + i_r19)("ngClass", i0.ɵɵpureFunction2(10, _c3, item_r18.active, item_r18.selected));
    i0.ɵɵattribute("href", item_r18.link, i0.ɵɵsanitizeUrl)("target", item_r18.target)("data-testid", "filter_" + ctx_r24.generateSubmenuItemTestID(item_r18));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(13, _c2, item_r18.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r18.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_4_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r36 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 29);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_4_a_1_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r36); const item_r18 = i0.ɵɵnextContext(3).$implicit; const ctx_r34 = i0.ɵɵnextContext(5); return ctx_r34.selectMenuItem(item_r18); });
    i0.ɵɵtext(1, "\n                      ");
    i0.ɵɵelementStart(2, "span", 23);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n                      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r37 = i0.ɵɵnextContext(3);
    const item_r18 = ctx_r37.$implicit;
    const i_r19 = ctx_r37.index;
    const ctx_r33 = i0.ɵɵnextContext(5);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link adapt-submenu__link_filter ", item_r18.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", item_r18.routerLink);
    i0.ɵɵproperty("adaptFocusMonitor", item_r18.label + "_" + item_r18.id || item_r18.label + "_" + i_r19)("ngClass", i0.ɵɵpureFunction2(9, _c3, item_r18.active, item_r18.selected));
    i0.ɵɵattribute("data-testid", "filter_" + ctx_r33.generateSubmenuItemTestID(item_r18));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(12, _c2, item_r18.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r18.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n                      ");
    i0.ɵɵtemplate(1, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_4_a_1_Template, 5, 14, "a", 28);
    i0.ɵɵtext(2, "\n                    ");
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r27 = i0.ɵɵreference(8);
    const item_r18 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", item_r18.routerLink)("ngIfElse", _r27);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    const _r41 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n                      ");
    i0.ɵɵelementStart(1, "button", 30);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_7_Template_button_click_1_listener($event) { i0.ɵɵrestoreView(_r41); const item_r18 = i0.ɵɵnextContext(2).$implicit; const ctx_r39 = i0.ɵɵnextContext(5); return ctx_r39.onActionClick($event, item_r18); });
    i0.ɵɵtext(2, "\n                      ");
    i0.ɵɵelementStart(3, "span", 23);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n                      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n                    ");
} if (rf & 2) {
    const ctx_r42 = i0.ɵɵnextContext(2);
    const item_r18 = ctx_r42.$implicit;
    const i_r19 = ctx_r42.index;
    const ctx_r28 = i0.ɵɵnextContext(5);
    i0.ɵɵadvance(1);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link adapt-submenu__link_filter btn btn-link ", item_r18.className, "");
    i0.ɵɵproperty("adaptFocusMonitor", item_r18.label + "_" + item_r18.id || item_r18.label + "_" + i_r19)("ngClass", i0.ɵɵpureFunction2(8, _c3, item_r18.active, item_r18.selected));
    i0.ɵɵattribute("data-testid", "filter_" + ctx_r28.generateSubmenuItemTestID(item_r18));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(11, _c2, item_r18.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r18.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                    ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_a_2_Template, 5, 15, "a", 24);
    i0.ɵɵtext(3, "\n\n                    ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_4_Template, 3, 2, "ng-template", null, 25, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(6, "\n\n                    ");
    i0.ɵɵtemplate(7, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_ng_template_7_Template, 7, 13, "ng-template", null, 26, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(9, "\n                  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r25 = i0.ɵɵreference(5);
    const item_r18 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r18.link)("ngIfElse", _r25);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ul_10_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
const _c4 = function (a0) { return { $implicit: a0 }; };
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ul_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "ul", 31);
    i0.ɵɵtext(1, "\n                    ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ul_10_ng_container_2_Template, 1, 0, "ng-container", 17);
    i0.ɵɵtext(3, "\n                  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r18 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵnextContext(2);
    const _r13 = i0.ɵɵreference(8);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", _r13)("ngTemplateOutletContext", i0.ɵɵpureFunction1(2, _c4, item_r18.items));
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "li", 19);
    i0.ɵɵtext(3, "\n                ");
    i0.ɵɵelementStart(4, "div", 20);
    i0.ɵɵtext(5, "\n                  ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_6_Template, 8, 9, "ng-container", 3);
    i0.ɵɵtext(7, "\n\n                  ");
    i0.ɵɵtemplate(8, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ng_container_8_Template, 10, 2, "ng-container", 3);
    i0.ɵɵtext(9, "\n                  ");
    i0.ɵɵtemplate(10, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_ul_10_Template, 4, 4, "ul", 21);
    i0.ɵɵtext(11, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r18 = ctx.$implicit;
    const i_r19 = ctx.index;
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("a-menu__item ", item_r18.listClassName, "");
    i0.ɵɵproperty("id", item_r18.label + "_" + item_r18.id || item_r18.label + "_" + i_r19);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", item_r18.disabled && !item_r18.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r18.disabled && !item_r18.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r18.items && item_r18.items.length > 0);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n            ");
    i0.ɵɵtemplate(1, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_ng_container_1_Template, 14, 7, "ng-container", 18);
    i0.ɵɵtext(2, "\n          ");
} if (rf & 2) {
    const nodes_r16 = ctx.$implicit;
    const ctx_r14 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngForOf", nodes_r16)("ngForTrackBy", ctx_r14.trackByFn);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_container_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "li", 9);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵelementStart(4, "ul", 14, 15);
    i0.ɵɵtext(6, "\n          ");
    i0.ɵɵtemplate(7, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_template_7_Template, 3, 2, "ng-template", null, 16, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(9, "\n          ");
    i0.ɵɵtemplate(10, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_ng_container_10_Template, 1, 0, "ng-container", 17);
    i0.ɵɵtext(11, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r13 = i0.ɵɵreference(8);
    const ctx_r9 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(10);
    i0.ɵɵproperty("ngTemplateOutlet", _r13)("ngTemplateOutletContext", i0.ɵɵpureFunction1(2, _c4, ctx_r9.filteredNodes));
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n      ");
    i0.ɵɵelementStart(1, "li", 32);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n    ");
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r11.filteringConfig.notFoundText);
} }
function AdaptSubmenuComponent_ul_0_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "li", 11);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵtemplate(5, AdaptSubmenuComponent_ul_0_ng_container_4_ng_container_5_Template, 14, 4, "ng-container", 12);
    i0.ɵɵtext(6, "\n\n    ");
    i0.ɵɵtemplate(7, AdaptSubmenuComponent_ul_0_ng_container_4_ng_template_7_Template, 4, 1, "ng-template", null, 13, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(9, "\n\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r10 = i0.ɵɵreference(8);
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r3.filteringConfig.resultsTitle);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r3.resultsFound)("ngIfElse", _r10);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_li_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 11);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r46 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r46.subMenuHeader);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_6_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "div", 38);
    i0.ɵɵtext(3, "\n                ");
    i0.ɵɵelementStart(4, "button", 39);
    i0.ɵɵtext(5, "\n                ");
    i0.ɵɵelementStart(6, "span", 23);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n                  ");
    i0.ɵɵelement(9, "span", 40);
    i0.ɵɵtext(10, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r53 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("node", item_r48)("theme", ctx_r53.theme)("makeItemActive", ctx_r53.makeItemActive);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link disabled-item disabled ", item_r48.className, "");
    i0.ɵɵattribute("data-testid", ctx_r53.generateSubmenuItemTestID(item_r48));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(9, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_6_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "button", 41);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵelementStart(4, "span", 23);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r54 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link disabled-item disabled ", item_r48.className, "");
    i0.ɵɵattribute("data-testid", ctx_r54.generateSubmenuItemTestID(item_r48));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(6, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_6_ng_container_2_Template, 13, 11, "ng-container", 3);
    i0.ɵɵtext(3, "\n\n            ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_6_ng_container_4_Template, 8, 8, "ng-container", 3);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r48.items && !item_r48.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r48.items && !item_r48.hide);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r65 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 45);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_a_2_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r65); const item_r48 = i0.ɵɵnextContext(3).$implicit; const ctx_r63 = i0.ɵɵnextContext(3); return item_r48.closeOnClick ? ctx_r63.closeMenu() : null; });
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "span", 23);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n              ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r60 = i0.ɵɵnextContext(3);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r48.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c3, item_r48.active, item_r48.selected));
    i0.ɵɵattribute("data-testid", ctx_r60.generateSubmenuItemTestID(item_r48))("href", item_r48.link, i0.ɵɵsanitizeUrl)("target", item_r48.target);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(12, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r70 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 46, 47);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_a_4_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r70); const item_r48 = i0.ɵɵnextContext(3).$implicit; const ctx_r68 = i0.ɵɵnextContext(3); return ctx_r68.selectMenuItem(item_r48); });
    i0.ɵɵtext(2, "\n              ");
    i0.ɵɵelementStart(3, "span", 23);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n              ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const _r67 = i0.ɵɵreference(1);
    const item_r48 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r61 = i0.ɵɵnextContext(3);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r48.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", item_r48.routerLink);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c3, item_r48.active, item_r48.selected));
    i0.ɵɵattribute("data-testid", ctx_r61.generateSubmenuItemTestID(item_r48))("aria-current", _r67.isActive ? "page" : undefined);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(12, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r74 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 48);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r74); const item_r48 = i0.ɵɵnextContext(3).$implicit; const ctx_r72 = i0.ɵɵnextContext(3); return ctx_r72.onActionClick($event, item_r48); });
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "span", 23);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n              ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r62 = i0.ɵɵnextContext(3);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link ", item_r48.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(7, _c3, item_r48.active, item_r48.selected));
    i0.ɵɵattribute("data-testid", ctx_r62.generateSubmenuItemTestID(item_r48));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(10, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_a_2_Template, 5, 14, "a", 42);
    i0.ɵɵtext(3, "\n\n              ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_a_4_Template, 6, 14, "a", 43);
    i0.ɵɵtext(5, "\n\n              ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_button_6_Template, 5, 12, "button", 44);
    i0.ɵɵtext(7, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r48.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r48.routerLink);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r48.action);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r84 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 50);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_a_4_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r84); const item_r48 = i0.ɵɵnextContext(3).$implicit; const ctx_r82 = i0.ɵɵnextContext(3); return item_r48.closeOnClick ? ctx_r82.closeMenu() : null; });
    i0.ɵɵtext(1, "\n                ");
    i0.ɵɵelementStart(2, "span", 23);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n                  ");
    i0.ɵɵelement(5, "span", 40);
    i0.ɵɵtext(6, "\n                ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r77 = i0.ɵɵnextContext(3);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r48.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c3, item_r48.active, item_r48.selected));
    i0.ɵɵattribute("href", item_r48.link, i0.ɵɵsanitizeUrl)("target", item_r48.target)("data-testid", ctx_r77.generateSubmenuItemTestID(item_r48));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(12, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_6_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r89 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 52);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_6_a_1_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r89); const item_r48 = i0.ɵɵnextContext(4).$implicit; const ctx_r87 = i0.ɵɵnextContext(3); return ctx_r87.selectMenuItem(item_r48); });
    i0.ɵɵtext(1, "\n                  ");
    i0.ɵɵelementStart(2, "span", 23);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n                    ");
    i0.ɵɵelement(5, "span", 40);
    i0.ɵɵtext(6, "\n                  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(4).$implicit;
    const ctx_r86 = i0.ɵɵnextContext(3);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r48.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", item_r48.routerLink);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(8, _c3, item_r48.active, item_r48.selected));
    i0.ɵɵattribute("data-testid", ctx_r86.generateSubmenuItemTestID(item_r48));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(11, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n                  ");
    i0.ɵɵtemplate(1, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_6_a_1_Template, 7, 13, "a", 51);
    i0.ɵɵtext(2, "\n                ");
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r80 = i0.ɵɵreference(10);
    const item_r48 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", item_r48.routerLink)("ngIfElse", _r80);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    const _r94 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n                  ");
    i0.ɵɵelementStart(1, "button", 53);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_9_Template_button_click_1_listener($event) { i0.ɵɵrestoreView(_r94); const item_r48 = i0.ɵɵnextContext(3).$implicit; const ctx_r92 = i0.ɵɵnextContext(3); return ctx_r92.onActionClick($event, item_r48); });
    i0.ɵɵtext(2, "\n                    ");
    i0.ɵɵelementStart(3, "span", 23);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n\n                    ");
    i0.ɵɵelement(6, "span", 40);
    i0.ɵɵtext(7, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n                ");
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r81 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link ", item_r48.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(7, _c3, item_r48.active, item_r48.selected));
    i0.ɵɵattribute("data-testid", ctx_r81.generateSubmenuItemTestID(item_r48));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(10, _c2, item_r48.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r48.label);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "div", 38);
    i0.ɵɵtext(3, "\n                ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_a_4_Template, 7, 14, "a", 49);
    i0.ɵɵtext(5, "\n\n                ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_6_Template, 3, 2, "ng-template", null, 25, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(8, "\n\n                ");
    i0.ɵɵtemplate(9, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_ng_template_9_Template, 9, 12, "ng-template", null, 26, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(11, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r78 = i0.ɵɵreference(7);
    const item_r48 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r59 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("node", item_r48)("theme", ctx_r59.theme)("makeItemActive", ctx_r59.makeItemActive);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r48.link)("ngIfElse", _r78);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_2_Template, 8, 3, "ng-container", 3);
    i0.ɵɵtext(3, "\n\n            ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_ng_container_4_Template, 13, 5, "ng-container", 3);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r48 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r48.items && !item_r48.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r48.items && !item_r48.hide);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_li_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 9);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelement(2, "div", 10);
    i0.ɵɵtext(3, "\n      ");
    i0.ɵɵelementEnd();
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "li", 35);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵelementStart(4, "div", 36);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_Template_div_click_4_listener($event) { return $event.stopPropagation(); });
    i0.ɵɵtext(5, "\n\n          ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_6_Template, 6, 2, "ng-container", 3);
    i0.ɵɵtext(7, "\n\n          ");
    i0.ɵɵtemplate(8, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_ng_container_8_Template, 6, 2, "ng-container", 3);
    i0.ɵɵtext(9, "\n\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n      ");
    i0.ɵɵtemplate(12, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_li_12_Template, 4, 0, "li", 37);
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r48 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("a-menu__item ", item_r48.listClassName, "");
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", item_r48.disabled);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r48.disabled);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", item_r48.separator);
} }
function AdaptSubmenuComponent_ul_0_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n    ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_6_li_2_Template, 2, 1, "li", 33);
    i0.ɵɵtext(3, "\n\n    ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_6_ng_container_4_Template, 14, 6, "ng-container", 34);
    i0.ɵɵtext(5, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r4.subMenuHeader);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r4.nodes);
} }
function AdaptSubmenuComponent_ul_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "ul", 2);
    i0.ɵɵtext(1, "\n\n  ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_0_ng_container_2_Template, 21, 6, "ng-container", 3);
    i0.ɵɵtext(3, "\n\n  ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_0_ng_container_4_Template, 10, 3, "ng-container", 3);
    i0.ɵɵtext(5, "\n\n  ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_0_ng_container_6_Template, 6, 2, "ng-container", 3);
    i0.ɵɵtext(7, "\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngClass", ctx_r0.getClass())("adaptFocusHolder", true)("focusFirstOnInit", true)("eventType", ctx_r0.eventType)("restoreLastFocusedElement", true);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0.filteringConfig && ctx_r0.filteringConfig.enabled && ctx_r0.level === 0);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0.filteringConfig && ctx_r0.filteringConfig.enabled && ctx_r0.filterQuery.length);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r0.filterQuery.length);
} }
const _c5 = function (a0) { return [a0]; };
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelementStart(2, "div", 38);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵelementStart(4, "div", 57);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_ng_container_2_Template_div_click_4_listener($event) { return $event.stopPropagation(); });
    i0.ɵɵtext(5, "\n                ");
    i0.ɵɵelementStart(6, "button", 58);
    i0.ɵɵtext(7, "\n                  ");
    i0.ɵɵelementStart(8, "span", 23);
    i0.ɵɵtext(9);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n                ");
    i0.ɵɵelement(12, "button", 59);
    i0.ɵɵtext(13, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n            ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(15, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r106 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("node", item_r100)("theme", ctx_r106.theme)("makeItemActive", ctx_r106.makeItemActive);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link disabled-item disabled  ", item_r100.className, "");
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("data-testid", ctx_r106.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(10, _c2, item_r100.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r100.label);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(12, _c5, ctx_r106.isRtl ? "d-icon-angle_left" : "d-icon-angle_right"));
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelementStart(2, "button", 60);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵelementStart(4, "span", 23);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n            ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r107 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link disabled-item disabled ", item_r100.className, "");
    i0.ɵɵattribute("data-testid", ctx_r107.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(6, _c2, item_r100.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r100.label);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n          ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_ng_container_2_Template, 16, 14, "ng-container", 3);
    i0.ɵɵtext(3, "\n\n          ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_ng_container_4_Template, 8, 8, "ng-container", 3);
    i0.ɵɵtext(5, "\n        ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r100.items);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r100.items);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r119 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 64);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_a_2_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r119); const item_r100 = i0.ɵɵnextContext(4).$implicit; const ctx_r117 = i0.ɵɵnextContext(2); return item_r100.closeOnClick ? ctx_r117.closeMenu() : null; });
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "span", 65);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(4).$implicit;
    const ctx_r114 = i0.ɵɵnextContext(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r100.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(8, _c3, item_r100.active, item_r100.selected));
    i0.ɵɵattribute("href", item_r100.link, i0.ɵɵsanitizeUrl)("target", item_r100.target)("data-testid", ctx_r114.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(item_r100.label);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r123 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 66);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_a_4_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r123); const item_r100 = i0.ɵɵnextContext(4).$implicit; const ctx_r121 = i0.ɵɵnextContext(2); return ctx_r121.selectMenuItem(item_r100); });
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "span", 65);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(4).$implicit;
    const ctx_r115 = i0.ɵɵnextContext(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r100.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", item_r100.routerLink);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(7, _c3, item_r100.active, item_r100.selected));
    i0.ɵɵattribute("data-testid", ctx_r115.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(item_r100.label);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r127 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 67);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r127); const item_r100 = i0.ɵɵnextContext(4).$implicit; const ctx_r125 = i0.ɵɵnextContext(2); return ctx_r125.onActionClick($event, item_r100); });
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "span", 65);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(4).$implicit;
    const ctx_r116 = i0.ɵɵnextContext(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link btn btn-link ", item_r100.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(6, _c3, item_r100.active, item_r100.selected));
    i0.ɵɵattribute("data-testid", ctx_r116.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(item_r100.label);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_a_2_Template, 5, 11, "a", 61);
    i0.ɵɵtext(3, "\n\n            ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_a_4_Template, 5, 10, "a", 62);
    i0.ɵɵtext(5, "\n\n            ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_button_6_Template, 5, 9, "button", 63);
    i0.ɵɵtext(7, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(3).$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r100.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r100.routerLink);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r100.action);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 69);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_div_4_Template_div_click_0_listener($event) { return $event.stopPropagation(); });
    i0.ɵɵtext(1, "\n                ");
    i0.ɵɵelementStart(2, "a");
    i0.ɵɵtext(3, "\n                  ");
    i0.ɵɵelementStart(4, "span", 23);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n                ");
    i0.ɵɵelement(8, "button", 59);
    i0.ɵɵtext(9, "\n              ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(4).$implicit;
    const ctx_r130 = i0.ɵɵnextContext(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r100.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(10, _c3, item_r100.active, item_r100.selected));
    i0.ɵɵattribute("data-testid", ctx_r130.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("href", item_r100.link, i0.ɵɵsanitizeUrl)("target", item_r100.target);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(13, _c2, item_r100.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r100.label);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(15, _c5, ctx_r130.isRtl ? "d-icon-angle_left" : "d-icon-angle_right"));
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_6_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r141 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 69);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_6_div_1_Template_div_click_0_listener($event) { return $event.stopPropagation(); });
    i0.ɵɵtext(1, "\n                  ");
    i0.ɵɵelementStart(2, "a", 70);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_6_div_1_Template_a_click_2_listener($event) { i0.ɵɵrestoreView(_r141); const item_r100 = i0.ɵɵnextContext(5).$implicit; const ctx_r139 = i0.ɵɵnextContext(2); $event.stopPropagation(); return ctx_r139.selectMenuItem(item_r100); });
    i0.ɵɵtext(3, "\n                    ");
    i0.ɵɵelementStart(4, "span", 23);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n                  ");
    i0.ɵɵelement(8, "button", 59);
    i0.ɵɵtext(9, "\n                ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(5).$implicit;
    const ctx_r137 = i0.ɵɵnextContext(2);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r100.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c3, item_r100.active, item_r100.selected));
    i0.ɵɵattribute("data-testid", ctx_r137.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(2);
    i0.ɵɵpropertyInterpolate("routerLink", item_r100.routerLink);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(12, _c2, item_r100.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r100.label);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(14, _c5, ctx_r137.isRtl ? "d-icon-angle_left" : "d-icon-angle_right"));
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n                ");
    i0.ɵɵtemplate(1, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_6_div_1_Template, 10, 16, "div", 68);
    i0.ɵɵtext(2, "\n              ");
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r133 = i0.ɵɵreference(10);
    const item_r100 = i0.ɵɵnextContext(3).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", item_r100.routerLink)("ngIfElse", _r133);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_9_Template(rf, ctx) { if (rf & 1) {
    const _r147 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n                ");
    i0.ɵɵelementStart(1, "div", 69);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_9_Template_div_click_1_listener($event) { return $event.stopPropagation(); });
    i0.ɵɵtext(2, "\n                  ");
    i0.ɵɵelementStart(3, "button", 71);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_9_Template_button_click_3_listener($event) { i0.ɵɵrestoreView(_r147); const item_r100 = i0.ɵɵnextContext(4).$implicit; const ctx_r145 = i0.ɵɵnextContext(2); $event.stopPropagation(); return ctx_r145.onActionClick($event, item_r100); });
    i0.ɵɵtext(4, "\n                    ");
    i0.ɵɵelementStart(5, "span", 23);
    i0.ɵɵtext(6);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n                  ");
    i0.ɵɵelement(9, "button", 59);
    i0.ɵɵtext(10, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n              ");
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(4).$implicit;
    const ctx_r134 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵclassMapInterpolate1("adapt-submenu__link ", item_r100.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(8, _c3, item_r100.active, item_r100.selected));
    i0.ɵɵattribute("data-testid", ctx_r134.generateSubmenuItemTestID(item_r100));
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(11, _c2, item_r100.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r100.label);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(13, _c5, ctx_r134.isRtl ? "d-icon-angle_left" : "d-icon-angle_right"));
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelementStart(2, "div", 38);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_div_4_Template, 10, 17, "div", 68);
    i0.ɵɵtext(5, "\n\n              ");
    i0.ɵɵtemplate(6, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_6_Template, 3, 2, "ng-template", null, 25, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(8, "\n\n              ");
    i0.ɵɵtemplate(9, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_ng_template_9_Template, 12, 15, "ng-template", null, 26, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(11, "\n            ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r131 = i0.ɵɵreference(7);
    const item_r100 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r113 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("node", item_r100)("theme", ctx_r113.theme)("makeItemActive", ctx_r113.makeItemActive);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r100.link)("ngIfElse", _r131);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n          ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_2_Template, 8, 3, "ng-container", 3);
    i0.ɵɵtext(3, "\n\n          ");
    i0.ɵɵtext(4, "\n          ");
    i0.ɵɵtemplate(5, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_ng_container_5_Template, 13, 5, "ng-container", 3);
    i0.ɵɵtext(6, "\n        ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r100.items);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", item_r100.items);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "div", 36);
    i0.ɵɵlistener("click", function AdaptSubmenuComponent_ul_2_ng_container_2_li_2_Template_div_click_2_listener($event) { return $event.stopPropagation(); });
    i0.ɵɵtext(3, "\n\n        ");
    i0.ɵɵtext(4, "\n        ");
    i0.ɵɵtemplate(5, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_5_Template, 6, 2, "ng-container", 3);
    i0.ɵɵtext(6, "\n\n        ");
    i0.ɵɵtext(7, "\n        ");
    i0.ɵɵtemplate(8, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_ng_container_8_Template, 7, 2, "ng-container", 3);
    i0.ɵɵtext(9, "\n\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r100 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-menu__item ", item_r100.listClassName, "");
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngIf", item_r100.disabled);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !item_r100.disabled);
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_li_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li", 72);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelement(2, "div", 10);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementEnd();
} }
function AdaptSubmenuComponent_ul_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n    ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_2_ng_container_2_li_2_Template, 11, 5, "li", 55);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵtemplate(4, AdaptSubmenuComponent_ul_2_ng_container_2_li_4_Template, 4, 0, "li", 56);
    i0.ɵɵtext(5, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r100 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r100.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r100.separator);
} }
function AdaptSubmenuComponent_ul_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "ul", 54);
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_2_ng_container_2_Template, 6, 2, "ng-container", 34);
    i0.ɵɵtext(3, "\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r1.nodes);
} }
const _c6 = ["contentPlacement"];
function AdaptMenuDirective_ng_template_3_Template(rf, ctx) { }
const _c7 = [[["", "adaptMenuAnchor", ""]]];
const _c8 = ["[adaptMenuAnchor]"];
let _submenuItemTestIdIndex = 1;
export class AdaptSubmenuComponent {
    constructor(_menuInnerService, _menuGlobalService, _elemRef, _renderer, _ngZone, deviceDetection, _focusHandle, _dir, _cd, _adaptTreeStructureFilteringService, _adaptRadarService) {
        this._menuInnerService = _menuInnerService;
        this._menuGlobalService = _menuGlobalService;
        this._elemRef = _elemRef;
        this._renderer = _renderer;
        this._ngZone = _ngZone;
        this.deviceDetection = deviceDetection;
        this._focusHandle = _focusHandle;
        this._dir = _dir;
        this._cd = _cd;
        this._adaptTreeStructureFilteringService = _adaptTreeStructureFilteringService;
        this._adaptRadarService = _adaptRadarService;
        this.isRtl = false;
        this.filterQuery = '';
        this.resultsFound = true;
        this.onFilteringDebounce = debounce(this.onFiltering.bind(this), 100);
        this.eventType = null;
        this.defaultHostClass = true;
        this.menuAnimation = true;
        this._shiftTop = 5;
        this._placementRootMenu = ['bottom-left', 'bottom-right', 'bottom-left'];
        this._placementSubmenu = ['right-top', 'left-top', 'right-top'];
        this.isRtl = this._dir.value === 'rtl';
    }
    get darkThemeClass() {
        return this.theme === 'dark';
    }
    get whiteThemeClass() {
        return this.theme === 'white';
    }
    get greyThemeClass() {
        return this.theme === 'grey';
    }
    get customThemeClass() {
        return this.theme === 'custom';
    }
    get isAnimationDisabled() {
        return this.level !== 0;
    }
    ngOnInit() {
        this.id = `adapt-submenu-${this.level}`;
        if (this.isRtl) {
            this._placementRootMenu = convertPlacement(this._placementRootMenu);
            this._placementSubmenu = convertPlacement(this._placementSubmenu);
        }
        if (this.isMobile) {
            this._menuInnerService.menuElementRef.next(this._elemRef);
        }
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    trackByFn(index, item) {
        return item.id || index;
    }
    ngAfterViewInit() {
        if (!this.isMobile && this._menuInnerService.appendToBody) {
            document.body.appendChild(this._elemRef.nativeElement);
        }
        if (this.focusMonitoring) {
            this._focusHandle.init(this.focusMonitoring);
        }
        if (this.anchor && !this.isMobile) {
            this._applyPosition();
        }
    }
    handleKeyDown(event) {
        if (event.currentTarget !== this._menuInnerService.currentActiveCascadeItem) {
            return;
        }
        const keyCode = getEventKeyCode(event);
        if (keyCode === ESCAPE) {
            this._menuInnerService.detachCascade();
        }
        if (keyCode === (this.isRtl ? RIGHT_ARROW : LEFT_ARROW)) {
            this.anchor.toggleOpen(event);
            return;
        }
        if (keyCode === (this.isRtl ? LEFT_ARROW : RIGHT_ARROW)) {
            return;
        }
        this._focusHandle.onKeyDown(event);
    }
    handleSubMenuLinkClick(item) {
        var _a;
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptMenu, {
            EventType: AdaptRadarAngularCustomEventType.Click,
            ParentComponent: (_a = this.analyticsMetadata) === null || _a === void 0 ? void 0 : _a.parentComponentName,
            Name: item.label,
            Disabled: item.disabled,
            Type: 'link',
            SiblingsOptionsCount: this._getSiblingsCount(),
            HiddenSiblingsOptionsCount: this._getHiddenSiblingsCount()
        });
        if (item.closeOnClick) {
            this.closeMenu();
        }
    }
    onActionClick($event, item) {
        this.selectMenuItem(item);
        if (item.action) {
            item.action($event, item.id);
        }
        if (item.closeOnClick) {
            this.closeMenu();
        }
    }
    selectMenuItem(item) {
        var _a;
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptMenu, {
            EventType: AdaptRadarAngularCustomEventType.Click,
            ParentComponent: (_a = this.analyticsMetadata) === null || _a === void 0 ? void 0 : _a.parentComponentName,
            Name: item.label,
            Disabled: item.disabled,
            Type: item.routerLink ? 'route' : 'action',
            SiblingsOptionsCount: this._getSiblingsCount(),
            HiddenSiblingsOptionsCount: this._getHiddenSiblingsCount()
        });
        if (item.disabled) {
            return;
        }
        else {
            if (this.makeItemActive) {
                this._menuGlobalService.selectMenuItem(item);
            }
            if (item.closeOnClick !== false) {
                this._menuInnerService.detachCascade();
            }
        }
    }
    getClass() {
        if (this.isMobile) {
            return;
        }
        if (this.level === 0) {
            return `a-dropdown-menu a-dropdown-menu__${this.theme}`;
        }
    }
    onFiltering() {
        this.filteredNodes = this._adaptTreeStructureFilteringService.filter(this.nodes, this.filterQuery, 'label');
        this.resultsFound = Boolean(this.filteredNodes && this.filteredNodes.length);
        // The main goal of this peace of code is to re-order QueryList of focusable items from Angular ListKeyManager
        // to have possibility to navigate by order from DOM
        // In case of tree structure it means that items will be navigated one by one no meter of nesting(like flat structure)
        this._ngZone.runOutsideAngular(() => {
            setTimeout(() => {
                if (this.filterResults && this.filterResults.nativeElement) {
                    const focusMonitoringArr = this.focusMonitoring.toArray();
                    const filteredItems = this.filterResults.nativeElement.querySelectorAll('li');
                    if (filteredItems && filteredItems.length) {
                        const filteredItemsArr = Array.prototype.slice.call(filteredItems);
                        const orderedListOfIds = filteredItemsArr.map(item => item.id);
                        focusMonitoringArr.sort((a, b) => orderedListOfIds.indexOf(a.optionId) - orderedListOfIds.indexOf(b.optionId));
                        this.focusMonitoring.reset(focusMonitoringArr);
                    }
                }
            }, 100);
        });
        SafeCdr.detectChanges(this._cd);
    }
    onClearFilter() {
        this.filterQuery = '';
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    isEventFrom(event) {
        return isEventFrom(event, this._elemRef.nativeElement);
    }
    closeMenu() {
        if (!this.isMobile) {
            this._menuGlobalService.closeMenu();
        }
        else {
            this._menuInnerService.detachCascade();
        }
    }
    _applyPosition() {
        positionElements(this.anchor._elemRef.nativeElement, this._elemRef.nativeElement, this.firstInit ? this._placementRootMenu : this._placementSubmenu, this._menuInnerService.appendToBody);
        if (!this.firstInit) {
            const top = getComputedStyle(this._elemRef.nativeElement).top;
            this._renderer.setStyle(this._elemRef.nativeElement, 'top', `${parseInt(top, 10) - this._shiftTop}px`);
        }
    }
    _getSiblingsCount() {
        return this.nodes ? this.nodes.filter(menuItem => !menuItem.hide).length : 0;
    }
    _getHiddenSiblingsCount() {
        return this.nodes ? this.nodes.filter(menuItem => menuItem.hide).length : 0;
    }
    generateSubmenuItemTestID(item) {
        if (!item.testID) {
            item.testID = `submenu-item-test-id-${_submenuItemTestIdIndex++}`;
        }
        return item.testID;
    }
}
AdaptSubmenuComponent.ɵfac = function AdaptSubmenuComponent_Factory(t) { return new (t || AdaptSubmenuComponent)(i0.ɵɵdirectiveInject(i1.MenuInnerService), i0.ɵɵdirectiveInject(i2.MenuGlobalService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i3.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i4.AdaptFocusMonitorService), i0.ɵɵdirectiveInject(i5.Directionality), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i6.AdaptTreeStructureFilteringService), i0.ɵɵdirectiveInject(i7.AdaptRadarAngularService)); };
AdaptSubmenuComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptSubmenuComponent, selectors: [["adapt-submenu"]], viewQuery: function AdaptSubmenuComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
        i0.ɵɵviewQuery(AdaptFocusMonitorDirective, 5, AdaptFocusMonitorDirective);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.filterResults = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.focusMonitoring = _t);
    } }, hostVars: 13, hostBindings: function AdaptSubmenuComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("keydown", function AdaptSubmenuComponent_keydown_HostBindingHandler($event) { return ctx.handleKeyDown($event); });
    } if (rf & 2) {
        i0.ɵɵhostProperty("id", ctx.id);
        i0.ɵɵsyntheticHostProperty("@.disabled", ctx.isAnimationDisabled)("@menuAnimation", ctx.menuAnimation);
        i0.ɵɵclassProp("adapt-submenu", ctx.defaultHostClass)("dark", ctx.darkThemeClass)("white", ctx.whiteThemeClass)("grey", ctx.greyThemeClass)("custom", ctx.customThemeClass);
    } }, features: [i0.ɵɵProvidersFeature([AdaptFocusMonitorService, AdaptTreeStructureFilteringService])], decls: 4, vars: 2, consts: [["class", "a-menu", "role", "listbox", 3, "ngClass", "adaptFocusHolder", "focusFirstOnInit", "eventType", "restoreLastFocusedElement", 4, "ngIf"], ["class", "a-menu", 4, "ngIf"], ["role", "listbox", 1, "a-menu", 3, "ngClass", "adaptFocusHolder", "focusFirstOnInit", "eventType", "restoreLastFocusedElement"], [4, "ngIf"], ["role", "option", 1, "a-dropdown__item", "a-menu__item", "a-filter__wrapper"], [1, "d-icon-search"], ["type", "text", "adaptFocusMonitor", "", 1, "form-control", "a-menu__filter", 3, "ngModel", "placeholder", "ngModelChange"], ["type", "button", 1, "adapt-search-clear", "d-icon-cross_adapt", "a-filter-clear", 3, "ngClass", "click"], [1, "sr-only"], [1, "a-menu__item"], [1, "dropdown-divider", "w-100"], [1, "a-menu__item", "dropdown-header"], [4, "ngIf", "ngIfElse"], ["noResultsFound", ""], [1, "a-menu__filter-results"], ["filterResults", ""], ["recursiveList", ""], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "id"], [1, "adapt-submenu-item"], ["class", "adapt-submenu__list_wrapper", 4, "ngIf"], ["type", "button", 3, "adaptFocusMonitor"], [1, "a-menu__text", 3, "ngClass"], [3, "adaptFocusMonitor", "class", "ngClass", "click", 4, "ngIf", "ngIfElse"], ["routerLink", ""], ["action", ""], [3, "adaptFocusMonitor", "ngClass", "click"], ["routerLinkActive", "selected", 3, "adaptFocusMonitor", "class", "routerLink", "ngClass", "click", 4, "ngIf", "ngIfElse"], ["routerLinkActive", "selected", 3, "adaptFocusMonitor", "routerLink", "ngClass", "click"], ["type", "button", 3, "adaptFocusMonitor", "ngClass", "click"], [1, "adapt-submenu__list_wrapper"], [1, "a-menu__item", "a-menu__item_not-found-text"], ["class", "a-menu__item dropdown-header", 4, "ngIf"], [4, "ngFor", "ngForOf"], ["role", "option"], [1, "adapt-submenu-item", 3, "click"], ["class", "a-menu__item", 4, "ngIf"], ["adaptMenu", "", 3, "node", "theme", "makeItemActive"], ["adaptMenuAnchor", "", "adaptFocusMonitor", "", "type", "button"], [1, "adapt-menu__link-toggle", "m-start-1"], ["type", "button", "adaptFocusMonitor", ""], ["adaptFocusMonitor", "", 3, "class", "ngClass", "click", 4, "ngIf"], ["adaptFocusMonitor", "", "routerLinkActive", "selected", 3, "class", "routerLink", "ngClass", "click", 4, "ngIf"], ["adaptFocusMonitor", "", "type", "button", 3, "class", "ngClass", "click", 4, "ngIf"], ["adaptFocusMonitor", "", 3, "ngClass", "click"], ["adaptFocusMonitor", "", "routerLinkActive", "selected", 3, "routerLink", "ngClass", "click"], ["routerLinkRef", "routerLinkActive"], ["adaptFocusMonitor", "", "type", "button", 3, "ngClass", "click"], ["adaptMenuAnchor", "", "adaptFocusMonitor", "", 3, "class", "ngClass", "click", 4, "ngIf", "ngIfElse"], ["adaptMenuAnchor", "", "adaptFocusMonitor", "", 3, "ngClass", "click"], ["adaptMenuAnchor", "", "adaptFocusMonitor", "", "routerLinkActive", "selected", 3, "class", "routerLink", "ngClass", "click", 4, "ngIf", "ngIfElse"], ["adaptMenuAnchor", "", "adaptFocusMonitor", "", "routerLinkActive", "selected", 3, "routerLink", "ngClass", "click"], ["adaptMenuAnchor", "", "adaptFocusMonitor", "", "type", "button", 3, "ngClass", "click"], [1, "a-menu"], [3, "class", 4, "ngIf"], ["class", "w-100", 4, "ngIf"], ["adaptMenuAnchor", "", 3, "click"], ["type", "button", 1, "btn", "btn-link"], ["type", "button", 1, "btn", "btn-link", 3, "ngClass"], ["type", "button"], [3, "class", "ngClass", "click", 4, "ngIf"], ["routerLinkActive", "selected", 3, "class", "routerLink", "ngClass", "click", 4, "ngIf"], ["type", "button", 3, "class", "ngClass", "click", 4, "ngIf"], [3, "ngClass", "click"], [1, "a-menu__text"], ["routerLinkActive", "selected", 3, "routerLink", "ngClass", "click"], ["type", "button", 3, "ngClass", "click"], ["adaptMenuAnchor", "", 3, "class", "ngClass", "click", 4, "ngIf", "ngIfElse"], ["adaptMenuAnchor", "", 3, "ngClass", "click"], ["routerLinkActive", "selected", 3, "routerLink", "click"], ["type", "button", 1, "btn", "btn-link", 3, "click"], [1, "w-100"]], template: function AdaptSubmenuComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, AdaptSubmenuComponent_ul_0_Template, 8, 8, "ul", 0);
        i0.ɵɵtext(1, "\n\n");
        i0.ɵɵtemplate(2, AdaptSubmenuComponent_ul_2_Template, 4, 1, "ul", 1);
        i0.ɵɵtext(3, "\n");
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", !ctx.isMobile);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.isMobile);
    } }, directives: function () { return [i8.NgIf, i8.NgClass, i9.AdaptFocusHolderDirective, i10.AdaptDragonInputHandlerDirective, i11.DefaultValueAccessor, i12.AdaptFocusMonitorDirective, i11.NgControlStatus, i11.NgModel, i8.NgTemplateOutlet, i8.NgForOf, i13.RouterLinkWithHref, i13.RouterLinkActive, AdaptMenuDirective, AdaptMenuAnchorDirective]; }, encapsulation: 2, data: { animation: [
            trigger('menuAnimation', [
                transition('void => *', [
                    useAnimation(ANIMATION_NAME.popIn)
                ]),
                transition('* => void', [
                    useAnimation(ANIMATION_NAME.popOut)
                ])
            ])
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSubmenuComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-submenu',
                templateUrl: './submenu.component.html',
                animations: [
                    trigger('menuAnimation', [
                        transition('void => *', [
                            useAnimation(ANIMATION_NAME.popIn)
                        ]),
                        transition('* => void', [
                            useAnimation(ANIMATION_NAME.popOut)
                        ])
                    ])
                ],
                providers: [AdaptFocusMonitorService, AdaptTreeStructureFilteringService]
            }]
    }], function () { return [{ type: i1.MenuInnerService }, { type: i2.MenuGlobalService }, { type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i0.NgZone }, { type: i3.AdaptDeviceDetectionService }, { type: i4.AdaptFocusMonitorService }, { type: i5.Directionality }, { type: i0.ChangeDetectorRef }, { type: i6.AdaptTreeStructureFilteringService }, { type: i7.AdaptRadarAngularService }]; }, { id: [{
            type: HostBinding,
            args: ['id']
        }], defaultHostClass: [{
            type: HostBinding,
            args: ['class.adapt-submenu']
        }], darkThemeClass: [{
            type: HostBinding,
            args: ['class.dark']
        }], whiteThemeClass: [{
            type: HostBinding,
            args: ['class.white']
        }], greyThemeClass: [{
            type: HostBinding,
            args: ['class.grey']
        }], customThemeClass: [{
            type: HostBinding,
            args: ['class.custom']
        }], isAnimationDisabled: [{
            type: HostBinding,
            args: ['@.disabled']
        }], menuAnimation: [{
            type: HostBinding,
            args: ['@menuAnimation']
        }], focusMonitoring: [{
            type: ViewChildren,
            args: [AdaptFocusMonitorDirective, { read: AdaptFocusMonitorDirective }]
        }], filterResults: [{
            type: ViewChild,
            args: ['filterResults', { static: true }]
        }], handleKeyDown: [{
            type: HostListener,
            args: ['keydown', ['$event']]
        }] }); })();
export class AdaptMenuAnchorDirective {
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    constructor(menu, _elemRef) {
        this.menu = menu;
        this._elemRef = _elemRef;
        this.defaultHostClass = true;
    }
    mouseEnter(e) {
        if (this.menu.isMobile || this.menu.firstInit || this.menu.node.expanded) {
            return;
        }
        this.menu.eventType = 'mouse';
        this.toggleOpen(e);
    }
    click(e) {
        if (this.menu.firstInit || this.menu.isMobile) {
            this.menu.eventType = getInteractionSource(e);
            this.toggleOpen(e);
        }
    }
    handleKeyDown(event) {
        const keyCode = getEventKeyCode(event);
        if ((keyCode === (this.menu.isRtl ? LEFT_ARROW : RIGHT_ARROW)) && !this.menu.firstInit) {
            this.menu.eventType = 'keyboard';
            this.menu.extendCascade();
        }
    }
    toggleOpen(e) {
        if (e.type === 'keydown') {
            e.stopPropagation();
        }
        this.menu.extendCascade();
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    isEventFrom(event) {
        return isEventFrom(event, this._elemRef.nativeElement);
    }
}
AdaptMenuAnchorDirective.ɵfac = function AdaptMenuAnchorDirective_Factory(t) { return new (t || AdaptMenuAnchorDirective)(i0.ɵɵdirectiveInject(forwardRef(() => AdaptMenuDirective)), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptMenuAnchorDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptMenuAnchorDirective, selectors: [["", "adaptMenuAnchor", ""]], hostVars: 2, hostBindings: function AdaptMenuAnchorDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("mouseenter", function AdaptMenuAnchorDirective_mouseenter_HostBindingHandler($event) { return ctx.mouseEnter($event); })("click", function AdaptMenuAnchorDirective_click_HostBindingHandler($event) { return ctx.click($event); })("keydown", function AdaptMenuAnchorDirective_keydown_HostBindingHandler($event) { return ctx.handleKeyDown($event); });
    } if (rf & 2) {
        i0.ɵɵclassProp("menu-toggle", ctx.defaultHostClass);
    } } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMenuAnchorDirective, [{
        type: Directive,
        args: [{
                selector: '[adaptMenuAnchor]',
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [forwardRef(() => AdaptMenuDirective)]
            }] }, { type: i0.ElementRef }]; }, { defaultHostClass: [{
            type: HostBinding,
            args: ['class.menu-toggle']
        }], mouseEnter: [{
            type: HostListener,
            args: ['mouseenter', ['$event']]
        }], click: [{
            type: HostListener,
            args: ['click', ['$event']]
        }], handleKeyDown: [{
            type: HostListener,
            args: ['keydown', ['$event']]
        }] }); })();
// eslint-disable-next-line @angular-eslint/component-class-suffix
export class AdaptMenuDirective {
    constructor(_componentFactoryResolver, _deviceDetector, _el, _injector, _renderer, _layerService, _menuInnerService, _menuGlobalService, _dir, _applicationRef) {
        this._componentFactoryResolver = _componentFactoryResolver;
        this._deviceDetector = _deviceDetector;
        this._el = _el;
        this._injector = _injector;
        this._renderer = _renderer;
        this._layerService = _layerService;
        this._menuInnerService = _menuInnerService;
        this._menuGlobalService = _menuGlobalService;
        this._dir = _dir;
        this._applicationRef = _applicationRef;
        this.analyticsMetadata = null;
        this.defaultHostClass = true;
        this.isRtl = false;
        this.eventType = null;
    }
    get cascade() {
        return this._menuInnerService.cascadeChain;
    }
    ngOnInit() {
        this.isRtl = this._dir.value === 'rtl';
        this.level = this.getCascadeLength();
        this.firstInit = (this.level === 0);
        this.cascadeId = this._menuInnerService.id;
        this.isMobile = this._menuInnerService.isMobile;
        this._popupWindowService = new PopupService(AdaptSubmenuComponent, this._injector, this.contentPlacement, this._renderer, this._componentFactoryResolver, this._applicationRef);
    }
    ngOnDestroy() {
        if (this.menuWindowRef) {
            this._popupWindowService.close();
            this.menuWindowRef = null;
        }
    }
    getCascadeLength() {
        return this.cascade ? this.cascade.subMenuItems.length : 0;
    }
    extendCascade() {
        this._toggleRootCascadeItem();
        this._toggleNestedCascadeItem();
    }
    _toggleRootCascadeItem() {
        if (!this.firstInit) {
            return;
        }
        if (this.cascade) {
            if (this.node.expanded) {
                this._menuInnerService.detachCascade();
                this.node.expanded = false;
                this._popupWindowService.close();
                this.menuWindowRef = null;
            }
            else {
                this._createCascadeItem();
            }
            return;
        }
        this._createCascadeItem();
    }
    _toggleNestedCascadeItem() {
        if (this.firstInit) {
            return;
        }
        if (!this.node.expanded && this._menuInnerService.cascadeChain.subMenuItems.length > this.level) {
            this._menuInnerService.destroyCascadeItems(this.level - 1);
            this._createCascadeItem();
            return;
        }
        if (!this.node.expanded && this._menuInnerService.cascadeChain.subMenuItems.length === this.level) {
            this._createCascadeItem();
            return;
        }
        this._menuInnerService.destroyCascadeItems(this.level - 1);
        this._popupWindowService.close();
        this.menuWindowRef = null;
    }
    _createCascadeItem() {
        if (this.menuWindowRef) {
            this._popupWindowService.close();
        }
        if (this.level === 0 && this.closePrevious) {
            // Close previously opened menu
            this._menuGlobalService.closeMenu();
        }
        this.menuWindowRef = this._popupWindowService.open();
        this.menuWindowRef.instance.nodes = this.node.items || this.node.subMenu;
        this.menuWindowRef.instance.filteringConfig = this.filteringConfig;
        this.menuWindowRef.instance.subMenuHeader = this.node.header;
        this.menuWindowRef.instance.level = this.level;
        this.menuWindowRef.instance.anchor = this._anchor;
        this.menuWindowRef.instance.firstInit = this.firstInit;
        this.menuWindowRef.instance.theme = this.theme;
        this.menuWindowRef.instance.isMobile = this.isMobile;
        this.menuWindowRef.instance.eventType = this.eventType;
        this.menuWindowRef.instance.makeItemActive = this.makeItemActive;
        this.menuWindowRef.instance.analyticsMetadata = this.analyticsMetadata;
        this.node.expanded = true;
        if (this.node.disabled && this.node.items) {
            this.node.items.forEach(element => {
                element.disabled = true;
            });
        }
        this._menuInnerService.addCascadeItem(this.node, this.menuWindowRef, this._anchor);
        if (this.closePrevious) {
            // Save newly opened menu so it can be closed later if other menu will be opened
            this._menuGlobalService.storeMenu(this.menuWindowRef, this.node);
        }
    }
}
AdaptMenuDirective.ɵfac = function AdaptMenuDirective_Factory(t) { return new (t || AdaptMenuDirective)(i0.ɵɵdirectiveInject(i0.ComponentFactoryResolver), i0.ɵɵdirectiveInject(i3.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.Injector), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i14.LayerService), i0.ɵɵdirectiveInject(i1.MenuInnerService), i0.ɵɵdirectiveInject(i2.MenuGlobalService), i0.ɵɵdirectiveInject(i5.Directionality), i0.ɵɵdirectiveInject(i0.ApplicationRef)); };
AdaptMenuDirective.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptMenuDirective, selectors: [["adapt-menu-renderer"], ["", "adaptMenu", ""]], contentQueries: function AdaptMenuDirective_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, AdaptMenuAnchorDirective, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx._anchor = _t.first);
    } }, viewQuery: function AdaptMenuDirective_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c6, 7, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.contentPlacement = _t.first);
    } }, hostVars: 2, hostBindings: function AdaptMenuDirective_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("submenu", ctx.defaultHostClass);
    } }, inputs: { node: "node", theme: "theme", subMenuHeader: "subMenuHeader", filteringConfig: "filteringConfig", closePrevious: "closePrevious", makeItemActive: "makeItemActive", analyticsMetadata: "analyticsMetadata" }, exportAs: ["adaptMenu"], ngContentSelectors: _c8, decls: 6, vars: 0, consts: [["contentPlacement", ""]], template: function AdaptMenuDirective_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c7);
        i0.ɵɵtext(0, "\n    ");
        i0.ɵɵprojection(1);
        i0.ɵɵtext(2, "\n\n    ");
        i0.ɵɵtemplate(3, AdaptMenuDirective_ng_template_3_Template, 0, 0, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(5, "\n  ");
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMenuDirective, [{
        type: Component,
        args: [{
                selector: 'adapt-menu-renderer, [adaptMenu]',
                exportAs: 'adaptMenu',
                template: `
    <ng-content select="[adaptMenuAnchor]"></ng-content>

    <ng-template #contentPlacement></ng-template>
  `,
            }]
    }], function () { return [{ type: i0.ComponentFactoryResolver }, { type: i3.AdaptDeviceDetectionService }, { type: i0.ElementRef }, { type: i0.Injector }, { type: i0.Renderer2 }, { type: i14.LayerService }, { type: i1.MenuInnerService }, { type: i2.MenuGlobalService }, { type: i5.Directionality }, { type: i0.ApplicationRef }]; }, { node: [{
            type: Input
        }], theme: [{
            type: Input
        }], subMenuHeader: [{
            type: Input
        }], filteringConfig: [{
            type: Input
        }], closePrevious: [{
            type: Input
        }], makeItemActive: [{
            type: Input
        }], analyticsMetadata: [{
            type: Input
        }], defaultHostClass: [{
            type: HostBinding,
            args: ['class.submenu']
        }], contentPlacement: [{
            type: ViewChild,
            args: ['contentPlacement', { read: ViewContainerRef, static: true }]
        }], _anchor: [{
            type: ContentChild,
            args: [AdaptMenuAnchorDirective]
        }] }); })();
