// @bmc-ux/adapt-radar/esm2015/lib/adapt-radar-products.model.js
import { Version } from '@angular/core';
import { isNil } from 'lodash-es';
export var AdaptRadarProductAreaName;
(function (AdaptRadarProductAreaName) {
    AdaptRadarProductAreaName["ITOM"] = "ITOM";
    AdaptRadarProductAreaName["Converge"] = "Converge";
    AdaptRadarProductAreaName["EndUser"] = "End User";
    AdaptRadarProductAreaName["ServiceManagement"] = "Service Management";
    AdaptRadarProductAreaName["Platform"] = "Platform";
    AdaptRadarProductAreaName["ADAPT"] = "ADAPT";
    AdaptRadarProductAreaName["IZOT"] = "IZOT";
})(AdaptRadarProductAreaName || (AdaptRadarProductAreaName = {}));
export var AdaptRadarSubProductNameLevel1;
(function (AdaptRadarSubProductNameLevel1) {
    // Platform
    AdaptRadarSubProductNameLevel1["Launcher"] = "Launcher";
    AdaptRadarSubProductNameLevel1["Manage"] = "Manage";
    AdaptRadarSubProductNameLevel1["Configure"] = "Configure";
    AdaptRadarSubProductNameLevel1["HelixSSOConfiguration"] = "Helix SSO Configuration";
    AdaptRadarSubProductNameLevel1["MultiCloudBroker"] = "Multi-Cloud Broker";
    AdaptRadarSubProductNameLevel1["Integrate"] = "Integrate";
    AdaptRadarSubProductNameLevel1["Design"] = "Design";
    // Service Management
    AdaptRadarSubProductNameLevel1["Telco"] = "Telco";
    AdaptRadarSubProductNameLevel1["HR"] = "HR";
    AdaptRadarSubProductNameLevel1["Finance"] = "Finance";
    AdaptRadarSubProductNameLevel1["Roles"] = "Roles";
    AdaptRadarSubProductNameLevel1["Notifications"] = "Notifications";
    AdaptRadarSubProductNameLevel1["Locations"] = "Locations";
    AdaptRadarSubProductNameLevel1["Reports"] = "Reports";
    AdaptRadarSubProductNameLevel1["Catalog"] = "Catalog";
    AdaptRadarSubProductNameLevel1["Administer"] = "Administer";
    AdaptRadarSubProductNameLevel1["ITSMInsights"] = "ITSM Insights";
    AdaptRadarSubProductNameLevel1["ServiceManagementManage"] = "Manage";
    // End Users
    AdaptRadarSubProductNameLevel1["Chatbot"] = "Chatbot";
    AdaptRadarSubProductNameLevel1["VirtualAssistant"] = "Virtual Assistant";
    AdaptRadarSubProductNameLevel1["DWPEndUser"] = "DWP End User";
    AdaptRadarSubProductNameLevel1["BWFPortal"] = "BWF Portal";
    // Converge
    AdaptRadarSubProductNameLevel1["SharedTicketConsole"] = "Shared Ticket Console";
    AdaptRadarSubProductNameLevel1["KnowledgeManagement"] = "Knowledge Management";
    AdaptRadarSubProductNameLevel1["TimelineCalendar"] = "Timeline Calendar";
    AdaptRadarSubProductNameLevel1["CMDB"] = "CMDB";
    // ITOM
    AdaptRadarSubProductNameLevel1["View"] = "View";
    AdaptRadarSubProductNameLevel1["Model"] = "Model";
    AdaptRadarSubProductNameLevel1["ITOMManage"] = "Manage";
    AdaptRadarSubProductNameLevel1["Workspace"] = "Workspace";
    AdaptRadarSubProductNameLevel1["ITOMAdminister"] = "Administer";
    AdaptRadarSubProductNameLevel1["Logs"] = "Logs";
    AdaptRadarSubProductNameLevel1["ITOMConfigure"] = "Configure";
    AdaptRadarSubProductNameLevel1["Dashboard"] = "Dashboard";
    AdaptRadarSubProductNameLevel1["ITOM"] = "ITOM";
    AdaptRadarSubProductNameLevel1["Authorize"] = "Authorize";
    AdaptRadarSubProductNameLevel1["Monitor"] = "Monitor";
    AdaptRadarSubProductNameLevel1["Policies"] = "Policies";
    // ADAPT
    AdaptRadarSubProductNameLevel1["Documentation"] = "Documentation";
    AdaptRadarSubProductNameLevel1["IconFinder"] = "Icon Finder";
    AdaptRadarSubProductNameLevel1["ADAPTTools"] = "ADAPT Tools";
    // IZOT
    AdaptRadarSubProductNameLevel1["AMIOps"] = "AMI Ops";
    AdaptRadarSubProductNameLevel1["Compuware"] = "Compuware";
    AdaptRadarSubProductNameLevel1["DB2"] = "DB2";
})(AdaptRadarSubProductNameLevel1 || (AdaptRadarSubProductNameLevel1 = {}));
export var AdaptRadarSubProductNameLevel2;
(function (AdaptRadarSubProductNameLevel2) {
    // Platform -> Manage
    AdaptRadarSubProductNameLevel2["Identities"] = "Identities";
    AdaptRadarSubProductNameLevel2["Roles"] = "Roles";
    // Platform -> Configure
    AdaptRadarSubProductNameLevel2["Tenants"] = "Tenants";
    AdaptRadarSubProductNameLevel2["Notifications"] = "Notifications";
    AdaptRadarSubProductNameLevel2["InnovationSuiteConfigurations"] = "Innovation Suite Configurations";
    AdaptRadarSubProductNameLevel2["SharedConfigurations"] = "Shared Configurations";
    // Platform -> Design
    AdaptRadarSubProductNameLevel2["DataModel"] = "Data Model";
    AdaptRadarSubProductNameLevel2["UI"] = "UI";
    AdaptRadarSubProductNameLevel2["RulesDesigner"] = "Rules Designer";
    AdaptRadarSubProductNameLevel2["ProcessDesigner"] = "Process Designer";
    // Service Management -> Catalog
    AdaptRadarSubProductNameLevel2["Services"] = "Services";
    // Service Management -> Administer
    AdaptRadarSubProductNameLevel2["Studio"] = "Studio";
    AdaptRadarSubProductNameLevel2["Features"] = "Features";
    AdaptRadarSubProductNameLevel2["Configure"] = "Configure";
    // Service Management -> Manage
    AdaptRadarSubProductNameLevel2["Incidents"] = "Incidents";
    AdaptRadarSubProductNameLevel2["Change"] = "Change";
    AdaptRadarSubProductNameLevel2["Problems"] = "Problems";
    AdaptRadarSubProductNameLevel2["Assets"] = "Assets";
    // End Users -> Virtual Assistant
    AdaptRadarSubProductNameLevel2["LiveChat"] = "Live Chat";
    // End Users -> BWF Portal
    AdaptRadarSubProductNameLevel2["Telco"] = "Telco";
    AdaptRadarSubProductNameLevel2["HR"] = "HR";
    AdaptRadarSubProductNameLevel2["Finance"] = "Finance";
    // Converge -> Shared Ticket Console
    AdaptRadarSubProductNameLevel2["SharedTicketConsoleITSM"] = "ITSM";
    AdaptRadarSubProductNameLevel2["BWF"] = "BWF";
    // Converge -> Knowledge Management
    AdaptRadarSubProductNameLevel2["HelixKnowledgeManagement"] = "Helix Knowledge Management";
    AdaptRadarSubProductNameLevel2["RemedyKnowledgeManagement"] = "Remedy Knowledge Management";
    AdaptRadarSubProductNameLevel2["BusinessWorkflowsKnowledgeManagement"] = "Business Workflows Knowledge Management";
    // Converge -> Timeline Calendar
    AdaptRadarSubProductNameLevel2["TimelineCalendarITSM"] = "ITSM";
    // Converge -> CMDB
    AdaptRadarSubProductNameLevel2["CMDB_ITSM"] = "ITSM";
    AdaptRadarSubProductNameLevel2["Discovery"] = "Discovery";
    AdaptRadarSubProductNameLevel2["OperationsManagement"] = "Operations Management";
    // ITOM	-> Manage
    AdaptRadarSubProductNameLevel2["EnrichmentSources"] = "Enrichment Sources";
    AdaptRadarSubProductNameLevel2["EnrichmentPolicies"] = "Enrichment Policies";
    AdaptRadarSubProductNameLevel2["AlertPolicies"] = "Alert Policies";
    AdaptRadarSubProductNameLevel2["DataMasking"] = "Data Masking";
    AdaptRadarSubProductNameLevel2["Entities"] = "Entities";
    AdaptRadarSubProductNameLevel2["Blueprints"] = "Blueprints";
    AdaptRadarSubProductNameLevel2["Situations"] = "Situations";
    AdaptRadarSubProductNameLevel2["Policies"] = "Policies";
    AdaptRadarSubProductNameLevel2["Connectors"] = "Connectors";
    // ITOM	-> Logs
    AdaptRadarSubProductNameLevel2["Explore"] = "Explore";
    AdaptRadarSubProductNameLevel2["Collect"] = "Collect";
    AdaptRadarSubProductNameLevel2["Archive"] = "Archive";
    // ITOM	-> View
    AdaptRadarSubProductNameLevel2["Requests"] = "Requests";
    AdaptRadarSubProductNameLevel2["Reports"] = "Reports";
    // ITOM	-> Policies
    AdaptRadarSubProductNameLevel2["Monitoring"] = "Monitoring";
    AdaptRadarSubProductNameLevel2["Blackout"] = "Blackout";
    AdaptRadarSubProductNameLevel2["Alarm"] = "Alarm";
    AdaptRadarSubProductNameLevel2["Even"] = "Even";
    // ADAPT -> ADAPT Tools
    AdaptRadarSubProductNameLevel2["SprintPlanner"] = "Sprint Planner";
})(AdaptRadarSubProductNameLevel2 || (AdaptRadarSubProductNameLevel2 = {}));
export var AdaptRadarSubProductNameLevel3;
(function (AdaptRadarSubProductNameLevel3) {
    // Platform -> Configure -> Innovation Suite Configurations
    AdaptRadarSubProductNameLevel3["DataManagement"] = "Data Management";
    AdaptRadarSubProductNameLevel3["Approval"] = "Approval";
    AdaptRadarSubProductNameLevel3["Assignment"] = "Assignment";
    // Platform -> Configure -> Shared Configurations
    AdaptRadarSubProductNameLevel3["ApprovalCentral"] = "Approval Central";
    AdaptRadarSubProductNameLevel3["Chatbot"] = "Chatbot";
    AdaptRadarSubProductNameLevel3["ChatOps"] = "Chat Ops";
    AdaptRadarSubProductNameLevel3["Foundation"] = "Foundation";
    AdaptRadarSubProductNameLevel3["LiveChatAgentConsole"] = "Live Chat Agent Console";
    // Converge -> Shared Ticket Console -> BWF
    AdaptRadarSubProductNameLevel3["Telco"] = "Telco";
    AdaptRadarSubProductNameLevel3["HR"] = "HR";
    AdaptRadarSubProductNameLevel3["Finance"] = "Finance";
})(AdaptRadarSubProductNameLevel3 || (AdaptRadarSubProductNameLevel3 = {}));
export var AdaptRadarHostingType;
(function (AdaptRadarHostingType) {
    AdaptRadarHostingType["OnPrem"] = "OnPrem";
    AdaptRadarHostingType["SaaS"] = "SaaS";
})(AdaptRadarHostingType || (AdaptRadarHostingType = {}));
export var AdaptRadarCustomerEnvironment;
(function (AdaptRadarCustomerEnvironment) {
    AdaptRadarCustomerEnvironment["Integration"] = "Integration";
    AdaptRadarCustomerEnvironment["QA"] = "QA";
    AdaptRadarCustomerEnvironment["Stage"] = "Stage";
    AdaptRadarCustomerEnvironment["Production"] = "Production";
})(AdaptRadarCustomerEnvironment || (AdaptRadarCustomerEnvironment = {}));
/**
 * Used to flatten "productInfo" property for better readability when data sent as a part of analytics event data
 *
 * @param productInfo non flatten version of "productInfo"
 */
export function flattenProductInfo(productInfo) {
    if (isNil(productInfo)) {
        return null;
    }
    const productArea = productInfo.productArea;
    const subProductLevel1 = productInfo.subProductLevel1;
    const subProductLevel2 = productInfo.subProductLevel2;
    const subProductLevel3 = productInfo.subProductLevel3;
    return {
        productAreaName: productArea.name,
        productAreaVersion: productArea.version instanceof Version
            ? productArea.version.full
            : productArea.version,
        productAreaHosting: productArea.hosting,
        productAreaCustomerEnvironment: productArea.customerEnvironment,
        subProductLevel1Name: subProductLevel1 === null || subProductLevel1 === void 0 ? void 0 : subProductLevel1.name,
        subProductLevel1Version: (subProductLevel1 === null || subProductLevel1 === void 0 ? void 0 : subProductLevel1.version) instanceof Version
            ? subProductLevel1 === null || subProductLevel1 === void 0 ? void 0 : subProductLevel1.version.full
            : subProductLevel1 === null || subProductLevel1 === void 0 ? void 0 : subProductLevel1.version,
        subProductLevel1Hosting: subProductLevel1 === null || subProductLevel1 === void 0 ? void 0 : subProductLevel1.hosting,
        subProductLevel1CustomerEnvironment: productArea.customerEnvironment,
        subProductLevel2Name: subProductLevel2 === null || subProductLevel2 === void 0 ? void 0 : subProductLevel2.name,
        subProductLevel2Version: (subProductLevel2 === null || subProductLevel2 === void 0 ? void 0 : subProductLevel2.version) instanceof Version
            ? subProductLevel2 === null || subProductLevel2 === void 0 ? void 0 : subProductLevel2.version.full
            : subProductLevel2 === null || subProductLevel2 === void 0 ? void 0 : subProductLevel2.version,
        subProductLevel2Hosting: subProductLevel2 === null || subProductLevel2 === void 0 ? void 0 : subProductLevel2.hosting,
        subProductLevel2CustomerEnvironment: productArea.customerEnvironment,
        subProductLevel3Name: subProductLevel3 === null || subProductLevel3 === void 0 ? void 0 : subProductLevel3.name,
        subProductLevel3Version: (subProductLevel3 === null || subProductLevel3 === void 0 ? void 0 : subProductLevel3.version) instanceof Version
            ? subProductLevel3 === null || subProductLevel3 === void 0 ? void 0 : subProductLevel3.version.full
            : subProductLevel3 === null || subProductLevel3 === void 0 ? void 0 : subProductLevel3.version,
        subProductLevel3Hosting: subProductLevel3 === null || subProductLevel3 === void 0 ? void 0 : subProductLevel3.hosting,
        subProductLevel3CustomerEnvironment: productArea.customerEnvironment,
    };
}
