// @bmc-ux/adapt-radar/esm2015/lib/adapt-radar-generic.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import { AdaptRadarService } from './adapt-radar.service';
import { AdaptRadarCustomEventName } from './adapt-radar.model';
import * as i0 from "@angular/core";
import * as i1 from "./adapt-radar.service";
import * as i2 from "./adapt-radar.model";
export class AdaptRadarGenericDirective {
    constructor(_adaptRadarService, _elem, _entityName) {
        this._adaptRadarService = _adaptRadarService;
        this._elem = _elem;
        this._entityName = _entityName;
        this.adaptRadarGroupName = null;
        this.adaptRadarDisableEventSending = this._checkIfDisabledFromSending();
    }
    ngAfterViewInit() {
        this._elementIntersectObserver = this._adaptRadarService.subscribeOnElementIntersect(this.adaptRadarDisableEventSending, this._elem, this._entityName);
    }
    ngOnDestroy() {
        this._unsubscribeFromIntersectionObserver();
    }
    _unsubscribeFromIntersectionObserver() {
        if (this._elementIntersectObserver) {
            this._elementIntersectObserver.disconnect();
            this._elementIntersectObserver = null;
        }
    }
    _checkIfDisabledFromSending() {
        var _a, _b;
        return !!((_b = (_a = this._adaptRadarService.config) === null || _a === void 0 ? void 0 : _a.disableEventSending) === null || _b === void 0 ? void 0 : _b.get(this._entityName));
    }
}
AdaptRadarGenericDirective.ɵfac = function AdaptRadarGenericDirective_Factory(t) { return new (t || AdaptRadarGenericDirective)(i0.ɵɵdirectiveInject(i1.AdaptRadarService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i2.AdaptRadarCustomEventName)); };
AdaptRadarGenericDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptRadarGenericDirective, inputs: { adaptRadarGroupName: "adaptRadarGroupName", adaptRadarDisableEventSending: "adaptRadarDisableEventSending" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRadarGenericDirective, [{
        type: Directive
    }], function () { return [{ type: i1.AdaptRadarService }, { type: i0.ElementRef }, { type: i2.AdaptRadarCustomEventName }]; }, { adaptRadarGroupName: [{
            type: Input
        }], adaptRadarDisableEventSending: [{
            type: Input
        }] }); })();
