// @bmc-ux/adapt-radar/esm2015/lib/adapt-radar.model.js
export var AdaptRadarStorageKey;
(function (AdaptRadarStorageKey) {
    AdaptRadarStorageKey["UserId"] = "adaptRadar.userId";
    AdaptRadarStorageKey["VersionsInfo"] = "adaptRadar.versionsInfo";
    AdaptRadarStorageKey["GainsightGlobalDataSent"] = "adaptRadar.gainsightGlobalDataSent";
})(AdaptRadarStorageKey || (AdaptRadarStorageKey = {}));
export var AdaptRadarSupportedProviders;
(function (AdaptRadarSupportedProviders) {
    AdaptRadarSupportedProviders["GoogleGlobalSiteTag"] = "ggst";
    AdaptRadarSupportedProviders["GoogleTagManager"] = "gtm";
    AdaptRadarSupportedProviders["Gainsight"] = "gainsight";
})(AdaptRadarSupportedProviders || (AdaptRadarSupportedProviders = {}));
// Allowed events names
export var AdaptRadarCustomEventName;
(function (AdaptRadarCustomEventName) {
    // Providers events
    AdaptRadarCustomEventName["PagePath"] = "page_path";
    AdaptRadarCustomEventName["GlobalContext"] = "globalContext";
    // Used to send 'global' data(browser info, libraries versions, etc.) on provider init stage
    AdaptRadarCustomEventName["AdaptGlobalData"] = "Adapt_Global_Data";
    // Used to send information when URL changed
    AdaptRadarCustomEventName["AdaptURLChanged"] = "Adapt_URL_Changed";
    AdaptRadarCustomEventName["AdaptGenericComponent"] = "Adapt_Generic_Component";
    // TODO: Need to find way how to deal with components that already extends from 3rd party component.
    // For now directly send event from component
    AdaptRadarCustomEventName["AdaptPagination"] = "Adapt_Pagination";
    AdaptRadarCustomEventName["AdaptSlider"] = "Adapt_Slider";
    AdaptRadarCustomEventName["AdaptToast"] = "Adapt_Toast";
    AdaptRadarCustomEventName["AdaptTree"] = "Adapt_Tree";
    // ADAPT Components names
    // TODO: Think should we split names for better readability and consistency.
    // TODO: Adapt_RxRadiobuttonGroup -> Adapt_Rx_Radiobutton_Group
    AdaptRadarCustomEventName["AdaptRxTextfield"] = "Adapt_RxTextfield";
    AdaptRadarCustomEventName["AdaptRxTextarea"] = "Adapt_RxTextarea";
    AdaptRadarCustomEventName["AdaptRxCounter"] = "Adapt_RxCounter";
    AdaptRadarCustomEventName["AdaptRxSearch"] = "Adapt_RxSearch";
    AdaptRadarCustomEventName["AdaptRxCheckbox"] = "Adapt_RxCheckbox";
    AdaptRadarCustomEventName["AdaptRxRadiobutton"] = "Adapt_RxRadiobutton";
    AdaptRadarCustomEventName["AdaptRxRadiobuttonGroup"] = "Adapt_RxRadiobuttonGroup";
    AdaptRadarCustomEventName["AdaptRxRating"] = "Adapt_RxRating_Default";
    AdaptRadarCustomEventName["AdaptRxRatingBinary"] = "Adapt_RxRating_Binary";
    AdaptRadarCustomEventName["AdaptRxRatingLike"] = "Adapt_RxRating_Like";
    AdaptRadarCustomEventName["AdaptRxRatingVoting"] = "Adapt_RxRating_Voting";
    AdaptRadarCustomEventName["AdaptRxSwitch"] = "Adapt_RxSwitch";
    AdaptRadarCustomEventName["AdaptRxSelect"] = "Adapt_RxSelect";
    AdaptRadarCustomEventName["AdaptRxSelectEditable"] = "Adapt_RxSelectEditable";
    AdaptRadarCustomEventName["AdaptRxDatetime"] = "Adapt_RxDateTime";
    AdaptRadarCustomEventName["AdaptRxDatetimeRange"] = "Adapt_RxDatetimeRange";
    AdaptRadarCustomEventName["AdaptRxTimeSlot"] = "Adapt_RxTimeSlot";
    AdaptRadarCustomEventName["AdaptRxListSelector"] = "Adapt_RxListSelector";
    AdaptRadarCustomEventName["AdaptRxListBuilder"] = "Adapt_RxListBuilder";
    AdaptRadarCustomEventName["AdaptRxTypeahead"] = "Adapt_RxTypeahead";
    AdaptRadarCustomEventName["AdaptRxUploader"] = "Adapt_RxUploader";
    AdaptRadarCustomEventName["AdaptRxFeedback"] = "Adapt_RxFeedback";
    AdaptRadarCustomEventName["AdaptAccordion"] = "Adapt_Accordion";
    AdaptRadarCustomEventName["AdaptSimpleFilter"] = "Adapt_Simple_Filter";
    AdaptRadarCustomEventName["AdaptFacetedFilter"] = "Adapt_Faceted_Filter";
    AdaptRadarCustomEventName["AdaptAdvancedFilter"] = "Adapt_Advanced_Filter";
    AdaptRadarCustomEventName["AdaptAgreement"] = "Adapt_Agreement";
    AdaptRadarCustomEventName["AdaptAlert"] = "Adapt_Alert";
    AdaptRadarCustomEventName["AdaptAlertAnimated"] = "Adapt_Alert_Animated";
    AdaptRadarCustomEventName["AdaptAnchorNavigator"] = "Adapt_Anchor_Navigator";
    AdaptRadarCustomEventName["AdaptBadge"] = "Adapt_Badge";
    AdaptRadarCustomEventName["AdaptBeard"] = "Adapt_Beard";
    AdaptRadarCustomEventName["AdaptBusy"] = "Adapt_Busy";
    AdaptRadarCustomEventName["AdaptButton"] = "Adapt_Button";
    AdaptRadarCustomEventName["AdaptButtonGroup"] = "Adapt_Button_Group";
    AdaptRadarCustomEventName["AdaptCardCarousel"] = "Adapt_Card_Carousel";
    AdaptRadarCustomEventName["AdaptCarousel"] = "Adapt_Carousel";
    AdaptRadarCustomEventName["AdaptClose"] = "Adapt_Close";
    AdaptRadarCustomEventName["AdaptCodeHighlighter"] = "Adapt_Code_Highlighter";
    AdaptRadarCustomEventName["AdaptCodeViewer"] = "Adapt_Code_Viewer";
    AdaptRadarCustomEventName["AdaptCollapse"] = "Adapt_Collapse";
    AdaptRadarCustomEventName["AdaptExpando"] = "AdaptExpando";
    AdaptRadarCustomEventName["AdaptColorPicker"] = "Adapt_Color_Picker";
    AdaptRadarCustomEventName["AdaptCrop"] = "Adapt_Crop";
    AdaptRadarCustomEventName["AdaptCustomActions"] = "Adapt_Custom_Actions";
    AdaptRadarCustomEventName["AdaptDockedPanel"] = "Adapt_Docked_Panel";
    AdaptRadarCustomEventName["AdaptDownloader"] = "Adapt_Downloader";
    AdaptRadarCustomEventName["AdaptDropdown"] = "Adapt_Dropdown";
    AdaptRadarCustomEventName["AdaptEmptyState"] = "Adapt_Empty_State";
    AdaptRadarCustomEventName["AdaptFocusHolder"] = "Adapt_Focus_Holder";
    AdaptRadarCustomEventName["AdaptFocusTracker"] = "Adapt_Focus_Tracker";
    AdaptRadarCustomEventName["AdaptIcon"] = "Adapt_Icon";
    AdaptRadarCustomEventName["AdaptInlineEditor"] = "Adapt_Inline_Editor";
    AdaptRadarCustomEventName["AdaptLoginPage"] = "Adapt_Login_Page";
    AdaptRadarCustomEventName["AdaptMention"] = "Adapt_Mention";
    AdaptRadarCustomEventName["AdaptMention2"] = "Adapt_Mention_2";
    AdaptRadarCustomEventName["AdaptDialog"] = "Adapt_Dialog";
    AdaptRadarCustomEventName["AdaptMenu"] = "Adapt_Menu";
    AdaptRadarCustomEventName["AdaptMetatag"] = "Adapt_Metatag";
    AdaptRadarCustomEventName["AdaptModal"] = "Adapt_Modal";
    AdaptRadarCustomEventName["AdaptMultiCard"] = "Adapt_Multi_Card";
    AdaptRadarCustomEventName["AdaptNavigation"] = "Adapt_Navigation";
    AdaptRadarCustomEventName["AdaptNavigationMenu"] = "Adapt_Navigation_Menu";
    AdaptRadarCustomEventName["AdaptNavigationAction"] = "Adapt_Navigation_Action";
    AdaptRadarCustomEventName["AdaptNavigationAlert"] = "Adapt_Navigation_Alert";
    AdaptRadarCustomEventName["AdaptNavigationHamburger"] = "Adapt_Navigation_Hamburger";
    AdaptRadarCustomEventName["AdaptNotFoundPage"] = "Adapt_Not_Found_Page";
    AdaptRadarCustomEventName["AdaptPopover"] = "Adapt_Popover";
    AdaptRadarCustomEventName["AdaptProfile"] = "Adapt_Profile";
    AdaptRadarCustomEventName["AdaptProgress"] = "Adapt_Progress";
    AdaptRadarCustomEventName["AdaptTime"] = "Adapt_Time";
    AdaptRadarCustomEventName["AdaptUploader"] = "Adapt_Uploader";
    AdaptRadarCustomEventName["AdaptSingleUploader"] = "Adapt_Single_Uploader";
    AdaptRadarCustomEventName["AdaptSidebar"] = "Adapt_Sidebar";
    AdaptRadarCustomEventName["AdaptSplit"] = "Adapt_Split";
    AdaptRadarCustomEventName["AdaptSteps"] = "Adapt_Steps";
    AdaptRadarCustomEventName["AdaptSubnav"] = "Adapt_Subnav";
    AdaptRadarCustomEventName["AdaptTabs"] = "Adapt_Tabs";
    AdaptRadarCustomEventName["AdaptTabsKebabMenu"] = "Adapt_Tabs_Kebab_Menu";
    AdaptRadarCustomEventName["AdaptTabsControls"] = "Adapt_Tabs_Controls";
    AdaptRadarCustomEventName["AdaptTag"] = "Adapt_Tag";
    AdaptRadarCustomEventName["AdaptTextField"] = "Adapt_Text_Field";
    AdaptRadarCustomEventName["AdaptTextFieldFormBuilder"] = "Adapt_Text_Field_Form_Builder";
    AdaptRadarCustomEventName["AdaptTooltip"] = "Adapt_Tooltip";
    AdaptRadarCustomEventName["AdaptTruncateFileName"] = "Adapt_Truncate_File_Name";
    AdaptRadarCustomEventName["AdaptTypeahead"] = "Adapt_Typeahead";
    AdaptRadarCustomEventName["AdaptWorkflowCarousel"] = "Adapt_Workflow_Carousel";
    AdaptRadarCustomEventName["AdaptIllustration"] = "Adapt_Illustration";
    AdaptRadarCustomEventName["AdaptRichTextEditor"] = "Adapt_RichTextEditor";
    AdaptRadarCustomEventName["AdaptAgreementAdminContent"] = "Adapt_Agreement_AdminContent";
    AdaptRadarCustomEventName["AdaptAgreementCard"] = "Adapt_Agreement_Card";
    AdaptRadarCustomEventName["AdaptAgreementContent"] = "Adapt_Agreement_Content";
    AdaptRadarCustomEventName["AdaptAgreementNotification"] = "Adapt_Agreement_Notification";
    AdaptRadarCustomEventName["AdaptAgreementUserModal"] = "Adapt_Agreement_UserModal";
    // Obsolete modules
    AdaptRadarCustomEventName["AdaptCheckbox2"] = "Adapt_Checkbox_2";
    AdaptRadarCustomEventName["AdaptCounter"] = "Adapt_Counter";
    AdaptRadarCustomEventName["AdaptDatetimeRange2"] = "Adapt_Datetime_Range_2";
    AdaptRadarCustomEventName["AdaptDatetime2"] = "Adapt_Datetime_2";
    AdaptRadarCustomEventName["AdaptListBuilder"] = "Adapt_List_Builder";
    AdaptRadarCustomEventName["AdaptList"] = "Adapt_List";
    AdaptRadarCustomEventName["AdaptRadioButton"] = "Adapt_Radio_Button";
    AdaptRadarCustomEventName["AdaptDropDownEditable"] = "Adapt_DropDown_Editable";
    AdaptRadarCustomEventName["AdaptRating"] = "Adapt_Rating";
    AdaptRadarCustomEventName["AdaptRatingBinary"] = "Adapt_Rating_Binary";
    AdaptRadarCustomEventName["AdaptRatingLike"] = "Adapt_Rating_Like";
    AdaptRadarCustomEventName["AdaptRatingVoting"] = "Adapt_Rating_Voting";
    AdaptRadarCustomEventName["AdaptSearch"] = "Adapt_Search";
    AdaptRadarCustomEventName["AdaptSelect"] = "Adapt_Select";
    AdaptRadarCustomEventName["AdaptSelectInline"] = "Adapt_Select_Inline";
    AdaptRadarCustomEventName["AdaptSwitcher"] = "Adapt_Switcher";
    AdaptRadarCustomEventName["AdaptTimeSlot"] = "Adapt_Time_Slot";
})(AdaptRadarCustomEventName || (AdaptRadarCustomEventName = {}));
// Generic events that can be applied across all components
export var AdaptRadarCustomEventType;
(function (AdaptRadarCustomEventType) {
    AdaptRadarCustomEventType["ErrorFired"] = "ErrorFired";
    AdaptRadarCustomEventType["Init"] = "Init";
    AdaptRadarCustomEventType["Focus"] = "Focus";
    AdaptRadarCustomEventType["Blur"] = "Blur";
    AdaptRadarCustomEventType["Click"] = "Click";
    AdaptRadarCustomEventType["Search"] = "Search";
    AdaptRadarCustomEventType["ValueChanged"] = "ValueChanged";
    AdaptRadarCustomEventType["Clear"] = "Clear";
    AdaptRadarCustomEventType["DragStart"] = "DragStart";
    AdaptRadarCustomEventType["Drop"] = "Drop";
})(AdaptRadarCustomEventType || (AdaptRadarCustomEventType = {}));
// Component specific events
export var AdaptRadarComponentEventType;
(function (AdaptRadarComponentEventType) {
    AdaptRadarComponentEventType["TabAdd"] = "TabAdd";
    AdaptRadarComponentEventType["TabDelete"] = "TabDelete";
})(AdaptRadarComponentEventType || (AdaptRadarComponentEventType = {}));
