// @bmc-ux/adapt-radar/esm2015/lib/adapt-radar-components.model.js
var AdaptNumberType;
(function (AdaptNumberType) {
    AdaptNumberType[AdaptNumberType["Integer"] = 0] = "Integer";
    AdaptNumberType[AdaptNumberType["Float"] = 1] = "Float";
    AdaptNumberType[AdaptNumberType["Scientific"] = 2] = "Scientific";
})(AdaptNumberType || (AdaptNumberType = {}));
/* rx-rating */
var RxRatingSize;
(function (RxRatingSize) {
    RxRatingSize[RxRatingSize["xs"] = 0] = "xs";
    RxRatingSize[RxRatingSize["sm"] = 1] = "sm";
    RxRatingSize[RxRatingSize["md"] = 2] = "md";
    RxRatingSize[RxRatingSize["lg"] = 3] = "lg";
})(RxRatingSize || (RxRatingSize = {}));
var RxRatingVotingType;
(function (RxRatingVotingType) {
    RxRatingVotingType[RxRatingVotingType["Star"] = 0] = "Star";
    RxRatingVotingType[RxRatingVotingType["Flag"] = 1] = "Flag";
    RxRatingVotingType[RxRatingVotingType["Like"] = 2] = "Like";
    RxRatingVotingType[RxRatingVotingType["Thanks"] = 3] = "Thanks";
})(RxRatingVotingType || (RxRatingVotingType = {}));
/* rx-datetime */
var RxDatetimePickerMode;
(function (RxDatetimePickerMode) {
    RxDatetimePickerMode[RxDatetimePickerMode["DateTime"] = 0] = "DateTime";
    RxDatetimePickerMode[RxDatetimePickerMode["Date"] = 1] = "Date";
    RxDatetimePickerMode[RxDatetimePickerMode["Time"] = 2] = "Time";
    RxDatetimePickerMode[RxDatetimePickerMode["YearMonth"] = 3] = "YearMonth";
    RxDatetimePickerMode[RxDatetimePickerMode["Year"] = 4] = "Year";
    RxDatetimePickerMode[RxDatetimePickerMode["Month"] = 5] = "Month";
})(RxDatetimePickerMode || (RxDatetimePickerMode = {}));
export {};
