// @bmc-ux/obsolete/esm2015/lib/time-slot/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptTimeSlotComponent } from './time-slot.component';
import { AdaptPopoverModule, AdaptTooltipModule, AdaptIconModule, AdaptButtonModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptTimeSlotModule {
}
AdaptTimeSlotModule.ɵfac = function AdaptTimeSlotModule_Factory(t) { return new (t || AdaptTimeSlotModule)(); };
AdaptTimeSlotModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTimeSlotModule });
AdaptTimeSlotModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptButtonModule, AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule.forRoot()]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTimeSlotModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptButtonModule, AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule.forRoot()],
                exports: [AdaptTimeSlotComponent],
                declarations: [AdaptTimeSlotComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTimeSlotModule, { declarations: [AdaptTimeSlotComponent], imports: [CommonModule, AdaptButtonModule, AdaptIconModule, AdaptTooltipModule, i1.AdaptPopoverModule], exports: [AdaptTimeSlotComponent] }); })();
export * from './time-slot.component';
