// @bmc-ux/adapt-angular/esm2015/modal/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { AdaptFocusHolderModule } from '../focus-holder/index';
import { AdaptAlertAnimatedModule } from '../alert-animated/index';
import { AdaptModalComponent } from './modal.component';
import { AdaptDialogComponent } from './dialog.component';
import { AdaptOverModalModule } from './over-modal';
import { AdaptModalService } from './modal.service';
import * as i0 from "@angular/core";
import * as i1 from "../focus-holder/focus-holder.directive";
import * as i2 from "@angular/common";
import * as i3 from "../alert-animated/alert-animated.component";
import * as i4 from "./over-modal";
/**
 * Modals module
 *
 * forRoot() - initialize
 */
export class AdaptModalModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptModalModule
        };
    }
}
AdaptModalModule.ɵfac = function AdaptModalModule_Factory(t) { return new (t || AdaptModalModule)(); };
AdaptModalModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptModalModule });
AdaptModalModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdaptModalService], imports: [[
            CommonModule,
            OverlayModule,
            AdaptOverModalModule,
            AdaptFocusHolderModule,
            AdaptAlertAnimatedModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptModalModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    OverlayModule,
                    AdaptOverModalModule,
                    AdaptFocusHolderModule,
                    AdaptAlertAnimatedModule
                ],
                declarations: [
                    AdaptModalComponent,
                    AdaptDialogComponent
                ],
                exports: [
                    AdaptModalComponent,
                    AdaptDialogComponent
                ],
                entryComponents: [
                    AdaptModalComponent,
                    AdaptDialogComponent
                ],
                providers: [AdaptModalService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptModalModule, { declarations: [AdaptModalComponent,
        AdaptDialogComponent], imports: [CommonModule,
        OverlayModule,
        AdaptOverModalModule,
        AdaptFocusHolderModule,
        AdaptAlertAnimatedModule], exports: [AdaptModalComponent,
        AdaptDialogComponent] }); })();
i0.ɵɵsetComponentScope(AdaptModalComponent, [i1.AdaptFocusHolderDirective, i2.NgClass, i2.NgIf, i3.AdaptAlertAnimatedComponent, i4.AdaptOverModalComponent], []);
i0.ɵɵsetComponentScope(AdaptDialogComponent, [i1.AdaptFocusHolderDirective, i2.NgClass, i2.NgIf, i3.AdaptAlertAnimatedComponent], []);
export * from './modal';
export * from './modal.config';
export * from './modal.service';
export * from './dialog.component';
export * from './modal.component';
export * from './over-modal';
export * from './modal.model';
export * from './modal.animation';
