// @bmc-ux/adapt-angular/esm2015/modal/modal.component.js
import { Component, ComponentFactoryResolver, ElementRef, HostListener, Optional, SkipSelf, ViewChild, ViewContainerRef } from '@angular/core';
import { AdaptTranslateService } from '../common/i18n/index';
import { AdaptRadarAngularService } from '../common/radar/adapt-radar';
import { ModalDirective } from './modal';
import { ActiveModalRef } from './modal.model';
import { ModalAnimation } from './modal.animation';
import * as i0 from "@angular/core";
import * as i1 from "./modal.model";
import * as i2 from "../common/i18n/index";
import * as i3 from "../common/radar/adapt-radar";
const _c0 = ["contentPlacement"];
function AdaptModalComponent_div_9_h5_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "h5", 13);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("id", ctx_r5.id + "_title");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("\n                ", ctx_r5._self.title, "\n              ");
} }
function AdaptModalComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    const _r7 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 10);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵtemplate(2, AdaptModalComponent_div_9_h5_2_Template, 2, 2, "h5", 11);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵelementStart(4, "button", 12);
    i0.ɵɵlistener("click", function AdaptModalComponent_div_9_Template_button_click_4_listener() { i0.ɵɵrestoreView(_r7); const ctx_r6 = i0.ɵɵnextContext(); return ctx_r6._self.dismiss(ctx_r6._self.reason.CLOSE_BTN); });
    i0.ɵɵtext(5, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r0._self.title);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("aria-label", ctx_r0.closeText);
} }
function AdaptModalComponent_ng_template_14_Template(rf, ctx) { }
function AdaptModalComponent_adapt_over_modal_21_Template(rf, ctx) { if (rf & 1) {
    const _r9 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "adapt-over-modal", 14);
    i0.ɵɵlistener("onClose", function AdaptModalComponent_adapt_over_modal_21_Template_adapt_over_modal_onClose_0_listener() { i0.ɵɵrestoreView(_r9); const ctx_r8 = i0.ɵɵnextContext(); return ctx_r8._self.omClose(); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext();
    i0.ɵɵproperty("config", ctx_r4.overModalConfig);
} }
/**
 * Adapt Modal Component
 */
export class AdaptModalComponent extends ModalDirective {
    constructor(_elRef, _componentFactoryResolver, _activeModalRef, _rootActiveModalRef, ts, _adaptRadarService) {
        // Call the parent constructor
        super(_elRef, _componentFactoryResolver, _activeModalRef, _rootActiveModalRef, ts, _adaptRadarService);
        this.ts = ts;
        // OverModal trigger
        this.omShow = false;
        /**
         * Close on backdrop click functional
         */
        this.hostMouseDownHandler = (target) => {
            this._mouseDownTarget = target;
        };
        this.hostMouseUpHandler = (target) => {
            if (target.classList.contains('modal-scroll-container') && this._mouseDownTarget === target) {
                this.dismiss(this.reason.BACKDROP_CLICK);
            }
        };
        this.context.openOverModal = (config) => {
            this.omOpen(config);
        };
        this.context.closeOverModal = () => {
            this.omClose();
        };
        this._self = this;
    }
    get modalDialogClass() {
        return '__is-modal modal-dialog modal-default modal-size-' + (this.config.size || 'default');
    }
    ngOnInit() {
        super.ngOnInit();
    }
    ngOnDestroy() {
        this.omClose();
    }
    /**
     * Open OverModal
     */
    omOpen(config) {
        this.overModalConfig = config;
        this.omShow = true;
    }
    /**
     * Close OverModal
     */
    omClose() {
        this.omShow = false;
    }
    // Implementation of Modal abstract method
    getContentPlacementRef() {
        return this.contentPlacement;
    }
}
AdaptModalComponent.ɵfac = function AdaptModalComponent_Factory(t) { return new (t || AdaptModalComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i0.ComponentFactoryResolver), i0.ɵɵdirectiveInject(i1.ActiveModalRef), i0.ɵɵdirectiveInject(i1.ActiveModalRef, 12), i0.ɵɵdirectiveInject(i2.AdaptTranslateService), i0.ɵɵdirectiveInject(i3.AdaptRadarAngularService)); };
AdaptModalComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptModalComponent, selectors: [["adapt-modal"]], viewQuery: function AdaptModalComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.contentPlacement = _t.first);
    } }, hostVars: 8, hostBindings: function AdaptModalComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵsyntheticHostListener("@modalAnimation.done", function AdaptModalComponent_animation_modalAnimation_done_HostBindingHandler($event) { return ctx._self.modalCloseAnimationEnd($event); });
        i0.ɵɵlistener("mousedown", function AdaptModalComponent_mousedown_HostBindingHandler($event) { return ctx.hostMouseDownHandler($event.target); })("mouseup", function AdaptModalComponent_mouseup_HostBindingHandler($event) { return ctx.hostMouseUpHandler($event.target); });
    } if (rf & 2) {
        i0.ɵɵsyntheticHostProperty("@modalAnimation", true)("@backdropAnimation", true);
        i0.ɵɵclassMap(ctx._self.config.customClass || "");
        i0.ɵɵstyleProp("position", "fixed")("z-index", 1050);
    } }, features: [i0.ɵɵProvidersFeature([ActiveModalRef]), i0.ɵɵInheritDefinitionFeature], decls: 25, vars: 13, consts: [["role", "dialog", "aria-modal", "true", 1, "modal", 3, "adaptFocusHolder"], [1, "modal-scroll-container"], [3, "ngClass"], [1, "modal-content"], ["class", "modal-header", 4, "ngIf"], [3, "adaptRadarDisableEventSending", "config", "isVisible", "alertClosed"], ["alertsSection", ""], ["contentPlacement", ""], [3, "config", "onClose", 4, "ngIf"], [1, "modal-backdrop"], [1, "modal-header"], ["class", "modal-title", 3, "id", 4, "ngIf"], ["type", "button", "role", "button", "data-dismiss", "modal", 1, "close", 3, "click"], [1, "modal-title", 3, "id"], [3, "config", "onClose"]], template: function AdaptModalComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtext(0, "\n    ");
        i0.ɵɵelementStart(1, "div", 0);
        i0.ɵɵtext(2, "\n      ");
        i0.ɵɵelementStart(3, "div", 1);
        i0.ɵɵtext(4, "\n        ");
        i0.ɵɵelementStart(5, "div", 2);
        i0.ɵɵtext(6, "\n          ");
        i0.ɵɵelementStart(7, "div", 3);
        i0.ɵɵtext(8, "\n            ");
        i0.ɵɵtemplate(9, AdaptModalComponent_div_9_Template, 7, 2, "div", 4);
        i0.ɵɵtext(10, "\n            ");
        i0.ɵɵelementStart(11, "adapt-alert-animated", 5, 6);
        i0.ɵɵlistener("alertClosed", function AdaptModalComponent_Template_adapt_alert_animated_alertClosed_11_listener($event) { return ctx.alertClosed.emit($event); });
        i0.ɵɵelementEnd();
        i0.ɵɵtext(13, "\n            ");
        i0.ɵɵtemplate(14, AdaptModalComponent_ng_template_14_Template, 0, 0, "ng-template", null, 7, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(16, "\n          ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(17, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(18, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(19, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(20, "\n\n    ");
        i0.ɵɵtemplate(21, AdaptModalComponent_adapt_over_modal_21_Template, 1, 1, "adapt-over-modal", 8);
        i0.ɵɵtext(22, "\n\n    ");
        i0.ɵɵelement(23, "div", 9);
        i0.ɵɵtext(24, "\n  ");
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("adaptFocusHolder", true);
        i0.ɵɵattribute("aria-labelledby", ctx._self.title ? ctx.id + "_title" : undefined)("data-testid", ctx._self.config.testID);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngClass", ctx.modalDialogClass);
        i0.ɵɵadvance(2);
        i0.ɵɵclassProp("footer-align_left", ctx._self.config.footerAlignLeft || ctx._self.configService.footerAlignLeft);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx._self.showHeader);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("adaptRadarDisableEventSending", true)("config", ctx.alerts)("isVisible", ctx.alerts.length > 0);
        i0.ɵɵadvance(10);
        i0.ɵɵproperty("ngIf", ctx._self.omShow);
        i0.ɵɵadvance(2);
        i0.ɵɵclassProp("show", !ctx._self.config.hideBackdrop);
    } }, styles: [".modal-header[_ngcontent-%COMP%]{min-height:34px}"], data: { animation: ModalAnimation } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptModalComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-modal',
                template: `
    <div class="modal"
         role="dialog"
         aria-modal="true"
         [attr.aria-labelledby]="_self.title ? id + '_title' : undefined"
         [attr.data-testid]="_self.config.testID"
         [adaptFocusHolder]="true">
      <div class="modal-scroll-container">
        <div [ngClass]="modalDialogClass">
          <div class="modal-content"
               [class.footer-align_left]="_self.config.footerAlignLeft || _self.configService.footerAlignLeft">
            <div class="modal-header" *ngIf="_self.showHeader">
              <h5 class="modal-title"
                  *ngIf="_self.title"
                  [id]="id + '_title'">
                {{_self.title}}
              </h5>
              <button type="button"
                      role="button"
                      data-dismiss="modal"
                      class="close"
                      [attr.aria-label]="closeText"
                      (click)="_self.dismiss(_self.reason.CLOSE_BTN)">
              </button>
            </div>
            <adapt-alert-animated #alertsSection
                                  [adaptRadarDisableEventSending]="true"
                                  [config]="alerts"
                                  [isVisible]="alerts.length > 0"
                                  (alertClosed)="alertClosed.emit($event)"></adapt-alert-animated>
            <ng-template #contentPlacement></ng-template>
          </div>
        </div>
      </div>
    </div>

    <adapt-over-modal [config]="overModalConfig" (onClose)="_self.omClose()" *ngIf="_self.omShow"></adapt-over-modal>

    <div class="modal-backdrop" [class.show]="!_self.config.hideBackdrop"></div>
  `,
                animations: ModalAnimation,
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {
                    '[class]': '_self.config.customClass || ""',
                    '[style.position]': '"fixed"',
                    '[style.zIndex]': '1050',
                    '[@modalAnimation]': 'true',
                    '[@backdropAnimation]': 'true',
                    '(@modalAnimation.done)': '_self.modalCloseAnimationEnd($event)'
                },
                providers: [ActiveModalRef],
                styleUrls: ['./modal.css']
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i0.ComponentFactoryResolver }, { type: i1.ActiveModalRef }, { type: i1.ActiveModalRef, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }, { type: i2.AdaptTranslateService }, { type: i3.AdaptRadarAngularService }]; }, { contentPlacement: [{
            type: ViewChild,
            args: ['contentPlacement', { read: ViewContainerRef, static: true }]
        }], hostMouseDownHandler: [{
            type: HostListener,
            args: ['mousedown', ['$event.target']]
        }], hostMouseUpHandler: [{
            type: HostListener,
            args: ['mouseup', ['$event.target']]
        }] }); })();
