// @bmc-ux/obsolete/esm2015/lib/counter/index.js
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { AdaptCounterComponent } from './counter.component';
import { AdaptRxLabelModule, AdaptDragonInputHandlerModule, AdaptPopoverModule, AdaptTooltipModule, AdaptIconModule, AdaptDeviceDetectionService } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptCounterModule {
}
AdaptCounterModule.ɵfac = function AdaptCounterModule_Factory(t) { return new (t || AdaptCounterModule)(); };
AdaptCounterModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCounterModule });
AdaptCounterModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdaptDeviceDetectionService], imports: [[
            CommonModule,
            FormsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCounterModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule
                ],
                declarations: [AdaptCounterComponent],
                exports: [AdaptCounterComponent],
                providers: [AdaptDeviceDetectionService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCounterModule, { declarations: [AdaptCounterComponent], imports: [CommonModule,
        FormsModule,
        AdaptIconModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptDragonInputHandlerModule,
        AdaptRxLabelModule], exports: [AdaptCounterComponent] }); })();
export * from './counter.component';
