// @bmc-ux/adapt-angular/esm2015/rx/form-control/index.js
import { NgModule } from '@angular/core';
import { AdaptRxLabelModule } from './rx-control-label/index';
import { AdaptRxFeedbackModule } from './rx-feedback/index';
import { AdaptRxValidatorsModule } from './rx-validators/index';
import * as i0 from "@angular/core";
export class AdaptRxFormControlModule {
}
AdaptRxFormControlModule.ɵfac = function AdaptRxFormControlModule_Factory(t) { return new (t || AdaptRxFormControlModule)(); };
AdaptRxFormControlModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxFormControlModule });
AdaptRxFormControlModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [], imports: [[
            AdaptRxLabelModule,
            AdaptRxFeedbackModule,
            AdaptRxValidatorsModule
        ], AdaptRxLabelModule,
        AdaptRxFeedbackModule,
        AdaptRxValidatorsModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxFormControlModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptRxLabelModule,
                    AdaptRxFeedbackModule,
                    AdaptRxValidatorsModule
                ],
                declarations: [],
                exports: [
                    AdaptRxLabelModule,
                    AdaptRxFeedbackModule,
                    AdaptRxValidatorsModule
                ],
                providers: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxFormControlModule, { imports: [AdaptRxLabelModule,
        AdaptRxFeedbackModule,
        AdaptRxValidatorsModule], exports: [AdaptRxLabelModule,
        AdaptRxFeedbackModule,
        AdaptRxValidatorsModule] }); })();
export * from './adapt-rx-component';
export * from './adapt-rx-form-control';
export * from './adapt-rx-form-control.types';
export * from './rx-control-label/index';
export * from './rx-feedback/index';
export * from './rx-validators/index';
