// @bmc-ux/adapt-angular/esm2015/rx/rx-switch/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxSwitchComponent } from './adapt-rx-switch.component';
import { AdaptIconModule } from '../../icon/index';
import { AdaptPopoverModule } from '../../popover/index';
import { AdaptTooltipModule } from '../../tooltip/index';
import * as i0 from "@angular/core";
export class AdaptRxSwitchModule {
}
AdaptRxSwitchModule.ɵfac = function AdaptRxSwitchModule_Factory(t) { return new (t || AdaptRxSwitchModule)(); };
AdaptRxSwitchModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxSwitchModule });
AdaptRxSwitchModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule,
            AdaptIconModule,
            AdaptPopoverModule,
            AdaptTooltipModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxSwitchModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule,
                    AdaptIconModule,
                    AdaptPopoverModule,
                    AdaptTooltipModule
                ],
                declarations: [
                    AdaptRxSwitchComponent
                ],
                exports: [
                    AdaptRxSwitchComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxSwitchModule, { declarations: [AdaptRxSwitchComponent], imports: [CommonModule,
        AdaptRxFormControlModule,
        AdaptIconModule,
        AdaptPopoverModule,
        AdaptTooltipModule], exports: [AdaptRxSwitchComponent] }); })();
export * from './adapt-rx-switch.component';
