// @bmc-ux/adapt-charts/esm2015/common/models/chart.model.js
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptChartAbstract {
}
AdaptChartAbstract.ɵfac = function AdaptChartAbstract_Factory(t) { return new (t || AdaptChartAbstract)(); };
AdaptChartAbstract.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptChartAbstract, inputs: { backgroundColor: "backgroundColor", width: "width", height: "height", header: "header", legend: "legend", suppressHeader: "suppressHeader", suppressLegend: "suppressLegend", tooltip: "tooltip", dataBuilder: "dataBuilder", noDataText: "noDataText" }, outputs: { clickChart: "clickChart" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptChartAbstract, [{
        type: Directive
    }], null, { backgroundColor: [{
            type: Input
        }], width: [{
            type: Input
        }], height: [{
            type: Input
        }], header: [{
            type: Input
        }], legend: [{
            type: Input
        }], suppressHeader: [{
            type: Input
        }], suppressLegend: [{
            type: Input
        }], tooltip: [{
            type: Input
        }], dataBuilder: [{
            type: Input
        }], noDataText: [{
            type: Input
        }], clickChart: [{
            type: Output
        }] }); })();
export var AdaptChartSliderType;
(function (AdaptChartSliderType) {
    AdaptChartSliderType["bar"] = "bar";
    AdaptChartSliderType["line"] = "line";
    AdaptChartSliderType["scatterplot"] = "scatterplot";
})(AdaptChartSliderType || (AdaptChartSliderType = {}));
