// @bmc-ux/adapt-charts/esm2015/common/models/stacked-chart.model.js
import { Directive, Input } from '@angular/core';
import { AdaptChartAbstract } from './chart.model';
import * as i0 from "@angular/core";
export var AdaptStackedChartType;
(function (AdaptStackedChartType) {
    AdaptStackedChartType["Bar"] = "bar";
    AdaptStackedChartType["Column"] = "column";
    AdaptStackedChartType["Area"] = "area";
})(AdaptStackedChartType || (AdaptStackedChartType = {}));
export class AdaptStackedChartAbstract extends AdaptChartAbstract {
}
AdaptStackedChartAbstract.ɵfac = /*@__PURE__*/ function () { let ɵAdaptStackedChartAbstract_BaseFactory; return function AdaptStackedChartAbstract_Factory(t) { return (ɵAdaptStackedChartAbstract_BaseFactory || (ɵAdaptStackedChartAbstract_BaseFactory = i0.ɵɵgetInheritedFactory(AdaptStackedChartAbstract)))(t || AdaptStackedChartAbstract); }; }();
AdaptStackedChartAbstract.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptStackedChartAbstract, inputs: { type: "type", xAxis: "xAxis", yAxis: "yAxis", series: "series", hideAxises: "hideAxises", seriesDisplay: "seriesDisplay", showSlider: "showSlider", slidePointsCount: "slidePointsCount" }, features: [i0.ɵɵInheritDefinitionFeature] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptStackedChartAbstract, [{
        type: Directive
    }], null, { type: [{
            type: Input
        }], xAxis: [{
            type: Input
        }], yAxis: [{
            type: Input
        }], series: [{
            type: Input
        }], hideAxises: [{
            type: Input
        }], seriesDisplay: [{
            type: Input
        }], showSlider: [{
            type: Input
        }], slidePointsCount: [{
            type: Input
        }] }); })();
