// @bmc-ux/adapt-angular/esm2015/rx/rx-radiobutton/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxRadiobuttonComponent } from './adapt-rx-radiobutton.component';
import { AdaptRxRadiobuttonGroupComponent } from './adapt-rx-radiobutton-group.component';
import * as i0 from "@angular/core";
export class AdaptRxRadiobuttonModule {
}
AdaptRxRadiobuttonModule.ɵfac = function AdaptRxRadiobuttonModule_Factory(t) { return new (t || AdaptRxRadiobuttonModule)(); };
AdaptRxRadiobuttonModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxRadiobuttonModule });
AdaptRxRadiobuttonModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxRadiobuttonModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule
                ],
                declarations: [
                    AdaptRxRadiobuttonComponent,
                    AdaptRxRadiobuttonGroupComponent
                ],
                exports: [
                    AdaptRxRadiobuttonComponent,
                    AdaptRxRadiobuttonGroupComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxRadiobuttonModule, { declarations: [AdaptRxRadiobuttonComponent,
        AdaptRxRadiobuttonGroupComponent], imports: [CommonModule,
        AdaptRxFormControlModule], exports: [AdaptRxRadiobuttonComponent,
        AdaptRxRadiobuttonGroupComponent] }); })();
export * from './adapt-rx-radiobutton.component';
export * from './adapt-rx-radiobutton-group.component';
