// @bmc-ux/obsolete/esm2015/lib/switcher/index.js
import { AdaptSwitcherComponent } from './switcher.component';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule } from '@bmc-ux/adapt-angular';
import { AdaptCheckbox2Module } from '../checkbox2/index';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptSwitcherModule {
}
AdaptSwitcherModule.ɵfac = function AdaptSwitcherModule_Factory(t) { return new (t || AdaptSwitcherModule)(); };
AdaptSwitcherModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSwitcherModule });
AdaptSwitcherModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptCheckbox2Module,
            AdaptPopoverModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSwitcherModule, [{
        type: NgModule,
        args: [{
                imports: [
                    FormsModule,
                    CommonModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptCheckbox2Module,
                    AdaptPopoverModule.forRoot()
                ],
                exports: [AdaptSwitcherComponent],
                declarations: [AdaptSwitcherComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSwitcherModule, { declarations: [AdaptSwitcherComponent], imports: [FormsModule,
        CommonModule,
        AdaptIconModule,
        AdaptTooltipModule,
        AdaptCheckbox2Module, i1.AdaptPopoverModule], exports: [AdaptSwitcherComponent] }); })();
export * from './switcher.component';
