// @bmc-ux/adapt-angular/esm2015/busy-loader/busy-config.js
export var LoaderInOutType;
(function (LoaderInOutType) {
    LoaderInOutType[LoaderInOutType["fly"] = 1] = "fly";
    LoaderInOutType[LoaderInOutType["fade"] = 2] = "fade";
})(LoaderInOutType || (LoaderInOutType = {}));
export const BUSY_CONFIG_DEFAULTS = {
    delay: 0,
    minDuration: 0,
    backdrop: true,
    message: 'adapt.busy.pleaseWait',
    wrapperClass: 'ng-busy',
    loaderType: 'app',
    overlayClass: '',
    sticky: false,
    offsetTop: '0',
    inOutType: LoaderInOutType.fly,
    percentageValue: null,
    loaderClass: ''
};
export class AdaptBusyConfig {
    constructor(config = {}) {
        for (const option in BUSY_CONFIG_DEFAULTS) {
            if (Object.prototype.hasOwnProperty.call(BUSY_CONFIG_DEFAULTS, option)) {
                this[option] = config[option] != null ? config[option] : BUSY_CONFIG_DEFAULTS[option];
            }
        }
    }
}
export class LoaderType {
}
LoaderType.APP = 'app';
LoaderType.PAGE = 'page';
LoaderType.FULL_PAGE = 'fullPage';
LoaderType.SECTION = 'section';
LoaderType.INLINE = 'inline';
LoaderType.LINE = 'lineLoader';
LoaderType.SPECIAL = 'specialLoader';
