// @bmc-ux/adapt-angular/esm2015/busy-loader/busy.service.js
/**
 * @file Service: Busy
 * @author yumao<yuzhang.lille@gmail.com>
 */
import { Injectable, Optional } from '@angular/core';
import { AdaptBusyConfig } from './busy-config';
import { AdaptTranslateService } from '../common/i18n';
import * as i0 from "@angular/core";
import * as i1 from "./busy-config";
import * as i2 from "../common/i18n";
export class AdaptBusyService {
    constructor(config, translateService) {
        if (config) {
            this.config = config;
        }
        else {
            this.config = new AdaptBusyConfig();
            // Default message should be translated
            this.config.message = translateService.getCurrentLanguage()[this.config.message];
        }
    }
}
AdaptBusyService.ɵfac = function AdaptBusyService_Factory(t) { return new (t || AdaptBusyService)(i0.ɵɵinject(i1.AdaptBusyConfig, 8), i0.ɵɵinject(i2.AdaptTranslateService)); };
AdaptBusyService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptBusyService, factory: AdaptBusyService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptBusyService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.AdaptBusyConfig, decorators: [{
                type: Optional
            }] }, { type: i2.AdaptTranslateService }]; }, null); })();
