// @bmc-ux/adapt-angular/esm2015/busy-loader/busy-tpl.component.js
import { Component, ElementRef, NgZone, ViewChild, ViewEncapsulation, ChangeDetectorRef } from '@angular/core';
import { animate, keyframes, style, transition, trigger } from '@angular/animations';
import { PromiseTrackerService } from './promise-tracker.service';
import { LoaderInOutType, LoaderType } from './busy-config';
import { ANIMATION_DURATION, ANIMATION_TIMING_FUNCTION } from '../common/animation';
import { fromEvent, ReplaySubject } from 'rxjs';
import { filter, takeUntil } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./promise-tracker.service";
import * as i2 from "@angular/common";
const _c0 = ["previewSection"];
const _c1 = ["previewContent"];
function AdaptBusyTplComponent_div_1_ng_container_7_ng_container_2_h1_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "h1", 10);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r11.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_7_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelement(2, "div", 8);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵtemplate(4, AdaptBusyTplComponent_div_1_ng_container_7_ng_container_2_h1_4_Template, 2, 1, "h1", 9);
    i0.ɵɵtext(5, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r8 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r8.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_7_ng_template_4__svg_text_29_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(0, "text", 22);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate2("", ctx_r12.message, " ", ctx_r12.determinate ? ctx_r12.value + "%" : "", "");
} }
function AdaptBusyTplComponent_div_1_ng_container_7_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n              ");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(1, "svg", 11);
    i0.ɵɵtext(2, "\n                ");
    i0.ɵɵelementStart(3, "defs");
    i0.ɵɵtext(4, "\n                  ");
    i0.ɵɵelementStart(5, "linearGradient", 12);
    i0.ɵɵtext(6, "\n                    ");
    i0.ɵɵelement(7, "stop", 13);
    i0.ɵɵtext(8, "\n                    ");
    i0.ɵɵelement(9, "stop", 14);
    i0.ɵɵtext(10, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n                  ");
    i0.ɵɵelementStart(12, "linearGradient", 15);
    i0.ɵɵtext(13, "\n                    ");
    i0.ɵɵelement(14, "stop", 16);
    i0.ɵɵtext(15, "\n                    ");
    i0.ɵɵelement(16, "stop", 14);
    i0.ɵɵtext(17, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n                  ");
    i0.ɵɵelementStart(19, "mask", 17);
    i0.ɵɵtext(20, "\n                    ");
    i0.ɵɵelement(21, "circle", 18);
    i0.ɵɵtext(22, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(23, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(24, "\n                ");
    i0.ɵɵelement(25, "rect", 19);
    i0.ɵɵtext(26, "\n                ");
    i0.ɵɵelement(27, "rect", 20);
    i0.ɵɵtext(28, "\n                ");
    i0.ɵɵtemplate(29, AdaptBusyTplComponent_div_1_ng_container_7_ng_template_4__svg_text_29_Template, 2, 2, "text", 21);
    i0.ɵɵtext(30, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(31, "\n            ");
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵattribute("width", ctx_r10.APP_LOADER_DIAMETER)("height", ctx_r10.APP_LOADER_DIAMETER)("viewBox", "0 0" + " " + ctx_r10.APP_LOADER_DIAMETER + " " + ctx_r10.APP_LOADER_DIAMETER);
    i0.ɵɵadvance(20);
    i0.ɵɵstyleProp("stroke-dashoffset", ctx_r10.strokeDashOffset(ctx_r10.APP_LOADER_DIAMETER, ctx_r10.APP_STROKE_WIDTH), "px")("stroke-dasharray", ctx_r10.strokeCircumference((ctx_r10.APP_LOADER_DIAMETER - ctx_r10.APP_STROKE_WIDTH) / 2), "px");
    i0.ɵɵattribute("r", ctx_r10.circleRadius(ctx_r10.APP_LOADER_DIAMETER, ctx_r10.APP_STROKE_WIDTH))("stroke-width", ctx_r10.APP_STROKE_WIDTH);
    i0.ɵɵadvance(4);
    i0.ɵɵattribute("width", ctx_r10.APP_LOADER_DIAMETER / 2)("height", ctx_r10.APP_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("width", ctx_r10.APP_LOADER_DIAMETER / 2)("height", ctx_r10.APP_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r10.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptBusyTplComponent_div_1_ng_container_7_ng_container_2_Template, 6, 1, "ng-container", 6);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptBusyTplComponent_div_1_ng_container_7_ng_template_4_Template, 32, 14, "ng-template", null, 7, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(6, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r9 = i0.ɵɵreference(5);
    const ctx_r2 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r2.determinate)("ngIfElse", _r9);
} }
function AdaptBusyTplComponent_div_1_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelement(2, "div", 23);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵelementStart(4, "div", 24);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r3.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_11_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelement(2, "div", 27);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵelementContainerEnd();
} }
function AdaptBusyTplComponent_div_1_ng_container_11_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n              ");
    i0.ɵɵelementStart(1, "div", 28);
    i0.ɵɵtext(2, "\n                ");
    i0.ɵɵelementStart(3, "div", 29);
    i0.ɵɵtext(4, "\n                  ");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(5, "svg");
    i0.ɵɵtext(6, "\n                    ");
    i0.ɵɵelement(7, "circle", 18);
    i0.ɵɵtext(8, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n                ");
    i0.ɵɵnamespaceHTML();
    i0.ɵɵelementStart(11, "div", 30);
    i0.ɵɵtext(12, "\n                  ");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(13, "svg");
    i0.ɵɵtext(14, "\n                    ");
    i0.ɵɵelement(15, "circle", 18);
    i0.ɵɵtext(16, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n                ");
    i0.ɵɵnamespaceHTML();
    i0.ɵɵelementStart(19, "div", 31);
    i0.ɵɵtext(20, "\n                  ");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(21, "svg");
    i0.ɵɵtext(22, "\n                    ");
    i0.ɵɵelement(23, "circle", 18);
    i0.ɵɵtext(24, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(26, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(27, "\n            ");
} if (rf & 2) {
    const ctx_r15 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(5);
    i0.ɵɵattribute("viewBox", "0 0" + " " + ctx_r15.SECTION_LOADER_DIAMETER + " " + ctx_r15.SECTION_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("r", ctx_r15.circleRadiusSectionLoader(ctx_r15.value))("fill-opacity", ctx_r15.circleOpacitySectionLoader(ctx_r15.value));
    i0.ɵɵadvance(6);
    i0.ɵɵattribute("viewBox", "0 0" + " " + ctx_r15.SECTION_LOADER_DIAMETER + " " + ctx_r15.SECTION_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("r", ctx_r15.circleRadiusSectionLoader(ctx_r15.value - ctx_r15.sectionLoaderRatio))("fill-opacity", ctx_r15.circleOpacitySectionLoader(ctx_r15.value - ctx_r15.sectionLoaderRatio));
    i0.ɵɵadvance(6);
    i0.ɵɵattribute("viewBox", "0 0" + " " + ctx_r15.SECTION_LOADER_DIAMETER + " " + ctx_r15.SECTION_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("r", ctx_r15.circleRadiusSectionLoader(ctx_r15.value - 2 * ctx_r15.sectionLoaderRatio))("fill-opacity", ctx_r15.circleOpacitySectionLoader(ctx_r15.value - 2 * ctx_r15.sectionLoaderRatio));
} }
function AdaptBusyTplComponent_div_1_ng_container_11_div_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 32);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate2("\n              ", ctx_r16.message, " ", ctx_r16.determinate ? ctx_r16.value + "%" : "", "\n            ");
} }
function AdaptBusyTplComponent_div_1_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptBusyTplComponent_div_1_ng_container_11_ng_container_2_Template, 4, 0, "ng-container", 6);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptBusyTplComponent_div_1_ng_container_11_ng_template_4_Template, 28, 9, "ng-template", null, 25, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(6, "\n            ");
    i0.ɵɵtemplate(7, AdaptBusyTplComponent_div_1_ng_container_11_div_7_Template, 2, 2, "div", 26);
    i0.ɵɵtext(8, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r14 = i0.ɵɵreference(5);
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r4.determinate)("ngIfElse", _r14);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngIf", ctx_r4.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_13_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelement(2, "div", 34);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵelementContainerEnd();
} }
function AdaptBusyTplComponent_div_1_ng_container_13_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n              ");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(1, "svg", 35);
    i0.ɵɵtext(2, "\n                ");
    i0.ɵɵelementStart(3, "circle", 36);
    i0.ɵɵtext(4, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n                ");
    i0.ɵɵelementStart(6, "circle", 18);
    i0.ɵɵtext(7, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n            ");
} if (rf & 2) {
    const ctx_r19 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵattribute("width", ctx_r19.INLINE_LOADER_DIAMETER)("height", ctx_r19.INLINE_LOADER_DIAMETER)("viewBox", "0 0" + " " + ctx_r19.INLINE_LOADER_DIAMETER + " " + ctx_r19.INLINE_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("r", ctx_r19.circleRadius(ctx_r19.INLINE_LOADER_DIAMETER, ctx_r19.INLINE_STROKE_WIDTH))("stroke-width", ctx_r19.INLINE_STROKE_WIDTH);
    i0.ɵɵadvance(3);
    i0.ɵɵstyleProp("stroke-dashoffset", ctx_r19.strokeDashOffset(ctx_r19.INLINE_LOADER_DIAMETER, ctx_r19.INLINE_STROKE_WIDTH), "px")("stroke-dasharray", ctx_r19.strokeCircumference((ctx_r19.INLINE_LOADER_DIAMETER - ctx_r19.INLINE_STROKE_WIDTH) / 2), "px");
    i0.ɵɵattribute("r", ctx_r19.circleRadius(ctx_r19.INLINE_LOADER_DIAMETER, ctx_r19.INLINE_STROKE_WIDTH))("stroke-width", ctx_r19.INLINE_STROKE_WIDTH);
} }
function AdaptBusyTplComponent_div_1_ng_container_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptBusyTplComponent_div_1_ng_container_13_ng_container_2_Template, 4, 0, "ng-container", 6);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptBusyTplComponent_div_1_ng_container_13_ng_template_4_Template, 10, 11, "ng-template", null, 33, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(6, "\n            ");
    i0.ɵɵelementStart(7, "span");
    i0.ɵɵtext(8);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r18 = i0.ɵɵreference(5);
    const ctx_r5 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r5.determinate)("ngIfElse", _r18);
    i0.ɵɵadvance(6);
    i0.ɵɵtextInterpolate2("", ctx_r5.message, " ", ctx_r5.determinate ? ctx_r5.value + "%" : "", "");
} }
function AdaptBusyTplComponent_div_1_ng_container_15_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                    ");
    i0.ɵɵelement(2, "div", 43);
    i0.ɵɵtext(3, "\n                    ");
    i0.ɵɵelement(4, "div", 44);
    i0.ɵɵtext(5, "\n                    ");
    i0.ɵɵelement(6, "div", 45);
    i0.ɵɵtext(7, "\n                  ");
    i0.ɵɵelementContainerEnd();
} }
function AdaptBusyTplComponent_div_1_ng_container_15_ng_template_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n                    ");
    i0.ɵɵelement(1, "div", 46);
    i0.ɵɵtext(2, "\n                  ");
} if (rf & 2) {
    const ctx_r23 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵstyleProp("width", ctx_r23.value, "%");
} }
function AdaptBusyTplComponent_div_1_ng_container_15_div_15_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 47);
    i0.ɵɵtext(1, "\n                  ");
    i0.ɵɵelementStart(2, "div", 48);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n                ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r24 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate2("", ctx_r24.message, " ", ctx_r24.determinate ? ctx_r24.value + "%" : "", "");
} }
function AdaptBusyTplComponent_div_1_ng_container_15_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelementStart(2, "div", 37, 38);
    i0.ɵɵtext(4, "\n              ");
    i0.ɵɵelementStart(5, "div", 39);
    i0.ɵɵtext(6, "\n                ");
    i0.ɵɵelementStart(7, "div", 40);
    i0.ɵɵtext(8, "\n                  ");
    i0.ɵɵtemplate(9, AdaptBusyTplComponent_div_1_ng_container_15_ng_container_9_Template, 8, 0, "ng-container", 6);
    i0.ɵɵtext(10, "\n\n                  ");
    i0.ɵɵtemplate(11, AdaptBusyTplComponent_div_1_ng_container_15_ng_template_11_Template, 3, 2, "ng-template", null, 41, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(13, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n\n                ");
    i0.ɵɵtemplate(15, AdaptBusyTplComponent_div_1_ng_container_15_div_15_Template, 5, 2, "div", 42);
    i0.ɵɵtext(16, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n            ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r22 = i0.ɵɵreference(12);
    const ctx_r6 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(5);
    i0.ɵɵclassProp("_fixed-top", ctx_r6.fixedPreview);
    i0.ɵɵproperty("ngStyle", ctx_r6.fixedPreview ? ctx_r6.objFixed : "");
    i0.ɵɵadvance(2);
    i0.ɵɵclassProp("determinate", !!ctx_r6.determinate);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r6.determinate)("ngIfElse", _r22);
    i0.ɵɵadvance(6);
    i0.ɵɵproperty("ngIf", ctx_r6.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_17_ng_container_2_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r28 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r28.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_17_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "div");
    i0.ɵɵtext(3, "\n                ");
    i0.ɵɵtemplate(4, AdaptBusyTplComponent_div_1_ng_container_17_ng_container_2_span_4_Template, 2, 1, "span", 50);
    i0.ɵɵtext(5, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r25 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("loader-special ", ctx_r25.loaderClass, "");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r25.message);
} }
function AdaptBusyTplComponent_div_1_ng_container_17_ng_template_4__svg_text_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(0, "text", 22);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r29 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate2("", ctx_r29.message, " ", ctx_r29.determinate ? ctx_r29.value + "%" : "", "");
} }
function AdaptBusyTplComponent_div_1_ng_container_17_ng_template_4_ng_template_19_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵnamespaceSVG();
} }
function AdaptBusyTplComponent_div_1_ng_container_17_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n              ");
    i0.ɵɵnamespaceSVG();
    i0.ɵɵelementStart(1, "svg", 51);
    i0.ɵɵtext(2, "\n                ");
    i0.ɵɵelementStart(3, "defs");
    i0.ɵɵtext(4, "\n                  ");
    i0.ɵɵelementStart(5, "pattern", 52);
    i0.ɵɵtext(6, "\n                    ");
    i0.ɵɵelement(7, "image", 53);
    i0.ɵɵtext(8, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n                ");
    i0.ɵɵelement(11, "circle", 18);
    i0.ɵɵtext(12, "\n                ");
    i0.ɵɵtemplate(13, AdaptBusyTplComponent_div_1_ng_container_17_ng_template_4__svg_text_13_Template, 2, 2, "text", 21);
    i0.ɵɵtext(14, "\n                ");
    i0.ɵɵelementStart(15, "circle", 18);
    i0.ɵɵtext(16, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n              ");
    i0.ɵɵtemplate(19, AdaptBusyTplComponent_div_1_ng_container_17_ng_template_4_ng_template_19_Template, 0, 0, "ng-template", 54);
    i0.ɵɵtext(20, "\n            ");
} if (rf & 2) {
    const ctx_r27 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵclassMapInterpolate1("loader-special__determinate ", ctx_r27.loaderClass, "");
    i0.ɵɵattribute("width", ctx_r27.SPECIAL_LOADER_DIAMETER)("height", ctx_r27.SPECIAL_LOADER_DIAMETER)("viewBox", "0 0" + " " + ctx_r27.SPECIAL_LOADER_DIAMETER + " " + ctx_r27.SPECIAL_LOADER_DIAMETER);
    i0.ɵɵadvance(4);
    i0.ɵɵattribute("viewBox", "0 0" + " " + ctx_r27.SPECIAL_LOADER_DIAMETER + " " + ctx_r27.SPECIAL_LOADER_DIAMETER);
    i0.ɵɵadvance(2);
    i0.ɵɵattribute("width", ctx_r27.SPECIAL_LOADER_DIAMETER)("height", ctx_r27.SPECIAL_LOADER_DIAMETER)("href", ctx_r27.loaderImg);
    i0.ɵɵadvance(4);
    i0.ɵɵattribute("r", ctx_r27.SPECIAL_LOADER_INNER_DIAMETER / 2)("style", ctx_r27.loaderImg ? "fill: url(#image)" : null, i0.ɵɵsanitizeStyle);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r27.message);
    i0.ɵɵadvance(2);
    i0.ɵɵstyleProp("stroke-dashoffset", ctx_r27.strokeDashOffset(ctx_r27.SPECIAL_LOADER_DIAMETER, ctx_r27.SPECIAL_STROKE_WIDTH), "px")("stroke-dasharray", ctx_r27.strokeCircumference((ctx_r27.SPECIAL_LOADER_DIAMETER - ctx_r27.SPECIAL_STROKE_WIDTH) / 2), "px");
    i0.ɵɵattribute("r", ctx_r27.circleRadius(ctx_r27.SPECIAL_LOADER_DIAMETER, ctx_r27.SPECIAL_STROKE_WIDTH))("stroke-width", ctx_r27.SPECIAL_STROKE_WIDTH);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r27.content);
} }
function AdaptBusyTplComponent_div_1_ng_container_17_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptBusyTplComponent_div_1_ng_container_17_ng_container_2_Template, 7, 4, "ng-container", 6);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptBusyTplComponent_div_1_ng_container_17_ng_template_4_Template, 21, 20, "ng-template", null, 49, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(6, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r26 = i0.ɵɵreference(5);
    const ctx_r7 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r7.determinate)("ngIfElse", _r26);
} }
function AdaptBusyTplComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 1, 2);
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵelementStart(3, "div", 3);
    i0.ɵɵtext(4, "\n        ");
    i0.ɵɵelementStart(5, "div", 4);
    i0.ɵɵtext(6, "\n          ");
    i0.ɵɵtemplate(7, AdaptBusyTplComponent_div_1_ng_container_7_Template, 7, 2, "ng-container", 5);
    i0.ɵɵtext(8, "\n          ");
    i0.ɵɵtemplate(9, AdaptBusyTplComponent_div_1_ng_container_9_Template, 7, 1, "ng-container", 5);
    i0.ɵɵtext(10, "\n          ");
    i0.ɵɵtemplate(11, AdaptBusyTplComponent_div_1_ng_container_11_Template, 9, 3, "ng-container", 5);
    i0.ɵɵtext(12, "\n          ");
    i0.ɵɵtemplate(13, AdaptBusyTplComponent_div_1_ng_container_13_Template, 10, 4, "ng-container", 5);
    i0.ɵɵtext(14, "\n          ");
    i0.ɵɵtemplate(15, AdaptBusyTplComponent_div_1_ng_container_15_Template, 19, 8, "ng-container", 5);
    i0.ɵɵtext(16, "\n          ");
    i0.ɵɵtemplate(17, AdaptBusyTplComponent_div_1_ng_container_17_Template, 7, 2, "ng-container", 5);
    i0.ɵɵtext(18, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(19, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(20, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵclassMap(ctx_r0.wrapperClass);
    i0.ɵɵproperty("ngClass", ctx_r0.getClass())("@flyInOut", ctx_r0.inOutType === ctx_r0.LoaderInOutType.fly && ctx_r0.type)("@fadeInOut", ctx_r0.inOutType === ctx_r0.LoaderInOutType.fade && ctx_r0.type);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngSwitch", ctx_r0.type);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.loaderType.APP);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.loaderType.PAGE);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.loaderType.SECTION);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.loaderType.INLINE);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.loaderType.LINE);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngSwitchCase", ctx_r0.loaderType.SPECIAL);
} }
const inactiveStyleForFlyInOut = style({
    opacity: 0,
    transform: 'translateY(-40px)'
});
const inactiveStyleForFadeInOut = style({
    opacity: 0,
});
const timingIn = ANIMATION_DURATION.fast + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionIn;
const timingOut = ANIMATION_DURATION.fast + ' ' + ANIMATION_TIMING_FUNCTION.base.transitionOut;
export class AdaptBusyTplComponent {
    constructor(tracker, zone, changeDetectorRef) {
        this.tracker = tracker;
        this.zone = zone;
        this.changeDetectorRef = changeDetectorRef;
        this.type = LoaderType.APP;
        this.LoaderInOutType = LoaderInOutType;
        this.inOutType = LoaderInOutType.fly;
        this.message = '';
        this.loaderType = LoaderType;
        this.fixedPreview = false;
        this.value = 0;
        this.isActive = false;
        this.sectionLoaderRatio = 100 / 3;
        this.sectionLoaderItems = 3;
        this.APP_LOADER_DIAMETER = 200;
        this.APP_STROKE_WIDTH = 4;
        this.SECTION_LOADER_DIAMETER = 11;
        this.INLINE_LOADER_DIAMETER = 15;
        this.INLINE_STROKE_WIDTH = 1;
        this.SPECIAL_LOADER_DIAMETER = 99;
        this.SPECIAL_LOADER_INNER_DIAMETER = 85;
        this.SPECIAL_STROKE_WIDTH = 3;
        this._fixedLoader = false;
        this._destroyed$ = new ReplaySubject(1);
        this._scrollHandler = () => {
            if (!this.previewSection || !this.previewContent) {
                return;
            }
            const { top, bottom } = this.previewContent.nativeElement.getBoundingClientRect();
            this.fixedPreview = Math.round(top - parseInt(this.offsetTop, 10)) < 0 && bottom > 0;
            if (this.fixedPreview) {
                this.zone.run(() => {
                    this.setWidth();
                });
            }
        };
    }
    ngOnInit() {
        this._fixedLoader = this.type === LoaderType.LINE && this.sticky;
        if (!this._fixedLoader) {
            return;
        }
        setTimeout(() => {
            this._scrollHandler();
        }, 0);
        this.zone.runOutsideAngular(() => {
            document.addEventListener('scroll', this._scrollHandler, true);
        });
    }
    ngAfterContentInit() {
        fromEvent(window, 'resize')
            .pipe(takeUntil(this._destroyed$), filter(() => this._fixedLoader && this.fixedPreview))
            .subscribe(() => {
            if (!this.previewSection || !this.previewContent) {
                return;
            }
            this.setWidth();
        });
        this.tracker.activeState$
            .pipe(takeUntil(this._destroyed$))
            .subscribe((value) => {
            this.isActive = value;
            this.changeDetectorRef.detectChanges();
        });
    }
    ngOnDestroy() {
        if (!this._fixedLoader) {
            return;
        }
        document.removeEventListener('scroll', this._scrollHandler, true);
        this._destroyed$.next(true);
        this._destroyed$.complete();
    }
    circleRadius(diameter, strokeWidth = 0) {
        return (diameter - strokeWidth) / 2;
    }
    strokeCircumference(radius) {
        return 2 * Math.PI * radius;
    }
    strokeDashOffset(diameter, strokeWidth) {
        const r = (diameter - strokeWidth) / 2;
        return this.strokeCircumference(r) * (100 - this.value) / 100;
    }
    circleRadiusSectionLoader(value) {
        const currentDiameter = (this.SECTION_LOADER_DIAMETER * value * this.sectionLoaderItems) / 100;
        return (currentDiameter > this.SECTION_LOADER_DIAMETER
            ? this.SECTION_LOADER_DIAMETER / 2
            : currentDiameter < 0 ? 0 : currentDiameter / 2);
    }
    circleOpacitySectionLoader(value) {
        return this.circleRadiusSectionLoader(value) / this.circleRadius(this.SECTION_LOADER_DIAMETER);
    }
    getClass() {
        return `ng-busy-${this.type}`;
    }
    setWidth() {
        const previewContentData = this.previewContent.nativeElement.getBoundingClientRect();
        this.objFixed = {
            top: this.offsetTop,
            left: `${previewContentData.left}px`,
            width: `${previewContentData.width}px`
        };
    }
}
AdaptBusyTplComponent.ɵfac = function AdaptBusyTplComponent_Factory(t) { return new (t || AdaptBusyTplComponent)(i0.ɵɵdirectiveInject(i1.PromiseTrackerService), i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef)); };
AdaptBusyTplComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptBusyTplComponent, selectors: [["ng-component"]], viewQuery: function AdaptBusyTplComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
        i0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.previewSection = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.previewContent = _t.first);
    } }, decls: 2, vars: 1, consts: [[3, "class", "ngClass", 4, "ngIf"], [3, "ngClass"], ["previewContent", ""], [1, "ng-busy-default-wrapper"], [1, "loader-container", 3, "ngSwitch"], [4, "ngSwitchCase"], [4, "ngIf", "ngIfElse"], ["appDeterminate", ""], [1, "loader-app"], ["class", "loader-app-title", 4, "ngIf"], [1, "loader-app-title"], ["preserveAspectRatio", "xMidYMid meet", "focusable", "false", "aria-hidden", "true", 1, "loader-app__determinate"], ["id", "gradient-first-half", "x1", "0", "y1", "50%", "x2", "0", "y2", "100%"], ["offset", "0%", 1, "gradient-color-initial"], ["offset", "100%", 1, "gradient-color-middle"], ["id", "gradient-second-half", "x1", "0%", "y1", "0%", "x2", "0%", "y2", "100%"], ["offset", "0%", 1, "gradient-color-final"], ["id", "myMask"], ["cx", "50%", "cy", "50%"], ["x", "100", "y", "0", "fill", "url(#gradient-first-half)", "mask", "url(#myMask)"], ["x", "0", "y", "0", "fill", "url(#gradient-second-half)", "mask", "url(#myMask)"], ["x", "50%", "y", "50%", "dominant-baseline", "central", "text-anchor", "middle", 4, "ngIf"], ["x", "50%", "y", "50%", "dominant-baseline", "central", "text-anchor", "middle"], [1, "loader-page"], [1, "loader-page-title"], ["sectionDeterminate", ""], ["class", "loader-section-title mt-3", 4, "ngIf"], [1, "loader-section"], [1, "loader-section__determinate"], [1, "dot_1"], [1, "dot_2"], [1, "dot_3"], [1, "loader-section-title", "mt-3"], ["inlineDeterminate", ""], [1, "loader-inline"], ["preserveAspectRatio", "xMidYMid meet", "focusable", "false", "aria-hidden", "true", 1, "loader-inline__determinate"], ["cx", "50%", "cy", "50%", 1, "loader-footprint"], [1, "a-loader"], ["previewSection", ""], [1, "a-loader_wrapper", 3, "ngStyle"], [1, "a-loader_bar", "adapt-alert-bar"], ["lineDeterminate", ""], ["class", "a-loader_message-wrapper", 4, "ngIf"], [1, "progress-bar-intermediate", "lay1"], [1, "progress-bar-intermediate", "lay2"], [1, "progress-bar-intermediate", "lay3"], [1, "progress-bar-determinate"], [1, "a-loader_message-wrapper"], [1, "a-loader_message"], ["specialDeterminate", ""], [4, "ngIf"], ["preserveAspectRatio", "xMidYMid meet", "focusable", "false", "aria-hidden", "true"], ["id", "image", "x", "0%", "y", "0%", "height", "100%", "width", "100%"], ["x", "0%", "y", "0%"], [3, "ngTemplateOutlet"]], template: function AdaptBusyTplComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtext(0, "\n    ");
        i0.ɵɵtemplate(1, AdaptBusyTplComponent_div_1_Template, 21, 12, "div", 0);
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isActive);
    } }, directives: [i2.NgIf, i2.NgClass, i2.NgSwitch, i2.NgSwitchCase, i2.NgStyle, i2.NgTemplateOutlet], styles: ["\n      .a-loader_wrapper._fixed-top {\n        position: fixed;\n      }\n    "], encapsulation: 2, data: { animation: [
            trigger('fadeInOut', getStyles(inactiveStyleForFadeInOut)),
            trigger('flyInOut', [
                ...getStyles(inactiveStyleForFlyInOut),
                // INLINE
                transition('void => inline', [
                    style({
                        opacity: 0,
                        transform: 'translateX(-100%)'
                    }),
                    animate(timingIn, keyframes([
                        style({ opacity: 0, transform: 'translateX(-100%)', offset: 0 }),
                        style({ opacity: 0, transform: 'translateX(-20%)', offset: 0.8 }),
                        style({ opacity: 1, transform: 'translateX(0%)', offset: 1.0 }),
                    ]))
                ]),
                transition('inline => void', [
                    style({
                        opacity: 1,
                        transform: 'translateX(0%)'
                    }),
                    animate(timingOut, keyframes([
                        style({ opacity: 1, transform: 'translateX(0%)', offset: 0 }),
                        style({ opacity: 0, transform: 'translateX(-30%)', offset: 0.3 }),
                        style({ opacity: 0, transform: 'translateX(-100%)', offset: 1.0 }),
                    ]))
                ])
            ])
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptBusyTplComponent, [{
        type: Component,
        args: [{
                template: `
    <div [class]="wrapperClass" [ngClass]="getClass()" *ngIf="isActive"
         [@flyInOut]="inOutType === LoaderInOutType.fly && type"
         [@fadeInOut]="inOutType === LoaderInOutType.fade && type" #previewContent>
      <div class="ng-busy-default-wrapper">
        <div class="loader-container" [ngSwitch]="type">
          <ng-container *ngSwitchCase="loaderType.APP">
            <ng-container *ngIf="!determinate; else appDeterminate">
              <div class="loader-app"></div>
              <h1 class="loader-app-title" *ngIf="message">{{message}}</h1>
            </ng-container>
            <ng-template #appDeterminate>
              <svg class="loader-app__determinate"
                   [attr.width]="APP_LOADER_DIAMETER"
                   [attr.height]="APP_LOADER_DIAMETER"
                   [attr.viewBox]="'0 0' + ' ' + APP_LOADER_DIAMETER + ' ' + APP_LOADER_DIAMETER"
                   preserveAspectRatio="xMidYMid meet"
                   focusable="false" aria-hidden="true">
                <defs>
                  <linearGradient id="gradient-first-half" x1="0" y1="50%" x2="0" y2="100%">
                    <stop class="gradient-color-initial" offset="0%"/>
                    <stop class="gradient-color-middle" offset="100%"/>
                  </linearGradient>
                  <linearGradient id="gradient-second-half" x1="0%" y1="0%" x2="0%" y2="100%">
                    <stop class="gradient-color-final" offset="0%"/>
                    <stop class="gradient-color-middle" offset="100%"/>
                  </linearGradient>
                  <mask id="myMask">
                    <circle cx="50%" cy="50%"
                            [attr.r]="circleRadius(APP_LOADER_DIAMETER, APP_STROKE_WIDTH)"
                            [attr.stroke-width]="APP_STROKE_WIDTH"
                            [style.stroke-dashoffset.px]="strokeDashOffset(APP_LOADER_DIAMETER, APP_STROKE_WIDTH)"
                            [style.stroke-dasharray.px]="strokeCircumference((APP_LOADER_DIAMETER - APP_STROKE_WIDTH) / 2)"/>
                  </mask>
                </defs>
                <rect x="100" y="0" [attr.width]="APP_LOADER_DIAMETER / 2" [attr.height]="APP_LOADER_DIAMETER" fill="url(#gradient-first-half)" mask="url(#myMask)"/>
                <rect x="0" y="0" [attr.width]="APP_LOADER_DIAMETER / 2" [attr.height]="APP_LOADER_DIAMETER" fill="url(#gradient-second-half)" mask="url(#myMask)"/>
                <text x="50%" y="50%" dominant-baseline="central" text-anchor="middle" *ngIf="message">{{message}} {{determinate ? value + '%' : ''}}</text>
              </svg>
            </ng-template>
          </ng-container>
          <ng-container *ngSwitchCase="loaderType.PAGE">
            <div class="loader-page"></div>
            <div class="loader-page-title">{{message}}</div>
          </ng-container>
          <ng-container *ngSwitchCase="loaderType.SECTION">
            <ng-container *ngIf="!determinate; else sectionDeterminate">
              <div class="loader-section"></div>
            </ng-container>
            <ng-template #sectionDeterminate>
              <div class="loader-section__determinate">
                <div class="dot_1">
                  <svg [attr.viewBox]="'0 0' + ' ' + SECTION_LOADER_DIAMETER + ' ' + SECTION_LOADER_DIAMETER">
                    <circle cx="50%" cy="50%"
                            [attr.r]="circleRadiusSectionLoader(value)"
                            [attr.fill-opacity]="circleOpacitySectionLoader(value)"></circle>
                  </svg>
                </div>
                <div class="dot_2">
                  <svg [attr.viewBox]="'0 0' + ' ' + SECTION_LOADER_DIAMETER + ' ' + SECTION_LOADER_DIAMETER">
                    <circle cx="50%" cy="50%"
                            [attr.r]="circleRadiusSectionLoader(value - sectionLoaderRatio)"
                            [attr.fill-opacity]="circleOpacitySectionLoader(value - sectionLoaderRatio)"></circle>
                  </svg>
                </div>
                <div class="dot_3">
                  <svg [attr.viewBox]="'0 0' + ' ' + SECTION_LOADER_DIAMETER + ' ' + SECTION_LOADER_DIAMETER">
                    <circle cx="50%" cy="50%"
                            [attr.r]="circleRadiusSectionLoader(value - 2 * sectionLoaderRatio)"
                            [attr.fill-opacity]="circleOpacitySectionLoader(value - 2 * sectionLoaderRatio)"></circle>
                  </svg>
                </div>
              </div>
            </ng-template>
            <div class="loader-section-title mt-3" *ngIf="message">
              {{message}} {{determinate ? value + '%' : ''}}
            </div>
          </ng-container>
          <ng-container *ngSwitchCase="loaderType.INLINE">
            <ng-container *ngIf="!determinate; else inlineDeterminate">
              <div class="loader-inline"></div>
            </ng-container>
            <ng-template #inlineDeterminate>
              <svg class="loader-inline__determinate"
                   [attr.width]="INLINE_LOADER_DIAMETER"
                   [attr.height]="INLINE_LOADER_DIAMETER"
                   [attr.viewBox]="'0 0' + ' ' + INLINE_LOADER_DIAMETER + ' ' + INLINE_LOADER_DIAMETER"
                   preserveAspectRatio="xMidYMid meet"
                   focusable="false" aria-hidden="true">
                <circle class="loader-footprint" cx="50%" cy="50%"
                        [attr.r]="circleRadius(INLINE_LOADER_DIAMETER, INLINE_STROKE_WIDTH)"
                        [attr.stroke-width]="INLINE_STROKE_WIDTH">
                </circle>
                <circle cx="50%" cy="50%"
                        [attr.r]="circleRadius(INLINE_LOADER_DIAMETER, INLINE_STROKE_WIDTH)"
                        [attr.stroke-width]="INLINE_STROKE_WIDTH"
                        [style.stroke-dashoffset.px]="strokeDashOffset(INLINE_LOADER_DIAMETER, INLINE_STROKE_WIDTH)"
                        [style.stroke-dasharray.px]="strokeCircumference((INLINE_LOADER_DIAMETER - INLINE_STROKE_WIDTH) / 2)">
                </circle>
              </svg>
            </ng-template>
            <span>{{message}} {{determinate ? value + '%' : ''}}</span>
          </ng-container>
          <ng-container *ngSwitchCase="loaderType.LINE">
            <div class="a-loader" #previewSection>
              <div class="a-loader_wrapper"
                   [class._fixed-top]="fixedPreview"
                   [ngStyle]="fixedPreview ? objFixed: ''">
                <div class="a-loader_bar adapt-alert-bar" [class.determinate]="!!determinate">
                  <ng-container *ngIf="!determinate; else lineDeterminate">
                    <div class="progress-bar-intermediate lay1"></div>
                    <div class="progress-bar-intermediate lay2"></div>
                    <div class="progress-bar-intermediate lay3"></div>
                  </ng-container>

                  <ng-template #lineDeterminate>
                    <div class="progress-bar-determinate" [style.width.%]="value"></div>
                  </ng-template>
                </div>

                <div class="a-loader_message-wrapper" *ngIf="message">
                  <div class="a-loader_message">{{message}} {{determinate ? value + '%' : ''}}</div>
                </div>
              </div>
            </div>
          </ng-container>
          <ng-container *ngSwitchCase="loaderType.SPECIAL">
            <ng-container *ngIf="!determinate; else specialDeterminate">
              <div class="loader-special {{loaderClass}}">
                <span *ngIf="message">{{message}}</span>
              </div>
            </ng-container>
            <ng-template #specialDeterminate>
              <svg class="loader-special__determinate {{loaderClass}}"
                   [attr.width]="SPECIAL_LOADER_DIAMETER"
                   [attr.height]="SPECIAL_LOADER_DIAMETER"
                   [attr.viewBox]="'0 0' + ' ' + SPECIAL_LOADER_DIAMETER + ' ' + SPECIAL_LOADER_DIAMETER"
                   preserveAspectRatio="xMidYMid meet"
                   focusable="false" aria-hidden="true">
                <defs>
                  <pattern id="image" x="0%" y="0%" height="100%" width="100%"
                           [attr.viewBox]="'0 0' + ' ' + SPECIAL_LOADER_DIAMETER + ' ' + SPECIAL_LOADER_DIAMETER">
                    <image x="0%" y="0%" [attr.width]="SPECIAL_LOADER_DIAMETER" [attr.height]="SPECIAL_LOADER_DIAMETER" [attr.href]="loaderImg"></image>
                  </pattern>
                </defs>
                <circle cx="50%" cy="50%" [attr.r]="SPECIAL_LOADER_INNER_DIAMETER / 2" [attr.style]="loaderImg ? 'fill: url(#image)' : null"></circle>
                <text x="50%" y="50%" dominant-baseline="central" text-anchor="middle" *ngIf="message">{{message}} {{determinate ? value + '%' : ''}}</text>
                <circle cx="50%" cy="50%"
                        [attr.r]="circleRadius(SPECIAL_LOADER_DIAMETER, SPECIAL_STROKE_WIDTH)"
                        [attr.stroke-width]="SPECIAL_STROKE_WIDTH"
                        [style.stroke-dashoffset.px]="strokeDashOffset(SPECIAL_LOADER_DIAMETER, SPECIAL_STROKE_WIDTH)"
                        [style.stroke-dasharray.px]="strokeCircumference((SPECIAL_LOADER_DIAMETER - SPECIAL_STROKE_WIDTH) / 2)">
                </circle>
              </svg>
              <ng-template [ngTemplateOutlet]="content"></ng-template>
            </ng-template>
          </ng-container>
        </div>
      </div>
    </div>`,
                animations: [
                    trigger('fadeInOut', getStyles(inactiveStyleForFadeInOut)),
                    trigger('flyInOut', [
                        ...getStyles(inactiveStyleForFlyInOut),
                        // INLINE
                        transition('void => inline', [
                            style({
                                opacity: 0,
                                transform: 'translateX(-100%)'
                            }),
                            animate(timingIn, keyframes([
                                style({ opacity: 0, transform: 'translateX(-100%)', offset: 0 }),
                                style({ opacity: 0, transform: 'translateX(-20%)', offset: 0.8 }),
                                style({ opacity: 1, transform: 'translateX(0%)', offset: 1.0 }),
                            ]))
                        ]),
                        transition('inline => void', [
                            style({
                                opacity: 1,
                                transform: 'translateX(0%)'
                            }),
                            animate(timingOut, keyframes([
                                style({ opacity: 1, transform: 'translateX(0%)', offset: 0 }),
                                style({ opacity: 0, transform: 'translateX(-30%)', offset: 0.3 }),
                                style({ opacity: 0, transform: 'translateX(-100%)', offset: 1.0 }),
                            ]))
                        ])
                    ])
                ],
                styles: [
                    `
      .a-loader_wrapper._fixed-top {
        position: fixed;
      }
    `
                ],
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.PromiseTrackerService }, { type: i0.NgZone }, { type: i0.ChangeDetectorRef }]; }, { previewSection: [{
            type: ViewChild,
            args: ['previewSection', { static: false }]
        }], previewContent: [{
            type: ViewChild,
            args: ['previewContent', { static: false }]
        }] }); })();
export function getStyles(inactiveStyle) {
    // APP
    return [
        transition('void => app', [
            inactiveStyle,
            animate(timingIn)
        ]),
        transition('app => void', [
            animate(timingOut, inactiveStyle)
        ]),
        // PAGE
        transition('void => page', [
            inactiveStyle,
            animate(timingIn)
        ]),
        transition('page => void', [
            animate(timingOut, inactiveStyle)
        ]),
        // SECTION
        transition('void => section', [
            inactiveStyle,
            animate(timingIn)
        ]),
        transition('section => void', [
            animate(timingOut, inactiveStyle)
        ]),
        // LINE
        transition('void => line', [
            inactiveStyle,
            animate(timingIn)
        ]),
        transition('line => void', [
            animate(timingOut, inactiveStyle)
        ])
    ];
}
