// @bmc-ux/adapt-angular/esm2015/busy-loader/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptBusyDirective } from './busy.directive';
import { AdaptBusyBackdropComponent } from './busy-backdrop.component';
import { AdaptBusyComponent } from './busy.component';
import { AdaptBusyConfig } from './busy-config';
import { AdaptBusyTplComponent } from './busy-tpl.component';
import { AdaptBusyHostDirective } from './busy-ad-host.directive';
import { AdaptInlineStandaloneDirective } from './busy-inline-standalone.directive';
import * as i0 from "@angular/core";
export class AdaptBusyModule {
    static forRoot(config) {
        return {
            ngModule: AdaptBusyModule,
            providers: [{ provide: AdaptBusyConfig, useValue: config }]
        };
    }
}
AdaptBusyModule.ɵfac = function AdaptBusyModule_Factory(t) { return new (t || AdaptBusyModule)(); };
AdaptBusyModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptBusyModule });
AdaptBusyModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptBusyModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptBusyDirective,
                    AdaptBusyComponent,
                    AdaptBusyBackdropComponent,
                    AdaptBusyTplComponent,
                    AdaptBusyHostDirective,
                    AdaptInlineStandaloneDirective
                ],
                exports: [AdaptBusyDirective, AdaptInlineStandaloneDirective],
                entryComponents: [
                    AdaptBusyComponent,
                    AdaptBusyBackdropComponent,
                    AdaptBusyTplComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptBusyModule, { declarations: [AdaptBusyDirective,
        AdaptBusyComponent,
        AdaptBusyBackdropComponent,
        AdaptBusyTplComponent,
        AdaptBusyHostDirective,
        AdaptInlineStandaloneDirective], imports: [CommonModule], exports: [AdaptBusyDirective, AdaptInlineStandaloneDirective] }); })();
export * from './busy-config';
export * from './busy-backdrop.component';
export * from './busy-ad-host.directive';
export * from './promise-tracker.service';
export * from './busy.component';
export * from './busy.service';
export * from './busy.directive';
export * from './busy-inline-standalone.directive';
