// @bmc-ux/adapt-angular/esm2015/common/uploader/services/send.file.service.js
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class SendFileService {
    constructor(http) {
        this.http = http;
        this.url = '/default-path';
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    sendFile(requestBody, uploadAsOneChunk = true, url) {
        const headers = new HttpHeaders({
            'Content-Type': 'multipart/form-data',
            'Accept': '*/*'
        });
        const formData = this.convertObjectToFormData(requestBody);
        return uploadAsOneChunk
            ? this.http.post(url || this.url, formData, {
                headers,
                reportProgress: true,
                observe: 'events',
                responseType: 'json'
            })
            : this.http.post(url || this.url, requestBody);
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    deleteFile(file, url) {
        return this.http.delete(url || this.url, file);
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line  */
    responseCallback(response, file) {
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    convertObjectToFormData(object) {
        const formData = new FormData();
        Object.keys(object).forEach(key => formData.append(key, object[key]));
        return formData;
    }
}
SendFileService.ɵfac = function SendFileService_Factory(t) { return new (t || SendFileService)(i0.ɵɵinject(i1.HttpClient)); };
SendFileService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: SendFileService, factory: SendFileService.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(SendFileService, [{
        type: Injectable
    }], function () { return [{ type: i1.HttpClient }]; }, null); })();
