// @bmc-ux/adapt-angular/esm2015/rx/rx-uploader/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxUploaderComponent } from './adapt-rx-uploader.component';
import { AdaptUploadItemModule } from '../../uploader-item/index';
import { AdaptFileSizePipeModule } from '../../common/uploader/pipes/file-size.pipe';
import { AdaptFileTypesPipeModule } from '../../common/uploader/pipes/file-type.pipe';
import { SendFileService } from '../../common/uploader/services/send.file.service';
import { AdaptTooltipModule } from '../../tooltip/index';
import { AdaptButtonModule } from '../../button/index';
import { AdaptAlertModule } from '../../alert/index';
import { AdaptPopoverModule } from '../../popover/index';
import { AdaptCollapseModule } from '../../collapse/index';
import { AdaptEmptyStateModule } from '../../empty-state/index';
import * as i0 from "@angular/core";
export class AdaptRxUploaderModule {
    static forRoot(config) {
        return {
            ngModule: AdaptRxUploaderModule,
            providers: [
                // TODO: investigate how service logic should work and rewrite -
                //  what exactly confusing: url (for post and delete methods) and/or service in the module?
                (config && config.uploader) ? config.uploader : { provide: 'UploaderService', useClass: SendFileService },
                { provide: SendFileService, useClass: SendFileService }
            ]
        };
    }
}
AdaptRxUploaderModule.ɵfac = function AdaptRxUploaderModule_Factory(t) { return new (t || AdaptRxUploaderModule)(); };
AdaptRxUploaderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxUploaderModule });
AdaptRxUploaderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule,
            AdaptButtonModule,
            AdaptTooltipModule,
            AdaptUploadItemModule,
            AdaptFileSizePipeModule,
            AdaptFileTypesPipeModule,
            AdaptAlertModule,
            AdaptPopoverModule,
            AdaptCollapseModule,
            AdaptEmptyStateModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxUploaderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule,
                    AdaptButtonModule,
                    AdaptTooltipModule,
                    AdaptUploadItemModule,
                    AdaptFileSizePipeModule,
                    AdaptFileTypesPipeModule,
                    AdaptAlertModule,
                    AdaptPopoverModule,
                    AdaptCollapseModule,
                    AdaptEmptyStateModule
                ],
                declarations: [
                    AdaptRxUploaderComponent
                ],
                exports: [
                    AdaptRxUploaderComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxUploaderModule, { declarations: [AdaptRxUploaderComponent], imports: [CommonModule,
        AdaptRxFormControlModule,
        AdaptButtonModule,
        AdaptTooltipModule,
        AdaptUploadItemModule,
        AdaptFileSizePipeModule,
        AdaptFileTypesPipeModule,
        AdaptAlertModule,
        AdaptPopoverModule,
        AdaptCollapseModule,
        AdaptEmptyStateModule], exports: [AdaptRxUploaderComponent] }); })();
export * from './adapt-rx-uploader.component';
export * from './adapt-rx-uploader-config.service';
