// @helix/platform/esm2015/process/designer/process-designer-page/process-designer-page.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ProcessDesignerModule } from '../process-designer';
import { ProcessDesignerPageComponent } from './process-designer-page.component';
import * as i0 from "@angular/core";
export class ProcessDesignerPageModule {
}
ProcessDesignerPageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ProcessDesignerPageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ProcessDesignerPageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ProcessDesignerPageModule, declarations: [ProcessDesignerPageComponent], imports: [CommonModule, ProcessDesignerModule], exports: [ProcessDesignerPageComponent] });
ProcessDesignerPageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ProcessDesignerPageModule, imports: [[CommonModule, ProcessDesignerModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ProcessDesignerPageModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [ProcessDesignerPageComponent],
                    imports: [CommonModule, ProcessDesignerModule],
                    exports: [ProcessDesignerPageComponent]
                }]
        }] });
