// @helix/platform/esm2015/shared/components/unknown-application/unknown-application.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RxUnknownApplicationComponent } from './unknown-application.component';
import * as i0 from "@angular/core";
export class RxUnknownApplicationModule {
}
RxUnknownApplicationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxUnknownApplicationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxUnknownApplicationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxUnknownApplicationModule, declarations: [RxUnknownApplicationComponent], imports: [CommonModule], exports: [RxUnknownApplicationComponent] });
RxUnknownApplicationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxUnknownApplicationModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxUnknownApplicationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxUnknownApplicationComponent],
                    imports: [CommonModule],
                    exports: [RxUnknownApplicationComponent]
                }]
        }] });
