// @helix/platform/esm2015/shared/components/iframe/iframe.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RxIframeComponent } from './iframe.component';
import * as i0 from "@angular/core";
export class RxIframeModule {
}
RxIframeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxIframeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxIframeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxIframeModule, declarations: [RxIframeComponent], imports: [CommonModule], exports: [RxIframeComponent] });
RxIframeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxIframeModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxIframeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [RxIframeComponent],
                    exports: [RxIframeComponent]
                }]
        }] });
