// @bmc-ux/adapt-angular/esm2015/rx/rx-counter/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptTooltipModule } from '../../tooltip/index';
import { AdaptDragonInputHandlerModule } from '../../dragon-input-handler/index';
import { AdaptRxCounterComponent } from './adapt-rx-counter.component';
import * as i0 from "@angular/core";
export class AdaptRxCounterModule {
}
AdaptRxCounterModule.ɵfac = function AdaptRxCounterModule_Factory(t) { return new (t || AdaptRxCounterModule)(); };
AdaptRxCounterModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxCounterModule });
AdaptRxCounterModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptTooltipModule,
            AdaptRxFormControlModule,
            AdaptDragonInputHandlerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxCounterModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptTooltipModule,
                    AdaptRxFormControlModule,
                    AdaptDragonInputHandlerModule
                ],
                declarations: [
                    AdaptRxCounterComponent
                ],
                exports: [
                    AdaptRxCounterComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxCounterModule, { declarations: [AdaptRxCounterComponent], imports: [CommonModule,
        AdaptTooltipModule,
        AdaptRxFormControlModule,
        AdaptDragonInputHandlerModule], exports: [AdaptRxCounterComponent] }); })();
export * from './adapt-rx-counter.component';
export * from './adapt-rx-counter.models';
