// @bmc-ux/adapt-angular/esm2015/highlight/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptHighlightDirective } from './highlight';
import * as i0 from "@angular/core";
export class AdaptHighlightModule {
}
AdaptHighlightModule.ɵfac = function AdaptHighlightModule_Factory(t) { return new (t || AdaptHighlightModule)(); };
AdaptHighlightModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptHighlightModule });
AdaptHighlightModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptHighlightModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptHighlightDirective],
                exports: [AdaptHighlightDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptHighlightModule, { declarations: [AdaptHighlightDirective], imports: [CommonModule], exports: [AdaptHighlightDirective] }); })();
export * from './highlight';
