// @bmc-ux/adapt-angular/esm2015/rx/rx-timeslot/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptButtonModule } from '../../button/index';
import { AdaptRxTimeSlotComponent } from './adapt-rx-timeslot.component';
import * as i0 from "@angular/core";
export class AdaptRxTimeSlotModule {
}
AdaptRxTimeSlotModule.ɵfac = function AdaptRxTimeSlotModule_Factory(t) { return new (t || AdaptRxTimeSlotModule)(); };
AdaptRxTimeSlotModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxTimeSlotModule });
AdaptRxTimeSlotModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule,
            AdaptButtonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxTimeSlotModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule,
                    AdaptButtonModule
                ],
                exports: [AdaptRxTimeSlotComponent],
                declarations: [AdaptRxTimeSlotComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxTimeSlotModule, { declarations: [AdaptRxTimeSlotComponent], imports: [CommonModule,
        AdaptRxFormControlModule,
        AdaptButtonModule], exports: [AdaptRxTimeSlotComponent] }); })();
export * from './adapt-rx-timeslot.component';
