// @helix/platform/esm2015/view/components/field-components/textarea-field/design/textarea-field-design.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptRxTextareaModule } from '@bmc-ux/adapt-angular';
import { TextareaFieldDesignComponent } from './textarea-field-design.component';
import * as i0 from "@angular/core";
export class TextareaFieldDesignModule {
}
TextareaFieldDesignModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: TextareaFieldDesignModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TextareaFieldDesignModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: TextareaFieldDesignModule, declarations: [TextareaFieldDesignComponent], imports: [CommonModule, AdaptRxTextareaModule, FormsModule] });
TextareaFieldDesignModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: TextareaFieldDesignModule, imports: [[CommonModule, AdaptRxTextareaModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: TextareaFieldDesignModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AdaptRxTextareaModule, FormsModule],
                    declarations: [TextareaFieldDesignComponent],
                    entryComponents: [TextareaFieldDesignComponent]
                }]
        }] });
