// @bmc-ux/adapt-angular/esm2015/color-picker/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptDropdownModule } from '../dropdown/index';
import { ColorComponent } from './color';
import { AdaptColorPickerComponent } from './color-picker.component';
import { AdaptDragonInputHandlerModule } from '../dragon-input-handler/index';
import { AdaptButtonGroupModule } from '../button-group/index';
import { AdaptTabsModule } from '../tabs/index';
import { AdaptEmptyStateModule } from '../empty-state/index';
import * as i0 from "@angular/core";
export class AdaptColorPickerModule {
}
AdaptColorPickerModule.ɵfac = function AdaptColorPickerModule_Factory(t) { return new (t || AdaptColorPickerModule)(); };
AdaptColorPickerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptColorPickerModule });
AdaptColorPickerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptDropdownModule,
            FormsModule,
            AdaptDragonInputHandlerModule,
            AdaptButtonGroupModule,
            AdaptTabsModule,
            AdaptEmptyStateModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptColorPickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptDropdownModule,
                    FormsModule,
                    AdaptDragonInputHandlerModule,
                    AdaptButtonGroupModule,
                    AdaptTabsModule,
                    AdaptEmptyStateModule
                ],
                declarations: [AdaptColorPickerComponent, ColorComponent],
                exports: [AdaptColorPickerComponent, ColorComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptColorPickerModule, { declarations: [AdaptColorPickerComponent, ColorComponent], imports: [CommonModule,
        AdaptDropdownModule,
        FormsModule,
        AdaptDragonInputHandlerModule,
        AdaptButtonGroupModule,
        AdaptTabsModule,
        AdaptEmptyStateModule], exports: [AdaptColorPickerComponent, ColorComponent] }); })();
export * from './color-picker.component';
export * from './color';
