// @bmc-ux/adapt-angular/esm2015/not-found/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptNotFoundPageComponent } from './not-found.component';
import { AdaptButtonModule } from '../button/index';
import { RouterModule } from '@angular/router';
import * as i0 from "@angular/core";
export class AdaptNotFoundModule {
}
AdaptNotFoundModule.ɵfac = function AdaptNotFoundModule_Factory(t) { return new (t || AdaptNotFoundModule)(); };
AdaptNotFoundModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptNotFoundModule });
AdaptNotFoundModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptButtonModule, RouterModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptNotFoundModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptButtonModule, RouterModule],
                exports: [AdaptNotFoundPageComponent],
                declarations: [AdaptNotFoundPageComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptNotFoundModule, { declarations: [AdaptNotFoundPageComponent], imports: [CommonModule, AdaptButtonModule, RouterModule], exports: [AdaptNotFoundPageComponent] }); })();
export * from './not-found.component';
export * from './not-found.model';
