// @bmc-ux/adapt-angular/esm2015/sidebar/sidebar-column/sidebar-column.component.js
import { Component, ContentChildren, HostBinding, QueryList, ViewEncapsulation } from '@angular/core';
import { AdaptSidebarItemComponent } from '../sidebar-item/sidebar-item.component';
import * as i0 from "@angular/core";
const _c0 = ["*"];
export class AdaptSidebarColumnComponent {
    constructor() {
        this.defaultClass = true;
    }
}
AdaptSidebarColumnComponent.ɵfac = function AdaptSidebarColumnComponent_Factory(t) { return new (t || AdaptSidebarColumnComponent)(); };
AdaptSidebarColumnComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptSidebarColumnComponent, selectors: [["adapt-sidebar-column"]], contentQueries: function AdaptSidebarColumnComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, AdaptSidebarItemComponent, 4);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.panels = _t);
    } }, hostVars: 2, hostBindings: function AdaptSidebarColumnComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("adapt-sidebar-column", ctx.defaultClass);
    } }, ngContentSelectors: _c0, decls: 5, vars: 0, consts: [[1, "adapt-sidebar-items"]], template: function AdaptSidebarColumnComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtext(0, "\n    ");
        i0.ɵɵelementStart(1, "div", 0);
        i0.ɵɵtext(2, "\n      ");
        i0.ɵɵprojection(3);
        i0.ɵɵtext(4, "\n    ");
        i0.ɵɵelementEnd();
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSidebarColumnComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-sidebar-column',
                template: `
    <div class="adapt-sidebar-items">
      <ng-content></ng-content>
    </div>`,
                encapsulation: ViewEncapsulation.None,
            }]
    }], null, { defaultClass: [{
            type: HostBinding,
            args: ['class.adapt-sidebar-column']
        }], panels: [{
            type: ContentChildren,
            args: [AdaptSidebarItemComponent]
        }] }); })();
