// @bmc-ux/obsolete/esm2015/lib/radiobutton/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptIconModule, AdaptTooltipModule } from '@bmc-ux/adapt-angular';
import { AdaptRadioButtonComponent } from './radiobutton.component';
import * as i0 from "@angular/core";
export class AdaptRadioButtonModule {
}
AdaptRadioButtonModule.ɵfac = function AdaptRadioButtonModule_Factory(t) { return new (t || AdaptRadioButtonModule)(); };
AdaptRadioButtonModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRadioButtonModule });
AdaptRadioButtonModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, FormsModule, AdaptIconModule, AdaptTooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRadioButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, AdaptIconModule, AdaptTooltipModule],
                exports: [AdaptRadioButtonComponent],
                declarations: [AdaptRadioButtonComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRadioButtonModule, { declarations: [AdaptRadioButtonComponent], imports: [CommonModule, FormsModule, AdaptIconModule, AdaptTooltipModule], exports: [AdaptRadioButtonComponent] }); })();
export * from './radiobutton.component';
