// @bmc-ux/adapt-angular/esm2015/carousel/carousel.component.js
import { AdaptDeviceDetectionService } from '../common/device-detection.service';
import { ANIMATION_NAME, ANIMATION_TIMING_FUNCTION } from '../common/animation';
import { AdaptTranslateService } from '../common/i18n/index';
import { Component, ContentChildren, ElementRef, HostListener, Input, NgZone, QueryList, ViewEncapsulation } from '@angular/core';
import { asapScheduler, ReplaySubject, timer } from 'rxjs';
import { take, takeUntil } from 'rxjs/operators';
import { animate, query, style, transition, trigger, useAnimation } from '@angular/animations';
import { Directionality } from '@angular/cdk/bidi';
import { SiemaService } from './siema.service';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "./siema.service";
import * as i2 from "../common/device-detection.service";
import * as i3 from "@angular/cdk/bidi";
import * as i4 from "../common/i18n/index";
import * as i5 from "../common/radar/adapt-radar";
import * as i6 from "@angular/common";
const _c0 = ["*"];
function AdaptCarouselComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵelement(2, "div", 5);
    i0.ɵɵtext(3, "\n");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    i0.ɵɵproperty("@backdropAnimation", undefined);
} }
function AdaptCarouselComponent_ng_container_11_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 7);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelementStart(2, "button", 8);
    i0.ɵɵlistener("click", function AdaptCarouselComponent_ng_container_11_div_2_Template_button_click_2_listener() { i0.ɵɵrestoreView(_r6); const ctx_r5 = i0.ɵɵnextContext(2); return ctx_r5.siema.prev(1); });
    i0.ɵɵtext(3, "\n          ");
    i0.ɵɵelement(4, "span", 9);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementStart(6, "span", 10);
    i0.ɵɵtext(7);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n        ");
    i0.ɵɵelementStart(10, "button", 11);
    i0.ɵɵlistener("click", function AdaptCarouselComponent_ng_container_11_div_2_Template_button_click_10_listener() { i0.ɵɵrestoreView(_r6); const ctx_r7 = i0.ɵɵnextContext(2); return ctx_r7.siema.next(1); });
    i0.ɵɵtext(11, "\n          ");
    i0.ɵɵelement(12, "span", 12);
    i0.ɵɵtext(13, "\n          ");
    i0.ɵɵelementStart(14, "span", 10);
    i0.ɵɵtext(15);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(16, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(7);
    i0.ɵɵtextInterpolate(ctx_r4.texts.previous);
    i0.ɵɵadvance(8);
    i0.ɵɵtextInterpolate(ctx_r4.texts.next);
} }
function AdaptCarouselComponent_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵtemplate(2, AdaptCarouselComponent_ng_container_11_div_2_Template, 18, 2, "div", 6);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r1.slides.toArray().length > 1);
} }
function AdaptCarouselComponent_div_13_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r12 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 15);
    i0.ɵɵlistener("click", function AdaptCarouselComponent_div_13_button_2_Template_button_click_0_listener() { const restoredCtx = i0.ɵɵrestoreView(_r12); const index_r10 = restoredCtx.index; const ctx_r11 = i0.ɵɵnextContext(2); return ctx_r11.onGoTo(index_r10); });
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelementStart(2, "span", 10);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const index_r10 = ctx.index;
    const ctx_r8 = i0.ɵɵnextContext(2);
    i0.ɵɵclassProp("is-active", ctx_r8.siema.currentSlide === index_r10);
    i0.ɵɵattribute("aria-pressed", ctx_r8.siema.currentSlide === index_r10);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(index_r10 + 1);
} }
const _c1 = function (a0) { return { "is-mobile": a0 }; };
function AdaptCarouselComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 13);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵtemplate(2, AdaptCarouselComponent_div_13_button_2_Template, 5, 4, "button", 14);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(2, _c1, ctx_r2.isMobile && !ctx_r2.isMobileVisible));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", ctx_r2.slides);
} }
function AdaptCarouselComponent_a_17_Template(rf, ctx) { if (rf & 1) {
    const _r14 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 16);
    i0.ɵɵlistener("tap", function AdaptCarouselComponent_a_17_Template_a_tap_0_listener($event) { i0.ɵɵrestoreView(_r14); const ctx_r13 = i0.ɵɵnextContext(); return ctx_r13.closeBackDrop($event); })("keyup.esc", function AdaptCarouselComponent_a_17_Template_a_keyup_esc_0_listener($event) { i0.ɵɵrestoreView(_r14); const ctx_r15 = i0.ɵɵnextContext(); return ctx_r15.closeBackDrop($event); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    i0.ɵɵproperty("@closeIconAnimation", undefined);
} }
const _c2 = [[["adapt-carousel-slide"]]];
const _c3 = ["adapt-carousel-slide"];
export class AdaptCarouselSlideComponent {
}
AdaptCarouselSlideComponent.ɵfac = function AdaptCarouselSlideComponent_Factory(t) { return new (t || AdaptCarouselSlideComponent)(); };
AdaptCarouselSlideComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptCarouselSlideComponent, selectors: [["adapt-carousel-slide"]], hostVars: 2, hostBindings: function AdaptCarouselSlideComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵclassProp("a-carousel__inner", true);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function AdaptCarouselSlideComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCarouselSlideComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-carousel-slide',
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {
                    '[class.a-carousel__inner]': 'true'
                },
                template: '<ng-content></ng-content>'
            }]
    }], null, null); })();
export class AdaptCarouselComponent extends AdaptRadarAngularGenericDirective {
    constructor(siema, dt, dir, _ngZone, ts, _adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptCarousel);
        this.siema = siema;
        this.dt = dt;
        this.dir = dir;
        this._ngZone = _ngZone;
        this.ts = ts;
        this.isMobileVisible = false;
        this.defaultTexts = {};
        /**
         * Full screen mode for mobile devices
         * @docs-default true
         */
        this.fullScreenMode = true;
        this.indicators = true;
        this.autoPlayInterval = 0;
        this._destroyed$ = new ReplaySubject(1);
        const translation = ts.getCurrentLanguage();
        this.defaultTexts.previous = translation['adapt.common.previous'];
        this.defaultTexts.next = translation['adapt.common.next'];
    }
    get isMobile() {
        return this.dt.deviceMobile();
    }
    onMouseOver($event) {
        var _a;
        if (!this.isMobile && !((_a = $event === null || $event === void 0 ? void 0 : $event.target) === null || _a === void 0 ? void 0 : _a.classList.contains('a-carousel__indicators-item'))) {
            this._disableAutoSlide();
        }
    }
    onMouseOut() {
        this._enableAutoSlide();
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        timer(0, asapScheduler)
            .pipe(takeUntil(this._destroyed$))
            .subscribe(() => {
            this._ngZone.runOutsideAngular(() => {
                this.siema.run(this.options);
            });
            this._enableAutoSlide();
        });
    }
    ngOnInit() {
        this.carouselClass = this.options.selector.replace('.', '');
        this.options.rtl = this.dir.value === 'rtl';
        this.carouselControls = !this.isMobile || (this.options.allowMobileControls && this.isMobile);
        this.texts = Object.assign(Object.assign({}, this.defaultTexts), this.texts);
    }
    ngOnDestroy() {
        this._destroyed$.next(true);
        this._destroyed$.complete();
        this.siema.destroy();
        this.siema = null;
        this._disableAutoSlide();
    }
    onGoTo(index) {
        this.siema.goTo(index);
    }
    onFullScreen(e) {
        const { type } = e;
        if (this.isMobile && this.fullScreenMode && type === 'tap') {
            this._ngZone.onStable
                .pipe(takeUntil(this._destroyed$), take(1))
                .subscribe(this.siema.resizeHandler.bind(this));
            this.isMobileVisible = true;
        }
    }
    closeBackDrop(e) {
        e.preventDefault();
        this._ngZone.onStable
            .pipe(takeUntil(this._destroyed$), take(1))
            .subscribe(this.siema.resizeHandler.bind(this));
        this.isMobileVisible = false;
    }
    _enableAutoSlide() {
        if (this.autoPlayInterval && (!this.autoPlaySubscription || this.autoPlaySubscription.closed)) {
            const _timer = timer(this.autoPlayInterval, this.autoPlayInterval);
            this.autoPlaySubscription = _timer.subscribe(() => this.siema.next(1));
        }
    }
    _disableAutoSlide() {
        if (this.autoPlayInterval && this.autoPlaySubscription) {
            this.autoPlaySubscription.unsubscribe();
        }
    }
}
AdaptCarouselComponent.ɵfac = function AdaptCarouselComponent_Factory(t) { return new (t || AdaptCarouselComponent)(i0.ɵɵdirectiveInject(i1.SiemaService), i0.ɵɵdirectiveInject(i2.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i3.Directionality), i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i4.AdaptTranslateService), i0.ɵɵdirectiveInject(i5.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptCarouselComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptCarouselComponent, selectors: [["adapt-carousel-component"]], contentQueries: function AdaptCarouselComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        i0.ɵɵcontentQuery(dirIndex, AdaptCarouselSlideComponent, 4);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.slides = _t);
    } }, hostBindings: function AdaptCarouselComponent_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("mouseover", function AdaptCarouselComponent_mouseover_HostBindingHandler($event) { return ctx.onMouseOver($event); })("mouseout", function AdaptCarouselComponent_mouseout_HostBindingHandler() { return ctx.onMouseOut(); });
    } }, inputs: { fullScreenMode: "fullScreenMode", indicators: "indicators", autoPlayInterval: "autoPlayInterval", options: "options", texts: "texts" }, features: [i0.ɵɵProvidersFeature([SiemaService]), i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c3, decls: 19, vars: 11, consts: [[4, "ngIf"], [1, "a-carousel"], ["data-adapt", "hammer-pan-y", 3, "tap"], ["class", "a-carousel__indicators", 3, "ngClass", 4, "ngIf"], ["class", "a-carousel-overlay__close d-icon-cross_adapt", 3, "tap", "keyup.esc", 4, "ngIf"], [1, "modal-backdrop", "show", "a-carousel-overlay"], ["class", "a-carousel__controls", 4, "ngIf"], [1, "a-carousel__controls"], ["type", "button", "role", "button", "tabindex", "0", 1, "a-carousel__controls-prev", 3, "click"], ["aria-hidden", "true", 1, "carousel-control-prev-icon"], [1, "sr-only"], ["type", "button", "role", "button", "tabindex", "0", 1, "a-carousel__controls-next", 3, "click"], ["aria-hidden", "true", 1, "carousel-control-next-icon"], [1, "a-carousel__indicators", 3, "ngClass"], ["type", "button", "role", "button", "class", "a-carousel__indicators-item", "tabindex", "0", 3, "is-active", "click", 4, "ngFor", "ngForOf"], ["type", "button", "role", "button", "tabindex", "0", 1, "a-carousel__indicators-item", 3, "click"], [1, "a-carousel-overlay__close", "d-icon-cross_adapt", 3, "tap", "keyup.esc"]], template: function AdaptCarouselComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c2);
        i0.ɵɵtemplate(0, AdaptCarouselComponent_div_0_Template, 4, 1, "div", 0);
        i0.ɵɵtext(1, "\n\n");
        i0.ɵɵelementStart(2, "div");
        i0.ɵɵtext(3, "\n  ");
        i0.ɵɵelementStart(4, "div", 1);
        i0.ɵɵtext(5, "\n    ");
        i0.ɵɵelementStart(6, "div", 2);
        i0.ɵɵlistener("tap", function AdaptCarouselComponent_Template_div_tap_6_listener($event) { return ctx.onFullScreen($event); });
        i0.ɵɵtext(7, "\n      ");
        i0.ɵɵprojection(8);
        i0.ɵɵtext(9, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(10, "\n\n    ");
        i0.ɵɵtemplate(11, AdaptCarouselComponent_ng_container_11_Template, 4, 1, "ng-container", 0);
        i0.ɵɵtext(12, "\n    ");
        i0.ɵɵtemplate(13, AdaptCarouselComponent_div_13_Template, 4, 4, "div", 3);
        i0.ɵɵtext(14, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(15, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(16, "\n");
        i0.ɵɵtemplate(17, AdaptCarouselComponent_a_17_Template, 1, 1, "a", 4);
        i0.ɵɵtext(18, "\n\n\n\n");
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", ctx.isMobileVisible);
        i0.ɵɵadvance(2);
        i0.ɵɵclassProp("a-carousel-wrapper", ctx.isMobileVisible);
        i0.ɵɵadvance(2);
        i0.ɵɵclassProp("a-carousel--mobile", ctx.isMobileVisible);
        i0.ɵɵadvance(2);
        i0.ɵɵclassMapInterpolate1("", ctx.carouselClass, " adapt-carousel-wrapper");
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", ctx.carouselControls);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.slides.length > 1 && ctx.indicators);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngIf", ctx.isMobileVisible);
    } }, directives: [i6.NgIf, i6.NgClass, i6.NgForOf], styles: [".a-carousel-wrapper, .a-carousel-overlay__close {\n      z-index: 1060\n    }\n\n    .a-carousel__controls-prev, .a-carousel__controls-next {\n      border: none;\n      background: transparent\n    }\n\n    .a-carousel__indicators-item {\n      border: none\n    }\n\n    .a-carousel__indicators-item.is-mouse-focused {\n      box-shadow: none !important;\n    }"], encapsulation: 2, data: { animation: [
            trigger('closeIconAnimation', [
                transition(':enter', [
                    useAnimation(ANIMATION_NAME.menuIn)
                ]),
                transition(':leave', [
                    useAnimation(ANIMATION_NAME.menuOut)
                ])
            ]),
            trigger('backdropAnimation', [
                transition('void => *', [
                    query('.modal-backdrop', [
                        useAnimation(ANIMATION_NAME.opacity)
                    ])
                ]),
                transition('* => void', [
                    query('.modal-backdrop', [
                        useAnimation(ANIMATION_NAME.opacity, {
                            params: {
                                from: '*',
                                to: '0'
                            }
                        })
                    ])
                ])
            ]),
            trigger('indicatorsAnimation', [
                transition('void => *', [
                    style({ transform: 'translate3d(0,100%,0)' }),
                    animate(`.25s ${ANIMATION_TIMING_FUNCTION.base.transitionIn}`, style({ transform: 'translate3d(0,0,0)' }))
                ]),
                transition('* => void', [
                    style({ transform: 'translate3d(0,0,0)' }),
                    animate(`.25s ${ANIMATION_TIMING_FUNCTION.base.transitionInOut}`, style({ transform: 'translate3d(0,100%,0)' }))
                ])
            ])
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCarouselComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-carousel-component',
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {},
                templateUrl: './carousel.component.html',
                encapsulation: ViewEncapsulation.None,
                animations: [
                    trigger('closeIconAnimation', [
                        transition(':enter', [
                            useAnimation(ANIMATION_NAME.menuIn)
                        ]),
                        transition(':leave', [
                            useAnimation(ANIMATION_NAME.menuOut)
                        ])
                    ]),
                    trigger('backdropAnimation', [
                        transition('void => *', [
                            query('.modal-backdrop', [
                                useAnimation(ANIMATION_NAME.opacity)
                            ])
                        ]),
                        transition('* => void', [
                            query('.modal-backdrop', [
                                useAnimation(ANIMATION_NAME.opacity, {
                                    params: {
                                        from: '*',
                                        to: '0'
                                    }
                                })
                            ])
                        ])
                    ]),
                    trigger('indicatorsAnimation', [
                        transition('void => *', [
                            style({ transform: 'translate3d(0,100%,0)' }),
                            animate(`.25s ${ANIMATION_TIMING_FUNCTION.base.transitionIn}`, style({ transform: 'translate3d(0,0,0)' }))
                        ]),
                        transition('* => void', [
                            style({ transform: 'translate3d(0,0,0)' }),
                            animate(`.25s ${ANIMATION_TIMING_FUNCTION.base.transitionInOut}`, style({ transform: 'translate3d(0,100%,0)' }))
                        ])
                    ])
                ],
                providers: [SiemaService],
                styles: [
                    `.a-carousel-wrapper, .a-carousel-overlay__close {
      z-index: 1060
    }

    .a-carousel__controls-prev, .a-carousel__controls-next {
      border: none;
      background: transparent
    }

    .a-carousel__indicators-item {
      border: none
    }

    .a-carousel__indicators-item.is-mouse-focused {
      box-shadow: none !important;
    }`
                ]
            }]
    }], function () { return [{ type: i1.SiemaService }, { type: i2.AdaptDeviceDetectionService }, { type: i3.Directionality }, { type: i0.NgZone }, { type: i4.AdaptTranslateService }, { type: i5.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { fullScreenMode: [{
            type: Input
        }], indicators: [{
            type: Input
        }], autoPlayInterval: [{
            type: Input
        }], options: [{
            type: Input
        }], texts: [{
            type: Input
        }], slides: [{
            type: ContentChildren,
            args: [AdaptCarouselSlideComponent]
        }], onMouseOver: [{
            type: HostListener,
            args: ['mouseover', ['$event']]
        }], onMouseOut: [{
            type: HostListener,
            args: ['mouseout']
        }] }); })();
