// @bmc-ux/adapt-angular/esm2015/carousel/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptCarouselComponent, AdaptCarouselSlideComponent } from './carousel.component';
import * as i0 from "@angular/core";
export class AdaptCarouselModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptCarouselModule
        };
    }
}
AdaptCarouselModule.ɵfac = function AdaptCarouselModule_Factory(t) { return new (t || AdaptCarouselModule)(); };
AdaptCarouselModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCarouselModule });
AdaptCarouselModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCarouselModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [AdaptCarouselComponent, AdaptCarouselSlideComponent],
                declarations: [AdaptCarouselComponent, AdaptCarouselSlideComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCarouselModule, { declarations: [AdaptCarouselComponent, AdaptCarouselSlideComponent], imports: [CommonModule], exports: [AdaptCarouselComponent, AdaptCarouselSlideComponent] }); })();
export * from './carousel.component';
export * from './carousel.model';
