// @helix/com-bmc-arsys-rx-innovationstudio/esm2015/lib/components/bundle-action-wizard/add-bundle-content-data-wizard-step/add-bundle-content-data-wizard-step.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { AdaptButtonModule, AdaptDropdownModule, AdaptIconModule, AdaptPopoverModule, AdaptRxCheckboxModule, AdaptRxSelectModule, AdaptRxSwitchModule } from '@bmc-ux/adapt-angular';
import { RxDefinitionModule } from '@helix/platform/shared/api';
import { ExpressionFormControlModule, RxDefinitionPickerModule } from '@helix/platform/shared/components';
import { RecordGridModule } from '@helix/platform/view/components';
import { TranslateModule } from '@ngx-translate/core';
import { AddBundleContentDataWizardStepComponent } from './add-bundle-content-data-wizard-step.component';
import * as i0 from "@angular/core";
export class AddBundleContentDataWizardStepModule {
}
/** @nocollapse */ AddBundleContentDataWizardStepModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AddBundleContentDataWizardStepModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ AddBundleContentDataWizardStepModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AddBundleContentDataWizardStepModule, declarations: [AddBundleContentDataWizardStepComponent], imports: [AdaptButtonModule,
        AdaptDropdownModule,
        AdaptIconModule,
        AdaptRxCheckboxModule,
        AdaptRxSelectModule,
        AdaptRxSwitchModule,
        CommonModule,
        ExpressionFormControlModule,
        FormsModule,
        ReactiveFormsModule,
        RecordGridModule,
        RxDefinitionModule,
        RxDefinitionPickerModule,
        TranslateModule,
        AdaptPopoverModule], exports: [AddBundleContentDataWizardStepComponent] });
/** @nocollapse */ AddBundleContentDataWizardStepModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AddBundleContentDataWizardStepModule, imports: [[
            AdaptButtonModule,
            AdaptDropdownModule,
            AdaptIconModule,
            AdaptRxCheckboxModule,
            AdaptRxSelectModule,
            AdaptRxSwitchModule,
            CommonModule,
            ExpressionFormControlModule,
            FormsModule,
            ReactiveFormsModule,
            RecordGridModule,
            RxDefinitionModule,
            RxDefinitionPickerModule,
            TranslateModule,
            AdaptPopoverModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AddBundleContentDataWizardStepModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        AdaptButtonModule,
                        AdaptDropdownModule,
                        AdaptIconModule,
                        AdaptRxCheckboxModule,
                        AdaptRxSelectModule,
                        AdaptRxSwitchModule,
                        CommonModule,
                        ExpressionFormControlModule,
                        FormsModule,
                        ReactiveFormsModule,
                        RecordGridModule,
                        RxDefinitionModule,
                        RxDefinitionPickerModule,
                        TranslateModule,
                        AdaptPopoverModule
                    ],
                    declarations: [AddBundleContentDataWizardStepComponent],
                    exports: [AddBundleContentDataWizardStepComponent]
                }]
        }] });
