// @bmc-ux/adapt-angular/esm2015/adapt.module.js
import { NgModule } from '@angular/core';
import { AdaptAccordionModule } from './accordion/index';
import { AdaptAgreementModule } from './agreement/index';
import { AdaptAlertModule } from './alert/index';
import { AdaptBadgeModule } from './badge/index';
import { AdaptBeardModule } from './beard/index';
import { AdaptButtonModule } from './button/index';
import { AdaptButtonGroupModule } from './button-group/index';
import { AdaptCardCarouselModule } from './card-carousel/index';
import { AdaptCarouselModule } from './carousel/index';
import { AdaptCloseModule } from './close/index';
import { AdaptCodeViewerModule } from './code-viewer/index';
import { AdaptColorPickerModule } from './color-picker/index';
import { AdaptCropModule } from './crop/index';
import { AdaptCustomActionsModule } from './custom-actions/index';
import { AdaptDockedPanelModule } from './docked-panel/index';
import { AdaptDownloadModule } from './downloader/index';
import { AdaptEmptyStateModule } from './empty-state/index';
import { AdaptIconModule } from './icon/index';
import { AdaptInlineEditorModule } from './inline-editor/index';
import { AdaptLoginPageModule } from './login/index';
import { AdaptMetatagModule } from './metatag/index';
import { AdaptModalModule } from './modal/index';
import { AdaptMention2Module } from './mention2/index';
import { AdaptMultiCardModule } from './multi-card/index';
import { AdaptNavigationModule } from './navigation/index';
import { AdaptNotFoundModule } from './not-found/index';
import { AdaptPaginationModule } from './pagination/index';
import { AdaptProfileModule } from './profile/index';
import { AdaptProgressModule } from './progress/index';
import { AdaptSidebarModule } from './sidebar/index';
import { AdaptSliderModule } from './slider/index';
import { AdaptSplitModule } from './split/index';
import { AdaptStepsModule } from './steps/index';
import { AdaptSubnavModule } from './subnav/index';
import { AdaptTabsModule } from './tabs/index';
import { AdaptTagModule } from './tag/index';
import { AdaptTimeModule } from './time/index';
import { AdaptToastModule } from './toast/index';
import { AdaptTreeModule } from './tree/index';
import { AdaptTruncatePipeModule } from './common/truncated.pipe';
import { AdaptSortByPipeModule } from './common/pipes/sortby.pipe';
import { AdaptWorkFlowModule } from './workflow/index';
import { AdaptMenuModule } from './menu/index';
import { AdaptBusyModule } from './busy-loader/index';
import { AdaptCollapseModule } from './collapse/index';
import { AdaptDropdownModule } from './dropdown/index';
import { AdaptFocusHolderModule } from './focus-holder/index';
import { AdaptFocusTrackerModule } from './focus-tracker/index';
import { AdaptPopoverModule } from './popover/index';
import { AdaptTabsControlsModule } from './tabs-controls/index';
import { AdaptTooltipModule } from './tooltip/index';
import { AdaptTruncateFileNameModule } from './truncate-file-name/index';
import { AdaptAnchorNavigatorModule } from './anchor-navigator/index';
import { AdaptAdvancedFilteringModule } from './advanced-filtering/index';
import { AdaptDragonInputHandlerModule } from './dragon-input-handler/index';
import { AdaptLifecycleHooksDirectiveModule } from './common/directives/index';
import { AdaptResizeModule } from './common/resize/index';
import { AdaptListKeyManagerModule } from './common/list-key-manager/index';
import { AdaptHighlightModule } from './highlight/index';
import { AdaptFocusKeyManagerModule } from './common/focus-key-manager/index';
import { AdaptRxFormsModule } from './rx/index';
import { AdaptAnimationDirectivesModule } from './common/animationDirectives/index';
import * as i0 from "@angular/core";
const ADAPT_MODULES = [
    AdaptAccordionModule,
    AdaptAdvancedFilteringModule,
    AdaptAgreementModule,
    AdaptAlertModule,
    AdaptAnchorNavigatorModule,
    AdaptAnimationDirectivesModule,
    AdaptBadgeModule,
    AdaptBeardModule,
    AdaptBusyModule,
    AdaptButtonModule,
    AdaptButtonGroupModule,
    AdaptCardCarouselModule,
    AdaptCarouselModule,
    AdaptCloseModule,
    AdaptCodeViewerModule,
    AdaptCollapseModule,
    AdaptColorPickerModule,
    AdaptCropModule,
    AdaptCustomActionsModule,
    AdaptDockedPanelModule,
    AdaptDownloadModule,
    AdaptDragonInputHandlerModule,
    AdaptDropdownModule,
    AdaptEmptyStateModule,
    AdaptFocusHolderModule,
    AdaptFocusTrackerModule,
    AdaptIconModule,
    AdaptInlineEditorModule,
    AdaptLoginPageModule,
    AdaptMention2Module,
    AdaptMenuModule,
    AdaptMetatagModule,
    AdaptModalModule,
    AdaptMultiCardModule,
    AdaptNavigationModule,
    AdaptNotFoundModule,
    AdaptPaginationModule,
    AdaptPopoverModule,
    AdaptProfileModule,
    AdaptProgressModule,
    AdaptRxFormsModule,
    AdaptSidebarModule,
    AdaptSliderModule,
    AdaptSplitModule,
    AdaptStepsModule,
    AdaptSubnavModule,
    AdaptTabsControlsModule,
    AdaptTabsModule,
    AdaptTagModule,
    AdaptTimeModule,
    AdaptToastModule,
    AdaptTooltipModule,
    AdaptTreeModule,
    AdaptTruncateFileNameModule,
    AdaptTruncatePipeModule,
    AdaptSortByPipeModule,
    AdaptWorkFlowModule,
    AdaptLifecycleHooksDirectiveModule,
    AdaptResizeModule,
    AdaptListKeyManagerModule,
    AdaptHighlightModule,
    AdaptFocusKeyManagerModule,
    AdaptHighlightModule
];
export class AdaptRootModule {
}
AdaptRootModule.ɵfac = function AdaptRootModule_Factory(t) { return new (t || AdaptRootModule)(); };
AdaptRootModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRootModule });
AdaptRootModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            AdaptAccordionModule,
            AdaptAdvancedFilteringModule,
            AdaptAgreementModule,
            AdaptAlertModule,
            AdaptAnchorNavigatorModule,
            AdaptAnimationDirectivesModule,
            AdaptBadgeModule,
            AdaptBeardModule,
            AdaptBusyModule,
            AdaptButtonModule,
            AdaptButtonGroupModule,
            AdaptCarouselModule,
            AdaptCloseModule,
            AdaptCollapseModule,
            AdaptColorPickerModule,
            AdaptCropModule,
            AdaptCustomActionsModule,
            AdaptDockedPanelModule,
            AdaptDownloadModule,
            AdaptDragonInputHandlerModule,
            AdaptDropdownModule,
            AdaptEmptyStateModule,
            AdaptFocusHolderModule,
            AdaptFocusTrackerModule,
            AdaptIconModule,
            AdaptInlineEditorModule,
            AdaptLoginPageModule,
            AdaptMention2Module,
            AdaptMenuModule,
            AdaptMetatagModule,
            AdaptModalModule,
            AdaptMultiCardModule,
            AdaptNavigationModule,
            AdaptNotFoundModule,
            AdaptPaginationModule,
            AdaptPopoverModule,
            AdaptProfileModule,
            AdaptRxFormsModule,
            AdaptSidebarModule,
            AdaptSliderModule,
            AdaptSplitModule,
            AdaptStepsModule,
            AdaptSubnavModule,
            AdaptTabsControlsModule,
            AdaptTabsModule,
            AdaptTagModule,
            AdaptTimeModule,
            AdaptToastModule,
            AdaptTooltipModule,
            AdaptTreeModule,
            AdaptTruncateFileNameModule,
            AdaptTruncatePipeModule,
            AdaptSortByPipeModule,
            AdaptWorkFlowModule,
            AdaptLifecycleHooksDirectiveModule,
            AdaptResizeModule,
            AdaptListKeyManagerModule,
            AdaptHighlightModule,
            AdaptFocusKeyManagerModule,
            AdaptHighlightModule
        ], AdaptAccordionModule,
        AdaptAdvancedFilteringModule,
        AdaptAgreementModule,
        AdaptAlertModule,
        AdaptAnchorNavigatorModule,
        AdaptAnimationDirectivesModule,
        AdaptBadgeModule,
        AdaptBeardModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptButtonGroupModule,
        AdaptCardCarouselModule,
        AdaptCarouselModule,
        AdaptCloseModule,
        AdaptCodeViewerModule,
        AdaptCollapseModule,
        AdaptColorPickerModule,
        AdaptCropModule,
        AdaptCustomActionsModule,
        AdaptDockedPanelModule,
        AdaptDownloadModule,
        AdaptDragonInputHandlerModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptFocusHolderModule,
        AdaptFocusTrackerModule,
        AdaptIconModule,
        AdaptInlineEditorModule,
        AdaptLoginPageModule,
        AdaptMention2Module,
        AdaptMenuModule,
        AdaptMetatagModule,
        AdaptModalModule,
        AdaptMultiCardModule,
        AdaptNavigationModule,
        AdaptNotFoundModule,
        AdaptPaginationModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptProgressModule,
        AdaptRxFormsModule,
        AdaptSidebarModule,
        AdaptSliderModule,
        AdaptSplitModule,
        AdaptStepsModule,
        AdaptSubnavModule,
        AdaptTabsControlsModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTimeModule,
        AdaptToastModule,
        AdaptTooltipModule,
        AdaptTreeModule,
        AdaptTruncateFileNameModule,
        AdaptTruncatePipeModule,
        AdaptSortByPipeModule,
        AdaptWorkFlowModule,
        AdaptLifecycleHooksDirectiveModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptHighlightModule,
        AdaptFocusKeyManagerModule,
        AdaptHighlightModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRootModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptAccordionModule,
                    AdaptAdvancedFilteringModule,
                    AdaptAgreementModule,
                    AdaptAlertModule,
                    AdaptAnchorNavigatorModule,
                    AdaptAnimationDirectivesModule,
                    AdaptBadgeModule,
                    AdaptBeardModule,
                    AdaptBusyModule,
                    AdaptButtonModule,
                    AdaptButtonGroupModule,
                    AdaptCarouselModule,
                    AdaptCloseModule,
                    AdaptCollapseModule,
                    AdaptColorPickerModule,
                    AdaptCropModule,
                    AdaptCustomActionsModule,
                    AdaptDockedPanelModule,
                    AdaptDownloadModule,
                    AdaptDragonInputHandlerModule,
                    AdaptDropdownModule,
                    AdaptEmptyStateModule,
                    AdaptFocusHolderModule,
                    AdaptFocusTrackerModule,
                    AdaptIconModule,
                    AdaptInlineEditorModule,
                    AdaptLoginPageModule,
                    AdaptMention2Module,
                    AdaptMenuModule,
                    AdaptMetatagModule,
                    AdaptModalModule,
                    AdaptMultiCardModule,
                    AdaptNavigationModule,
                    AdaptNotFoundModule,
                    AdaptPaginationModule,
                    AdaptPopoverModule,
                    AdaptProfileModule,
                    AdaptRxFormsModule,
                    AdaptSidebarModule,
                    AdaptSliderModule,
                    AdaptSplitModule,
                    AdaptStepsModule,
                    AdaptSubnavModule,
                    AdaptTabsControlsModule,
                    AdaptTabsModule,
                    AdaptTagModule,
                    AdaptTimeModule,
                    AdaptToastModule,
                    AdaptTooltipModule,
                    AdaptTreeModule,
                    AdaptTruncateFileNameModule,
                    AdaptTruncatePipeModule,
                    AdaptSortByPipeModule,
                    AdaptWorkFlowModule,
                    AdaptLifecycleHooksDirectiveModule,
                    AdaptResizeModule,
                    AdaptListKeyManagerModule,
                    AdaptHighlightModule,
                    AdaptFocusKeyManagerModule,
                    AdaptHighlightModule
                ],
                exports: ADAPT_MODULES
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRootModule, { imports: [AdaptAccordionModule,
        AdaptAdvancedFilteringModule,
        AdaptAgreementModule,
        AdaptAlertModule,
        AdaptAnchorNavigatorModule,
        AdaptAnimationDirectivesModule,
        AdaptBadgeModule,
        AdaptBeardModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptButtonGroupModule,
        AdaptCarouselModule,
        AdaptCloseModule,
        AdaptCollapseModule,
        AdaptColorPickerModule,
        AdaptCropModule,
        AdaptCustomActionsModule,
        AdaptDockedPanelModule,
        AdaptDownloadModule,
        AdaptDragonInputHandlerModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptFocusHolderModule,
        AdaptFocusTrackerModule,
        AdaptIconModule,
        AdaptInlineEditorModule,
        AdaptLoginPageModule,
        AdaptMention2Module,
        AdaptMenuModule,
        AdaptMetatagModule,
        AdaptModalModule,
        AdaptMultiCardModule,
        AdaptNavigationModule,
        AdaptNotFoundModule,
        AdaptPaginationModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptRxFormsModule,
        AdaptSidebarModule,
        AdaptSliderModule,
        AdaptSplitModule,
        AdaptStepsModule,
        AdaptSubnavModule,
        AdaptTabsControlsModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTimeModule,
        AdaptToastModule,
        AdaptTooltipModule,
        AdaptTreeModule,
        AdaptTruncateFileNameModule,
        AdaptTruncatePipeModule,
        AdaptSortByPipeModule,
        AdaptWorkFlowModule,
        AdaptLifecycleHooksDirectiveModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptHighlightModule,
        AdaptFocusKeyManagerModule,
        AdaptHighlightModule], exports: [AdaptAccordionModule,
        AdaptAdvancedFilteringModule,
        AdaptAgreementModule,
        AdaptAlertModule,
        AdaptAnchorNavigatorModule,
        AdaptAnimationDirectivesModule,
        AdaptBadgeModule,
        AdaptBeardModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptButtonGroupModule,
        AdaptCardCarouselModule,
        AdaptCarouselModule,
        AdaptCloseModule,
        AdaptCodeViewerModule,
        AdaptCollapseModule,
        AdaptColorPickerModule,
        AdaptCropModule,
        AdaptCustomActionsModule,
        AdaptDockedPanelModule,
        AdaptDownloadModule,
        AdaptDragonInputHandlerModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptFocusHolderModule,
        AdaptFocusTrackerModule,
        AdaptIconModule,
        AdaptInlineEditorModule,
        AdaptLoginPageModule,
        AdaptMention2Module,
        AdaptMenuModule,
        AdaptMetatagModule,
        AdaptModalModule,
        AdaptMultiCardModule,
        AdaptNavigationModule,
        AdaptNotFoundModule,
        AdaptPaginationModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptProgressModule,
        AdaptRxFormsModule,
        AdaptSidebarModule,
        AdaptSliderModule,
        AdaptSplitModule,
        AdaptStepsModule,
        AdaptSubnavModule,
        AdaptTabsControlsModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTimeModule,
        AdaptToastModule,
        AdaptTooltipModule,
        AdaptTreeModule,
        AdaptTruncateFileNameModule,
        AdaptTruncatePipeModule,
        AdaptSortByPipeModule,
        AdaptWorkFlowModule,
        AdaptLifecycleHooksDirectiveModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptHighlightModule,
        AdaptFocusKeyManagerModule,
        AdaptHighlightModule] }); })();
export class AdaptModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return { ngModule: AdaptRootModule };
    }
}
AdaptModule.ɵfac = function AdaptModule_Factory(t) { return new (t || AdaptModule)(); };
AdaptModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptModule });
AdaptModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [ADAPT_MODULES, AdaptAccordionModule,
        AdaptAdvancedFilteringModule,
        AdaptAgreementModule,
        AdaptAlertModule,
        AdaptAnchorNavigatorModule,
        AdaptAnimationDirectivesModule,
        AdaptBadgeModule,
        AdaptBeardModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptButtonGroupModule,
        AdaptCardCarouselModule,
        AdaptCarouselModule,
        AdaptCloseModule,
        AdaptCodeViewerModule,
        AdaptCollapseModule,
        AdaptColorPickerModule,
        AdaptCropModule,
        AdaptCustomActionsModule,
        AdaptDockedPanelModule,
        AdaptDownloadModule,
        AdaptDragonInputHandlerModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptFocusHolderModule,
        AdaptFocusTrackerModule,
        AdaptIconModule,
        AdaptInlineEditorModule,
        AdaptLoginPageModule,
        AdaptMention2Module,
        AdaptMenuModule,
        AdaptMetatagModule,
        AdaptModalModule,
        AdaptMultiCardModule,
        AdaptNavigationModule,
        AdaptNotFoundModule,
        AdaptPaginationModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptProgressModule,
        AdaptRxFormsModule,
        AdaptSidebarModule,
        AdaptSliderModule,
        AdaptSplitModule,
        AdaptStepsModule,
        AdaptSubnavModule,
        AdaptTabsControlsModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTimeModule,
        AdaptToastModule,
        AdaptTooltipModule,
        AdaptTreeModule,
        AdaptTruncateFileNameModule,
        AdaptTruncatePipeModule,
        AdaptSortByPipeModule,
        AdaptWorkFlowModule,
        AdaptLifecycleHooksDirectiveModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptHighlightModule,
        AdaptFocusKeyManagerModule,
        AdaptHighlightModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptModule, [{
        type: NgModule,
        args: [{
                imports: ADAPT_MODULES,
                exports: ADAPT_MODULES
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptModule, { imports: [AdaptAccordionModule,
        AdaptAdvancedFilteringModule,
        AdaptAgreementModule,
        AdaptAlertModule,
        AdaptAnchorNavigatorModule,
        AdaptAnimationDirectivesModule,
        AdaptBadgeModule,
        AdaptBeardModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptButtonGroupModule,
        AdaptCardCarouselModule,
        AdaptCarouselModule,
        AdaptCloseModule,
        AdaptCodeViewerModule,
        AdaptCollapseModule,
        AdaptColorPickerModule,
        AdaptCropModule,
        AdaptCustomActionsModule,
        AdaptDockedPanelModule,
        AdaptDownloadModule,
        AdaptDragonInputHandlerModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptFocusHolderModule,
        AdaptFocusTrackerModule,
        AdaptIconModule,
        AdaptInlineEditorModule,
        AdaptLoginPageModule,
        AdaptMention2Module,
        AdaptMenuModule,
        AdaptMetatagModule,
        AdaptModalModule,
        AdaptMultiCardModule,
        AdaptNavigationModule,
        AdaptNotFoundModule,
        AdaptPaginationModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptProgressModule,
        AdaptRxFormsModule,
        AdaptSidebarModule,
        AdaptSliderModule,
        AdaptSplitModule,
        AdaptStepsModule,
        AdaptSubnavModule,
        AdaptTabsControlsModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTimeModule,
        AdaptToastModule,
        AdaptTooltipModule,
        AdaptTreeModule,
        AdaptTruncateFileNameModule,
        AdaptTruncatePipeModule,
        AdaptSortByPipeModule,
        AdaptWorkFlowModule,
        AdaptLifecycleHooksDirectiveModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptHighlightModule,
        AdaptFocusKeyManagerModule,
        AdaptHighlightModule], exports: [AdaptAccordionModule,
        AdaptAdvancedFilteringModule,
        AdaptAgreementModule,
        AdaptAlertModule,
        AdaptAnchorNavigatorModule,
        AdaptAnimationDirectivesModule,
        AdaptBadgeModule,
        AdaptBeardModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptButtonGroupModule,
        AdaptCardCarouselModule,
        AdaptCarouselModule,
        AdaptCloseModule,
        AdaptCodeViewerModule,
        AdaptCollapseModule,
        AdaptColorPickerModule,
        AdaptCropModule,
        AdaptCustomActionsModule,
        AdaptDockedPanelModule,
        AdaptDownloadModule,
        AdaptDragonInputHandlerModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptFocusHolderModule,
        AdaptFocusTrackerModule,
        AdaptIconModule,
        AdaptInlineEditorModule,
        AdaptLoginPageModule,
        AdaptMention2Module,
        AdaptMenuModule,
        AdaptMetatagModule,
        AdaptModalModule,
        AdaptMultiCardModule,
        AdaptNavigationModule,
        AdaptNotFoundModule,
        AdaptPaginationModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptProgressModule,
        AdaptRxFormsModule,
        AdaptSidebarModule,
        AdaptSliderModule,
        AdaptSplitModule,
        AdaptStepsModule,
        AdaptSubnavModule,
        AdaptTabsControlsModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTimeModule,
        AdaptToastModule,
        AdaptTooltipModule,
        AdaptTreeModule,
        AdaptTruncateFileNameModule,
        AdaptTruncatePipeModule,
        AdaptSortByPipeModule,
        AdaptWorkFlowModule,
        AdaptLifecycleHooksDirectiveModule,
        AdaptResizeModule,
        AdaptListKeyManagerModule,
        AdaptHighlightModule,
        AdaptFocusKeyManagerModule,
        AdaptHighlightModule] }); })();
