// @bmc-ux/obsolete/esm2015/lib/select/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HammerModule } from '@angular/platform-browser';
import { AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule, AdaptBusyModule, AdaptDragonInputHandlerModule, AdaptCustomActionsModule, AdaptFocusMonitorDirectiveModule } from '@bmc-ux/adapt-angular';
import { AdaptSelectSearchFilterModule } from './select.pipe';
import { AdaptSelectComponent } from './select.component';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptSelectModule {
}
AdaptSelectModule.ɵfac = function AdaptSelectModule_Factory(t) { return new (t || AdaptSelectModule)(); };
AdaptSelectModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSelectModule });
AdaptSelectModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            HammerModule,
            ReactiveFormsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptBusyModule,
            AdaptSelectSearchFilterModule,
            AdaptFocusMonitorDirectiveModule,
            AdaptDragonInputHandlerModule,
            AdaptCustomActionsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    HammerModule,
                    ReactiveFormsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptBusyModule,
                    AdaptSelectSearchFilterModule,
                    AdaptFocusMonitorDirectiveModule,
                    AdaptDragonInputHandlerModule,
                    AdaptCustomActionsModule
                ],
                exports: [AdaptSelectComponent],
                declarations: [AdaptSelectComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSelectModule, { declarations: [AdaptSelectComponent], imports: [CommonModule,
        FormsModule,
        HammerModule,
        ReactiveFormsModule,
        AdaptIconModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptBusyModule,
        AdaptSelectSearchFilterModule,
        AdaptFocusMonitorDirectiveModule,
        AdaptDragonInputHandlerModule,
        AdaptCustomActionsModule], exports: [AdaptSelectComponent] }); })();
export * from './select.component';
export * from './select.pipe';
export * from './select.model';
