// @bmc-ux/obsolete/esm2015/lib/select/select.pipe.js
import { NgModule, Pipe } from '@angular/core';
import { CommonModule } from '@angular/common';
import { of } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * Search filter by name and badgeText
 */
export class AdaptSelectSearchFilter {
    transform(options, args) {
        return options.filter((option) => {
            return option.name && option.name.toLowerCase().indexOf((args || '').toLowerCase()) > -1
                || option.badgeText && option.badgeText.toLowerCase().indexOf((args || '').toLowerCase()) > -1;
        });
    }
}
AdaptSelectSearchFilter.ɵfac = function AdaptSelectSearchFilter_Factory(t) { return new (t || AdaptSelectSearchFilter)(); };
AdaptSelectSearchFilter.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "searchFilter", type: AdaptSelectSearchFilter, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSelectSearchFilter, [{
        type: Pipe,
        args: [{
                name: 'searchFilter'
            }]
    }], null, null); })();
/**
 * Search filter by name
 */
export class AdaptSelectSearchFilterAsync {
    transform(options, query) {
        return of(options).pipe(map(data => {
            if (!query) {
                return data;
            }
            return data.filter((option) => option.name.toLowerCase().indexOf((query || '').toLowerCase()) > -1);
        }));
    }
}
AdaptSelectSearchFilterAsync.ɵfac = function AdaptSelectSearchFilterAsync_Factory(t) { return new (t || AdaptSelectSearchFilterAsync)(); };
AdaptSelectSearchFilterAsync.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "searchFilterAsync", type: AdaptSelectSearchFilterAsync, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSelectSearchFilterAsync, [{
        type: Pipe,
        args: [{
                name: 'searchFilterAsync'
            }]
    }], null, null); })();
export class AdaptSelectSearchFilterModule {
}
AdaptSelectSearchFilterModule.ɵfac = function AdaptSelectSearchFilterModule_Factory(t) { return new (t || AdaptSelectSearchFilterModule)(); };
AdaptSelectSearchFilterModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSelectSearchFilterModule });
AdaptSelectSearchFilterModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSelectSearchFilterModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptSelectSearchFilter,
                    AdaptSelectSearchFilterAsync
                ],
                exports: [
                    AdaptSelectSearchFilter,
                    AdaptSelectSearchFilterAsync
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSelectSearchFilterModule, { declarations: [AdaptSelectSearchFilter, AdaptSelectSearchFilterAsync], imports: [CommonModule], exports: [AdaptSelectSearchFilter, AdaptSelectSearchFilterAsync] }); })();
