// @bmc-ux/adapt-angular/esm2015/tree/index.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { AdaptCustomActionsModule } from '../custom-actions/index';
import { AdaptDragonInputHandlerModule } from '../dragon-input-handler/index';
import { AdaptHighlightModule } from './../highlight/index';
import { AdaptLazyLoadModule } from '../common/lazy-load/index';
import { AdaptRxSearchModule } from '../rx/rx-search/index';
import { AdaptTreeComponent } from './tree.component';
import { AdaptTreeNodeComponent } from './tree-node.component';
import { AdaptTreeNodeTemplateDirective } from './tree-node-template.directive';
import { AdaptTreeNodeTemplateLoaderComponent } from './tree-template.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../rx/rx-search/adapt-rx-search.component";
import * as i3 from "@angular/forms";
import * as i4 from "../custom-actions/adapt-custom-actions.component";
import * as i5 from "../common/lazy-load/lazy-load-anchor.directive";
export class AdaptTreeModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptTreeModule
        };
    }
}
AdaptTreeModule.ɵfac = function AdaptTreeModule_Factory(t) { return new (t || AdaptTreeModule)(); };
AdaptTreeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTreeModule });
AdaptTreeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptDragonInputHandlerModule,
            AdaptLazyLoadModule,
            AdaptRxSearchModule,
            AdaptHighlightModule,
            AdaptCustomActionsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTreeModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptDragonInputHandlerModule,
                    AdaptLazyLoadModule,
                    AdaptRxSearchModule,
                    AdaptHighlightModule,
                    AdaptCustomActionsModule
                ],
                exports: [
                    AdaptTreeComponent,
                    AdaptTreeNodeTemplateDirective
                ],
                declarations: [
                    AdaptTreeComponent,
                    AdaptTreeNodeComponent,
                    AdaptTreeNodeTemplateDirective,
                    AdaptTreeNodeTemplateLoaderComponent
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTreeModule, { declarations: [AdaptTreeComponent,
        AdaptTreeNodeComponent,
        AdaptTreeNodeTemplateDirective,
        AdaptTreeNodeTemplateLoaderComponent], imports: [CommonModule,
        FormsModule,
        AdaptDragonInputHandlerModule,
        AdaptLazyLoadModule,
        AdaptRxSearchModule,
        AdaptHighlightModule,
        AdaptCustomActionsModule], exports: [AdaptTreeComponent,
        AdaptTreeNodeTemplateDirective] }); })();
i0.ɵɵsetComponentScope(AdaptTreeComponent, [i1.NgIf, i1.NgClass, i1.NgStyle, i2.AdaptRxSearchComponent, i3.NgControlStatus, i3.NgModel, i4.AdaptCustomActionsComponent, i1.NgForOf, i1.NgTemplateOutlet, AdaptTreeNodeComponent, i5.AdaptLazyLoadAnchorDirective], []);
export * from './tree.component';
export * from './tree-node.component';
export * from './tree-template.component';
export * from './tree-node-template.directive';
export * from './tree.model';
