// @bmc-ux/adapt-angular/esm2015/tree/tree-template.component.js
import { Component, Inject, ViewEncapsulation, forwardRef } from '@angular/core';
import { UITreeNode, Tree } from 'primeng/tree';
import * as i0 from "@angular/core";
import * as i1 from "primeng/tree";
export class AdaptTreeNodeTemplateLoaderComponent extends UITreeNode {
    constructor(tree) {
        super(tree);
        this.tree = tree;
    }
}
AdaptTreeNodeTemplateLoaderComponent.ɵfac = function AdaptTreeNodeTemplateLoaderComponent_Factory(t) { return new (t || AdaptTreeNodeTemplateLoaderComponent)(i0.ɵɵdirectiveInject(forwardRef(() => Tree))); };
AdaptTreeNodeTemplateLoaderComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptTreeNodeTemplateLoaderComponent, selectors: [["adapt-tree-node-loader"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function AdaptTreeNodeTemplateLoaderComponent_Template(rf, ctx) { }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTreeNodeTemplateLoaderComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-tree-node-loader',
                template: '',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.Tree, decorators: [{
                type: Inject,
                args: [forwardRef(() => Tree)]
            }] }]; }, null); })();
