// @bmc-ux/adapt-angular/esm2015/login/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptLoginPageComponent } from './login.component';
import * as i0 from "@angular/core";
export class AdaptLoginPageModule {
}
AdaptLoginPageModule.ɵfac = function AdaptLoginPageModule_Factory(t) { return new (t || AdaptLoginPageModule)(); };
AdaptLoginPageModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptLoginPageModule });
AdaptLoginPageModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptLoginPageModule, [{
        type: NgModule,
        args: [{
                imports: [
                    FormsModule,
                    CommonModule
                ],
                exports: [AdaptLoginPageComponent],
                declarations: [AdaptLoginPageComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptLoginPageModule, { declarations: [AdaptLoginPageComponent], imports: [FormsModule,
        CommonModule], exports: [AdaptLoginPageComponent] }); })();
export * from './login.component';
