// @bmc-ux/adapt-angular/esm2015/login/login.component.js
import { Component, ElementRef, Input } from '@angular/core';
import { AdaptTranslateService } from '../common/i18n/index';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "../common/i18n/index";
import * as i2 from "../common/radar/adapt-radar";
import * as i3 from "@angular/common";
const _c0 = [[["", 8, "form"]]];
const _c1 = [".form"];
export class AdaptLoginPageComponent extends AdaptRadarAngularGenericDirective {
    constructor(el, ts, _adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptLoginPage);
        this.el = el;
        this.ts = ts;
        this.powered = 'powered by';
        /**
         * String with classes for setting-up the application logo.
         * When the value is '', the logo will be hidden. For Helix logo use: 'logo-bmc-helix logo-light'
         * @docs-default logo logo-light
         */
        this.logoClass = 'logo logo-light';
        this._footerConfigDefault = {
            link: {
                productLinkText: 'www.bmc.com',
                productUrl: '//www.bmc.com'
            }
        };
        this.element = el.nativeElement;
        this.powered = ts.getCurrentLanguage()['adapt.login.powered'];
    }
    /**
     * Product name text
     * @docs-default
     */
    get productName() {
        return this._productName;
    }
    set productName(value) {
        this._productName = value;
    }
    ngOnInit() {
        this.footerConfig = Object.assign(Object.assign({}, this._footerConfigDefault), this.footerConfig);
    }
}
AdaptLoginPageComponent.ɵfac = function AdaptLoginPageComponent_Factory(t) { return new (t || AdaptLoginPageComponent)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.AdaptTranslateService), i0.ɵɵdirectiveInject(i2.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptLoginPageComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptLoginPageComponent, selectors: [["adapt-login"]], hostVars: 4, hostBindings: function AdaptLoginPageComponent_HostBindings(rf, ctx) { if (rf & 2) {
        i0.ɵɵstyleProp("display", "flex");
        i0.ɵɵclassProp("adapt-login", true);
    } }, inputs: { logoClass: "logoClass", footerConfig: "footerConfig", productName: "productName" }, features: [i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c1, decls: 58, vars: 13, consts: [[1, "wrapper-login", "bg-helix"], [1, "product"], [1, "container"], [1, "product__link"], ["role", "img"], [1, "product__logo_name"], [1, "product__logo_version"], [1, "form-section", "container"], [1, "footer"], [1, "container", "footer__container"], [1, "footer__copy"], [1, "footer__copy-span"], ["target", "_blank", 1, "footer__copy-link", 3, "href", "title"], [1, "footer__powered"], ["href", "//www.bmc.com", "target", "_blank", "title", "powered by BMC", 1, "footer__powered-link"], [1, "footer__powered-title"], ["role", "img", 1, "footer__powered-logo", "logo-light", 3, "ngClass"]], template: function AdaptLoginPageComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef(_c0);
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtext(1, "\n  ");
        i0.ɵɵelementStart(2, "div", 1);
        i0.ɵɵtext(3, "\n    ");
        i0.ɵɵelementStart(4, "div", 2);
        i0.ɵɵtext(5, "\n      ");
        i0.ɵɵelementStart(6, "div", 3);
        i0.ɵɵtext(7, "\n        ");
        i0.ɵɵelement(8, "div", 4);
        i0.ɵɵtext(9, "\n        ");
        i0.ɵɵelementStart(10, "div", 5);
        i0.ɵɵtext(11);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(12, "\n        ");
        i0.ɵɵelementStart(13, "div", 6);
        i0.ɵɵtext(14);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(15, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(16, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(17, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(18, "\n\n  ");
        i0.ɵɵelementStart(19, "div", 7);
        i0.ɵɵtext(20, "\n    ");
        i0.ɵɵprojection(21);
        i0.ɵɵtext(22, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(23, "\n\n  ");
        i0.ɵɵelementStart(24, "footer", 8);
        i0.ɵɵtext(25, "\n    ");
        i0.ɵɵelementStart(26, "div", 9);
        i0.ɵɵtext(27, "\n      ");
        i0.ɵɵelementStart(28, "div", 10);
        i0.ɵɵtext(29, "\n            ");
        i0.ɵɵelementStart(30, "span", 11);
        i0.ɵɵtext(31, "\n              ");
        i0.ɵɵelementStart(32, "a", 12);
        i0.ɵɵtext(33);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(34, "\n            ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(35, "\n\n        ");
        i0.ɵɵelementStart(36, "span", 11);
        i0.ɵɵtext(37);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(38, "\n        ");
        i0.ɵɵelementStart(39, "span", 11);
        i0.ɵɵtext(40);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(41, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(42, "\n\n      ");
        i0.ɵɵelementStart(43, "div", 13);
        i0.ɵɵtext(44, "\n        ");
        i0.ɵɵelementStart(45, "a", 14);
        i0.ɵɵtext(46, "\n          ");
        i0.ɵɵelementStart(47, "span", 15);
        i0.ɵɵtext(48);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(49, "\n          ");
        i0.ɵɵelementStart(50, "span", 16);
        i0.ɵɵtext(51, "\n              ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(52, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(53, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(54, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(55, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(56, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(57, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(8);
        i0.ɵɵclassMap("product__logo " + ctx.logoClass);
        i0.ɵɵattribute("aria-label", ctx.productName + " logo");
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.productName);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.footerConfig.version);
        i0.ɵɵadvance(18);
        i0.ɵɵproperty("href", ctx.footerConfig.link.productUrl, i0.ɵɵsanitizeUrl)("title", ctx.footerConfig.link.productLinkText);
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate1("\n                ", ctx.footerConfig.link.productLinkText, "");
        i0.ɵɵadvance(4);
        i0.ɵɵtextInterpolate(ctx.footerConfig.content.copyright);
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.footerConfig.content.info);
        i0.ɵɵadvance(8);
        i0.ɵɵtextInterpolate(ctx.powered);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngClass", ctx.footerConfig.content.helixLogo ? "logo-bmc-helix" : "logo-bmc");
        i0.ɵɵattribute("aria-label", ctx.footerConfig.content.helixLogo ? "BMC Helix" : "BMC");
    } }, directives: [i3.NgClass], styles: [".wrapper-login[_ngcontent-%COMP%] {\n      height: auto;\n    }"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptLoginPageComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-login',
                // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                host: {
                    '[class.adapt-login]': 'true',
                    '[style.display]': '"flex"'
                },
                templateUrl: './login.component.html',
                styles: [`
    .wrapper-login {
      height: auto;
    }
  `]
            }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.AdaptTranslateService }, { type: i2.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { logoClass: [{
            type: Input
        }], footerConfig: [{
            type: Input
        }], productName: [{
            type: Input
        }] }); })();
