// @helix/platform/esm2015/view/components/image/runtime/image.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ImageComponent } from './image.component';
import * as i0 from "@angular/core";
export class ImageModule {
}
ImageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ImageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ImageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ImageModule, declarations: [ImageComponent], imports: [CommonModule], exports: [ImageComponent] });
ImageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ImageModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ImageModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [ImageComponent],
                    entryComponents: [ImageComponent],
                    exports: [ImageComponent]
                }]
        }] });
