// @bmc-ux/adapt-angular/esm2015/dragon-input-handler/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptDragonInputHandlerDirective } from './dragon-input-handler.directive';
import * as i0 from "@angular/core";
export class AdaptDragonInputHandlerModule {
}
AdaptDragonInputHandlerModule.ɵfac = function AdaptDragonInputHandlerModule_Factory(t) { return new (t || AdaptDragonInputHandlerModule)(); };
AdaptDragonInputHandlerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptDragonInputHandlerModule });
AdaptDragonInputHandlerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDragonInputHandlerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptDragonInputHandlerDirective],
                exports: [AdaptDragonInputHandlerDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptDragonInputHandlerModule, { declarations: [AdaptDragonInputHandlerDirective], imports: [CommonModule], exports: [AdaptDragonInputHandlerDirective] }); })();
export * from './dragon-input-handler.directive';
