// @bmc-ux/obsolete/esm2015/lib/mention/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptMentionDirective } from './mention.directive';
import { AdaptMentionListComponent } from './mention-list/mention-list.component';
import { AdaptDropdownModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptMentionModule {
}
AdaptMentionModule.ɵfac = function AdaptMentionModule_Factory(t) { return new (t || AdaptMentionModule)(); };
AdaptMentionModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptMentionModule });
AdaptMentionModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule,
            AdaptDropdownModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMentionModule, [{
        type: NgModule,
        args: [{
                imports: [
                    FormsModule,
                    CommonModule,
                    AdaptDropdownModule.forRoot()
                ],
                exports: [AdaptMentionDirective],
                declarations: [AdaptMentionDirective, AdaptMentionListComponent],
                entryComponents: [AdaptMentionListComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptMentionModule, { declarations: [AdaptMentionDirective, AdaptMentionListComponent], imports: [FormsModule,
        CommonModule, i1.AdaptDropdownModule], exports: [AdaptMentionDirective] }); })();
export * from './mention.model';
export * from './mention.directive';
