// @bmc-ux/adapt-radar/esm2015/lib/providers/gainsight/gainsight.model.js
/**
 * The list of api methods that are used by the "aptrinsic" JavaScript call are as follows with links to customer facing documentation where available:
 *
 * 'track' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Use_Custom_Event_API
 * 'identify' - https://support.gainsight.com/PX/API_for_Developers/01About/Track_your_Users_and_Accounts#The_Identify_Call
 * 'get'
 * 'users'
 * 'set' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Use_Global_Context_API
 * 'initWithExternalState'
 * 'carousel'
 * 'config' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Use_Config_API
 * 'login'
 * 'lead' - https://support.gainsight.com/PX/Install_PX/Track_Users_on_Your_Corporate_Site_Using_PX#Customizing_Lead_Tracking
 * 'account'
 * 'checklog'
 * 'event'
 * 'setOnce' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Use_Global_Context_API#Global_Context
 * 'remove' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Use_Global_Context_API#Remove
 * 'pageView' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Using_The_Page_View_API
 * 'send' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Gainsight_PX_User_Feedback_API
 * 'bot' - https://support.gainsight.com/PX/Integrations/Knowledge_Center_Integrations/Zendesk_Knowledge_Base_Integration#Add_Label_Filters_to_Zendesk_Knowledge_Base
 * 'kcb' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Work_with_Knowledge_Center_Bot_API#Knowledge_Center_Bot_API_Calls
 * 'reset' - https://support.gainsight.com/PX/API_for_Developers/01About/Work_with_Gainsight_PX_Web_SDK#Reset.2FLogout
 * 'addListener' - https://support.gainsight.com/PX/API_for_Developers/02Usage_of_Different_APIs/Javascript_Event_Listener_API
 */
export var AdaptRadarGainsightEventName;
(function (AdaptRadarGainsightEventName) {
    AdaptRadarGainsightEventName["Track"] = "track";
    AdaptRadarGainsightEventName["Identify"] = "identify";
    AdaptRadarGainsightEventName["Get"] = "get";
    AdaptRadarGainsightEventName["Users"] = "users";
    AdaptRadarGainsightEventName["Set"] = "set";
    AdaptRadarGainsightEventName["SetOnce"] = "setOnce";
    AdaptRadarGainsightEventName["InitWithExternalState"] = "initWithExternalState";
    AdaptRadarGainsightEventName["Carousel"] = "carousel";
    AdaptRadarGainsightEventName["Config"] = "config";
    AdaptRadarGainsightEventName["Login"] = "login";
    AdaptRadarGainsightEventName["Lead"] = "lead";
    AdaptRadarGainsightEventName["Account"] = "account";
    AdaptRadarGainsightEventName["Checklog"] = "checklog";
    AdaptRadarGainsightEventName["Event"] = "event";
    AdaptRadarGainsightEventName["Remove"] = "remove";
    AdaptRadarGainsightEventName["PageView"] = "pageView";
    AdaptRadarGainsightEventName["Send"] = "send";
    AdaptRadarGainsightEventName["Bot"] = "bot";
    AdaptRadarGainsightEventName["Kcb"] = "kcb";
    AdaptRadarGainsightEventName["Reset"] = "reset";
    AdaptRadarGainsightEventName["AddListener"] = "addListener";
})(AdaptRadarGainsightEventName || (AdaptRadarGainsightEventName = {}));
export var AdaptRadarGainsightIdentifyActionName;
(function (AdaptRadarGainsightIdentifyActionName) {
    AdaptRadarGainsightIdentifyActionName["UserIdentify"] = "userIdentify";
})(AdaptRadarGainsightIdentifyActionName || (AdaptRadarGainsightIdentifyActionName = {}));
export var AdaptRadarGainsightConfigName;
(function (AdaptRadarGainsightConfigName) {
    AdaptRadarGainsightConfigName["EnableTag"] = "enableTag";
})(AdaptRadarGainsightConfigName || (AdaptRadarGainsightConfigName = {}));
