// @bmc-ux/adapt-angular/esm2015/common/directives/adapt-lifecycle-hooks.directive.js
import { EventEmitter, Directive, Output, NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptLifecycleHooksDirective {
    constructor() {
        /* eslint-disable-next-line @angular-eslint/no-output-on-prefix */
        this.onInit = new EventEmitter();
        this.afterViewInit = new EventEmitter();
        /* eslint-disable-next-line @angular-eslint/no-output-on-prefix */
        this.onDestroy = new EventEmitter();
    }
    ngOnInit() {
        this.onInit.emit();
    }
    ngAfterViewInit() {
        this.afterViewInit.emit();
    }
    ngOnDestroy() {
        this.onDestroy.emit();
    }
}
AdaptLifecycleHooksDirective.ɵfac = function AdaptLifecycleHooksDirective_Factory(t) { return new (t || AdaptLifecycleHooksDirective)(); };
AdaptLifecycleHooksDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptLifecycleHooksDirective, selectors: [["", "adaptLifecycleHooks", ""]], outputs: { onInit: "onInit", afterViewInit: "afterViewInit", onDestroy: "onDestroy" }, exportAs: ["adaptLifecycleHooks"] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptLifecycleHooksDirective, [{
        type: Directive,
        args: [{
                selector: '[adaptLifecycleHooks]',
                exportAs: 'adaptLifecycleHooks',
            }]
    }], null, { onInit: [{
            type: Output
        }], afterViewInit: [{
            type: Output
        }], onDestroy: [{
            type: Output
        }] }); })();
export class AdaptLifecycleHooksDirectiveModule {
}
AdaptLifecycleHooksDirectiveModule.ɵfac = function AdaptLifecycleHooksDirectiveModule_Factory(t) { return new (t || AdaptLifecycleHooksDirectiveModule)(); };
AdaptLifecycleHooksDirectiveModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptLifecycleHooksDirectiveModule });
AdaptLifecycleHooksDirectiveModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptLifecycleHooksDirectiveModule, [{
        type: NgModule,
        args: [{
                declarations: [AdaptLifecycleHooksDirective],
                exports: [AdaptLifecycleHooksDirective],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptLifecycleHooksDirectiveModule, { declarations: [AdaptLifecycleHooksDirective], exports: [AdaptLifecycleHooksDirective] }); })();
