// @bmc-ux/adapt-angular/esm2015/rx/rx-rating/adapt-rx-rating-binary/adapt-rx-rating-binary.component.js
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
import { transition, trigger, useAnimation } from '@angular/animations';
import { cloneDeep } from 'lodash-es';
import { ANIMATION_NAME, ANIMATION_TIMING_FUNCTION } from '../../../common/animation';
import { AdaptDeviceDetectionService } from '../../../common/device-detection.service';
import { getInteractionSource } from '../../../common/interaction-source';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularCustomEventType, AdaptRadarAngularService, AdaptRadarAngularGenericDirective } from '../../../common/radar/adapt-radar';
import { AdaptRxRatingConfigService } from '../adapt-rx-rating-config.service';
import { RxRatingBinaryAction, RxRatingBinaryType } from '../adapt-rx-rating.types';
import * as i0 from "@angular/core";
import * as i1 from "../adapt-rx-rating-config.service";
import * as i2 from "../../../common/radar/adapt-radar";
import * as i3 from "../../../common/device-detection.service";
import * as i4 from "@angular/common";
const _c0 = ["ratingContainerRef"];
const _c1 = function (a0) { return { "d-icon-thumb_up_adapt": a0 }; };
const _c2 = function (a0) { return { "d-icon-thumb_down_adapt": a0 }; };
function AdaptRxRatingBinaryComponent_div_2_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵlistener("@ratingBlockAnimation.start", function AdaptRxRatingBinaryComponent_div_2_div_1_Template_div_animation_ratingBlockAnimation_start_0_listener() { i0.ɵɵrestoreView(_r6); const ctx_r5 = i0.ɵɵnextContext(2); return ctx_r5.ratingBlockAnimationStateChange(false); })("@ratingBlockAnimation.done", function AdaptRxRatingBinaryComponent_div_2_div_1_Template_div_animation_ratingBlockAnimation_done_0_listener() { i0.ɵɵrestoreView(_r6); const ctx_r7 = i0.ɵɵnextContext(2); return ctx_r7.ratingBlockAnimationStateChange(true); });
    i0.ɵɵelementStart(1, "button", 6);
    i0.ɵɵlistener("click", function AdaptRxRatingBinaryComponent_div_2_div_1_Template_button_click_1_listener($event) { i0.ɵɵrestoreView(_r6); const ctx_r8 = i0.ɵɵnextContext(2); return ctx_r8.processAction(ctx_r8.UI_ACTION.Like, $event); });
    i0.ɵɵelementStart(2, "span");
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "button", 6);
    i0.ɵɵlistener("click", function AdaptRxRatingBinaryComponent_div_2_div_1_Template_button_click_4_listener($event) { i0.ɵɵrestoreView(_r6); const ctx_r9 = i0.ɵɵnextContext(2); return ctx_r9.processAction(ctx_r9.UI_ACTION.Dislike, $event); });
    i0.ɵɵelementStart(5, "span");
    i0.ɵɵtext(6);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("@ratingBlockAnimation", undefined);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(11, _c1, ctx_r3.type === ctx_r3.UI_TYPE.Icon));
    i0.ɵɵattribute("data-testid", ctx_r3.testID + "_item_like");
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("sr-only", ctx_r3.type === ctx_r3.UI_TYPE.Icon);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r3.uiTexts.btnTextLike);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(13, _c2, ctx_r3.type === ctx_r3.UI_TYPE.Icon));
    i0.ɵɵattribute("data-testid", ctx_r3.testID + "_item_dislike");
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("sr-only", ctx_r3.type === ctx_r3.UI_TYPE.Icon);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r3.uiTexts.btnTextDislike);
} }
function AdaptRxRatingBinaryComponent_div_2_ng_container_2_span_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r10.showMessage);
} }
function AdaptRxRatingBinaryComponent_div_2_ng_container_2_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1);
    i0.ɵɵelementStart(2, "span", 9);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵelement(4, "span", 7);
    i0.ɵɵprojection(5);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1(" ", ctx_r11.uiTexts.textMessage, " ");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate3("", ctx_r11.uiTexts.quotesLeft, "", ctx_r11.showBtnValue, "", ctx_r11.uiTexts.quotesRight, ".");
} }
function AdaptRxRatingBinaryComponent_div_2_ng_container_2_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r14 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 10);
    i0.ɵɵlistener("click", function AdaptRxRatingBinaryComponent_div_2_ng_container_2_button_4_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r14); const ctx_r13 = i0.ɵɵnextContext(3); return ctx_r13.processAction(ctx_r13.UI_ACTION.Undo, $event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = i0.ɵɵnextContext(3);
    i0.ɵɵattribute("data-testid", ctx_r12.testID + "_item_undo");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r12.uiTexts.btnTextUndo);
} }
function AdaptRxRatingBinaryComponent_div_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtemplate(1, AdaptRxRatingBinaryComponent_div_2_ng_container_2_span_1_Template, 2, 1, "span", 5);
    i0.ɵɵtemplate(2, AdaptRxRatingBinaryComponent_div_2_ng_container_2_span_2_Template, 6, 4, "span", 5);
    i0.ɵɵelement(3, "span", 7);
    i0.ɵɵtemplate(4, AdaptRxRatingBinaryComponent_div_2_ng_container_2_button_4_Template, 2, 2, "button", 8);
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r4.type === ctx_r4.UI_TYPE.Icon);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r4.type === ctx_r4.UI_TYPE.Text);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r4.uiTexts.btnTextUndo !== "");
} }
function AdaptRxRatingBinaryComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 4);
    i0.ɵɵtemplate(1, AdaptRxRatingBinaryComponent_div_2_div_1_Template, 7, 15, "div", 5);
    i0.ɵɵtemplate(2, AdaptRxRatingBinaryComponent_div_2_ng_container_2_Template, 5, 3, "ng-container", 5);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r1.toggle);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r1.toggle && ctx_r1.showUndo);
} }
function AdaptRxRatingBinaryComponent_div_3_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r18 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 17);
    i0.ɵɵlistener("@ratingBlockAnimation.start", function AdaptRxRatingBinaryComponent_div_3_button_6_Template_button_animation_ratingBlockAnimation_start_0_listener() { i0.ɵɵrestoreView(_r18); const ctx_r17 = i0.ɵɵnextContext(2); return ctx_r17.ratingBlockAnimationStateChange(false); })("@ratingBlockAnimation.done", function AdaptRxRatingBinaryComponent_div_3_button_6_Template_button_animation_ratingBlockAnimation_done_0_listener() { i0.ɵɵrestoreView(_r18); const ctx_r19 = i0.ɵɵnextContext(2); return ctx_r19.ratingBlockAnimationStateChange(true); })("click", function AdaptRxRatingBinaryComponent_div_3_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r18); const ctx_r20 = i0.ɵɵnextContext(2); return ctx_r20.processAction(ctx_r20.UI_ACTION.Decrease, $event); });
    i0.ɵɵelementStart(1, "span", 13);
    i0.ɵɵtext(2);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("@ratingBlockAnimation", undefined);
    i0.ɵɵattribute("data-testid", ctx_r15.testID + "_item_decrease");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r15.uiTexts.btnTextDislike);
} }
function AdaptRxRatingBinaryComponent_div_3_div_7_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r23 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 10);
    i0.ɵɵlistener("click", function AdaptRxRatingBinaryComponent_div_3_div_7_button_1_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r23); const ctx_r22 = i0.ɵɵnextContext(3); return ctx_r22.processAction(ctx_r22.UI_ACTION.Undo, $event); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r21 = i0.ɵɵnextContext(3);
    i0.ɵɵattribute("data-testid", ctx_r21.testID + "_item_undo");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r21.uiTexts.btnTextUndo);
} }
function AdaptRxRatingBinaryComponent_div_3_div_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 18);
    i0.ɵɵtemplate(1, AdaptRxRatingBinaryComponent_div_3_div_7_button_1_Template, 2, 2, "button", 8);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r16.uiTexts.btnTextUndo !== "");
} }
function AdaptRxRatingBinaryComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r25 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "div", 11);
    i0.ɵɵelementStart(1, "button", 12);
    i0.ɵɵlistener("click", function AdaptRxRatingBinaryComponent_div_3_Template_button_click_1_listener($event) { i0.ɵɵrestoreView(_r25); const ctx_r24 = i0.ɵɵnextContext(); return ctx_r24.processAction(ctx_r24.UI_ACTION.Increase, $event); });
    i0.ɵɵelementStart(2, "span", 13);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵelementEnd();
    i0.ɵɵelementStart(4, "div", 14);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtemplate(6, AdaptRxRatingBinaryComponent_div_3_button_6_Template, 3, 3, "button", 15);
    i0.ɵɵtemplate(7, AdaptRxRatingBinaryComponent_div_3_div_7_Template, 2, 1, "div", 16);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("a-rating_btn-hidden", !ctx_r2.toggle);
    i0.ɵɵproperty("@ratingBlockAnimation", undefined);
    i0.ɵɵattribute("data-testid", ctx_r2.testID + "_item_increase");
    i0.ɵɵadvance(2);
    i0.ɵɵtextInterpolate(ctx_r2.uiTexts.btnTextLike);
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("font-weight-semibold", !ctx_r2.toggle);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r2.value);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", ctx_r2.toggle);
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngIf", !ctx_r2.toggle && ctx_r2.showUndo);
} }
const _c3 = ["*"];
export class AdaptRxRatingBinaryComponent extends AdaptRadarAngularGenericDirective {
    constructor(_renderer, rxRatingConfigService, _adaptRadarService, _deviceDetectionService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptRxRatingBinary);
        this._renderer = _renderer;
        this._adaptRadarService = _adaptRadarService;
        this._deviceDetectionService = _deviceDetectionService;
        this.UI_TYPE = RxRatingBinaryType;
        this.UI_ACTION = RxRatingBinaryAction;
        this.toggle = true;
        this.isMobile = false;
        this.showUndo = false;
        /**
         * The mode of component
         * @docs-default RxRatingBinaryType.Text
         */
        this.type = RxRatingBinaryType.Text;
        /**
         * Value for RxRatingBinaryType.Count mode
         * @docs-default 0
         */
        this.value = 0;
        /**
         * @docs-default
         */
        this.analyticsMetadata = null;
        /* eslint-disable @angular-eslint/no-output-on-prefix */
        /**
         * An event fired when a user selects a new rating.
         */
        this.onRateChange = new EventEmitter();
        this.uiTexts = cloneDeep(rxRatingConfigService.texts.binary);
        this.isMobile = this._deviceDetectionService.isMobile();
    }
    /**
     * UI Texts
     * @docs-default AdaptRxRatingConfigService.texts.binary
     */
    set texts(newTexts) {
        Object.assign(this.uiTexts, newTexts);
    }
    processAction(action, event) {
        var _a;
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptRxRatingBinary, {
            GroupName: this.adaptRadarGroupName,
            EventType: AdaptRadarAngularCustomEventType.Click,
            Label: (_a = this.analyticsMetadata) === null || _a === void 0 ? void 0 : _a.label,
            RatingEntered: !!this.value,
            RatingValue: this.value,
            SubmitMethod: getInteractionSource(event, this.isMobile)
        });
        if (action !== RxRatingBinaryAction.Undo) {
            this._setHostStyles();
        }
        this._toggleState();
        switch (action) {
            case RxRatingBinaryAction.Like:
                this.showMessage = this.uiTexts.likeMessage;
                this.showBtnValue = this.uiTexts.btnTextLike;
                this.onRateChange.emit({
                    type: RxRatingBinaryAction.Like,
                    value: this.value
                });
                break;
            case RxRatingBinaryAction.Dislike:
                this.showMessage = this.uiTexts.dislikeMessage;
                this.showBtnValue = this.uiTexts.btnTextDislike;
                this.onRateChange.emit({
                    type: RxRatingBinaryAction.Dislike,
                    value: this.value
                });
                break;
            case RxRatingBinaryAction.Increase:
                this._savedValue = this.value;
                this.value++;
                this.onRateChange.emit({
                    type: RxRatingBinaryAction.Increase,
                    value: this.value
                });
                break;
            case RxRatingBinaryAction.Decrease:
                this._savedValue = this.value;
                this.value--;
                this.onRateChange.emit({
                    type: RxRatingBinaryAction.Decrease,
                    value: this.value
                });
                break;
            case RxRatingBinaryAction.Undo:
                this.value = this._savedValue;
                this.onRateChange.emit({
                    type: RxRatingBinaryAction.Undo,
                    value: this.value
                });
                break;
        }
    }
    ratingBlockAnimationStateChange(ratingBlockAnimationState) {
        this.showUndo = ratingBlockAnimationState;
    }
    _setHostStyles() {
        if (this.ratingContainerRef) {
            const { nativeElement } = this.ratingContainerRef;
            const { offsetHeight } = nativeElement;
            this._renderer.setStyle(nativeElement, 'minHeight', `${offsetHeight}px`);
        }
    }
    _toggleState() {
        this.toggle = !this.toggle;
    }
}
AdaptRxRatingBinaryComponent.ɵfac = function AdaptRxRatingBinaryComponent_Factory(t) { return new (t || AdaptRxRatingBinaryComponent)(i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i1.AdaptRxRatingConfigService), i0.ɵɵdirectiveInject(i2.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i3.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptRxRatingBinaryComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptRxRatingBinaryComponent, selectors: [["adapt-rx-rating-binary"]], viewQuery: function AdaptRxRatingBinaryComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.ratingContainerRef = _t.first);
    } }, inputs: { type: "type", value: "value", texts: "texts", testID: "testID", analyticsMetadata: "analyticsMetadata" }, outputs: { onRateChange: "onRateChange" }, features: [i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c3, decls: 4, vars: 2, consts: [[1, "a-rating-binary"], ["ratingContainerRef", ""], ["class", "btn-group align-items-center", 4, "ngIf"], ["class", "a-rating-group", 4, "ngIf"], [1, "btn-group", "align-items-center"], [4, "ngIf"], ["type", "button", 1, "btn", "btn-secondary", 3, "ngClass", "click"], [1, "letter-space"], ["type", "button", "class", "btn btn-link p-0", 3, "click", 4, "ngIf"], [1, "font-weight-semibold"], ["type", "button", 1, "btn", "btn-link", "p-0", 3, "click"], [1, "a-rating-group"], ["type", "button", 1, "btn", "btn-secondary", "d-icon-triangle_up", "a-rating_btn", 3, "click"], [1, "sr-only"], [1, "a-rating-group__value"], ["type", "button", "class", "btn btn-secondary d-icon-triangle_down", 3, "click", 4, "ngIf"], ["class", "a-rating__undo-container", 4, "ngIf"], ["type", "button", 1, "btn", "btn-secondary", "d-icon-triangle_down", 3, "click"], [1, "a-rating__undo-container"]], template: function AdaptRxRatingBinaryComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵelementStart(0, "div", 0, 1);
        i0.ɵɵtemplate(2, AdaptRxRatingBinaryComponent_div_2_Template, 3, 2, "div", 2);
        i0.ɵɵtemplate(3, AdaptRxRatingBinaryComponent_div_3_Template, 8, 10, "div", 3);
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.type !== ctx.UI_TYPE.Count);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.type === ctx.UI_TYPE.Count);
    } }, directives: [i4.NgIf, i4.NgClass], encapsulation: 2, data: { animation: [
            trigger('ratingBlockAnimation', [
                transition('void => *', [
                    useAnimation(ANIMATION_NAME.opacity, {
                        params: {
                            timingFunction: ANIMATION_TIMING_FUNCTION.base.transitionIn
                        }
                    })
                ]),
                transition('* => void', [
                    useAnimation(ANIMATION_NAME.opacity, {
                        params: {
                            timingFunction: ANIMATION_TIMING_FUNCTION.base.transitionOut,
                            from: '*',
                            to: 0
                        }
                    })
                ])
            ])
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxRatingBinaryComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-rx-rating-binary',
                templateUrl: './adapt-rx-rating-binary.component.html',
                preserveWhitespaces: false,
                animations: [
                    trigger('ratingBlockAnimation', [
                        transition('void => *', [
                            useAnimation(ANIMATION_NAME.opacity, {
                                params: {
                                    timingFunction: ANIMATION_TIMING_FUNCTION.base.transitionIn
                                }
                            })
                        ]),
                        transition('* => void', [
                            useAnimation(ANIMATION_NAME.opacity, {
                                params: {
                                    timingFunction: ANIMATION_TIMING_FUNCTION.base.transitionOut,
                                    from: '*',
                                    to: 0
                                }
                            })
                        ])
                    ])
                ]
            }]
    }], function () { return [{ type: i0.Renderer2 }, { type: i1.AdaptRxRatingConfigService }, { type: i2.AdaptRadarAngularService }, { type: i3.AdaptDeviceDetectionService }, { type: i0.ElementRef }]; }, { type: [{
            type: Input
        }], value: [{
            type: Input
        }], texts: [{
            type: Input
        }], testID: [{
            type: Input
        }], analyticsMetadata: [{
            type: Input
        }], onRateChange: [{
            type: Output
        }], ratingContainerRef: [{
            type: ViewChild,
            args: ['ratingContainerRef', { static: true }]
        }] }); })();
