// @bmc-ux/adapt-angular/esm2015/docked-panel/docked-panel.config.js
import { Injectable } from '@angular/core';
import { DockedPanelDirection, DockedPanelFooterAlign } from './docked-panel.model';
import * as i0 from "@angular/core";
/**
 * DockedPanel configuration service
 */
export class AdaptDockedPanelConfig {
    constructor() {
        this.direction = DockedPanelDirection.DEFAULT;
        this.title = '';
        this.content = '';
        this.floatFooter = false;
        this.footerAlign = DockedPanelFooterAlign.DEFAULT;
        this.hideBackdrop = false;
        this.blockDismiss = true;
        this.blockKeyboard = false;
        this.size = 'default';
    }
}
AdaptDockedPanelConfig.ɵfac = function AdaptDockedPanelConfig_Factory(t) { return new (t || AdaptDockedPanelConfig)(); };
AdaptDockedPanelConfig.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptDockedPanelConfig, factory: AdaptDockedPanelConfig.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDockedPanelConfig, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
