// @bmc-ux/adapt-angular/esm2015/docked-panel/docked-panel.model.js
import { merge } from 'lodash-es';
import { PopupDeferred } from '../common/popup-deferred';
export var DockedPanelDirection;
(function (DockedPanelDirection) {
    DockedPanelDirection["DEFAULT"] = "right";
    DockedPanelDirection["LEFT"] = "left";
    DockedPanelDirection["RIGHT"] = "right";
})(DockedPanelDirection || (DockedPanelDirection = {}));
export var DockedPanelFooterAlign;
(function (DockedPanelFooterAlign) {
    DockedPanelFooterAlign["DEFAULT"] = "auto";
    DockedPanelFooterAlign["AUTO"] = "auto";
    DockedPanelFooterAlign["REVERSE"] = "reverse";
    DockedPanelFooterAlign["LEFT"] = "left";
    DockedPanelFooterAlign["RIGHT"] = "right";
})(DockedPanelFooterAlign || (DockedPanelFooterAlign = {}));
export class DockedPanelInstance {
    constructor(config) {
        this.inactive = false;
        this.overModalOpen = false;
        this.id = DockedPanelInstance.getNewId();
        this.config = config;
    }
    static getNewId() {
        return 'docked-panel-' + DockedPanelInstance.idCounter++;
    }
    updateConfig(config) {
        this.config = merge(this.config, config);
    }
}
DockedPanelInstance.idCounter = 1;
export class DockedDeferred extends PopupDeferred {
}
