// @bmc-ux/adapt-angular/esm2015/docked-panel/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptDockedPanelComponent } from './docked-panel.component';
import { AdaptOverModalModule } from '../modal/over-modal';
import { AdaptFocusHolderModule } from '../focus-holder/index';
import { OverlayModule } from '@angular/cdk/overlay';
import { AdaptAlertAnimatedModule } from '../alert-animated/index';
import { AdaptModalModule } from '../modal/index';
import { AdaptDockedPanelContainerComponent } from './docked-panel.container';
import { AdaptDockedPanelService } from './docked-panel.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class AdaptDockedPanelModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptDockedPanelModule
        };
    }
}
AdaptDockedPanelModule.ɵfac = function AdaptDockedPanelModule_Factory(t) { return new (t || AdaptDockedPanelModule)(); };
AdaptDockedPanelModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptDockedPanelModule });
AdaptDockedPanelModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdaptDockedPanelService], imports: [[
            CommonModule,
            AdaptOverModalModule,
            AdaptAlertAnimatedModule,
            AdaptModalModule,
            OverlayModule,
            AdaptFocusHolderModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDockedPanelModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptOverModalModule,
                    AdaptAlertAnimatedModule,
                    AdaptModalModule,
                    OverlayModule,
                    AdaptFocusHolderModule
                ],
                declarations: [
                    AdaptDockedPanelContainerComponent,
                    AdaptDockedPanelComponent
                ],
                exports: [
                    AdaptDockedPanelContainerComponent,
                    AdaptDockedPanelComponent
                ],
                entryComponents: [
                    AdaptDockedPanelContainerComponent
                ],
                providers: [AdaptDockedPanelService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptDockedPanelModule, { declarations: [AdaptDockedPanelContainerComponent,
        AdaptDockedPanelComponent], imports: [CommonModule,
        AdaptOverModalModule,
        AdaptAlertAnimatedModule,
        AdaptModalModule,
        OverlayModule,
        AdaptFocusHolderModule], exports: [AdaptDockedPanelContainerComponent,
        AdaptDockedPanelComponent] }); })();
i0.ɵɵsetComponentScope(AdaptDockedPanelContainerComponent, [i1.NgForOf, AdaptDockedPanelComponent, i1.NgIf], []);
export * from './docked-panel.component';
export * from './docked-panel.container';
export * from './docked-panel.config';
export * from './docked-panel.model';
export * from './docked-panel.service';
export * from './docked-panel.context';
