// @bmc-ux/adapt-angular/esm2015/rx/form-control/rx-validators/index.js
import { NgModule } from '@angular/core';
import { AdaptEmailValidatorDirective, AdaptFloatNumberValidatorDirective, AdaptIntegerNumberValidatorDirective, AdaptMaxLengthValidatorDirective, AdaptMaxValidatorDirective, AdaptMinLengthValidatorDirective, AdaptMinValidatorDirective, AdaptPatternValidatorDirective, AdaptRangeValidatorDirective, AdaptRequiredTrueValidatorDirective, AdaptRequiredValidatorDirective, AdaptRxDatetimeFormatValidatorDirective, AdaptRxDatetimeRangeFormatValidatorDirective, AdaptRxDatetimeRangeOrderValidatorDirective, AdaptScientificNumberValidatorDirective } from './adapt-validators';
import * as i0 from "@angular/core";
export class AdaptRxValidatorsModule {
}
AdaptRxValidatorsModule.ɵfac = function AdaptRxValidatorsModule_Factory(t) { return new (t || AdaptRxValidatorsModule)(); };
AdaptRxValidatorsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxValidatorsModule });
AdaptRxValidatorsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxValidatorsModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    AdaptEmailValidatorDirective,
                    AdaptFloatNumberValidatorDirective,
                    AdaptIntegerNumberValidatorDirective,
                    AdaptMaxLengthValidatorDirective,
                    AdaptMaxValidatorDirective,
                    AdaptMinLengthValidatorDirective,
                    AdaptMinValidatorDirective,
                    AdaptPatternValidatorDirective,
                    AdaptRangeValidatorDirective,
                    AdaptRequiredTrueValidatorDirective,
                    AdaptRequiredValidatorDirective,
                    AdaptRxDatetimeFormatValidatorDirective,
                    AdaptRxDatetimeRangeFormatValidatorDirective,
                    AdaptRxDatetimeRangeOrderValidatorDirective,
                    AdaptScientificNumberValidatorDirective
                ],
                exports: [
                    AdaptEmailValidatorDirective,
                    AdaptFloatNumberValidatorDirective,
                    AdaptIntegerNumberValidatorDirective,
                    AdaptMaxLengthValidatorDirective,
                    AdaptMaxValidatorDirective,
                    AdaptMinLengthValidatorDirective,
                    AdaptMinValidatorDirective,
                    AdaptPatternValidatorDirective,
                    AdaptRangeValidatorDirective,
                    AdaptRequiredTrueValidatorDirective,
                    AdaptRequiredValidatorDirective,
                    AdaptRxDatetimeFormatValidatorDirective,
                    AdaptRxDatetimeRangeFormatValidatorDirective,
                    AdaptRxDatetimeRangeOrderValidatorDirective,
                    AdaptScientificNumberValidatorDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxValidatorsModule, { declarations: [AdaptEmailValidatorDirective,
        AdaptFloatNumberValidatorDirective,
        AdaptIntegerNumberValidatorDirective,
        AdaptMaxLengthValidatorDirective,
        AdaptMaxValidatorDirective,
        AdaptMinLengthValidatorDirective,
        AdaptMinValidatorDirective,
        AdaptPatternValidatorDirective,
        AdaptRangeValidatorDirective,
        AdaptRequiredTrueValidatorDirective,
        AdaptRequiredValidatorDirective,
        AdaptRxDatetimeFormatValidatorDirective,
        AdaptRxDatetimeRangeFormatValidatorDirective,
        AdaptRxDatetimeRangeOrderValidatorDirective,
        AdaptScientificNumberValidatorDirective], exports: [AdaptEmailValidatorDirective,
        AdaptFloatNumberValidatorDirective,
        AdaptIntegerNumberValidatorDirective,
        AdaptMaxLengthValidatorDirective,
        AdaptMaxValidatorDirective,
        AdaptMinLengthValidatorDirective,
        AdaptMinValidatorDirective,
        AdaptPatternValidatorDirective,
        AdaptRangeValidatorDirective,
        AdaptRequiredTrueValidatorDirective,
        AdaptRequiredValidatorDirective,
        AdaptRxDatetimeFormatValidatorDirective,
        AdaptRxDatetimeRangeFormatValidatorDirective,
        AdaptRxDatetimeRangeOrderValidatorDirective,
        AdaptScientificNumberValidatorDirective] }); })();
export * from './adapt-validators';
export * from './adapt-validators.model';
