// @bmc-ux/adapt-angular/esm2015/rx/rx-rating/adapt-rx-rating.types.js
export var RxRatingSize;
(function (RxRatingSize) {
    RxRatingSize[RxRatingSize["xs"] = 0] = "xs";
    RxRatingSize[RxRatingSize["sm"] = 1] = "sm";
    RxRatingSize[RxRatingSize["md"] = 2] = "md";
    RxRatingSize[RxRatingSize["lg"] = 3] = "lg";
})(RxRatingSize || (RxRatingSize = {}));
export var RxRatingBinaryType;
(function (RxRatingBinaryType) {
    RxRatingBinaryType[RxRatingBinaryType["Icon"] = 0] = "Icon";
    RxRatingBinaryType[RxRatingBinaryType["Text"] = 1] = "Text";
    RxRatingBinaryType[RxRatingBinaryType["Count"] = 2] = "Count";
})(RxRatingBinaryType || (RxRatingBinaryType = {}));
export var RxRatingBinaryAction;
(function (RxRatingBinaryAction) {
    RxRatingBinaryAction[RxRatingBinaryAction["Like"] = 0] = "Like";
    RxRatingBinaryAction[RxRatingBinaryAction["Dislike"] = 1] = "Dislike";
    RxRatingBinaryAction[RxRatingBinaryAction["Increase"] = 2] = "Increase";
    RxRatingBinaryAction[RxRatingBinaryAction["Decrease"] = 3] = "Decrease";
    RxRatingBinaryAction[RxRatingBinaryAction["Undo"] = 4] = "Undo";
})(RxRatingBinaryAction || (RxRatingBinaryAction = {}));
export var RxRatingLikeScale;
(function (RxRatingLikeScale) {
    RxRatingLikeScale[RxRatingLikeScale["ScalePoint5"] = 0] = "ScalePoint5";
    RxRatingLikeScale[RxRatingLikeScale["ScalePoint3"] = 1] = "ScalePoint3";
})(RxRatingLikeScale || (RxRatingLikeScale = {}));
export var RxRatingLikeItem;
(function (RxRatingLikeItem) {
    RxRatingLikeItem[RxRatingLikeItem["Terrible"] = 0] = "Terrible";
    RxRatingLikeItem[RxRatingLikeItem["Disliked"] = 1] = "Disliked";
    RxRatingLikeItem[RxRatingLikeItem["Ok"] = 2] = "Ok";
    RxRatingLikeItem[RxRatingLikeItem["Good"] = 3] = "Good";
    RxRatingLikeItem[RxRatingLikeItem["Excellent"] = 4] = "Excellent";
})(RxRatingLikeItem || (RxRatingLikeItem = {}));
export var RxRatingVotingType;
(function (RxRatingVotingType) {
    RxRatingVotingType[RxRatingVotingType["Star"] = 0] = "Star";
    RxRatingVotingType[RxRatingVotingType["Flag"] = 1] = "Flag";
    RxRatingVotingType[RxRatingVotingType["Like"] = 2] = "Like";
    RxRatingVotingType[RxRatingVotingType["Thanks"] = 3] = "Thanks";
})(RxRatingVotingType || (RxRatingVotingType = {}));
