// @bmc-ux/adapt-angular/esm2015/rx/rx-rating/adapt-rx-rating-config.service.js
import { Injectable } from '@angular/core';
import { AdaptTranslateService } from '../../common/i18n/index';
import * as i0 from "@angular/core";
import * as i1 from "../../common/i18n/index";
export class AdaptRxRatingConfigService {
    constructor(_ts) {
        this._ts = _ts;
    }
    get texts() {
        const translation = this._ts.getCurrentLanguage();
        return {
            rating: {
                tooltip: [
                    translation['adapt.rating.like.l1'],
                    translation['adapt.rating.like.l2'],
                    translation['adapt.rating.like.l3'],
                    translation['adapt.rating.like.l4'],
                    translation['adapt.rating.like.l5']
                ]
            },
            binary: {
                btnTextLike: translation['adapt.rating.binary.left'],
                btnTextDislike: translation['adapt.rating.binary.right'],
                btnTextUndo: translation['adapt.common.undo'],
                quotesLeft: translation['adapt.common.doubleQuotesLeft'],
                quotesRight: translation['adapt.common.doubleQuotesRight'],
                likeMessage: translation['adapt.rating.binary.like'],
                dislikeMessage: translation['adapt.rating.binary.dislike'],
                textMessage: translation['adapt.rating.binary.confirm']
            },
            like: {
                tooltipTerrible: translation['adapt.rating.like.l1'],
                tooltipDisliked: translation['adapt.rating.like.l2'],
                tooltipOk: translation['adapt.rating.like.l3'],
                tooltipGood: translation['adapt.rating.like.l4'],
                tooltipExcellent: translation['adapt.rating.like.l5']
            },
            voting: {
                thanksText: translation['adapt.rating.vote.thank'],
                thanksAbbrMessage: translation['adapt.rating.vote.confirm']
            }
        };
    }
}
AdaptRxRatingConfigService.ɵfac = function AdaptRxRatingConfigService_Factory(t) { return new (t || AdaptRxRatingConfigService)(i0.ɵɵinject(i1.AdaptTranslateService)); };
AdaptRxRatingConfigService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptRxRatingConfigService, factory: AdaptRxRatingConfigService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxRatingConfigService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.AdaptTranslateService }]; }, null); })();
