// @bmc-ux/adapt-angular/esm2015/rx/rx-rating/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptTooltipModule } from '../../tooltip/index';
import { AdaptIconModule } from '../../icon/index';
import { AdaptRxRatingComponent } from './adapt-rx-rating.component';
import { AdaptRxRatingVotingComponent } from './adapt-rx-rating-voting/adapt-rx-rating-voting.component';
import { AdaptRxRatingLikeComponent } from './adapt-rx-rating-like/adapt-rx-rating-like.component';
import { AdaptRxRatingBinaryComponent } from './adapt-rx-rating-binary/adapt-rx-rating-binary.component';
import * as i0 from "@angular/core";
export class AdaptRxRatingModule {
}
AdaptRxRatingModule.ɵfac = function AdaptRxRatingModule_Factory(t) { return new (t || AdaptRxRatingModule)(); };
AdaptRxRatingModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxRatingModule });
AdaptRxRatingModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule,
            AdaptTooltipModule,
            AdaptIconModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxRatingModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule,
                    AdaptTooltipModule,
                    AdaptIconModule
                ],
                exports: [
                    AdaptRxRatingBinaryComponent,
                    AdaptRxRatingComponent,
                    AdaptRxRatingLikeComponent,
                    AdaptRxRatingVotingComponent,
                ],
                declarations: [
                    AdaptRxRatingBinaryComponent,
                    AdaptRxRatingComponent,
                    AdaptRxRatingLikeComponent,
                    AdaptRxRatingVotingComponent,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxRatingModule, { declarations: [AdaptRxRatingBinaryComponent,
        AdaptRxRatingComponent,
        AdaptRxRatingLikeComponent,
        AdaptRxRatingVotingComponent], imports: [CommonModule,
        AdaptRxFormControlModule,
        AdaptTooltipModule,
        AdaptIconModule], exports: [AdaptRxRatingBinaryComponent,
        AdaptRxRatingComponent,
        AdaptRxRatingLikeComponent,
        AdaptRxRatingVotingComponent] }); })();
export * from './adapt-rx-rating-binary/adapt-rx-rating-binary.component';
export * from './adapt-rx-rating-config.service';
export * from './adapt-rx-rating-like/adapt-rx-rating-like.component';
export * from './adapt-rx-rating-voting/adapt-rx-rating-voting.component';
export * from './adapt-rx-rating.component';
export * from './adapt-rx-rating.types';
