// @bmc-ux/adapt-angular/esm2015/common/lazy-load/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptLazyLoadAnchorDirective } from './lazy-load-anchor.directive';
import * as i0 from "@angular/core";
export class AdaptLazyLoadModule {
}
AdaptLazyLoadModule.ɵfac = function AdaptLazyLoadModule_Factory(t) { return new (t || AdaptLazyLoadModule)(); };
AdaptLazyLoadModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptLazyLoadModule });
AdaptLazyLoadModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptLazyLoadModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptLazyLoadAnchorDirective],
                exports: [AdaptLazyLoadAnchorDirective],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptLazyLoadModule, { declarations: [AdaptLazyLoadAnchorDirective], imports: [CommonModule], exports: [AdaptLazyLoadAnchorDirective] }); })();
export * from './lazy-load-anchor.directive';
