// @bmc-ux/adapt-angular/esm2015/badge/badge.component.js
import { Component, ElementRef, Input } from '@angular/core';
import { AdaptRadarAngularService, AdaptRadarAngularGenericDirective, AdaptRadarAngularCustomEventName } from '../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "../common/radar/adapt-radar";
import * as i2 from "@angular/common";
function AdaptBadgeComponent_div_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 2);
    i0.ɵɵtext(1, "\u00A0");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("classList", "badge-dot badge-dot-" + ctx_r0.alertVariant);
} }
const _c0 = ["*"];
export class AdaptBadgeComponent extends AdaptRadarAngularGenericDirective {
    constructor(_adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptBadge);
        /**
         * Enables pulse animation
         * @docs-default false
         */
        this.animate = false;
        /**
         * Show/hide alert on badge
         * @since v10.1.0
         * @docs-default false
         */
        this.showAlert = false;
        /**
         * Type of badge: primary, secondary, info, danger, warning, success
         * @docs-default primary
         */
        this.variant = 'primary';
        /**
         * Type of alert: primary, secondary, info, danger, warning, success
         * @since v10.1.0
         * @docs-default info
         */
        this.alertVariant = 'info';
        /**
         * Set custom class for the badge
         * @docs-default
         */
        this.customCls = '';
    }
}
AdaptBadgeComponent.ɵfac = function AdaptBadgeComponent_Factory(t) { return new (t || AdaptBadgeComponent)(i0.ɵɵdirectiveInject(i1.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptBadgeComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptBadgeComponent, selectors: [["adapt-badge"]], inputs: { animate: "animate", showAlert: "showAlert", variant: "variant", alertVariant: "alertVariant", customCls: "customCls" }, features: [i0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c0, decls: 13, vars: 7, consts: [["tabindex", "0", 3, "classList"], [1, "hop", "hop-danger", 3, "classList"], [3, "classList"], [3, "classList", 4, "ngIf"]], template: function AdaptBadgeComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵprojectionDef();
        i0.ɵɵtext(0, "\n    ");
        i0.ɵɵelementStart(1, "div", 0);
        i0.ɵɵtext(2, "\n      ");
        i0.ɵɵelementStart(3, "div", 1);
        i0.ɵɵtext(4, "\n        ");
        i0.ɵɵelementStart(5, "div", 2);
        i0.ɵɵtext(6, "\n          ");
        i0.ɵɵtemplate(7, AdaptBadgeComponent_div_7_Template, 2, 1, "div", 3);
        i0.ɵɵtext(8, "\n          ");
        i0.ɵɵprojection(9);
        i0.ɵɵtext(10, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(11, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(12, "\n    ");
        i0.ɵɵelementEnd();
    } if (rf & 2) {
        i0.ɵɵadvance(1);
        i0.ɵɵclassProp("badge-animation", ctx.animate);
        i0.ɵɵproperty("classList", "badge-pulse" + " " + ctx.customCls);
        i0.ɵɵattribute("role", ctx.animate ? "alert" : "none");
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("classList", "hop hop-" + ctx.variant);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("classList", "badge badge-" + ctx.variant);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.showAlert);
    } }, directives: [i2.NgIf], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptBadgeComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-badge',
                template: `
    <div tabindex="0"
         [attr.role]="animate ? 'alert' : 'none'"
         [class.badge-animation]="animate"
         [classList]="'badge-pulse' + ' ' + customCls">
      <div [classList]="'hop hop-' + variant"
           class="hop hop-danger">
        <div [classList]="'badge badge-' + variant">
          <div *ngIf="showAlert" [classList]="'badge-dot badge-dot-' + alertVariant">&nbsp;</div>
          <ng-content></ng-content>
        </div>
      </div>
    </div>`
            }]
    }], function () { return [{ type: i1.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { animate: [{
            type: Input
        }], showAlert: [{
            type: Input
        }], variant: [{
            type: Input
        }], alertVariant: [{
            type: Input
        }], customCls: [{
            type: Input
        }] }); })();
