// @bmc-ux/adapt-angular/esm2015/badge/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptBadgeComponent } from './badge.component';
import * as i0 from "@angular/core";
export class AdaptBadgeModule {
}
AdaptBadgeModule.ɵfac = function AdaptBadgeModule_Factory(t) { return new (t || AdaptBadgeModule)(); };
AdaptBadgeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptBadgeModule });
AdaptBadgeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptBadgeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptBadgeComponent],
                exports: [AdaptBadgeComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptBadgeModule, { declarations: [AdaptBadgeComponent], imports: [CommonModule], exports: [AdaptBadgeComponent] }); })();
export * from './badge.component';
export * from './badge.model';
