// @bmc-ux/adapt-angular/esm2015/uploader-item/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptUploaderItemComponent } from './uploader-item.component';
import { AdaptTooltipModule } from '../tooltip/index';
import { AdaptBusyModule } from '../busy-loader/index';
import { AdaptTruncateFileNameModule } from '../truncate-file-name/index';
import * as i0 from "@angular/core";
export class AdaptUploadItemModule {
}
AdaptUploadItemModule.ɵfac = function AdaptUploadItemModule_Factory(t) { return new (t || AdaptUploadItemModule)(); };
AdaptUploadItemModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptUploadItemModule });
AdaptUploadItemModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptTooltipModule,
            AdaptBusyModule,
            AdaptTruncateFileNameModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptUploadItemModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptTooltipModule,
                    AdaptBusyModule,
                    AdaptTruncateFileNameModule
                ],
                declarations: [
                    AdaptUploaderItemComponent
                ],
                exports: [AdaptUploaderItemComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptUploadItemModule, { declarations: [AdaptUploaderItemComponent], imports: [CommonModule,
        AdaptTooltipModule,
        AdaptBusyModule,
        AdaptTruncateFileNameModule], exports: [AdaptUploaderItemComponent] }); })();
export * from './uploader-item.component';
