// @helix/platform/esm2015/web-api/designer/web-api-designer-page/web-api-designer-page.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RxWebApiDesignerModule } from '../web-api-designer';
import { RxWebApiDesignerPageComponent } from './web-api-designer-page.component';
import * as i0 from "@angular/core";
export class RxWebApiDesignerPageModule {
}
RxWebApiDesignerPageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxWebApiDesignerPageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxWebApiDesignerPageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxWebApiDesignerPageModule, declarations: [RxWebApiDesignerPageComponent], imports: [CommonModule, RxWebApiDesignerModule], exports: [RxWebApiDesignerPageComponent] });
RxWebApiDesignerPageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxWebApiDesignerPageModule, imports: [[CommonModule, RxWebApiDesignerModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxWebApiDesignerPageModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxWebApiDesignerPageComponent],
                    imports: [CommonModule, RxWebApiDesignerModule],
                    exports: [RxWebApiDesignerPageComponent]
                }]
        }] });
