// @bmc-ux/adapt-angular/esm2015/focus-holder/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptFocusHolderDirective } from './focus-holder.directive';
import * as i0 from "@angular/core";
export class AdaptFocusHolderModule {
}
AdaptFocusHolderModule.ɵfac = function AdaptFocusHolderModule_Factory(t) { return new (t || AdaptFocusHolderModule)(); };
AdaptFocusHolderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptFocusHolderModule });
AdaptFocusHolderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFocusHolderModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptFocusHolderDirective],
                exports: [AdaptFocusHolderDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptFocusHolderModule, { declarations: [AdaptFocusHolderDirective], imports: [CommonModule], exports: [AdaptFocusHolderDirective] }); })();
export * from './focus-holder.directive';
