// @helix/platform/esm2015/named-list/designer/named-list-designer-page/named-list-designer-page.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RxNamedListDesignerModule } from '../named-list-designer';
import { RxNamedListDesignerPageComponent } from './named-list-designer-page.component';
import * as i0 from "@angular/core";
export class RxNamedListDesignerPageModule {
}
RxNamedListDesignerPageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxNamedListDesignerPageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxNamedListDesignerPageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxNamedListDesignerPageModule, declarations: [RxNamedListDesignerPageComponent], imports: [CommonModule, RxNamedListDesignerModule], exports: [RxNamedListDesignerPageComponent] });
RxNamedListDesignerPageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxNamedListDesignerPageModule, imports: [[CommonModule, RxNamedListDesignerModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxNamedListDesignerPageModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxNamedListDesignerPageComponent],
                    imports: [CommonModule, RxNamedListDesignerModule],
                    exports: [RxNamedListDesignerPageComponent]
                }]
        }] });
