// @bmc-ux/adapt-angular/esm2015/common/animationDirectives/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptTextSlideDirective } from './text-slide.directive';
import * as i0 from "@angular/core";
export class AdaptAnimationDirectivesModule {
}
AdaptAnimationDirectivesModule.ɵfac = function AdaptAnimationDirectivesModule_Factory(t) { return new (t || AdaptAnimationDirectivesModule)(); };
AdaptAnimationDirectivesModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAnimationDirectivesModule });
AdaptAnimationDirectivesModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAnimationDirectivesModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptTextSlideDirective
                ],
                exports: [
                    AdaptTextSlideDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAnimationDirectivesModule, { declarations: [AdaptTextSlideDirective], imports: [CommonModule], exports: [AdaptTextSlideDirective] }); })();
export * from './text-slide.directive';
