// @bmc-ux/adapt-angular/esm2015/tabs-controls/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptTabsControlsDirective } from './tabs-controls.directive';
import * as i0 from "@angular/core";
export class AdaptTabsControlsModule {
}
AdaptTabsControlsModule.ɵfac = function AdaptTabsControlsModule_Factory(t) { return new (t || AdaptTabsControlsModule)(); };
AdaptTabsControlsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTabsControlsModule });
AdaptTabsControlsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTabsControlsModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptTabsControlsDirective],
                exports: [AdaptTabsControlsDirective]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTabsControlsModule, { declarations: [AdaptTabsControlsDirective], imports: [CommonModule], exports: [AdaptTabsControlsDirective] }); })();
export * from './tabs-controls.directive';
export * from './tabs-controls.enum';
