// @bmc-ux/obsolete/esm2015/lib/typeahead/typeahed.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { AdaptTypeaheadWindowComponent } from './typeahead-window';
import { AdaptTypeaheadDirective } from './typeahead';
import { AdaptTypeaheadModel } from './typeahead.config';
import { AdaptDeviceDetectionService, AdaptHighlightModule, ARIA_LIVE_DELAY, DEFAULT_ARIA_LIVE_DELAY, Live } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
export class AdaptTypeaheadSubModule {
    static forRoot() {
        return {
            ngModule: AdaptTypeaheadSubModule,
            providers: [
                Live,
                AdaptTypeaheadModel, {
                    provide: ARIA_LIVE_DELAY,
                    useValue: DEFAULT_ARIA_LIVE_DELAY
                },
                AdaptDeviceDetectionService
            ]
        };
    }
}
AdaptTypeaheadSubModule.ɵfac = function AdaptTypeaheadSubModule_Factory(t) { return new (t || AdaptTypeaheadSubModule)(); };
AdaptTypeaheadSubModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTypeaheadSubModule });
AdaptTypeaheadSubModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, ScrollingModule, AdaptHighlightModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTypeaheadSubModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    AdaptTypeaheadDirective,
                    AdaptTypeaheadWindowComponent
                ],
                exports: [AdaptTypeaheadDirective,],
                imports: [CommonModule, ScrollingModule, AdaptHighlightModule],
                entryComponents: [AdaptTypeaheadWindowComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTypeaheadSubModule, { declarations: [AdaptTypeaheadDirective,
        AdaptTypeaheadWindowComponent], imports: [CommonModule, ScrollingModule, AdaptHighlightModule], exports: [AdaptTypeaheadDirective] }); })();
export * from './typeahead-window';
export * from './typeahead.config';
export * from './typeahead';
