// @bmc-ux/adapt-angular/esm2015/tabs/index.js
import { DragDropModule } from '@angular/cdk/drag-drop';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptBadgeModule } from '../badge/index';
import { AdaptDropdownModule } from '../dropdown/index';
import { AdaptTabsControlsModule } from '../tabs-controls/index';
import { AdaptTabsPanelComponent } from './tabs-panel.component';
import { AdaptTabsComponent } from './tabs.component';
import * as i0 from "@angular/core";
const ADAPT_TABS_DIRECTIVES = [
    AdaptTabsComponent,
    AdaptTabsPanelComponent
];
export class AdaptTabsModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptTabsModule
        };
    }
}
AdaptTabsModule.ɵfac = function AdaptTabsModule_Factory(t) { return new (t || AdaptTabsModule)(); };
AdaptTabsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTabsModule });
AdaptTabsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptDropdownModule,
            AdaptTabsControlsModule,
            DragDropModule,
            AdaptBadgeModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTabsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptDropdownModule,
                    AdaptTabsControlsModule,
                    DragDropModule,
                    AdaptBadgeModule
                ],
                exports: ADAPT_TABS_DIRECTIVES,
                declarations: [...ADAPT_TABS_DIRECTIVES]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTabsModule, { declarations: [AdaptTabsComponent,
        AdaptTabsPanelComponent], imports: [CommonModule,
        AdaptDropdownModule,
        AdaptTabsControlsModule,
        DragDropModule,
        AdaptBadgeModule], exports: [AdaptTabsComponent,
        AdaptTabsPanelComponent] }); })();
export * from './tabs.component';
export * from './tabs-panel.component';
