// @bmc-ux/adapt-angular/esm2015/agreement/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptRxCheckboxModule, AdaptRxSwitchModule } from '../rx/index';
import { AdaptAlertModule } from '../alert/index';
import { AdaptCollapseModule } from '../collapse/index';
import { AdaptButtonModule } from '../button/index';
import { AdaptModalModule } from '../modal/index';
import { AdaptToastModule } from '../toast/index';
import { AdaptBadgeModule } from '../badge/index';
import { AdaptEmptyStateModule } from '../empty-state/index';
import { AdaptAgreementCardComponent } from './agreement-card/agreement-card.component';
import { AdaptAgreementAdminModalComponent } from './agreement-admin-modal/agreement-admin-modal.component';
import { AdaptAgreementAdminContentComponent } from './agreement-admin-content/agreement-admin-content.component';
import { AdaptAgreementUserModalComponent } from './agreement-user-modal/agreement-user-modal.component';
import { AdaptAgreementNotificationComponent } from './agreement-notification/agreement-notification.component';
import { AdaptAgreementContentComponent } from './agreement-content/agreement-content.component';
import * as i0 from "@angular/core";
export class AdaptAgreementModule {
}
AdaptAgreementModule.ɵfac = function AdaptAgreementModule_Factory(t) { return new (t || AdaptAgreementModule)(); };
AdaptAgreementModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAgreementModule });
AdaptAgreementModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptRxSwitchModule,
            AdaptAlertModule,
            AdaptCollapseModule,
            AdaptButtonModule,
            AdaptToastModule,
            AdaptModalModule,
            AdaptRxCheckboxModule,
            AdaptBadgeModule,
            AdaptEmptyStateModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAgreementModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptRxSwitchModule,
                    AdaptAlertModule,
                    AdaptCollapseModule,
                    AdaptButtonModule,
                    AdaptToastModule,
                    AdaptModalModule,
                    AdaptRxCheckboxModule,
                    AdaptBadgeModule,
                    AdaptEmptyStateModule
                ],
                declarations: [
                    AdaptAgreementContentComponent,
                    AdaptAgreementCardComponent,
                    AdaptAgreementAdminModalComponent,
                    AdaptAgreementAdminContentComponent,
                    AdaptAgreementUserModalComponent,
                    AdaptAgreementNotificationComponent
                ],
                exports: [
                    AdaptAgreementContentComponent,
                    AdaptAgreementCardComponent,
                    AdaptAgreementAdminModalComponent,
                    AdaptAgreementAdminContentComponent,
                    AdaptAgreementUserModalComponent,
                    AdaptAgreementNotificationComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAgreementModule, { declarations: [AdaptAgreementContentComponent,
        AdaptAgreementCardComponent,
        AdaptAgreementAdminModalComponent,
        AdaptAgreementAdminContentComponent,
        AdaptAgreementUserModalComponent,
        AdaptAgreementNotificationComponent], imports: [CommonModule,
        FormsModule,
        AdaptRxSwitchModule,
        AdaptAlertModule,
        AdaptCollapseModule,
        AdaptButtonModule,
        AdaptToastModule,
        AdaptModalModule,
        AdaptRxCheckboxModule,
        AdaptBadgeModule,
        AdaptEmptyStateModule], exports: [AdaptAgreementContentComponent,
        AdaptAgreementCardComponent,
        AdaptAgreementAdminModalComponent,
        AdaptAgreementAdminContentComponent,
        AdaptAgreementUserModalComponent,
        AdaptAgreementNotificationComponent] }); })();
export * from './agreement-card/agreement-card.component';
export * from './agreement-admin-modal/agreement-admin-modal.component';
export * from './agreement-admin-content/agreement-admin-content.component';
export * from './agreement-user-modal/agreement-user-modal.component';
export * from './agreement-notification/agreement-notification.component';
export * from './agreement-content/agreement-content.component';
export * from './argeement.model';
