// @helix/platform/esm2015/view/shell-designer-page/components/shell/controls/shell-global-search-records-control/shell-global-search-records-control.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RxShellGlobalSearchRecordsControlComponent } from './shell-global-search-records-control.component';
import { AdaptButtonModule } from '@bmc-ux/adapt-angular';
import { RxShellGlobalSearchRecordsModalModule } from './shell-global-search-records-modal/shell-global-search-records-modal.module';
import * as i0 from "@angular/core";
export class RxShellGlobalSearchRecordsControlModule {
}
RxShellGlobalSearchRecordsControlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxShellGlobalSearchRecordsControlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxShellGlobalSearchRecordsControlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxShellGlobalSearchRecordsControlModule, declarations: [RxShellGlobalSearchRecordsControlComponent], imports: [CommonModule, AdaptButtonModule, RxShellGlobalSearchRecordsModalModule], exports: [RxShellGlobalSearchRecordsControlComponent] });
RxShellGlobalSearchRecordsControlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxShellGlobalSearchRecordsControlModule, imports: [[CommonModule, AdaptButtonModule, RxShellGlobalSearchRecordsModalModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxShellGlobalSearchRecordsControlModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxShellGlobalSearchRecordsControlComponent],
                    exports: [RxShellGlobalSearchRecordsControlComponent],
                    entryComponents: [RxShellGlobalSearchRecordsControlComponent],
                    imports: [CommonModule, AdaptButtonModule, RxShellGlobalSearchRecordsModalModule]
                }]
        }] });
