// @bmc-ux/adapt-angular/esm2015/metatag/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptCloseModule } from '../close/index';
import { AdaptDragonInputHandlerModule } from '../dragon-input-handler/index';
import { AdaptIconModule } from '../icon/index';
import { AdaptPopoverModule } from '../popover/index';
import { AdaptRxLabelModule } from '../rx/form-control/index';
import { AdaptRxTypeaheadModule } from '../rx/rx-typeahead/index';
import { AdaptTooltipModule } from '../tooltip/index';
import { AdaptTruncatePipe } from './../common/truncated.pipe';
import { AdaptMetatagComponent } from './metatag.component';
import * as i0 from "@angular/core";
export class AdaptMetatagModule {
}
AdaptMetatagModule.ɵfac = function AdaptMetatagModule_Factory(t) { return new (t || AdaptMetatagModule)(); };
AdaptMetatagModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptMetatagModule });
AdaptMetatagModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdaptTruncatePipe], imports: [[
            CommonModule,
            FormsModule,
            AdaptCloseModule,
            AdaptDragonInputHandlerModule,
            AdaptIconModule,
            AdaptPopoverModule,
            AdaptRxLabelModule,
            AdaptRxTypeaheadModule,
            AdaptTooltipModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMetatagModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptCloseModule,
                    AdaptDragonInputHandlerModule,
                    AdaptIconModule,
                    AdaptPopoverModule,
                    AdaptRxLabelModule,
                    AdaptRxTypeaheadModule,
                    AdaptTooltipModule
                ],
                declarations: [AdaptMetatagComponent],
                exports: [AdaptMetatagComponent],
                providers: [AdaptTruncatePipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptMetatagModule, { declarations: [AdaptMetatagComponent], imports: [CommonModule,
        FormsModule,
        AdaptCloseModule,
        AdaptDragonInputHandlerModule,
        AdaptIconModule,
        AdaptPopoverModule,
        AdaptRxLabelModule,
        AdaptRxTypeaheadModule,
        AdaptTooltipModule], exports: [AdaptMetatagComponent] }); })();
export * from './metatag.component';
export * from './metatag.types';
