// @bmc-ux/adapt-angular/esm2015/sidebar/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptSidebarComponent } from './sidebar.component';
import { AdaptSidebarItemComponent } from './sidebar-item/sidebar-item.component';
import { AdaptSidebarColumnComponent } from './sidebar-column/sidebar-column.component';
import { AdaptCloseModule } from '../close/index';
import { AdaptTooltipModule } from '../tooltip/index';
import * as i0 from "@angular/core";
export class AdaptSidebarModule {
}
AdaptSidebarModule.ɵfac = function AdaptSidebarModule_Factory(t) { return new (t || AdaptSidebarModule)(); };
AdaptSidebarModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSidebarModule });
AdaptSidebarModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptCloseModule, AdaptTooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSidebarModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptCloseModule, AdaptTooltipModule],
                declarations: [AdaptSidebarComponent, AdaptSidebarItemComponent, AdaptSidebarColumnComponent],
                exports: [AdaptSidebarComponent, AdaptSidebarItemComponent, AdaptSidebarColumnComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSidebarModule, { declarations: [AdaptSidebarComponent, AdaptSidebarItemComponent, AdaptSidebarColumnComponent], imports: [CommonModule, AdaptCloseModule, AdaptTooltipModule], exports: [AdaptSidebarComponent, AdaptSidebarItemComponent, AdaptSidebarColumnComponent] }); })();
export * from './sidebar.component';
export * from './sidebar-item/sidebar-item.component';
export * from './sidebar-column/sidebar-column.component';
