// @bmc-ux/obsolete/esm2015/lib/search/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptSearchComponent } from './search.component';
import { AdaptDragonInputHandlerModule, AdaptButtonModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
export class AdaptSearchModule {
}
AdaptSearchModule.ɵfac = function AdaptSearchModule_Factory(t) { return new (t || AdaptSearchModule)(); };
AdaptSearchModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSearchModule });
AdaptSearchModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, FormsModule, AdaptDragonInputHandlerModule, AdaptButtonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSearchModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, AdaptDragonInputHandlerModule, AdaptButtonModule],
                declarations: [AdaptSearchComponent],
                exports: [AdaptSearchComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSearchModule, { declarations: [AdaptSearchComponent], imports: [CommonModule, FormsModule, AdaptDragonInputHandlerModule, AdaptButtonModule], exports: [AdaptSearchComponent] }); })();
export * from './search.component';
