// @bmc-ux/adapt-angular/esm2015/downloader/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { IconManager } from '../common/uploader/services/icons-manager';
import { AdaptTruncateFileNameModule } from '../truncate-file-name/index';
import { AdaptDownloaderComponent } from './downloader.component';
import { AdaptTooltipModule } from '../tooltip/index';
import * as i0 from "@angular/core";
export class AdaptDownloadModule {
}
AdaptDownloadModule.ɵfac = function AdaptDownloadModule_Factory(t) { return new (t || AdaptDownloadModule)(); };
AdaptDownloadModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptDownloadModule });
AdaptDownloadModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [IconManager], imports: [[
            CommonModule,
            AdaptTooltipModule,
            AdaptTruncateFileNameModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDownloadModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptTooltipModule,
                    AdaptTruncateFileNameModule
                ],
                declarations: [
                    AdaptDownloaderComponent
                ],
                exports: [AdaptDownloaderComponent],
                providers: [IconManager]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptDownloadModule, { declarations: [AdaptDownloaderComponent], imports: [CommonModule,
        AdaptTooltipModule,
        AdaptTruncateFileNameModule], exports: [AdaptDownloaderComponent] }); })();
export * from './downloader.component';
export * from './downloader.model';
