// @bmc-ux/obsolete/esm2015/lib/datetime2/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptIconModule, AdaptDropdownModule, AdaptTooltipModule, AdaptPopoverModule, AdaptAnimationDirectivesModule, AdaptRxLabelModule, AdaptDragonInputHandlerModule, AdaptIconConfig } from '@bmc-ux/adapt-angular';
import { ai_angle_left } from '@bmc-ux/dpl-iconfont';
import { AdaptDatetime2CalendarComponent, AdaptDatetime2ClockComponent, AdaptDatetime2PickerComponent } from './datetime2';
import { AdaptDatetime2Component } from './datetime2.component';
import { AdaptDatetimeRange2Component } from './datetime-range2.component';
import { AdaptDatetime2YearPickerComponent } from './datetime-year-picker.component';
import { AdaptDatetime2MonthPickerComponent } from './datetime-month-picker.component';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
import * as i2 from "@angular/common";
export class AdaptDatetime2Module {
    constructor(iconConfig) {
        iconConfig.registerIcons([ai_angle_left]);
    }
}
AdaptDatetime2Module.ɵfac = function AdaptDatetime2Module_Factory(t) { return new (t || AdaptDatetime2Module)(i0.ɵɵinject(i1.AdaptIconConfig)); };
AdaptDatetime2Module.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptDatetime2Module });
AdaptDatetime2Module.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptDropdownModule.forRoot(),
            AdaptIconModule,
            AdaptPopoverModule.forRoot(),
            AdaptTooltipModule,
            AdaptAnimationDirectivesModule,
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDatetime2Module, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptDropdownModule.forRoot(),
                    AdaptIconModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptTooltipModule,
                    AdaptAnimationDirectivesModule,
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule
                ],
                declarations: [
                    AdaptDatetime2CalendarComponent,
                    AdaptDatetime2ClockComponent,
                    AdaptDatetime2Component,
                    AdaptDatetime2PickerComponent,
                    AdaptDatetimeRange2Component,
                    AdaptDatetime2MonthPickerComponent,
                    AdaptDatetime2YearPickerComponent
                ],
                exports: [
                    AdaptDatetime2CalendarComponent,
                    AdaptDatetime2ClockComponent,
                    AdaptDatetime2Component,
                    AdaptDatetime2PickerComponent,
                    AdaptDatetimeRange2Component,
                    AdaptDatetime2MonthPickerComponent,
                    AdaptDatetime2YearPickerComponent
                ]
            }]
    }], function () { return [{ type: i1.AdaptIconConfig }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptDatetime2Module, { declarations: [AdaptDatetime2CalendarComponent,
        AdaptDatetime2ClockComponent,
        AdaptDatetime2Component,
        AdaptDatetime2PickerComponent,
        AdaptDatetimeRange2Component,
        AdaptDatetime2MonthPickerComponent,
        AdaptDatetime2YearPickerComponent], imports: [CommonModule,
        FormsModule, i1.AdaptDropdownModule, AdaptIconModule, i1.AdaptPopoverModule, AdaptTooltipModule,
        AdaptAnimationDirectivesModule,
        AdaptDragonInputHandlerModule,
        AdaptRxLabelModule], exports: [AdaptDatetime2CalendarComponent,
        AdaptDatetime2ClockComponent,
        AdaptDatetime2Component,
        AdaptDatetime2PickerComponent,
        AdaptDatetimeRange2Component,
        AdaptDatetime2MonthPickerComponent,
        AdaptDatetime2YearPickerComponent] }); })();
i0.ɵɵsetComponentScope(AdaptDatetime2Component, [i1.AdaptDropdownDirective, i1.AdaptDropdownAnchorDirective, i1.AdaptRxControlLabelComponent, i2.NgIf, i1.AdaptDragonInputHandlerDirective, i1.AdaptDropdownMenuTemplateDirective, i2.NgClass, AdaptDatetime2PickerComponent, i2.NgForOf], []);
i0.ɵɵsetComponentScope(AdaptDatetimeRange2Component, [i1.AdaptDropdownDirective, i1.AdaptDropdownAnchorDirective, i1.AdaptRxControlLabelComponent, i2.NgIf, i1.AdaptDragonInputHandlerDirective, i1.AdaptDropdownMenuTemplateDirective, i2.NgClass, AdaptDatetime2PickerComponent, i2.NgForOf], []);
export * from './datetime2';
export * from './datetime2.model';
export { AdaptDatetime2Component } from './datetime2.component';
export { AdaptDatetimeRange2Component } from './datetime-range2.component';
export * from './datetime-year-picker.component';
export * from './datetime-month-picker.component';
