// @bmc-ux/adapt-angular/esm2015/alert-animated/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptAlertAnimatedComponent } from './alert-animated.component';
import { AdaptTruncatePipeModule } from '../common/truncated.pipe';
import * as i0 from "@angular/core";
export class AdaptAlertAnimatedModule {
}
AdaptAlertAnimatedModule.ɵfac = function AdaptAlertAnimatedModule_Factory(t) { return new (t || AdaptAlertAnimatedModule)(); };
AdaptAlertAnimatedModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAlertAnimatedModule });
AdaptAlertAnimatedModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptTruncatePipeModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAlertAnimatedModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptTruncatePipeModule],
                exports: [AdaptAlertAnimatedComponent],
                declarations: [AdaptAlertAnimatedComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAlertAnimatedModule, { declarations: [AdaptAlertAnimatedComponent], imports: [CommonModule, AdaptTruncatePipeModule], exports: [AdaptAlertAnimatedComponent] }); })();
export * from './alert-animated.component';
export * from './adapt-alert.model';
