// @bmc-ux/adapt-angular/esm2015/time/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { AdaptDropdownModule } from '../dropdown/index';
import { ClockComponent } from './clock.component';
import { AdaptTimeComponent } from './time.component';
import { AdaptIconModule } from '../icon/index';
import { AdaptTooltipModule } from '../tooltip/index';
import { AdaptPopoverModule } from '../popover/index';
import { AdaptDragonInputHandlerModule } from '../dragon-input-handler/index';
import { AdaptRxLabelModule } from '../rx/form-control/index';
import * as i0 from "@angular/core";
export class AdaptTimeModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptTimeModule
        };
    }
}
AdaptTimeModule.ɵfac = function AdaptTimeModule_Factory(t) { return new (t || AdaptTimeModule)(); };
AdaptTimeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTimeModule });
AdaptTimeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptDropdownModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule,
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTimeModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptDropdownModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule,
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule
                ],
                declarations: [ClockComponent, AdaptTimeComponent],
                exports: [ClockComponent, AdaptTimeComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTimeModule, { declarations: [ClockComponent, AdaptTimeComponent], imports: [CommonModule,
        FormsModule,
        AdaptDropdownModule,
        AdaptIconModule,
        AdaptTooltipModule,
        AdaptPopoverModule,
        AdaptDragonInputHandlerModule,
        AdaptRxLabelModule], exports: [ClockComponent, AdaptTimeComponent] }); })();
export * from './clock.component';
export * from './time.component';
