// @bmc-ux/adapt-angular/esm2015/common/pipes/sortby.pipe.js
import { NgModule, Pipe } from '@angular/core';
import { isNil, isString, toNumber, toString } from 'lodash-es';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class AdaptSortByPipe {
    transform(items, comparator, sortType, sortOrder = 'ASC') {
        if (isNil(items)) {
            return items;
        }
        else {
            const itemsCopy = [...items];
            if (isString(comparator)) {
                switch (sortType) {
                    case 'string':
                        itemsCopy.sort((a, b) => {
                            // undefined and null values goes to bottom\top
                            if (isNil(a[comparator])) {
                                return sortOrder === 'DESC' ? -1 : 1;
                            }
                            else if (isNil(b[comparator])) {
                                return sortOrder === 'DESC' ? 1 : -1;
                            }
                            return sortOrder === 'DESC'
                                ? b[comparator].localeCompare(a[comparator])
                                : a[comparator].localeCompare(b[comparator]);
                        });
                        break;
                    case 'number':
                        itemsCopy.sort((a, b) => {
                            // undefined and null values goes to bottom\top
                            if (isNil(a[comparator])) {
                                return sortOrder === 'DESC' ? -1 : 1;
                            }
                            else if (isNil(b[comparator])) {
                                return sortOrder === 'DESC' ? 1 : -1;
                            }
                            return sortOrder === 'DESC'
                                ? b[comparator] - a[comparator]
                                : a[comparator] - b[comparator];
                        });
                        break;
                    case 'boolean':
                        itemsCopy.sort((a, b) => {
                            // undefined and null values goes to bottom\top
                            if (isNil(a[comparator])) {
                                return sortOrder === 'DESC' ? -1 : 1;
                            }
                            else if (isNil(b[comparator])) {
                                return sortOrder === 'DESC' ? 1 : -1;
                            }
                            return sortOrder === 'DESC'
                                ? toNumber(!!a[comparator]) - toNumber(!!b[comparator])
                                : toNumber(!!b[comparator]) - toNumber(!!a[comparator]);
                        });
                        break;
                    default:
                        itemsCopy.sort((a, b) => {
                            if (isNil(a[comparator])) {
                                return sortOrder === 'DESC' ? -1 : 1;
                            }
                            else if (isNil(b[comparator])) {
                                return sortOrder === 'DESC' ? 1 : -1;
                            }
                            else {
                                if (toString(a[comparator]) < toString(b[comparator])) {
                                    return sortOrder === 'DESC' ? 1 : -1;
                                }
                                if (toString(a[comparator]) > toString(b[comparator])) {
                                    return sortOrder === 'DESC' ? -1 : 1;
                                }
                            }
                            return 0;
                        });
                }
            }
            else {
                itemsCopy.sort(comparator);
            }
            return itemsCopy;
        }
    }
}
AdaptSortByPipe.ɵfac = function AdaptSortByPipe_Factory(t) { return new (t || AdaptSortByPipe)(); };
AdaptSortByPipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "sortBy", type: AdaptSortByPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSortByPipe, [{
        type: Pipe,
        args: [{
                name: 'sortBy'
            }]
    }], null, null); })();
export class AdaptSortByPipeModule {
}
AdaptSortByPipeModule.ɵfac = function AdaptSortByPipeModule_Factory(t) { return new (t || AdaptSortByPipeModule)(); };
AdaptSortByPipeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptSortByPipeModule });
AdaptSortByPipeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptSortByPipeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptSortByPipe],
                exports: [AdaptSortByPipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptSortByPipeModule, { declarations: [AdaptSortByPipe], imports: [CommonModule], exports: [AdaptSortByPipe] }); })();
