// @helix/platform/esm2015/shared/components/user-preferences/user-preferences.constants.js
export const RX_USER_PREFERENCES = {
    supportedLocales: [
        {
            id: 'en',
            name: 'English'
        },
        {
            id: 'fr',
            name: 'français'
        },
        {
            id: 'de',
            name: 'Deutsch'
        },
        {
            id: 'it',
            name: 'italiano'
        },
        {
            id: 'zh-cn',
            name: '中文'
        },
        {
            id: 'ko',
            name: '한국어'
        },
        {
            id: 'pt-br',
            name: 'português'
        },
        {
            id: 'ru',
            name: 'русский'
        },
        {
            id: 'ja',
            name: '日本語'
        },
        {
            id: 'es',
            name: 'Español'
        },
        {
            id: 'nl',
            name: 'Nederlands'
        },
        {
            id: 'sv',
            name: 'svenska'
        },
        {
            id: 'da',
            name: 'dansk'
        },
        {
            id: 'no',
            name: 'norsk'
        }
    ],
    userPreference: {
        recordDefinitionName: 'User Preference',
        fieldIds: {
            loginName: 70061,
            name: 70062,
            value: 70063
        }
    },
    preferenceNames: {
        locale: 'Preferred-User-Locale'
    }
};
