// @bmc-ux/adapt-angular/esm2015/rx/rx-textarea/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxTextareaComponent } from './adapt-rx-textarea.component';
import * as i0 from "@angular/core";
export class AdaptRxTextareaModule {
}
AdaptRxTextareaModule.ɵfac = function AdaptRxTextareaModule_Factory(t) { return new (t || AdaptRxTextareaModule)(); };
AdaptRxTextareaModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxTextareaModule });
AdaptRxTextareaModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            AdaptRxFormControlModule,
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxTextareaModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptRxFormControlModule,
                    CommonModule
                ],
                declarations: [
                    AdaptRxTextareaComponent
                ],
                exports: [
                    AdaptRxTextareaComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxTextareaModule, { declarations: [AdaptRxTextareaComponent], imports: [AdaptRxFormControlModule,
        CommonModule], exports: [AdaptRxTextareaComponent] }); })();
export * from './adapt-rx-textarea.component';
export * from './adapt-rx-textarea-config.service';
