// @bmc-ux/obsolete/esm2015/lib/dropdown-editable/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptDropDownEditableComponent } from './dropdown-editable.component';
import { FormsModule } from '@angular/forms';
import { AdaptDropdownModule, AdaptDragonInputHandlerModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptDropdownEditableModule {
}
AdaptDropdownEditableModule.ɵfac = function AdaptDropdownEditableModule_Factory(t) { return new (t || AdaptDropdownEditableModule)(); };
AdaptDropdownEditableModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptDropdownEditableModule });
AdaptDropdownEditableModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptDropdownModule.forRoot(),
            AdaptDragonInputHandlerModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDropdownEditableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptDropdownModule.forRoot(),
                    AdaptDragonInputHandlerModule
                ],
                declarations: [AdaptDropDownEditableComponent],
                exports: [AdaptDropDownEditableComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptDropdownEditableModule, { declarations: [AdaptDropDownEditableComponent], imports: [CommonModule,
        FormsModule, i1.AdaptDropdownModule, AdaptDragonInputHandlerModule], exports: [AdaptDropDownEditableComponent] }); })();
export * from './dropdown-editable.component';
