// @bmc-ux/adapt-angular/esm2015/popover/index.js
import { AdaptPopoverDirective, AdaptPopoverWindow2Component, AdaptPopoverWindowComponent } from './popover.directive';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import * as i0 from "@angular/core";
export class AdaptPopoverModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptPopoverModule
        };
    }
}
AdaptPopoverModule.ɵfac = function AdaptPopoverModule_Factory(t) { return new (t || AdaptPopoverModule)(); };
AdaptPopoverModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptPopoverModule });
AdaptPopoverModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, OverlayModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptPopoverModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, OverlayModule],
                declarations: [AdaptPopoverDirective, AdaptPopoverWindowComponent, AdaptPopoverWindow2Component],
                exports: [AdaptPopoverDirective],
                entryComponents: [AdaptPopoverWindowComponent, AdaptPopoverWindow2Component]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptPopoverModule, { declarations: [AdaptPopoverDirective, AdaptPopoverWindowComponent, AdaptPopoverWindow2Component], imports: [CommonModule, OverlayModule], exports: [AdaptPopoverDirective] }); })();
export * from './popover.config';
export { AdaptPopoverDirective, AdaptPopoverWindowComponent, AdaptPopoverWindow2Component } from './popover.directive';
