// @helix/platform/esm2015/shared/api/expression-operator/expression-operator.enum.js
export var ExpressionOperator;
(function (ExpressionOperator) {
    ExpressionOperator["LeftGrouping"] = "(";
    ExpressionOperator["RightGrouping"] = ")";
    ExpressionOperator["DoubleQuote"] = "\"";
    ExpressionOperator["Add"] = "+";
    ExpressionOperator["Subtract"] = "-";
    ExpressionOperator["Multiply"] = "*";
    ExpressionOperator["Divide"] = "/";
    ExpressionOperator["Remainder"] = "%";
    ExpressionOperator["GreaterThan"] = ">";
    ExpressionOperator["LessThan"] = "<";
    ExpressionOperator["Equal"] = "=";
    ExpressionOperator["NotEqual"] = "!=";
    ExpressionOperator["GreaterThanOrEqual"] = ">=";
    ExpressionOperator["LessThanOrEqual"] = "<=";
    ExpressionOperator["In"] = "IN";
    ExpressionOperator["Like"] = "LIKE";
    ExpressionOperator["Contains"] = "CONTAINS";
    ExpressionOperator["And"] = "AND";
    ExpressionOperator["Or"] = "OR";
    ExpressionOperator["Not"] = "NOT";
    ExpressionOperator["Null"] = "NULL";
    ExpressionOperator["NullExpression"] = "$NULL$";
})(ExpressionOperator || (ExpressionOperator = {}));
