// @bmc-ux/adapt-angular/esm2015/common/scroll/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptScrollToElementDirective } from './scroll.directive';
import { AdaptAnimateScrollService } from './scroll.service';
import * as i0 from "@angular/core";
export class AdaptScrollModule {
}
AdaptScrollModule.ɵfac = function AdaptScrollModule_Factory(t) { return new (t || AdaptScrollModule)(); };
AdaptScrollModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptScrollModule });
AdaptScrollModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [
        AdaptAnimateScrollService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptScrollModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    AdaptScrollToElementDirective
                ],
                exports: [
                    AdaptScrollToElementDirective
                ],
                providers: [
                    AdaptAnimateScrollService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptScrollModule, { declarations: [AdaptScrollToElementDirective], imports: [CommonModule], exports: [AdaptScrollToElementDirective] }); })();
export * from './scroll.directive';
export * from './scroll.service';
