// @bmc-ux/adapt-angular/esm2015/common/scroll/scroll.directive.js
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { AdaptAnimateScrollService } from './scroll.service';
import * as i0 from "@angular/core";
import * as i1 from "./scroll.service";
export class AdaptScrollToElementDirective {
    constructor(scrollToService, elem) {
        this.scrollToService = scrollToService;
        this.elem = elem;
        this.hash = undefined;
        this.hash = elem.nativeElement.hash.slice(1);
    }
    onMouseClick() {
        this.scrollToService.scrollToElement(this.hash, this.scrollDuration, 0);
    }
}
AdaptScrollToElementDirective.ɵfac = function AdaptScrollToElementDirective_Factory(t) { return new (t || AdaptScrollToElementDirective)(i0.ɵɵdirectiveInject(i1.AdaptAnimateScrollService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptScrollToElementDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptScrollToElementDirective, selectors: [["", "adaptScrollToElement", ""]], hostBindings: function AdaptScrollToElementDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("mousedown", function AdaptScrollToElementDirective_mousedown_HostBindingHandler() { return ctx.onMouseClick(); });
    } }, inputs: { scrollDuration: "scrollDuration" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptScrollToElementDirective, [{
        type: Directive,
        args: [{
                selector: '[adaptScrollToElement]'
            }]
    }], function () { return [{ type: i1.AdaptAnimateScrollService }, { type: i0.ElementRef }]; }, { scrollDuration: [{
            type: Input
        }], onMouseClick: [{
            type: HostListener,
            args: ['mousedown']
        }] }); })();
