// @bmc-ux/adapt-angular/esm2015/advanced-filtering/index.js
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HammerModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { AdaptAccordionModule } from '../accordion/index';
import { AdaptBusyModule } from '../busy-loader/index';
import { AdaptButtonModule } from '../button/index';
import { AdaptCustomActionsModule } from '../custom-actions/index';
import { AdaptDropdownModule } from '../dropdown/index';
import { AdaptListKeyManagerModule } from '../common/list-key-manager/index';
import { AdaptMetatagModule } from '../metatag/index';
import { AdaptSubnavModule } from '../subnav/index';
import { AdaptTabsModule } from '../tabs/index';
import { AdaptTagModule } from '../tag/index';
import { AdaptTreeModule } from '../tree/index';
import { AdaptRxCounterModule, AdaptRxDatetimeModule, AdaptRxLabelModule, AdaptRxRadiobuttonModule, AdaptRxSelectModule, AdaptRxTextfieldModule } from '../rx/index';
import { AdaptAdvancedFilterComponent, AdaptAdvancedFilterNameValidator } from './advanced-filter/advanced-filter.component';
import { AdaptFacetedFilterComponent } from './faceted-filter/faceted-filter.component';
import { AdaptFilterControlsComponent } from './advanced-filter/filter-controls/filter-controls.component';
import { AdaptFilterTagsComponent } from './filter-tags/adapt-filter-tags';
import { AdaptSimpleFilterComponent } from './simple-filter/simple-filter.component';
import { AdvancedFilterDataTypesConfigsService } from './advanced-filter/advanced-filter-configs.service';
import { AdaptEmptyStateModule } from '../empty-state/index';
import { AdaptHighlightModule } from '../highlight/index';
import * as i0 from "@angular/core";
export class AdaptAdvancedFilteringModule {
}
AdaptAdvancedFilteringModule.ɵfac = function AdaptAdvancedFilteringModule_Factory(t) { return new (t || AdaptAdvancedFilteringModule)(); };
AdaptAdvancedFilteringModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAdvancedFilteringModule });
AdaptAdvancedFilteringModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdvancedFilterDataTypesConfigsService], imports: [[
            CommonModule,
            FormsModule,
            HammerModule,
            AdaptAccordionModule,
            AdaptBusyModule,
            AdaptButtonModule,
            AdaptCustomActionsModule,
            AdaptDropdownModule,
            AdaptListKeyManagerModule,
            AdaptMetatagModule,
            AdaptRxCounterModule,
            AdaptRxDatetimeModule,
            AdaptRxLabelModule,
            AdaptRxRadiobuttonModule,
            AdaptRxSelectModule,
            AdaptRxTextfieldModule,
            AdaptTabsModule,
            AdaptTagModule,
            AdaptTreeModule,
            AdaptSubnavModule,
            AdaptEmptyStateModule,
            AdaptHighlightModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAdvancedFilteringModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    HammerModule,
                    AdaptAccordionModule,
                    AdaptBusyModule,
                    AdaptButtonModule,
                    AdaptCustomActionsModule,
                    AdaptDropdownModule,
                    AdaptListKeyManagerModule,
                    AdaptMetatagModule,
                    AdaptRxCounterModule,
                    AdaptRxDatetimeModule,
                    AdaptRxLabelModule,
                    AdaptRxRadiobuttonModule,
                    AdaptRxSelectModule,
                    AdaptRxTextfieldModule,
                    AdaptTabsModule,
                    AdaptTagModule,
                    AdaptTreeModule,
                    AdaptSubnavModule,
                    AdaptEmptyStateModule,
                    AdaptHighlightModule
                ],
                declarations: [
                    AdaptAdvancedFilterComponent,
                    AdaptAdvancedFilterNameValidator,
                    AdaptFacetedFilterComponent,
                    AdaptFilterControlsComponent,
                    AdaptSimpleFilterComponent,
                    AdaptFilterTagsComponent
                ],
                exports: [
                    AdaptAdvancedFilterComponent,
                    AdaptFacetedFilterComponent,
                    AdaptFilterControlsComponent,
                    AdaptSimpleFilterComponent
                ],
                providers: [AdvancedFilterDataTypesConfigsService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAdvancedFilteringModule, { declarations: [AdaptAdvancedFilterComponent,
        AdaptAdvancedFilterNameValidator,
        AdaptFacetedFilterComponent,
        AdaptFilterControlsComponent,
        AdaptSimpleFilterComponent,
        AdaptFilterTagsComponent], imports: [CommonModule,
        FormsModule,
        HammerModule,
        AdaptAccordionModule,
        AdaptBusyModule,
        AdaptButtonModule,
        AdaptCustomActionsModule,
        AdaptDropdownModule,
        AdaptListKeyManagerModule,
        AdaptMetatagModule,
        AdaptRxCounterModule,
        AdaptRxDatetimeModule,
        AdaptRxLabelModule,
        AdaptRxRadiobuttonModule,
        AdaptRxSelectModule,
        AdaptRxTextfieldModule,
        AdaptTabsModule,
        AdaptTagModule,
        AdaptTreeModule,
        AdaptSubnavModule,
        AdaptEmptyStateModule,
        AdaptHighlightModule], exports: [AdaptAdvancedFilterComponent,
        AdaptFacetedFilterComponent,
        AdaptFilterControlsComponent,
        AdaptSimpleFilterComponent] }); })();
export * from './simple-filter/index';
export * from './faceted-filter/index';
export * from './advanced-filter/index';
export * from './advanced-filtering-utils';
