// @bmc-ux/adapt-angular/esm2015/rx/rx-checkbox/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxCheckboxComponent } from './adapt-rx-checkbox.component';
import * as i0 from "@angular/core";
export class AdaptRxCheckboxModule {
}
AdaptRxCheckboxModule.ɵfac = function AdaptRxCheckboxModule_Factory(t) { return new (t || AdaptRxCheckboxModule)(); };
AdaptRxCheckboxModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxCheckboxModule });
AdaptRxCheckboxModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxCheckboxModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule
                ],
                declarations: [
                    AdaptRxCheckboxComponent
                ],
                exports: [
                    AdaptRxCheckboxComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxCheckboxModule, { declarations: [AdaptRxCheckboxComponent], imports: [CommonModule,
        AdaptRxFormControlModule], exports: [AdaptRxCheckboxComponent] }); })();
export * from './adapt-rx-checkbox.component';
