// @bmc-ux/adapt-angular/esm2015/rx/rx-search/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxSearchComponent } from './adapt-rx-search.component';
import * as i0 from "@angular/core";
export class AdaptRxSearchModule {
}
AdaptRxSearchModule.ɵfac = function AdaptRxSearchModule_Factory(t) { return new (t || AdaptRxSearchModule)(); };
AdaptRxSearchModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxSearchModule });
AdaptRxSearchModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            AdaptRxFormControlModule,
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxSearchModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptRxFormControlModule,
                    CommonModule
                ],
                declarations: [
                    AdaptRxSearchComponent
                ],
                exports: [
                    AdaptRxSearchComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxSearchModule, { declarations: [AdaptRxSearchComponent], imports: [AdaptRxFormControlModule,
        CommonModule], exports: [AdaptRxSearchComponent] }); })();
export * from './adapt-rx-search.component';
export * from './adapt-rx-search-config.service';
