// @bmc-ux/adapt-angular/esm2015/navigation/index.js
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AdaptAlertAnimatedModule } from '../alert-animated/index';
import { AdaptModalModule } from '../modal/index';
import { AdaptCloseModule } from '../close/index';
import { AdaptDockedPanelModule } from '../docked-panel/index';
import { AdaptDropdownModule } from '../dropdown/index';
import { AdaptMenuModule } from '../menu/index';
import { AdaptPopoverModule } from '../popover/index';
import { AdaptProfileModule } from '../profile/index';
import { AdaptRxSearchModule } from '../rx/rx-search/index';
import { AdaptRxTextfieldModule } from '../rx/rx-textfield/index';
import { AdaptTooltipModule } from '../tooltip/index';
import { AdaptNavActionsComponent, AdaptNavActionsItemComponent, AdaptNavActionsSubitemComponent } from './nav.actions.component';
import { AdaptNavTextSearchPipe, AdaptNavHamburgerPipe } from './navigation.pipe';
import { AdaptNavigationComponent } from './navigation.component';
import { AdaptNavigationService } from './navigation.service';
import { AdaptNavBadgeCounterComponent } from './badge-counter/nav-badge-counter.component';
import { AdaptNavActionsSwictherComponent } from './nav-actions-switcher/nav-actions-swicther.component';
import * as i0 from "@angular/core";
export class AdaptNavigationModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptNavigationModule
        };
    }
}
AdaptNavigationModule.ɵfac = function AdaptNavigationModule_Factory(t) { return new (t || AdaptNavigationModule)(); };
AdaptNavigationModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptNavigationModule });
AdaptNavigationModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdaptNavigationService], imports: [[
            CommonModule,
            FormsModule,
            RouterModule,
            AdaptAlertAnimatedModule,
            AdaptModalModule,
            AdaptCloseModule,
            AdaptDockedPanelModule,
            AdaptDropdownModule,
            AdaptMenuModule,
            AdaptPopoverModule,
            AdaptProfileModule,
            AdaptRxSearchModule,
            AdaptRxTextfieldModule,
            AdaptTooltipModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptNavigationModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    RouterModule,
                    AdaptAlertAnimatedModule,
                    AdaptModalModule,
                    AdaptCloseModule,
                    AdaptDockedPanelModule,
                    AdaptDropdownModule,
                    AdaptMenuModule,
                    AdaptPopoverModule,
                    AdaptProfileModule,
                    AdaptRxSearchModule,
                    AdaptRxTextfieldModule,
                    AdaptTooltipModule
                ],
                exports: [AdaptNavigationComponent],
                declarations: [
                    AdaptNavActionsComponent,
                    AdaptNavActionsItemComponent,
                    AdaptNavActionsSubitemComponent,
                    AdaptNavBadgeCounterComponent,
                    AdaptNavHamburgerPipe,
                    AdaptNavTextSearchPipe,
                    AdaptNavigationComponent,
                    AdaptNavActionsSwictherComponent
                ],
                providers: [AdaptNavigationService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptNavigationModule, { declarations: [AdaptNavActionsComponent,
        AdaptNavActionsItemComponent,
        AdaptNavActionsSubitemComponent,
        AdaptNavBadgeCounterComponent,
        AdaptNavHamburgerPipe,
        AdaptNavTextSearchPipe,
        AdaptNavigationComponent,
        AdaptNavActionsSwictherComponent], imports: [CommonModule,
        FormsModule,
        RouterModule,
        AdaptAlertAnimatedModule,
        AdaptModalModule,
        AdaptCloseModule,
        AdaptDockedPanelModule,
        AdaptDropdownModule,
        AdaptMenuModule,
        AdaptPopoverModule,
        AdaptProfileModule,
        AdaptRxSearchModule,
        AdaptRxTextfieldModule,
        AdaptTooltipModule], exports: [AdaptNavigationComponent] }); })();
export { AdaptNavigationComponent } from './navigation.component';
export { AdaptNavActionsComponent, AdaptNavActionsItemComponent, AdaptNavActionsSubitemComponent } from './nav.actions.component';
export { AdaptNavigationService } from './navigation.service';
export * from './navigation.config';
