// @bmc-ux/adapt-angular/esm2015/navigation/navigation.config.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptSwitcherModel {
}
export class AdaptNavigationMenuItem {
}
export class AdaptNavigationSubMenu {
}
export class AdaptNavigationSubMenuItem {
}
export class AdaptProfileConfig {
}
AdaptProfileConfig.ɵfac = function AdaptProfileConfig_Factory(t) { return new (t || AdaptProfileConfig)(); };
AdaptProfileConfig.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptProfileConfig, factory: AdaptProfileConfig.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptProfileConfig, [{
        type: Injectable
    }], null, null); })();
export class AdaptActionsConfig {
}
AdaptActionsConfig.ɵfac = function AdaptActionsConfig_Factory(t) { return new (t || AdaptActionsConfig)(); };
AdaptActionsConfig.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptActionsConfig, factory: AdaptActionsConfig.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptActionsConfig, [{
        type: Injectable
    }], null, null); })();
export class AdaptActionItem {
}
