// @bmc-ux/adapt-angular/esm2015/navigation/navigation.component.js
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, NgZone, Output, QueryList, Renderer2, TemplateRef, ViewChild, ViewChildren } from '@angular/core';
import { transition, trigger, useAnimation } from '@angular/animations';
import { asyncScheduler, noop, ReplaySubject, Subject } from 'rxjs';
import { takeUntil, throttleTime } from 'rxjs/operators';
import { cloneDeep, isArray } from 'lodash-es';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularCustomEventType, AdaptRadarAngularService, AdaptRadarAngularGenericDirective } from '../common/radar/adapt-radar';
import { AdaptDockedPanelService } from '../docked-panel/index';
import { AdaptDropdownDirective } from '../dropdown/index';
import { AdaptProfileComponent } from '../profile/profile.component';
import { AdaptTranslateService } from '../common/i18n/index';
import { MenuGlobalService } from '../menu/menu-global.service';
import { ANIMATION_NAME } from '../common/animation';
import { AdaptDeviceDetectionService } from '../common/device-detection.service';
import { ResizeService } from '../common/resize/resize.service';
import { SafeCdr } from '../common/safe-cdr';
import { checkURL } from '../common/check-url';
import { isDefined, maxDepth } from '../common/utilities';
import { AdaptNavActionsComponent } from './nav.actions.component';
import { AdaptNavigationService } from './navigation.service';
import * as i0 from "@angular/core";
import * as i1 from "../docked-panel/index";
import * as i2 from "../common/resize/resize.service";
import * as i3 from "../common/device-detection.service";
import * as i4 from "./navigation.service";
import * as i5 from "../menu/menu-global.service";
import * as i6 from "../common/radar/adapt-radar";
import * as i7 from "../common/i18n/index";
import * as i8 from "@angular/router";
import * as i9 from "@angular/common";
import * as i10 from "./nav.actions.component";
import * as i11 from "./nav-actions-switcher/nav-actions-swicther.component";
import * as i12 from "../alert-animated/alert-animated.component";
import * as i13 from "../menu/menu.component";
import * as i14 from "../profile/profile.component";
import * as i15 from "../close/close.component";
import * as i16 from "../rx/rx-textfield/adapt-rx-textfield.component";
import * as i17 from "@angular/forms";
import * as i18 from "./navigation.pipe";
const _c0 = ["menuEl"];
const _c1 = ["logoEl"];
const _c2 = ["hamburgerBtn"];
const _c3 = ["hamburgerLogo"];
const _c4 = ["alertsSection"];
const _c5 = ["actionsEl"];
const _c6 = ["profileEl"];
const _c7 = ["hamburgerContent"];
const _c8 = ["hamburgerMenuTemplate"];
const _c9 = ["loaderSection"];
const _c10 = ["floatSection"];
const _c11 = ["floatWrapper"];
const _c12 = ["navElShadow"];
const _c13 = ["navElContentShadow"];
const _c14 = ["profileElShadow"];
const _c15 = ["actionsElShadow"];
const _c16 = ["menuElShadow"];
const _c17 = ["hamburgerBtnElShadow"];
const _c18 = ["logoElShadow"];
const _c19 = function (a0) { return { "active": a0 }; };
function AdaptNavigationComponent_li_32_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "a", 45);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r25 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-menu__link text-primary ", item_r25.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(5, _c19, item_r25.active));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r25.label);
} }
const _c20 = function (a0) { return { "m-start-2": a0 }; };
function AdaptNavigationComponent_li_32_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "div", 45);
    i0.ɵɵtext(3, "\n                ");
    i0.ɵɵelementStart(4, "span", 46);
    i0.ɵɵtext(5);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n                ");
    i0.ɵɵelement(7, "span", 47);
    i0.ɵɵtext(8, "\n              ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r25 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵclassMapInterpolate1("adapt-menu__link btn btn-link text-primary ", item_r25.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(6, _c19, item_r25.active));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c20, item_r25.className));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r25.label);
} }
function AdaptNavigationComponent_li_32_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_li_32_ng_container_2_Template, 5, 7, "ng-container", 26);
    i0.ɵɵtext(3, "\n\n            ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_li_32_div_4_Template, 10, 10, "div", 26);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r25 = ctx.$implicit;
    i0.ɵɵclassMapInterpolate1("a-menu__item ", item_r25.listClassName, "");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r25.items && !item_r25.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r25.items && !item_r25.hide);
} }
function AdaptNavigationComponent_ng_container_40_a_2_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 51);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const act_r31 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(act_r31.count);
} }
function AdaptNavigationComponent_ng_container_40_a_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a");
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_container_40_a_2_span_2_Template, 2, 1, "span", 50);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const act_r31 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassProp("btn", act_r31.action)("btn-link", act_r31.action);
    i0.ɵɵattribute("class", "a-action " + act_r31.className);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", act_r31.count);
} }
function AdaptNavigationComponent_ng_container_40_button_4_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 54);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const act_r31 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(act_r31.title);
} }
function AdaptNavigationComponent_ng_container_40_button_4_span_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 51);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const act_r31 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(act_r31.count);
} }
function AdaptNavigationComponent_ng_container_40_button_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "button", 52);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_container_40_button_4_span_2_Template, 2, 1, "span", 53);
    i0.ɵɵtext(3, "\n              ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_container_40_button_4_span_4_Template, 2, 1, "span", 50);
    i0.ɵɵtext(5, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const act_r31 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-action btn btn-link ", act_r31.className, " dropdown-toggle");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", act_r31.title);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", act_r31.count);
} }
function AdaptNavigationComponent_ng_container_40_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_container_40_a_2_Template, 4, 6, "a", 48);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_container_40_button_4_Template, 6, 5, "button", 49);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const act_r31 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !act_r31.subMenu && !act_r31.popover && !act_r31.switcher);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !(!act_r31.subMenu && !act_r31.popover) || act_r31.switcher);
} }
const _c21 = function (a0) { return { "a-profile__avatar--imaged": a0 }; };
const _c22 = function (a0) { return { "background-image": a0 }; };
function AdaptNavigationComponent_div_45_span_7_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 61);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r44 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(2, _c21, ctx_r44.profile.image))("ngStyle", i0.ɵɵpureFunction1(4, _c22, ctx_r44.profile.image ? "url(" + ctx_r44.profile.image + ")" : "none"));
} }
function AdaptNavigationComponent_div_45_span_9_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 62);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelementStart(2, "span", 63);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n              ");
    i0.ɵɵelementStart(5, "span", 64);
    i0.ɵɵtext(6);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r45 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r45.profile.desc);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r45.profile.userName);
} }
const _c23 = function (a0, a1) { return { "a-profile--long": a0, "a-profile--small": a1 }; };
const _c24 = function (a0) { return { "a-profile__menu--small": a0 }; };
function AdaptNavigationComponent_div_45_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 55, 56);
    i0.ɵɵtext(2, "\n          ");
    i0.ɵɵelementStart(3, "div", 57);
    i0.ɵɵtext(4, "\n            ");
    i0.ɵɵelementStart(5, "div", 58);
    i0.ɵɵtext(6, "\n            ");
    i0.ɵɵtemplate(7, AdaptNavigationComponent_div_45_span_7_Template, 2, 6, "span", 59);
    i0.ɵɵtext(8, "\n              ");
    i0.ɵɵtemplate(9, AdaptNavigationComponent_div_45_span_9_Template, 8, 2, "span", 60);
    i0.ɵɵtext(10, "\n            ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n          ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(12, "\n        ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = i0.ɵɵnextContext();
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(4, _c23, ctx_r7.profile.typeLong, !ctx_r7.profile.typeLong));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(7, _c24, !ctx_r7.profile.typeLong));
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r7.profile.hideAvatar);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r7.profile.typeLong);
} }
function AdaptNavigationComponent_ng_container_57_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_button_60_Template(rf, ctx) { if (rf & 1) {
    const _r48 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 65, 66);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_button_60_Template_button_click_0_listener() { i0.ɵɵrestoreView(_r48); const ctx_r47 = i0.ɵɵnextContext(); return ctx_r47.openHamburger(); });
    i0.ɵɵtext(2, "\n        ");
    i0.ɵɵelementStart(3, "span", 67);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = i0.ɵɵnextContext();
    i0.ɵɵattribute("aria-expanded", ctx_r9.hamburgerMenuOpen);
    i0.ɵɵadvance(4);
    i0.ɵɵtextInterpolate(ctx_r9.metadata.menuText);
} }
const _c25 = function () { return { simpleLogo: true }; };
function AdaptNavigationComponent_div_64_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 68, 69);
    i0.ɵɵtext(2, "\n        ");
    i0.ɵɵelementContainer(3, 70);
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r17 = i0.ɵɵreference(86);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", _r17)("ngTemplateOutletContext", i0.ɵɵpureFunction0(2, _c25));
} }
function AdaptNavigationComponent_ul_66_li_3_div_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵelement(2, "adapt-menu", 72);
    i0.ɵɵtext(3, "\n          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r52 = i0.ɵɵnextContext().$implicit;
    const ctx_r54 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("cascadeItem", item_r52)("adaptRadarDisableEventSending", true)("headerTheme", ctx_r54.theme)("bodyTheme", ctx_r54.theme)("connectMenus", true)("filteringConfig", item_r52.filteringConfig)("analyticsMetadata", ctx_r54.getAnalyticsMetadata());
} }
function AdaptNavigationComponent_ul_66_li_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n          ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ul_66_li_3_div_2_Template, 4, 7, "div", 26);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r52 = ctx.$implicit;
    i0.ɵɵclassMapInterpolate1("a-menu__item ", item_r52.listClassName, "");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r52.hide);
} }
function AdaptNavigationComponent_ul_66_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "ul", 11, 71);
    i0.ɵɵtext(2, "\n        ");
    i0.ɵɵtemplate(3, AdaptNavigationComponent_ul_66_li_3_Template, 4, 4, "li", 13);
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = i0.ɵɵnextContext();
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngForOf", ctx_r11.menu)("ngForTrackBy", ctx_r11.trackByFnMenu);
} }
function AdaptNavigationComponent_ng_container_71_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0, null, 74);
} }
const _c26 = function () { return { type: "normal" }; };
function AdaptNavigationComponent_ng_container_71_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n          ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_container_71_ng_container_2_ng_container_2_Template, 2, 0, "ng-container", 73);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r56 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r56.customProfile)("ngTemplateOutletContext", i0.ɵɵpureFunction0(2, _c26));
} }
function AdaptNavigationComponent_ng_container_71_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n          ");
    i0.ɵɵelement(2, "adapt-profile", 75, 74);
    i0.ɵɵtext(4, "\n        ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r57 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("profile", ctx_r57.profile)("adaptRadarDisableEventSending", true);
} }
function AdaptNavigationComponent_ng_container_71_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n        ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_container_71_ng_container_2_Template, 4, 3, "ng-container", 26);
    i0.ɵɵtext(3, "\n        ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_container_71_ng_container_4_Template, 5, 2, "ng-container", 26);
    i0.ɵɵtext(5, "\n\n      ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r13 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r13.customProfilePresent);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r13.customProfilePresent);
} }
function AdaptNavigationComponent_ng_template_82_ng_container_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r74 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 86);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_a_2_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r74); const item_r65 = i0.ɵɵnextContext(2).$implicit; const ctx_r72 = i0.ɵɵnextContext(2); return ctx_r72.handleMenuLinkSelection(item_r65); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r65 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r69 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(5, _c19, item_r65.active));
    i0.ɵɵattribute("href", item_r65.link, i0.ɵɵsanitizeUrl)("data-testid", ctx_r69.generateMenuItemTestID(item_r65) + "_hamburger")("target", ctx_r69.getTarget(item_r65));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", item_r65.label, "\n                ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r78 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 87);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_a_4_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r78); const item_r65 = i0.ɵɵnextContext(2).$implicit; const ctx_r76 = i0.ɵɵnextContext(2); return ctx_r76.handleMenuRouteSelection(item_r65); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r65 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r70 = i0.ɵɵnextContext(2);
    i0.ɵɵpropertyInterpolate("routerLink", item_r65.routerLink);
    i0.ɵɵattribute("data-testid", ctx_r70.generateMenuItemTestID(item_r65) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", item_r65.label, "\n                ");
} }
const _c27 = function (a0, a1) { return { "active": a0, "selected": a1 }; };
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r82 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 88);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r82); const item_r65 = i0.ɵɵnextContext(2).$implicit; const ctx_r80 = i0.ɵɵnextContext(2); return ctx_r80.handleMenuActionSelection($event, item_r65); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r65 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r71 = i0.ɵɵnextContext(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(3, _c27, item_r65.active, item_r65.selected));
    i0.ɵɵattribute("data-testid", ctx_r71.generateMenuItemTestID(item_r65) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", item_r65.label, "\n                ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_a_2_Template, 2, 7, "a", 83);
    i0.ɵɵtext(3, "\n\n                ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_a_4_Template, 2, 3, "a", 84);
    i0.ɵɵtext(5, "\n\n                ");
    i0.ɵɵtemplate(6, AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_button_6_Template, 2, 6, "button", 85);
    i0.ɵɵtext(7, "\n              ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r65 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r65.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r65.routerLink);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r65.action);
} }
const _c28 = function (a0) { return { "adapt-search-clear-visible": a0 }; };
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_22_Template(rf, ctx) { if (rf & 1) {
    const _r89 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "li", 95);
    i0.ɵɵtext(1, "\n                    ");
    i0.ɵɵelement(2, "div", 96);
    i0.ɵɵtext(3, "\n                    ");
    i0.ɵɵelementStart(4, "div", 97);
    i0.ɵɵtext(5, "\n                      ");
    i0.ɵɵelementStart(6, "adapt-rx-textfield", 98, 99);
    i0.ɵɵlistener("ngModelChange", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_22_Template_adapt_rx_textfield_ngModelChange_6_listener($event) { i0.ɵɵrestoreView(_r89); const ctx_r88 = i0.ɵɵnextContext(4); return ctx_r88.filtering.name = $event; });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n                      ");
    i0.ɵɵelementStart(9, "button", 100);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_22_Template_button_click_9_listener() { i0.ɵɵrestoreView(_r89); const _r87 = i0.ɵɵreference(7); const ctx_r90 = i0.ɵɵnextContext(4); return ctx_r90.clearFilter(_r87.inputRef.nativeElement); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n                    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n                    ");
    i0.ɵɵelement(12, "div", 96);
    i0.ɵɵtext(13, "\n                  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r65 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r85 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(6);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r85.filtering.name)("placeholder", item_r65.filteringConfig == null ? null : item_r65.filteringConfig.placeholder);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(4, _c28, !!ctx_r85.filtering.name.length));
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r103 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 105);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_a_2_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r103); const sub_r92 = i0.ɵɵnextContext(3).$implicit; const ctx_r101 = i0.ɵɵnextContext(4); return ctx_r101.handleMenuLinkSelection(sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r98 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link ", sub_r92.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c19, sub_r92.active));
    i0.ɵɵattribute("href", sub_r92.link, i0.ɵɵsanitizeUrl)("data-testid", ctx_r98.generateMenuItemTestID(sub_r92) + "_hamburger")("target", ctx_r98.getTarget(sub_r92));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                        ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r107 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 106);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_a_4_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r107); const sub_r92 = i0.ɵɵnextContext(3).$implicit; const ctx_r105 = i0.ɵɵnextContext(4); return ctx_r105.handleMenuRouteSelection(sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r99 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link ", sub_r92.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", sub_r92.routerLink);
    i0.ɵɵattribute("data-testid", ctx_r99.generateMenuItemTestID(sub_r92) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                        ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_button_6_Template(rf, ctx) { if (rf & 1) {
    const _r111 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 107);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_button_6_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r111); const sub_r92 = i0.ɵɵnextContext(3).$implicit; const ctx_r109 = i0.ɵɵnextContext(4); return ctx_r109.handleMenuActionSelection($event, sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r100 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link btn-link ", sub_r92.className, "");
    i0.ɵɵattribute("data-testid", ctx_r100.generateMenuItemTestID(sub_r92) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                        ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                        ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_a_2_Template, 2, 10, "a", 102);
    i0.ɵɵtext(3, "\n\n                        ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_a_4_Template, 2, 6, "a", 103);
    i0.ɵɵtext(5, "\n\n                        ");
    i0.ɵɵtemplate(6, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_button_6_Template, 2, 5, "button", 104);
    i0.ɵɵtext(7, "\n\n                      ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.routerLink);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.action);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r97 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link disabled ", sub_r92.className, "");
    i0.ɵɵattribute("data-testid", ctx_r97.generateMenuItemTestID(sub_r92) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                      ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                      ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_ng_container_2_Template, 8, 3, "ng-container", 26);
    i0.ɵɵtext(3, "\n\n                      ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_div_4_Template, 2, 5, "div", 101);
    i0.ɵɵtext(5, "\n\n                    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !sub_r92.disabled);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.disabled);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_span_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 45);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r116 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-text ", sub_r92.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(6, _c19, sub_r92.active));
    i0.ɵɵattribute("data-testid", ctx_r116.generateMenuItemTestID(sub_r92) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(sub_r92.label);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r124 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 112);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_button_4_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r124); const sub_r92 = i0.ɵɵnextContext(2).$implicit; const ctx_r122 = i0.ɵɵnextContext(4); return ctx_r122.handleMenuActionSelection($event, sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r117 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link menu-parent btn-link ", sub_r92.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(6, _c19, sub_r92.active));
    i0.ɵɵattribute("data-testid", ctx_r117.generateMenuItemTestID(sub_r92) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                      ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_a_6_Template(rf, ctx) { if (rf & 1) {
    const _r128 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 105);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_a_6_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r128); const sub_r92 = i0.ɵɵnextContext(2).$implicit; const ctx_r126 = i0.ɵɵnextContext(4); return ctx_r126.handleMenuLinkSelection(sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r118 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link menu-parent ", sub_r92.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c19, sub_r92.active));
    i0.ɵɵattribute("href", sub_r92.link, i0.ɵɵsanitizeUrl)("data-testid", ctx_r118.generateMenuItemTestID(sub_r92) + "_hamburger")("target", ctx_r118.getTarget(sub_r92));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                      ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_a_8_Template(rf, ctx) { if (rf & 1) {
    const _r132 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 113);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_a_8_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r132); const sub_r92 = i0.ɵɵnextContext(2).$implicit; const ctx_r130 = i0.ɵɵnextContext(4); return ctx_r130.handleMenuRouteSelection(sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r119 = i0.ɵɵnextContext(4);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link menu-parent ", sub_r92.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", sub_r92.routerLink);
    i0.ɵɵattribute("data-testid", ctx_r119.generateMenuItemTestID(sub_r92) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", sub_r92.label, "\n                      ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r144 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 105);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_a_2_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r144); const s_r134 = i0.ɵɵnextContext(3).$implicit; const ctx_r142 = i0.ɵɵnextContext(6); return ctx_r142.handleMenuLinkSelection(s_r134); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const s_r134 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r139 = i0.ɵɵnextContext(6);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link a-hamburger__sublink ", s_r134.className, "");
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(8, _c19, s_r134.active));
    i0.ɵɵattribute("href", s_r134.link, i0.ɵɵsanitizeUrl)("data-testid", ctx_r139.generateMenuItemTestID(s_r134) + "_hamburger")("target", ctx_r139.getTarget(s_r134));
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", s_r134.label, "\n                              ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_a_4_Template(rf, ctx) { if (rf & 1) {
    const _r148 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "a", 106);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_a_4_Template_a_click_0_listener() { i0.ɵɵrestoreView(_r148); const sub_r92 = i0.ɵɵnextContext(5).$implicit; const ctx_r146 = i0.ɵɵnextContext(4); return ctx_r146.handleMenuRouteSelection(sub_r92); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const s_r134 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r140 = i0.ɵɵnextContext(6);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link a-hamburger__sublink ", s_r134.className, "");
    i0.ɵɵpropertyInterpolate("routerLink", s_r134.routerLink);
    i0.ɵɵattribute("data-testid", ctx_r140.generateMenuItemTestID(s_r134) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", s_r134.label, "\n                              ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_button_7_Template(rf, ctx) { if (rf & 1) {
    const _r152 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 107);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_button_7_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r152); const s_r134 = i0.ɵɵnextContext(3).$implicit; const ctx_r150 = i0.ɵɵnextContext(6); return ctx_r150.handleMenuActionSelection($event, s_r134, true, false); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const s_r134 = i0.ɵɵnextContext(3).$implicit;
    const ctx_r141 = i0.ɵɵnextContext(6);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link a-hamburger__sublink btn-link ", s_r134.className, "");
    i0.ɵɵattribute("data-testid", ctx_r141.generateMenuItemTestID(s_r134) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", s_r134.label, "\n                              ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                              ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_a_2_Template, 2, 10, "a", 102);
    i0.ɵɵtext(3, "\n\n                              ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_a_4_Template, 2, 6, "a", 103);
    i0.ɵɵtext(5, "\n\n                              ");
    i0.ɵɵtext(6, "\n                              ");
    i0.ɵɵtemplate(7, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_button_7_Template, 2, 5, "button", 104);
    i0.ɵɵtext(8, "\n                            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const s_r134 = i0.ɵɵnextContext(2).$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", s_r134.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", s_r134.routerLink);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", s_r134.action);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_div_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div");
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const s_r134 = i0.ɵɵnextContext(2).$implicit;
    const ctx_r138 = i0.ɵɵnextContext(6);
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-link a-hamburger__sublink disabled ", s_r134.className, "");
    i0.ɵɵattribute("data-testid", ctx_r138.generateMenuItemTestID(s_r134) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("", s_r134.label, "\n                            ");
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n\n                            ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_ng_container_2_Template, 9, 3, "ng-container", 26);
    i0.ɵɵtext(3, "\n\n                            ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_div_4_Template, 2, 5, "div", 101);
    i0.ɵɵtext(5, "\n                          ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const s_r134 = i0.ɵɵnextContext().$implicit;
    i0.ɵɵclassMapInterpolate1("a-hamburger__submenu-item ", s_r134.listClassName, "");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !s_r134.disabled);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", s_r134.disabled && !s_r134.hide);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n                          ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_li_2_Template, 6, 5, "li", 101);
    i0.ɵɵtext(3, "\n                        ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const s_r134 = ctx.$implicit;
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !s_r134.hide);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n\n                      ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_span_2_Template, 2, 8, "span", 108);
    i0.ɵɵtext(3, "\n\n                      ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_button_4_Template, 2, 8, "button", 109);
    i0.ɵɵtext(5, "\n\n                      ");
    i0.ɵɵtemplate(6, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_a_6_Template, 2, 10, "a", 102);
    i0.ɵɵtext(7, "\n\n                      ");
    i0.ɵɵtemplate(8, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_a_8_Template, 2, 6, "a", 110);
    i0.ɵɵtext(9, "\n\n                      ");
    i0.ɵɵelementStart(10, "ul", 111);
    i0.ɵɵtext(11, "\n\n                        ");
    i0.ɵɵtemplate(12, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_ng_container_12_Template, 4, 1, "ng-container", 16);
    i0.ɵɵpipe(13, "navFilterSubItems");
    i0.ɵɵtext(14, "\n\n                      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(15, "\n                    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const sub_r92 = i0.ɵɵnextContext().$implicit;
    const ctx_r95 = i0.ɵɵnextContext(4);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !sub_r92.routerLink && !sub_r92.link && !sub_r92.action);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.action);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.link);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.routerLink);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngForOf", i0.ɵɵpipeBind2(13, 6, sub_r92.items, ctx_r95.filtering.name))("ngForTrackBy", ctx_r95.trackByFnMenu);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n\n                    ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_2_Template, 6, 2, "ng-container", 26);
    i0.ɵɵtext(3, "\n\n                    ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_ng_container_4_Template, 16, 9, "ng-container", 26);
    i0.ɵɵtext(5, "\n\n                  ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const sub_r92 = ctx.$implicit;
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-item ", sub_r92.listClassName, "");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !sub_r92.items && !sub_r92.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", sub_r92.items && !sub_r92.hide);
} }
function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    const _r160 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n                ");
    i0.ɵɵelementStart(2, "a", 89);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_Template_a_click_2_listener() { i0.ɵɵrestoreView(_r160); const i_r66 = i0.ɵɵnextContext().index; const ctx_r158 = i0.ɵɵnextContext(2); return ctx_r158.showSubmenu("submenu_" + i_r66); })("keypress", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_Template_a_keypress_2_listener() { i0.ɵɵrestoreView(_r160); const i_r66 = i0.ɵɵnextContext().index; const ctx_r161 = i0.ɵɵnextContext(2); return ctx_r161.showSubmenu("submenu_" + i_r66); });
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n\n                ");
    i0.ɵɵelementStart(5, "ul", 90);
    i0.ɵɵtext(6, "\n                  ");
    i0.ɵɵelementStart(7, "li", 91);
    i0.ɵɵtext(8, "\n                    ");
    i0.ɵɵelementStart(9, "button", 92);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_Template_button_click_9_listener() { i0.ɵɵrestoreView(_r160); const i_r66 = i0.ɵɵnextContext().index; const ctx_r163 = i0.ɵɵnextContext(2); return ctx_r163.showSubmenu("submenu_" + i_r66); });
    i0.ɵɵtext(10);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n                    ");
    i0.ɵɵelementStart(12, "adapt-close", 80);
    i0.ɵɵlistener("close", function AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_Template_adapt_close_close_12_listener() { i0.ɵɵrestoreView(_r160); const ctx_r165 = i0.ɵɵnextContext(3); return ctx_r165.closeHamburger(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n                    ");
    i0.ɵɵelementStart(14, "div", 93);
    i0.ɵɵtext(15, "\n                      ");
    i0.ɵɵelementStart(16, "span");
    i0.ɵɵtext(17);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n                    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(19, "\n                  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(20, "\n\n                  ");
    i0.ɵɵtext(21, "\n                  ");
    i0.ɵɵtemplate(22, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_22_Template, 14, 6, "li", 94);
    i0.ɵɵtext(23, "\n                  ");
    i0.ɵɵtemplate(24, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_li_24_Template, 6, 5, "li", 13);
    i0.ɵɵpipe(25, "navFilter");
    i0.ɵɵtext(26, "\n                ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(27, "\n              ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r166 = i0.ɵɵnextContext();
    const item_r65 = ctx_r166.$implicit;
    const i_r66 = ctx_r166.index;
    const ctx_r68 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(14, _c19, item_r65.active));
    i0.ɵɵattribute("data-testid", ctx_r68.generateMenuItemTestID(item_r65) + "_hamburger");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(item_r65.label);
    i0.ɵɵadvance(2);
    i0.ɵɵpropertyInterpolate("id", "submenu_" + i_r66);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate1("\n                      ", ctx_r68.metadata.backBtn, "\n                    ");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("inverse", ctx_r68.themeClass === "");
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(item_r65.label);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngIf", item_r65.filteringConfig == null ? null : item_r65.filteringConfig.enabled);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngForOf", i0.ɵɵpipeBind2(25, 11, item_r65.items, ctx_r68.filtering.name))("ngForTrackBy", ctx_r68.trackByFnMenu);
} }
function AdaptNavigationComponent_ng_template_82_li_20_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "li");
    i0.ɵɵtext(1, "\n\n              ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_li_20_ng_container_2_Template, 8, 3, "ng-container", 26);
    i0.ɵɵtext(3, "\n\n              ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_li_20_ng_container_4_Template, 28, 16, "ng-container", 26);
    i0.ɵɵtext(5, "\n\n            ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r65 = ctx.$implicit;
    i0.ɵɵclassMapInterpolate1("a-hamburger__menu-item ", item_r65.listClassName, "");
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !item_r65.items && !item_r65.hide);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", item_r65.items && !item_r65.hide);
} }
function AdaptNavigationComponent_ng_template_82_ng_container_23_ng_container_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
const _c29 = function () { return { type: "hamburger" }; };
function AdaptNavigationComponent_ng_template_82_ng_container_23_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_ng_container_23_ng_container_2_ng_container_2_Template, 1, 0, "ng-container", 73);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r167 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngTemplateOutlet", ctx_r167.customProfile)("ngTemplateOutletContext", i0.ɵɵpureFunction0(2, _c29));
} }
function AdaptNavigationComponent_ng_template_82_ng_container_23_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n              ");
    i0.ɵɵelement(2, "adapt-profile", 114, 74);
    i0.ɵɵtext(4, "\n            ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r168 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("profile", ctx_r168.profile)("type", "hamburger")("closeHamburger", ctx_r168.closeHamburger);
} }
function AdaptNavigationComponent_ng_template_82_ng_container_23_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n            ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_82_ng_container_23_ng_container_2_Template, 4, 3, "ng-container", 26);
    i0.ɵɵtext(3, "\n            ");
    i0.ɵɵtemplate(4, AdaptNavigationComponent_ng_template_82_ng_container_23_ng_container_4_Template, 5, 3, "ng-container", 26);
    i0.ɵɵtext(5, "\n\n          ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r64 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r64.customProfilePresent);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r64.customProfilePresent);
} }
const _c30 = function () { return { ham: true }; };
function AdaptNavigationComponent_ng_template_82_Template(rf, ctx) { if (rf & 1) {
    const _r172 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n    ");
    i0.ɵɵelementStart(1, "div", 76);
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵelementStart(3, "div");
    i0.ɵɵtext(4, "\n\n        ");
    i0.ɵɵelementStart(5, "div", 77, 78);
    i0.ɵɵtext(7, "\n          ");
    i0.ɵɵelementStart(8, "div", 79);
    i0.ɵɵtext(9, "\n            ");
    i0.ɵɵelementStart(10, "adapt-close", 80);
    i0.ɵɵlistener("close", function AdaptNavigationComponent_ng_template_82_Template_adapt_close_close_10_listener() { i0.ɵɵrestoreView(_r172); const ctx_r171 = i0.ɵɵnextContext(); return ctx_r171.closeHamburger(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n            ");
    i0.ɵɵelementStart(12, "div", 81);
    i0.ɵɵtext(13, "\n              ");
    i0.ɵɵtemplate(14, AdaptNavigationComponent_ng_template_82_ng_container_14_Template, 1, 0, "ng-container", 73);
    i0.ɵɵtext(15, "\n            ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(16, "\n          ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(17, "\n\n          ");
    i0.ɵɵelementStart(18, "ul", 82);
    i0.ɵɵtext(19, "\n            ");
    i0.ɵɵtemplate(20, AdaptNavigationComponent_ng_template_82_li_20_Template, 6, 5, "li", 13);
    i0.ɵɵtext(21, "\n          ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(22, "\n\n          ");
    i0.ɵɵtemplate(23, AdaptNavigationComponent_ng_template_82_ng_container_23_Template, 6, 2, "ng-container", 26);
    i0.ɵɵtext(24, "\n\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(25, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(26, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(27, "\n  ");
} if (rf & 2) {
    const ctx_r16 = i0.ɵɵnextContext();
    const _r17 = i0.ɵɵreference(86);
    i0.ɵɵadvance(3);
    i0.ɵɵclassMap("a-hamburger__container  " + ctx_r16.themeClass + " " + ctx_r16.iphoneCSSclass);
    i0.ɵɵadvance(7);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("inverse", ctx_r16.themeClass === "");
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngTemplateOutlet", _r17)("ngTemplateOutletContext", i0.ɵɵpureFunction0(10, _c30));
    i0.ɵɵadvance(6);
    i0.ɵɵproperty("ngForOf", ctx_r16.menu)("ngForTrackBy", ctx_r16.trackByFnMenu);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", ctx_r16.profile && ctx_r16.hamburgerMode);
} }
function AdaptNavigationComponent_ng_template_85_a_2_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_ng_template_85_a_2_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 121, 122);
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵtemplate(3, AdaptNavigationComponent_ng_template_85_a_2_ng_container_3_Template, 1, 0, "ng-container", 123);
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const justLogo_r173 = i0.ɵɵnextContext().simpleLogo;
    const _r181 = i0.ɵɵreference(18);
    const _r179 = i0.ɵɵreference(15);
    const ctx_r175 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("href", ctx_r175.metadata.link, i0.ɵɵsanitizeUrl);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", justLogo_r173 ? _r181 : _r179);
} }
function AdaptNavigationComponent_ng_template_85_a_5_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_ng_template_85_a_5_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "a", 124, 122);
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵtemplate(3, AdaptNavigationComponent_ng_template_85_a_5_ng_container_3_Template, 1, 0, "ng-container", 123);
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const justLogo_r173 = i0.ɵɵnextContext().simpleLogo;
    const _r181 = i0.ɵɵreference(18);
    const _r179 = i0.ɵɵreference(15);
    const ctx_r176 = i0.ɵɵnextContext();
    i0.ɵɵpropertyInterpolate("routerLink", ctx_r176.metadata.route);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", justLogo_r173 ? _r181 : _r179);
} }
function AdaptNavigationComponent_ng_template_85_button_8_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_ng_template_85_button_8_Template(rf, ctx) { if (rf & 1) {
    const _r192 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "button", 125, 122);
    i0.ɵɵlistener("click", function AdaptNavigationComponent_ng_template_85_button_8_Template_button_click_0_listener($event) { i0.ɵɵrestoreView(_r192); const ctx_r191 = i0.ɵɵnextContext(2); return ctx_r191.metadata.action($event); });
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵtemplate(3, AdaptNavigationComponent_ng_template_85_button_8_ng_container_3_Template, 1, 0, "ng-container", 123);
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const justLogo_r173 = i0.ɵɵnextContext().simpleLogo;
    const _r181 = i0.ɵɵreference(18);
    const _r179 = i0.ɵɵreference(15);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", justLogo_r173 ? _r181 : _r179);
} }
function AdaptNavigationComponent_ng_template_85_span_11_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_ng_template_85_span_11_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span", 126, 122);
    i0.ɵɵtext(2, "\n        ");
    i0.ɵɵtemplate(3, AdaptNavigationComponent_ng_template_85_span_11_ng_container_3_Template, 1, 0, "ng-container", 123);
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const justLogo_r173 = i0.ɵɵnextContext().simpleLogo;
    const _r181 = i0.ɵɵreference(18);
    const _r179 = i0.ɵɵreference(15);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngTemplateOutlet", justLogo_r173 ? _r181 : _r179);
} }
function AdaptNavigationComponent_ng_template_85_ng_template_14_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainer(0);
} }
function AdaptNavigationComponent_ng_template_85_ng_template_14_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n      ");
    i0.ɵɵtemplate(1, AdaptNavigationComponent_ng_template_85_ng_template_14_ng_container_1_Template, 1, 0, "ng-container", 123);
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵelementStart(3, "span", 9);
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n    ");
} if (rf & 2) {
    i0.ɵɵnextContext();
    const _r181 = i0.ɵɵreference(18);
    const ctx_r180 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵproperty("ngTemplateOutlet", _r181);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r180.metadata.productName);
} }
function AdaptNavigationComponent_ng_template_85_ng_template_17_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n      ");
    i0.ɵɵelementStart(1, "span", 8);
    i0.ɵɵtext(2, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n    ");
} if (rf & 2) {
    const ctx_r182 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(1);
    i0.ɵɵclassProp("logo-light", ctx_r182.themeClass === "" || ctx_r182.textColor === "#FFFFFF");
    i0.ɵɵproperty("ngClass", ctx_r182.metadata.logoClass || "logo");
    i0.ɵɵattribute("role", ctx_r182.metadata.ariaLabelForLogo ? "img" : null)("aria-label", ctx_r182.metadata.ariaLabelForLogo);
} }
function AdaptNavigationComponent_ng_template_85_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵtext(0, "\n    ");
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵtemplate(2, AdaptNavigationComponent_ng_template_85_a_2_Template, 5, 2, "a", 115);
    i0.ɵɵtext(3, "\n\n    ");
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵtemplate(5, AdaptNavigationComponent_ng_template_85_a_5_Template, 5, 2, "a", 116);
    i0.ɵɵtext(6, "\n\n    ");
    i0.ɵɵtext(7, "\n    ");
    i0.ɵɵtemplate(8, AdaptNavigationComponent_ng_template_85_button_8_Template, 5, 1, "button", 117);
    i0.ɵɵtext(9, "\n\n    ");
    i0.ɵɵtext(10, "\n    ");
    i0.ɵɵtemplate(11, AdaptNavigationComponent_ng_template_85_span_11_Template, 5, 1, "span", 118);
    i0.ɵɵtext(12, "\n\n    ");
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵtemplate(14, AdaptNavigationComponent_ng_template_85_ng_template_14_Template, 6, 2, "ng-template", null, 119, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(16, "\n\n    ");
    i0.ɵɵtemplate(17, AdaptNavigationComponent_ng_template_85_ng_template_17_Template, 4, 5, "ng-template", null, 120, i0.ɵɵtemplateRefExtractor);
    i0.ɵɵtext(19, "\n\n  ");
} if (rf & 2) {
    const ctx_r18 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r18.metadata.link);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", ctx_r18.metadata.route);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", ctx_r18.metadata.action);
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("ngIf", !(ctx_r18.metadata.action || ctx_r18.metadata.link || ctx_r18.metadata.route));
} }
function AdaptNavigationComponent_div_103_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 127);
    i0.ɵɵtext(1, "\n          ");
    i0.ɵɵelement(2, "div", 128);
    i0.ɵɵtext(3, "\n          ");
    i0.ɵɵelement(4, "div", 129);
    i0.ɵɵtext(5, "\n          ");
    i0.ɵɵelement(6, "div", 130);
    i0.ɵɵtext(7, "\n        ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    i0.ɵɵproperty("@loaderAnimation", undefined);
} }
function AdaptNavigationComponent_div_107_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "div", 131);
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r24 = i0.ɵɵnextContext();
    i0.ɵɵproperty("@loaderMessageAnimation", undefined);
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate1("\n            ", ctx_r24.loadingMessage, "\n          ");
} }
const _c31 = function (a0) { return { "container": a0 }; };
export class AdaptNavigationComponent extends AdaptRadarAngularGenericDirective {
    constructor(dps, el, _resizeService, deviceDetection, _navService, renderer, zone, _menuGlobalService, _changeDetectorRef, _adaptRadarService, ts, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptNavigation);
        this.el = el;
        this._resizeService = _resizeService;
        this.deviceDetection = deviceDetection;
        this._navService = _navService;
        this.renderer = renderer;
        this.zone = zone;
        this._menuGlobalService = _menuGlobalService;
        this._changeDetectorRef = _changeDetectorRef;
        this._adaptRadarService = _adaptRadarService;
        this.alerts = [];
        this.loadingRun = false;
        this.loadingMessage = '';
        this.floating = false;
        this.bizarreZone = false;
        this.customProfilePresent = false;
        this.iphoneCSSclass = '';
        this.margin = '';
        // TODO: Refactor this if possible
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        this.filtering = {
            name: ''
        };
        /**
         * DockedPanel open indicator
         */
        this.hamburgerMenuOpen = false;
        /* eslint-enable @typescript-eslint/no-explicit-any */
        /**
         * current active id of the Menu
         */
        this.currentActiveId = null;
        /**
         * hamburger mode - is collapsed mode
         */
        this.hamburgerMode = false;
        this.kebabMode = false;
        this.defaultTexts = {};
        this.alertClosed = new EventEmitter();
        this._themeClass = '';
        this._theme = '';
        this._destroyed$ = new ReplaySubject(1);
        this._menuTestIdIndex = 1;
        this.hamburgerMode = deviceDetection.isMobile();
        this.hamburgerService = dps;
        this.container = false;
        this.gapWidth = 20;
        this.setDefaultTexts(ts.getCurrentLanguage());
    }
    /**
     * Input which defines menu items
     */
    set menu(value) {
        if (isArray(value)) {
            this._menu = cloneDeep(value);
            this._menu.forEach(menuItem => this._extendProps(menuItem));
        }
    }
    get menu() {
        return this._menu;
    }
    set theme(v) {
        this._theme = v;
        this.themeClass = v;
    }
    get theme() {
        return this._theme;
    }
    set themeClass(v) {
        this.textColor = getComputedStyle(document.documentElement).getPropertyValue('--nav-links-color');
        switch (v.toLowerCase()) {
            case 'grey':
                this._themeClass = 'a-nav__grey';
                break;
            case 'white':
                this._themeClass = 'a-nav__white';
                break;
            case '':
                this._themeClass = '';
                break;
            case 'dark':
                this._themeClass = '';
                break;
            default:
                this._themeClass = 'a-nav__custom-color';
        }
    }
    get themeClass() {
        return this._themeClass;
    }
    ngOnInit() {
        var _a, _b, _c;
        if (this.config) {
            this.container = this.config.container || false;
            this.gapWidth = this.config.gapWidth || 20;
        }
        this.iphoneCSSclass = this.deviceDetection.deviceIphone() ? 'h-100' : '';
        this.customProfilePresent = !!this.customProfile;
        // create copy of menu and convert it to Tree
        this.menuModel = {
            subMenu: this.menu
        };
        this.addParent(this.menu, this.menuModel);
        this._navService.loaderState
            .pipe(takeUntil(this._destroyed$))
            .subscribe(this._loaderSubscriptionHandler.bind(this));
        this._navService.alertsState
            .pipe(takeUntil(this._destroyed$))
            .subscribe(this._alertSubscriptionHandler.bind(this));
        this.zone.runOutsideAngular(() => {
            this._removeDocumentScrollListener = this.renderer.listen('document', 'scroll', this._checkFloating.bind(this));
        });
        this.metadata = Object.assign(Object.assign({}, this.defaultTexts), this.metadata);
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptNavigation, {
            EventType: AdaptRadarAngularCustomEventType.Init,
            NavigationColor: this.theme,
            Logo: this.metadata.logoClass,
            MenuCount: (_a = this.menu) === null || _a === void 0 ? void 0 : _a.length,
            ActionCount: (_b = this.actions) === null || _b === void 0 ? void 0 : _b.length,
            HamburgerMode: this.hamburgerMode,
            Title: this.metadata.productName,
            ActionsLoaded: (_c = this.actions) === null || _c === void 0 ? void 0 : _c.map(action => action.name || action.title)
        });
    }
    setDefaultTexts(translation) {
        this.defaultTexts.skipText = translation['adapt.navigation.skip'];
        this.defaultTexts.menuText = translation['adapt.navigation.menu'];
        this.defaultTexts.backBtn = translation['adapt.common.back'];
        this.defaultTexts.switcher = {
            notFoundText: translation['adapt.navigation.notFound'],
            filter: {
                placeholder: translation['adapt.common.search'],
                btnClearText: translation['adapt.common.clear'],
                btnSearchText: translation['adapt.common.search'],
                labelText: ''
            }
        };
    }
    alertsHeightChanged(height) {
        this.margin = height + 'px';
        if (this.bizarreZone) {
            this.floatSection.nativeElement.classList.add('a-margin-transition');
        }
        this.setMargin();
    }
    setMargin() {
        if (!this.floating) {
            this.renderer.setStyle(this.floatSection.nativeElement, 'margin-bottom', '-2px');
        }
        else {
            this.renderer.setStyle(this.floatSection.nativeElement, 'margin-bottom', this.margin);
        }
    }
    /**
     * Recursive. Makes a Tree from the Menu object
     * parent - parent Node
     */
    addParent(subArray, parent) {
        for (const subArr of subArray) {
            subArr.parent = parent;
            if (subArr.items) {
                this.addParent(subArr.items, subArr);
            }
        }
    }
    // TODO: Refactor this if possible
    /* eslint-disable @typescript-eslint/no-explicit-any */
    /**
     * Recursive. Search Menu tree for the Node with given Id
     */
    searchTree(element, matchingId) {
        /* eslint-enable @typescript-eslint/no-explicit-any */
        if (parseInt(element.id, 10) === matchingId) {
            return element;
        }
        else if (element.subMenu !== null) {
            let i;
            let result = null;
            if (element.subMenu && element.subMenu.length) {
                for (i = 0; result === null && i < element.subMenu.length; i++) {
                    result = this.searchTree(element.subMenu[i], matchingId);
                }
            }
            return result;
        }
        return null;
    }
    /**
     * add resize listener to global nav
     */
    ngAfterContentInit() {
        const navResize$ = new Subject();
        navResize$
            .pipe(takeUntil(this._destroyed$), throttleTime(100, asyncScheduler, { trailing: true }))
            .subscribe(() => {
            this.zone.run(() => {
                this.checkForHamburger();
                SafeCdr.detectChanges(this._changeDetectorRef);
            });
        });
        this._resizeService.addResizeEventListener(this.el.nativeElement, navResize$.next.bind(navResize$));
        this._resizeService.addResizeEventListener(this.navElContentShadow.nativeElement, navResize$.next.bind(navResize$));
    }
    ngOnChanges(changes) {
        if (changes.metadata) {
            this.metadata = Object.assign(Object.assign({}, this.defaultTexts), this.metadata);
        }
        if (changes.menu) {
            // need to check of menu is loaded afterwards and need to rebuild the tree
            this.menuModel = {
                subMenu: this.menu
            };
        }
    }
    /**
     * in this hook we are saving width of menu element to determine later if it fits in header or not
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.hamburgerMenuConfig = {
            content: this.hamburgerMenuTemplate,
            direction: 'left',
            blockDismiss: false
        };
        // Throw 'inverted' prop to the inner profile component
        if (this.profileEl) {
            this.profileEl.profile.inverted = !!this._themeClass;
        }
        this._verifyLogo();
    }
    /**
     * remove resize listener
     */
    ngOnDestroy() {
        this._resizeService.removeResizeEventListener(this.el.nativeElement);
        this._resizeService.removeResizeEventListener(this.navElContentShadow.nativeElement);
        if (this._removeDocumentScrollListener) {
            this._removeDocumentScrollListener();
        }
        this._destroyed$.next(true);
        this._destroyed$.complete();
    }
    /**
     * check hamburger mode
     */
    checkForHamburger() {
        var _a, _b, _c;
        const isHamburgerMode = this.navElContentShadow.nativeElement.offsetWidth + this.gapWidth > this.navElShadow.nativeElement.offsetWidth;
        const navContainerWidth = this.navElShadow.nativeElement.offsetWidth;
        // If hamburger mode changed send event to analytics system
        if (isHamburgerMode !== this.hamburgerMode) {
            this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptNavigationHamburger, {
                EventType: AdaptRadarAngularCustomEventType.ValueChanged,
                HamburgerMode: this.hamburgerMode,
                ContainerSize: navContainerWidth,
                MenuCount: (_a = this.menu) === null || _a === void 0 ? void 0 : _a.length,
                ActionCount: (_b = this.actions) === null || _b === void 0 ? void 0 : _b.length,
                ProfilePresent: !!this.profile
            });
        }
        this.hamburgerMode = isHamburgerMode;
        this.kebabMode = this.hamburgerMode ?
            this.navElContentShadow.nativeElement.offsetWidth -
                this.menuElShadow.nativeElement.offsetWidth -
                (((_c = this.profileElShadow) === null || _c === void 0 ? void 0 : _c.nativeElement.offsetWidth) || 0) -
                this.logoElShadow.nativeElement.offsetWidth +
                this.logoEl.nativeElement.offsetWidth +
                this.gapWidth > navContainerWidth : false;
    }
    showSubmenu(id) {
        this.hamburgerContent.nativeElement.querySelector('#' + id).classList.toggle('show');
        this.hamburgerContent.nativeElement.classList.toggle('submenu');
        this.filtering.name = '';
    }
    setFocusToAnchor(el) {
        el.focus();
    }
    /**
     * Close all submenus and opens current
     */
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    toggleSubmenu(event) {
        this.closeAllSubmenu();
        event.target.classList.toggle('submenuOpen');
    }
    /**
     * Traverses all opened submenus and closes them
     */
    closeAllSubmenu() {
        const elements = this.menuEl.nativeElement.querySelectorAll('.submenuOpen');
        // TODO: Refactor this if possible
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        Array.prototype.map.call(elements, function (item) {
            item.classList.remove('submenuOpen');
        });
        this.filtering.name = '';
    }
    /**
     * clears the filter
     */
    clearFilter(searchInputElement) {
        this.filtering.name = '';
        searchInputElement.focus();
    }
    handleMenuLinkSelection(item) {
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptNavigationMenu, {
            EventType: AdaptRadarAngularCustomEventType.Click,
            Name: item.label,
            Disabled: item.disabled,
            Type: 'link',
            MaxDepth: maxDepth(this.menu, 'subMenu'),
            SiblingsOptionsCount: this._getSiblingsCount(),
            HiddenSiblingsOptionsCount: this._getHiddenSiblingsCount()
        });
    }
    handleMenuRouteSelection(item, isAction = false, closeHamburger = true) {
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptNavigationMenu, {
            EventType: AdaptRadarAngularCustomEventType.Click,
            Name: item.label,
            Disabled: item.disabled,
            Type: 'route',
            MaxDepth: maxDepth(this.menu, 'subMenu'),
            SiblingsOptionsCount: this._getSiblingsCount(),
            HiddenSiblingsOptionsCount: this._getHiddenSiblingsCount()
        });
        this.selectMenuItem(item.id, isAction);
        if (closeHamburger) {
            this.closeHamburger();
        }
    }
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    handleMenuActionSelection(event, item, invokeSelectMenuItem = false, isAction = false, closeHamburger = true) {
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptNavigationMenu, {
            EventType: AdaptRadarAngularCustomEventType.Click,
            Name: item.label,
            Disabled: item.disabled,
            Type: 'action',
            MaxDepth: maxDepth(this.menu, 'subMenu'),
            SiblingsOptionsCount: this._getSiblingsCount(),
            HiddenSiblingsOptionsCount: this._getHiddenSiblingsCount()
        });
        if (invokeSelectMenuItem) {
            this.selectMenuItem(item.id, isAction);
        }
        item.action(event, item.id);
        if (closeHamburger) {
            this.closeHamburger();
        }
    }
    // TODO: Refactor this if possible
    /* eslint-disable @typescript-eslint/no-explicit-any */
    /**
     * Set active state to the Menu item and all parent Node in the Tree
     */
    selectMenuItem(id, isAction = false) {
        /* eslint-enable @typescript-eslint/no-explicit-any */
        if (!isAction) {
            this._menuGlobalService.removeActiveState();
        }
        if (isDefined(id)) {
            // remove all previous active
            let oldNode = this.searchTree(this.menuModel, this.currentActiveId);
            while (oldNode) {
                oldNode.active = false;
                oldNode = oldNode.parent;
            }
            this.currentActiveId = id;
            let node = this.searchTree(this.menuModel, id);
            while (node) {
                node.active = true;
                node = node.parent;
            }
        }
        // close all open dropdowns
        if (this.allDropdownsDirectives) {
            this.allDropdownsDirectives.forEach((item) => {
                item.close();
            });
        }
    }
    /**
     * open menu
     */
    openHamburger() {
        this.hamburgerService.open(this.hamburgerMenuConfig)
            .then(() => {
            this.hamburgerMenuOpen = false;
        }, () => {
            this.hamburgerMenuOpen = false;
        });
        this.hamburgerMenuOpen = true;
    }
    /**
     * closes menu
     */
    closeHamburger() {
        this.hamburgerService.closeAllDockedPanels();
    }
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    getTarget(item) {
        return item.target ? item.target : '_self';
    }
    getAnalyticsMetadata() {
        return {
            parentComponentName: AdaptRadarAngularCustomEventName.AdaptNavigation,
            siblingsCount: this._getSiblingsCount(),
            hiddenSiblingsCount: this._getHiddenSiblingsCount()
        };
    }
    trackByFnMenu(index, item) {
        return item.id;
    }
    trackByFnActions(index, item) {
        return item;
    }
    _extendProps(item) {
        item.label = item.label || item.name;
        item.routerLink = item.routerLink || item.route;
        item.items = item.items || item.subMenu;
        if (item.items) {
            item.items.forEach(i => this._extendProps(i));
        }
    }
    /**
     * Checks whether provided url() data type is valid and image can be loaded.
     * Otherwise, replace with the default bmc logo
     */
    _verifyLogo() {
        if (this.logoEl) {
            const logoElem = this.logoEl.nativeElement.querySelector('.a-product__logo');
            const logoBgImage = getComputedStyle(logoElem).backgroundImage;
            if (!logoBgImage.includes('url')) {
                return;
            }
            // retrieve a string between parentheses
            const parenthesesRegExp = /(?:\(['"]?)(.*?)(?:['"]?\))/;
            const url = parenthesesRegExp.exec(logoBgImage)[1];
            checkURL(url)
                .pipe(takeUntil(this._destroyed$))
                .subscribe(noop, () => {
                this.logoEl.nativeElement.querySelector('.a-product__logo').classList.add('logo-bmc');
                this.logoElShadow.nativeElement.querySelector('.a-product__logo').classList.add('logo-bmc');
            });
        }
    }
    /**
     * Checks alerts position according to the content scroll
     */
    _checkFloating() {
        const { top } = this.floatSection.nativeElement.getBoundingClientRect();
        this.floating = top <= 0;
        // since this function is out of Angular detection zone, need to change class manually
        // from bmc-ux/adapt-angular#972
        if (this.floating) {
            this.floatWrapper.nativeElement.classList.add('a-floating');
        }
        else {
            this.floatWrapper.nativeElement.classList.remove('a-floating');
        }
        // there a zone of 145px under navigation, after scrolling it browser doesn't redraw upper part
        // eslint-disable-next-line yoda
        this.bizarreZone = 0 > top && top > -145;
        this.floatSection.nativeElement.classList.remove('a-margin-transition');
        this.setMargin();
    }
    ;
    _loaderSubscriptionHandler(animationState) {
        this.loadingRun = animationState.show;
        this.loadingMessage = animationState.message;
        this._changeDetectorRef.detectChanges();
    }
    _alertSubscriptionHandler(newAlert) {
        this._adaptRadarService.sendCustomEventForAllProviders(AdaptRadarAngularCustomEventName.AdaptNavigationAlert, {
            EventType: AdaptRadarAngularCustomEventType.Init,
            Type: newAlert.type,
            Title: newAlert.title
        });
        this.alerts.unshift(newAlert);
    }
    _getSiblingsCount() {
        return this.menu ? this.menu.filter(menuItem => !menuItem.hide).length : 0;
    }
    _getHiddenSiblingsCount() {
        return this.menu ? this.menu.filter(menuItem => menuItem.hide).length : 0;
    }
    generateMenuItemTestID(item) {
        if (!item.testID) {
            item.testID = `navigation-menu-item-test-id-${this._menuTestIdIndex++}`;
        }
        return item.testID;
    }
}
AdaptNavigationComponent.ɵfac = function AdaptNavigationComponent_Factory(t) { return new (t || AdaptNavigationComponent)(i0.ɵɵdirectiveInject(i1.AdaptDockedPanelService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i2.ResizeService), i0.ɵɵdirectiveInject(i3.AdaptDeviceDetectionService), i0.ɵɵdirectiveInject(i4.AdaptNavigationService), i0.ɵɵdirectiveInject(i0.Renderer2), i0.ɵɵdirectiveInject(i0.NgZone), i0.ɵɵdirectiveInject(i5.MenuGlobalService), i0.ɵɵdirectiveInject(i0.ChangeDetectorRef), i0.ɵɵdirectiveInject(i6.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i7.AdaptTranslateService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptNavigationComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptNavigationComponent, selectors: [["adapt-navigation"]], viewQuery: function AdaptNavigationComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
        i0.ɵɵviewQuery(_c1, 5);
        i0.ɵɵviewQuery(_c2, 5);
        i0.ɵɵviewQuery(_c3, 5);
        i0.ɵɵviewQuery(_c4, 7);
        i0.ɵɵviewQuery(_c5, 7);
        i0.ɵɵviewQuery(_c6, 5);
        i0.ɵɵviewQuery(_c7, 5);
        i0.ɵɵviewQuery(_c8, 7);
        i0.ɵɵviewQuery(_c9, 7);
        i0.ɵɵviewQuery(_c10, 7);
        i0.ɵɵviewQuery(_c11, 7);
        i0.ɵɵviewQuery(_c12, 7);
        i0.ɵɵviewQuery(_c13, 7);
        i0.ɵɵviewQuery(_c14, 5);
        i0.ɵɵviewQuery(_c15, 7);
        i0.ɵɵviewQuery(_c16, 7);
        i0.ɵɵviewQuery(_c17, 7);
        i0.ɵɵviewQuery(_c18, 7);
        i0.ɵɵviewQuery(AdaptDropdownDirective, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.menuEl = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.logoEl = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.hamburgerBtn = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.hamburgerLogo = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.alertsSection = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.actionsEl = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.profileEl = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.hamburgerContent = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.hamburgerMenuTemplate = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.loaderSection = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.floatSection = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.floatWrapper = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.navElShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.navElContentShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.profileElShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.actionsElShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.menuElShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.hamburgerBtnElShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.logoElShadow = _t.first);
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.allDropdownsDirectives = _t);
    } }, inputs: { menu: "menu", metadata: "metadata", config: "config", kebabView: "kebabView", gapWidth: "gapWidth", container: "container", actions: "actions", profile: "profile", customProfile: "customProfile", appSwitcherEnabled: "appSwitcherEnabled", appSwitcherMetaData: "appSwitcherMetaData", theme: "theme", checkForHamburger: "checkForHamburger", selectMenuItem: "selectMenuItem", closeHamburger: "closeHamburger" }, outputs: { alertClosed: "alertClosed" }, features: [i0.ɵɵInheritDefinitionFeature, i0.ɵɵNgOnChangesFeature], decls: 114, vars: 40, consts: [["role", "link", "queryParamsHandling", "preserve", 1, "accessibility-start-of-content", 3, "routerLink", "fragment", "click"], ["aria-hidden", "true", 1, "a-shadow", "d-flex", "flex-nowrap"], [1, "a-nav", "flex-grow-0", "flex-shrink-1", "w-100", 3, "ngClass"], ["navElShadow", ""], [1, "d-flex", "flex-grow-1", "flex-shrink-0"], ["navElContentShadow", ""], ["type", "button", "disabled", "", 1, "py-0", "pl-0", "a-product", "btn-link"], ["logoElShadow", ""], [1, "a-product__logo", 3, "ngClass"], [1, "a-product__name"], [1, "a-sep"], [1, "a-menu"], ["menuElShadow", ""], [3, "class", 4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "a-actions"], ["actionsElShadow", ""], [4, "ngFor", "ngForOf", "ngForTrackBy"], ["class", "adapt-profile", 4, "ngIf"], [1, "a-nav", 3, "ngClass"], [1, "navigation-wrapper"], [4, "ngIf", "ngIfThen"], ["class", "a-hamburger", "type", "button", 3, "click", 4, "ngIf"], ["class", "pl-1 pr-1 a-product", 4, "ngIf"], ["class", "a-menu", 4, "ngIf"], [1, "a-actions", 3, "actions", "kebabMode", "kebabView", "texts"], ["actionsEl", ""], [4, "ngIf"], [3, "switcherDisplay", "texts", "appSwitcherMetaData"], [1, "gradient"], ["tabindex", "-1", 1, "accessibility-start-of-content", 3, "id"], ["startContent", ""], ["hamburgerMenuTemplate", ""], ["logoSection", ""], [1, "a-header_divider"], ["floatSection", ""], [1, "a-floating-wrapper"], ["floatWrapper", ""], [3, "adaptRadarDisableEventSending", "isContainer", "config", "isVisible", "alertClosed", "heightChanged"], ["alertsSection", ""], [1, "a-loader"], ["loaderSection", ""], [1, "a-loader_wrapper"], ["class", "a-loader_bar adapt-alert-bar", 4, "ngIf"], [1, "a-loader_message-wrapper"], ["class", "a-loader_message", 4, "ngIf"], [3, "ngClass"], [1, "a-menu__text", 3, "ngClass"], [1, "adapt-menu__link-toggle", "m-start-1"], [3, "btn", "btn-link", 4, "ngIf"], ["disabled", "", 3, "class", 4, "ngIf"], ["class", "a-action__counter", 4, "ngIf"], [1, "a-action__counter"], ["disabled", ""], ["class", "a-action__title p-end-1", 4, "ngIf"], [1, "a-action__title", "p-end-1"], [1, "adapt-profile"], ["profileElShadow", ""], [1, "a-profile", 3, "ngClass"], [1, "a-profile__menu", "dropdown-toggle", 2, "position", "relative", 3, "ngClass"], ["class", "a-profile__avatar", 3, "ngClass", "ngStyle", 4, "ngIf"], ["class", "menu-profile__name", 4, "ngIf"], [1, "a-profile__avatar", 3, "ngClass", "ngStyle"], [1, "menu-profile__name"], [1, "menu-profile__signed"], [1, "menu-profile__username"], ["type", "button", 1, "a-hamburger", 3, "click"], ["hamburgerBtn", ""], [1, "sr-only"], [1, "pl-1", "pr-1", "a-product"], ["hamburgerLogo", ""], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["menuEl", ""], [3, "cascadeItem", "adaptRadarDisableEventSending", "headerTheme", "bodyTheme", "connectMenus", "filteringConfig", "analyticsMetadata"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], ["profileEl", ""], [3, "profile", "adaptRadarDisableEventSending"], [1, "a-hamburger__wrapper"], [1, "a-hamburger__content"], ["hamburgerContent", ""], [2, "padding", "0 15px 15px"], [1, "a-hamburger__close", 3, "adaptRadarDisableEventSending", "inverse", "close"], [1, "m-end-5"], [1, "a-hamburger__menu"], ["class", "a-hamburger__menu-link", 3, "ngClass", "click", 4, "ngIf"], ["class", "a-hamburger__menu-link", "routerLinkActive", "active", 3, "routerLink", "click", 4, "ngIf"], ["type", "button", "class", "a-hamburger__menu-link btn-link", 3, "ngClass", "click", 4, "ngIf"], [1, "a-hamburger__menu-link", 3, "ngClass", "click"], ["routerLinkActive", "active", 1, "a-hamburger__menu-link", 3, "routerLink", "click"], ["type", "button", 1, "a-hamburger__menu-link", "btn-link", 3, "ngClass", "click"], ["tabindex", "0", 1, "a-hamburger__menu-link", "a-dropdown__submenu-handler", 3, "ngClass", "click", "keypress"], [1, "a-hamburger__menu", "a-hamburger__submenu", 3, "id"], [2, "height", "2.5rem"], ["type", "button", "aria-label", "Close", 1, "submenu-close", 3, "click"], [1, "a-hamburger__submenu-category"], ["class", "a-hamburger__search-wrapper", 4, "ngIf"], [1, "a-hamburger__search-wrapper"], [1, "dropdown-divider"], [1, "d-flex"], [1, "d-icon-search", "a-hamburger__search", 3, "adaptRadarDisableEventSending", "ngModel", "placeholder", "ngModelChange"], ["adaptRxTextfieldComponent", ""], ["type", "button", 1, "adapt-search-clear", "d-icon-cross_adapt", "a-filter-clear", 3, "ngClass", "click"], [3, "class", 4, "ngIf"], [3, "class", "ngClass", "click", 4, "ngIf"], ["routerLinkActive", "active", 3, "class", "routerLink", "click", 4, "ngIf"], ["type", "button", 3, "class", "click", 4, "ngIf"], [3, "ngClass", "click"], ["routerLinkActive", "active", 3, "routerLink", "click"], ["type", "button", 3, "click"], [3, "ngClass", "class", 4, "ngIf"], ["type", "button", 3, "class", "ngClass", "click", 4, "ngIf"], ["routerLinkActive", "selected", 3, "class", "routerLink", "click", 4, "ngIf"], [1, "a-hamburger__submenu-results"], ["type", "button", 3, "ngClass", "click"], ["routerLinkActive", "selected", 3, "routerLink", "click"], [3, "profile", "type", "closeHamburger"], ["class", "a-product", "target", "_blank", 3, "href", 4, "ngIf"], ["class", "a-product", 3, "routerLink", 4, "ngIf"], ["class", "a-product btn-link", "type", "button", 3, "click", 4, "ngIf"], ["class", "a-product", 4, "ngIf"], ["logoAndProductTmpl", ""], ["justLogoTmpl", ""], ["target", "_blank", 1, "a-product", 3, "href"], ["logoEl", ""], [4, "ngTemplateOutlet"], [1, "a-product", 3, "routerLink"], ["type", "button", 1, "a-product", "btn-link", 3, "click"], [1, "a-product"], [1, "a-loader_bar", "adapt-alert-bar"], [1, "progress-bar-intermediate", "lay1"], [1, "progress-bar-intermediate", "lay2"], [1, "progress-bar-intermediate", "lay3"], [1, "a-loader_message"]], template: function AdaptNavigationComponent_Template(rf, ctx) { if (rf & 1) {
        const _r198 = i0.ɵɵgetCurrentView();
        i0.ɵɵtext(0, "\n");
        i0.ɵɵelementStart(1, "header");
        i0.ɵɵtext(2, "\n  ");
        i0.ɵɵelementStart(3, "a", 0);
        i0.ɵɵlistener("click", function AdaptNavigationComponent_Template_a_click_3_listener() { i0.ɵɵrestoreView(_r198); const _r14 = i0.ɵɵreference(80); return ctx.setFocusToAnchor(_r14); });
        i0.ɵɵtext(4);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(5, "\n\n  ");
        i0.ɵɵtext(6, "\n  ");
        i0.ɵɵelementStart(7, "div", 1);
        i0.ɵɵtext(8, "\n    ");
        i0.ɵɵelementStart(9, "nav", 2, 3);
        i0.ɵɵtext(11, "\n      ");
        i0.ɵɵelementStart(12, "div", 4, 5);
        i0.ɵɵtext(14, "\n        ");
        i0.ɵɵtext(15, "\n        ");
        i0.ɵɵelementStart(16, "button", 6, 7);
        i0.ɵɵtext(18, "\n            ");
        i0.ɵɵelement(19, "span", 8);
        i0.ɵɵtext(20, "\n          ");
        i0.ɵɵelementStart(21, "span", 9);
        i0.ɵɵtext(22);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(23, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(24, "\n        ");
        i0.ɵɵtext(25, "\n        \n        ");
        i0.ɵɵelement(26, "div", 10);
        i0.ɵɵtext(27, "\n\n        ");
        i0.ɵɵtext(28, "\n        ");
        i0.ɵɵelementStart(29, "ul", 11, 12);
        i0.ɵɵtext(31, "\n          ");
        i0.ɵɵtemplate(32, AdaptNavigationComponent_li_32_Template, 6, 5, "li", 13);
        i0.ɵɵtext(33, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(34, "\n        ");
        i0.ɵɵtext(35, "\n\n        ");
        i0.ɵɵtext(36, "\n        ");
        i0.ɵɵelementStart(37, "div", 14, 15);
        i0.ɵɵtext(39, "\n          ");
        i0.ɵɵtemplate(40, AdaptNavigationComponent_ng_container_40_Template, 6, 2, "ng-container", 16);
        i0.ɵɵtext(41, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(42, "\n        ");
        i0.ɵɵtext(43, "\n\n        ");
        i0.ɵɵtext(44, "\n        ");
        i0.ɵɵtemplate(45, AdaptNavigationComponent_div_45_Template, 13, 9, "div", 17);
        i0.ɵɵtext(46, "\n        ");
        i0.ɵɵtext(47, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(48, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(49, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(50, "\n  ");
        i0.ɵɵtext(51, "\n\n  ");
        i0.ɵɵelementStart(52, "nav", 18);
        i0.ɵɵtext(53, "\n    ");
        i0.ɵɵelementStart(54, "div", 19);
        i0.ɵɵtext(55, "\n      ");
        i0.ɵɵtext(56, "\n      ");
        i0.ɵɵtemplate(57, AdaptNavigationComponent_ng_container_57_Template, 1, 0, "ng-container", 20);
        i0.ɵɵtext(58, "\n\n      ");
        i0.ɵɵtext(59, "\n      ");
        i0.ɵɵtemplate(60, AdaptNavigationComponent_button_60_Template, 6, 2, "button", 21);
        i0.ɵɵtext(61, "\n\n      ");
        i0.ɵɵelement(62, "div", 10);
        i0.ɵɵtext(63, "\n\n      ");
        i0.ɵɵtemplate(64, AdaptNavigationComponent_div_64_Template, 5, 3, "div", 22);
        i0.ɵɵtext(65, "\n\n      ");
        i0.ɵɵtemplate(66, AdaptNavigationComponent_ul_66_Template, 5, 2, "ul", 23);
        i0.ɵɵtext(67, "\n\n      ");
        i0.ɵɵelement(68, "adapt-nav-actions", 24, 25);
        i0.ɵɵtext(70, "\n\n      ");
        i0.ɵɵtemplate(71, AdaptNavigationComponent_ng_container_71_Template, 6, 2, "ng-container", 26);
        i0.ɵɵtext(72, "\n\n      ");
        i0.ɵɵelement(73, "adapt-nav-actions-switcher", 27);
        i0.ɵɵtext(74, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(75, "\n    ");
        i0.ɵɵelement(76, "div", 28);
        i0.ɵɵtext(77, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(78, "\n\n  ");
        i0.ɵɵelement(79, "div", 29, 30);
        i0.ɵɵtext(81, "\n\n  ");
        i0.ɵɵtemplate(82, AdaptNavigationComponent_ng_template_82_Template, 28, 11, "ng-template", null, 31, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(84, "\n\n  ");
        i0.ɵɵtemplate(85, AdaptNavigationComponent_ng_template_85_Template, 20, 4, "ng-template", null, 32, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(87, "\n\n  ");
        i0.ɵɵtext(88, "\n  ");
        i0.ɵɵelement(89, "div", 33, 34);
        i0.ɵɵtext(91, "\n\n  ");
        i0.ɵɵelementStart(92, "div", 35, 36);
        i0.ɵɵtext(94, "\n\n    ");
        i0.ɵɵelementStart(95, "adapt-alert-animated", 37, 38);
        i0.ɵɵlistener("alertClosed", function AdaptNavigationComponent_Template_adapt_alert_animated_alertClosed_95_listener($event) { return ctx.alertClosed.emit($event); })("heightChanged", function AdaptNavigationComponent_Template_adapt_alert_animated_heightChanged_95_listener($event) { return ctx.alertsHeightChanged($event); });
        i0.ɵɵelementEnd();
        i0.ɵɵtext(97, "\n\n    ");
        i0.ɵɵelementStart(98, "div", 39, 40);
        i0.ɵɵtext(100, "\n      ");
        i0.ɵɵelementStart(101, "div", 41);
        i0.ɵɵtext(102, "\n        ");
        i0.ɵɵtemplate(103, AdaptNavigationComponent_div_103_Template, 8, 1, "div", 42);
        i0.ɵɵtext(104, "\n        ");
        i0.ɵɵelementStart(105, "div", 43);
        i0.ɵɵtext(106, "\n          ");
        i0.ɵɵtemplate(107, AdaptNavigationComponent_div_107_Template, 2, 2, "div", 44);
        i0.ɵɵtext(108, "\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(109, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(110, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(111, "\n\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(112, "\n\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(113, "\n");
    } if (rf & 2) {
        const _r17 = i0.ɵɵreference(86);
        i0.ɵɵadvance(1);
        i0.ɵɵclassMap("a-header " + ctx.themeClass);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("fragment", ctx.metadata.skipHref || "start-of-content");
        i0.ɵɵadvance(1);
        i0.ɵɵtextInterpolate(ctx.metadata.skipText);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(36, _c31, ctx.container));
        i0.ɵɵadvance(10);
        i0.ɵɵclassProp("logo-light", ctx.themeClass === "" || ctx.textColor === "#FFFFFF");
        i0.ɵɵproperty("ngClass", ctx.metadata.logoClass || "logo");
        i0.ɵɵadvance(3);
        i0.ɵɵtextInterpolate(ctx.metadata.productName);
        i0.ɵɵadvance(10);
        i0.ɵɵproperty("ngForOf", ctx.menu)("ngForTrackBy", ctx.trackByFnMenu);
        i0.ɵɵadvance(8);
        i0.ɵɵproperty("ngForOf", ctx.actions)("ngForTrackBy", ctx.trackByFnActions);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", ctx.profile);
        i0.ɵɵadvance(7);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction1(38, _c31, ctx.container));
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", !ctx.hamburgerMode)("ngIfThen", _r17);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.hamburgerMode);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngIf", ctx.hamburgerMode && !ctx.metadata.hideMobileLogo);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx.hamburgerMode);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("actions", ctx.actions)("kebabMode", ctx.kebabMode)("kebabView", ctx.kebabView)("texts", ctx.metadata.switcher);
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("ngIf", ctx.profile && !ctx.hamburgerMode);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("switcherDisplay", ctx.appSwitcherEnabled)("texts", ctx.metadata.switcher)("appSwitcherMetaData", ctx.appSwitcherMetaData);
        i0.ɵɵadvance(6);
        i0.ɵɵpropertyInterpolate("id", ctx.metadata.skipHref || "start-of-content");
        i0.ɵɵadvance(16);
        i0.ɵɵproperty("adaptRadarDisableEventSending", true)("isContainer", ctx.container)("config", ctx.alerts)("isVisible", ctx.alerts.length > 0);
        i0.ɵɵadvance(8);
        i0.ɵɵproperty("ngIf", ctx.loadingRun);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngIf", ctx.loadingMessage);
    } }, directives: [i8.RouterLinkWithHref, i9.NgClass, i9.NgForOf, i9.NgIf, i10.AdaptNavActionsComponent, i11.AdaptNavActionsSwictherComponent, i12.AdaptAlertAnimatedComponent, i9.NgStyle, i9.NgTemplateOutlet, i13.AdaptMenuComponent, i14.AdaptProfileComponent, i15.AdaptCloseComponent, i8.RouterLinkActive, i16.AdaptRxTextfieldComponent, i17.NgControlStatus, i17.NgModel], pipes: [i18.AdaptNavHamburgerPipe, i18.AdaptNavTextSearchPipe], styles: [".adapt-profile[_ngcontent-%COMP%] {\n      display: inline-block;\n      margin-right: -10px;\n      margin-left: auto;\n      -webkit-box-flex: 0;\n      -ms-flex: 0 0 auto;\n      flex: 0 0 auto;\n    }\n\n    .a-actions[_ngcontent-%COMP%]    + .adapt-profile[_ngcontent-%COMP%] {\n      margin-left: 0;\n    }\n\n    .navigation-wrapper[_ngcontent-%COMP%] {\n        width: 100%;\n        display: flex;\n    }"], data: { animation: [
            trigger('loaderAnimation', [
                transition('void => *', [
                    useAnimation(ANIMATION_NAME.height)
                ]),
                transition('* => void', [
                    useAnimation(ANIMATION_NAME.opacity, {
                        params: {
                            from: '*',
                            to: '0'
                        }
                    })
                ])
            ]),
            trigger('loaderMessageAnimation', [
                transition('void => *', [
                    useAnimation(ANIMATION_NAME.moveIn, { params: { translateY: -50 } })
                ]),
                transition('* => void', [
                    useAnimation(ANIMATION_NAME.moveOut, { params: { translateY: -50 } })
                ])
            ])
        ] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptNavigationComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-navigation',
                templateUrl: './navigation.component.html',
                styles: [`
    .adapt-profile {
      display: inline-block;
      margin-right: -10px;
      margin-left: auto;
      -webkit-box-flex: 0;
      -ms-flex: 0 0 auto;
      flex: 0 0 auto;
    }

    .a-actions + .adapt-profile {
      margin-left: 0;
    }

    .navigation-wrapper {
        width: 100%;
        display: flex;
    }
    `
                ],
                animations: [
                    trigger('loaderAnimation', [
                        transition('void => *', [
                            useAnimation(ANIMATION_NAME.height)
                        ]),
                        transition('* => void', [
                            useAnimation(ANIMATION_NAME.opacity, {
                                params: {
                                    from: '*',
                                    to: '0'
                                }
                            })
                        ])
                    ]),
                    trigger('loaderMessageAnimation', [
                        transition('void => *', [
                            useAnimation(ANIMATION_NAME.moveIn, { params: { translateY: -50 } })
                        ]),
                        transition('* => void', [
                            useAnimation(ANIMATION_NAME.moveOut, { params: { translateY: -50 } })
                        ])
                    ])
                ]
            }]
    }], function () { return [{ type: i1.AdaptDockedPanelService }, { type: i0.ElementRef }, { type: i2.ResizeService }, { type: i3.AdaptDeviceDetectionService }, { type: i4.AdaptNavigationService }, { type: i0.Renderer2 }, { type: i0.NgZone }, { type: i5.MenuGlobalService }, { type: i0.ChangeDetectorRef }, { type: i6.AdaptRadarAngularService }, { type: i7.AdaptTranslateService }, { type: i0.ElementRef }]; }, { allDropdownsDirectives: [{
            type: ViewChildren,
            args: [AdaptDropdownDirective]
        }], menuEl: [{
            type: ViewChild,
            args: ['menuEl', { static: false }]
        }], logoEl: [{
            type: ViewChild,
            args: ['logoEl', { static: false }]
        }], hamburgerBtn: [{
            type: ViewChild,
            args: ['hamburgerBtn', { static: false }]
        }], hamburgerLogo: [{
            type: ViewChild,
            args: ['hamburgerLogo', { static: false }]
        }], alertsSection: [{
            type: ViewChild,
            args: ['alertsSection', { static: true }]
        }], actionsEl: [{
            type: ViewChild,
            args: ['actionsEl', { static: true }]
        }], profileEl: [{
            type: ViewChild,
            args: ['profileEl', { static: false }]
        }], hamburgerContent: [{
            type: ViewChild,
            args: ['hamburgerContent', { static: false }]
        }], hamburgerMenuTemplate: [{
            type: ViewChild,
            args: ['hamburgerMenuTemplate', { static: true }]
        }], loaderSection: [{
            type: ViewChild,
            args: ['loaderSection', { static: true }]
        }], floatSection: [{
            type: ViewChild,
            args: ['floatSection', { static: true }]
        }], floatWrapper: [{
            type: ViewChild,
            args: ['floatWrapper', { static: true }]
        }], navElShadow: [{
            type: ViewChild,
            args: ['navElShadow', { static: true }]
        }], navElContentShadow: [{
            type: ViewChild,
            args: ['navElContentShadow', { static: true }]
        }], profileElShadow: [{
            type: ViewChild,
            args: ['profileElShadow', { static: false }]
        }], actionsElShadow: [{
            type: ViewChild,
            args: ['actionsElShadow', { static: true }]
        }], menuElShadow: [{
            type: ViewChild,
            args: ['menuElShadow', { static: true }]
        }], hamburgerBtnElShadow: [{
            type: ViewChild,
            args: ['hamburgerBtnElShadow', { static: true }]
        }], logoElShadow: [{
            type: ViewChild,
            args: ['logoElShadow', { static: true }]
        }], menu: [{
            type: Input
        }], metadata: [{
            type: Input
        }], config: [{
            type: Input
        }], kebabView: [{
            type: Input
        }], gapWidth: [{
            type: Input
        }], container: [{
            type: Input
        }], actions: [{
            type: Input
        }], profile: [{
            type: Input
        }], customProfile: [{
            type: Input
        }], appSwitcherEnabled: [{
            type: Input
        }], appSwitcherMetaData: [{
            type: Input
        }], theme: [{
            type: Input
        }], alertClosed: [{
            type: Output
        }], checkForHamburger: [{
            type: Input
        }], selectMenuItem: [{
            type: Input
        }], closeHamburger: [{
            type: Input
        }] }); })();
