// @bmc-ux/adapt-radar/esm2015/lib/providers/providers-scripts.js
/* Gainsight */
export const gainsightInitScript = (id, customSourceScript) => `
  (function(n,t,a,e){var i="aptrinsic";n[i]=n[i]||function(){
   (n[i].q=n[i].q||[]).push(arguments)},n[i].p=e;
   var r=t.createElement("script");r.async=!0,r.src=a+"?a="+e;
   var c=t.getElementsByTagName("script")[0];c.parentNode.insertBefore(r,c)
  })(window,document, "${customSourceScript || gainsightSourceScript()}", "${id}");
`;
// Use own BMC cdn to provide Gainsight script to avoid problems with security like SRI
export const gainsightSourceScript = () => 'https://documents.bmc.com/products/docs/gainsight/main/aptrinsic.js';
// export const gainsightSourceScript = (): string => 'https://web-sdk.aptrinsic.com/api/aptrinsic.js';
/* Google Site Tag */
export const googleGlobalSiteTagInitScript = (id) => `
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', '${id}');
`;
export const googleGlobalSiteTagSourceScript = (id) => `https://www.googletagmanager.com/gtag/js?id=${id}`;
/* Google Tag Manager */
export const googleTagManagerInitScript = (id) => `
  (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':new Date().getTime(),event:'gtm.js'});
  var f=d.getElementsByTagName(s)[0],j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;
  j.src='https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);})(window,document,'script','dataLayer','${id}');
`;
