// @helix/platform/esm2015/view/components/field-components/extension-container/design/extension-container-design.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ExtensionContainerDesignComponent } from './extension-container-design.component';
import { AdaptEmptyStateModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
export class ExtensionContainerDesignModule {
}
ExtensionContainerDesignModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ExtensionContainerDesignModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ExtensionContainerDesignModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ExtensionContainerDesignModule, declarations: [ExtensionContainerDesignComponent], imports: [CommonModule, AdaptEmptyStateModule] });
ExtensionContainerDesignModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ExtensionContainerDesignModule, imports: [[CommonModule, AdaptEmptyStateModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: ExtensionContainerDesignModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, AdaptEmptyStateModule],
                    declarations: [ExtensionContainerDesignComponent],
                    entryComponents: [ExtensionContainerDesignComponent]
                }]
        }] });
