// @bmc-ux/adapt-angular/esm2015/progress/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptProgressComponent } from './progress.component';
import { AdaptTooltipModule } from '../tooltip/index';
import * as i0 from "@angular/core";
export class AdaptProgressModule {
}
AdaptProgressModule.ɵfac = function AdaptProgressModule_Factory(t) { return new (t || AdaptProgressModule)(); };
AdaptProgressModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptProgressModule });
AdaptProgressModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, AdaptTooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptProgressModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, AdaptTooltipModule],
                exports: [AdaptProgressComponent],
                declarations: [AdaptProgressComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptProgressModule, { declarations: [AdaptProgressComponent], imports: [CommonModule, AdaptTooltipModule], exports: [AdaptProgressComponent] }); })();
export * from './progress.config';
export * from './progress.component';
