// @bmc-ux/adapt-angular/esm2015/rx/index.js
import { NgModule } from '@angular/core';
import { AdaptRxCheckboxModule } from './rx-checkbox/index';
import { AdaptRxCounterModule } from './rx-counter/index';
import { AdaptRxDatetimeModule } from './rx-datetime/index';
import { AdaptRxFormControlModule } from './form-control/index';
import { AdaptRxListBuilderModule } from './rx-list-builder/index';
import { AdaptRxListSelectorModule } from './rx-list-selector/index';
import { AdaptRxRadiobuttonModule } from './rx-radiobutton/index';
import { AdaptRxRatingModule } from './rx-rating/index';
import { AdaptRxSearchModule } from './rx-search/index';
import { AdaptRxSelectEditableModule } from './rx-select-editable/index';
import { AdaptRxSelectModule } from './rx-select/index';
import { AdaptRxSwitchModule } from './rx-switch/index';
import { AdaptRxTextareaModule } from './rx-textarea/index';
import { AdaptRxTextfieldModule } from './rx-textfield/index';
import { AdaptRxTimeSlotModule } from './rx-timeslot/index';
import { AdaptRxTypeaheadModule } from './rx-typeahead/index';
import { AdaptRxUploaderModule } from './rx-uploader/index';
import * as i0 from "@angular/core";
import * as i1 from "./rx-uploader/index";
export class AdaptRxFormsModule {
}
AdaptRxFormsModule.ɵfac = function AdaptRxFormsModule_Factory(t) { return new (t || AdaptRxFormsModule)(); };
AdaptRxFormsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxFormsModule });
AdaptRxFormsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [], imports: [[
            AdaptRxCheckboxModule,
            AdaptRxCounterModule,
            AdaptRxDatetimeModule,
            AdaptRxFormControlModule,
            AdaptRxListBuilderModule,
            AdaptRxListSelectorModule,
            AdaptRxRadiobuttonModule,
            AdaptRxRatingModule,
            AdaptRxSearchModule,
            AdaptRxSelectEditableModule,
            AdaptRxSelectModule,
            AdaptRxSwitchModule,
            AdaptRxTextareaModule,
            AdaptRxTextfieldModule,
            AdaptRxTimeSlotModule,
            AdaptRxTypeaheadModule,
            AdaptRxUploaderModule.forRoot()
        ], AdaptRxCheckboxModule,
        AdaptRxCounterModule,
        AdaptRxDatetimeModule,
        AdaptRxFormControlModule,
        AdaptRxListBuilderModule,
        AdaptRxListSelectorModule,
        AdaptRxRadiobuttonModule,
        AdaptRxRatingModule,
        AdaptRxSearchModule,
        AdaptRxSelectEditableModule,
        AdaptRxSelectModule,
        AdaptRxSwitchModule,
        AdaptRxTextareaModule,
        AdaptRxTextfieldModule,
        AdaptRxTimeSlotModule,
        AdaptRxTypeaheadModule,
        AdaptRxUploaderModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxFormsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptRxCheckboxModule,
                    AdaptRxCounterModule,
                    AdaptRxDatetimeModule,
                    AdaptRxFormControlModule,
                    AdaptRxListBuilderModule,
                    AdaptRxListSelectorModule,
                    AdaptRxRadiobuttonModule,
                    AdaptRxRatingModule,
                    AdaptRxSearchModule,
                    AdaptRxSelectEditableModule,
                    AdaptRxSelectModule,
                    AdaptRxSwitchModule,
                    AdaptRxTextareaModule,
                    AdaptRxTextfieldModule,
                    AdaptRxTimeSlotModule,
                    AdaptRxTypeaheadModule,
                    AdaptRxUploaderModule.forRoot()
                ],
                declarations: [],
                exports: [
                    AdaptRxCheckboxModule,
                    AdaptRxCounterModule,
                    AdaptRxDatetimeModule,
                    AdaptRxFormControlModule,
                    AdaptRxListBuilderModule,
                    AdaptRxListSelectorModule,
                    AdaptRxRadiobuttonModule,
                    AdaptRxRatingModule,
                    AdaptRxSearchModule,
                    AdaptRxSelectEditableModule,
                    AdaptRxSelectModule,
                    AdaptRxSwitchModule,
                    AdaptRxTextareaModule,
                    AdaptRxTextfieldModule,
                    AdaptRxTimeSlotModule,
                    AdaptRxTypeaheadModule,
                    AdaptRxUploaderModule
                ],
                providers: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxFormsModule, { imports: [AdaptRxCheckboxModule,
        AdaptRxCounterModule,
        AdaptRxDatetimeModule,
        AdaptRxFormControlModule,
        AdaptRxListBuilderModule,
        AdaptRxListSelectorModule,
        AdaptRxRadiobuttonModule,
        AdaptRxRatingModule,
        AdaptRxSearchModule,
        AdaptRxSelectEditableModule,
        AdaptRxSelectModule,
        AdaptRxSwitchModule,
        AdaptRxTextareaModule,
        AdaptRxTextfieldModule,
        AdaptRxTimeSlotModule,
        AdaptRxTypeaheadModule, i1.AdaptRxUploaderModule], exports: [AdaptRxCheckboxModule,
        AdaptRxCounterModule,
        AdaptRxDatetimeModule,
        AdaptRxFormControlModule,
        AdaptRxListBuilderModule,
        AdaptRxListSelectorModule,
        AdaptRxRadiobuttonModule,
        AdaptRxRatingModule,
        AdaptRxSearchModule,
        AdaptRxSelectEditableModule,
        AdaptRxSelectModule,
        AdaptRxSwitchModule,
        AdaptRxTextareaModule,
        AdaptRxTextfieldModule,
        AdaptRxTimeSlotModule,
        AdaptRxTypeaheadModule,
        AdaptRxUploaderModule] }); })();
export * from './form-control/index';
export * from './rx-checkbox/index';
export * from './rx-counter/index';
export * from './rx-datetime/index';
export * from './rx-list-builder/index';
export * from './rx-list-selector/index';
export * from './rx-radiobutton/index';
export * from './rx-rating/index';
export * from './rx-search/index';
export * from './rx-select-editable/index';
export * from './rx-select/index';
export * from './rx-switch/index';
export * from './rx-textarea/index';
export * from './rx-textfield/index';
export * from './rx-timeslot/index';
export * from './rx-typeahead/index';
export * from './rx-uploader/index';
