// @bmc-ux/adapt-angular/esm2015/custom-actions/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptCustomActionsComponent } from './adapt-custom-actions.component';
import { AdaptFocusMonitorDirectiveModule } from '../common/focus-monitor.directive';
import * as i0 from "@angular/core";
export class AdaptCustomActionsModule {
}
AdaptCustomActionsModule.ɵfac = function AdaptCustomActionsModule_Factory(t) { return new (t || AdaptCustomActionsModule)(); };
AdaptCustomActionsModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptCustomActionsModule });
AdaptCustomActionsModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptFocusMonitorDirectiveModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptCustomActionsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptFocusMonitorDirectiveModule
                ],
                exports: [
                    AdaptCustomActionsComponent
                ],
                declarations: [
                    AdaptCustomActionsComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptCustomActionsModule, { declarations: [AdaptCustomActionsComponent], imports: [CommonModule,
        AdaptFocusMonitorDirectiveModule], exports: [AdaptCustomActionsComponent] }); })();
export * from './adapt-custom-actions.component';
export * from './adapt-custom-actions.model';
