// @bmc-ux/adapt-angular/esm2015/dropdown/dropdown.model.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptDropdownConfig {
    constructor() {
        this.autoClose = true;
        this.placement = 'auto';
    }
}
AdaptDropdownConfig.ɵfac = function AdaptDropdownConfig_Factory(t) { return new (t || AdaptDropdownConfig)(); };
AdaptDropdownConfig.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptDropdownConfig, factory: AdaptDropdownConfig.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDropdownConfig, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
export var AdaptDropdownCloseReason;
(function (AdaptDropdownCloseReason) {
    AdaptDropdownCloseReason[AdaptDropdownCloseReason["Esc"] = 0] = "Esc";
    AdaptDropdownCloseReason[AdaptDropdownCloseReason["ClickInside"] = 1] = "ClickInside";
    AdaptDropdownCloseReason[AdaptDropdownCloseReason["ClickOutside"] = 2] = "ClickOutside";
    AdaptDropdownCloseReason[AdaptDropdownCloseReason["Destroy"] = 3] = "Destroy";
})(AdaptDropdownCloseReason || (AdaptDropdownCloseReason = {}));
