// @bmc-ux/adapt-angular/esm2015/dropdown/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { AdaptFocusHolderModule } from '../focus-holder/index';
import { AdaptCloseModule } from '../close/index';
import { AdaptDropdownAnchorDirective, AdaptDropdownDirective, AdaptDropdownMenuDirective, AdaptDropdownMenuTemplateComponent, AdaptDropdownMenuTemplateDirective, AdaptDropdownToggleDirective, AdaptDropdownWindowComponent } from './dropdown.component';
import { AdaptDropdownWindow2Component } from './dropdown-window2.component';
import * as i0 from "@angular/core";
const ADAPT_DROPDOWN_DIRECTIVES = [
    AdaptDropdownAnchorDirective,
    AdaptDropdownDirective,
    AdaptDropdownMenuDirective,
    AdaptDropdownMenuTemplateDirective,
    AdaptDropdownMenuTemplateComponent,
    AdaptDropdownToggleDirective,
    AdaptDropdownWindow2Component,
    AdaptDropdownWindowComponent
];
export class AdaptDropdownModule {
    /**
     * Importing with '.forRoot()' is no longer necessary, you can simply import the module.
     * Will be removed in a short while
     *
     * @deprecated 11.16.0
     */
    static forRoot() {
        return {
            ngModule: AdaptDropdownModule
        };
    }
}
AdaptDropdownModule.ɵfac = function AdaptDropdownModule_Factory(t) { return new (t || AdaptDropdownModule)(); };
AdaptDropdownModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptDropdownModule });
AdaptDropdownModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, OverlayModule, AdaptFocusHolderModule, AdaptCloseModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptDropdownModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, OverlayModule, AdaptFocusHolderModule, AdaptCloseModule],
                declarations: ADAPT_DROPDOWN_DIRECTIVES,
                exports: ADAPT_DROPDOWN_DIRECTIVES,
                entryComponents: [AdaptDropdownWindowComponent, AdaptDropdownWindow2Component]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptDropdownModule, { declarations: [AdaptDropdownAnchorDirective,
        AdaptDropdownDirective,
        AdaptDropdownMenuDirective,
        AdaptDropdownMenuTemplateDirective,
        AdaptDropdownMenuTemplateComponent,
        AdaptDropdownToggleDirective,
        AdaptDropdownWindow2Component,
        AdaptDropdownWindowComponent], imports: [CommonModule, OverlayModule, AdaptFocusHolderModule, AdaptCloseModule], exports: [AdaptDropdownAnchorDirective,
        AdaptDropdownDirective,
        AdaptDropdownMenuDirective,
        AdaptDropdownMenuTemplateDirective,
        AdaptDropdownMenuTemplateComponent,
        AdaptDropdownToggleDirective,
        AdaptDropdownWindow2Component,
        AdaptDropdownWindowComponent] }); })();
export * from './dropdown.component';
export * from './dropdown-window2.component';
export * from './dropdown.model';
