// @bmc-ux/adapt-angular/esm2015/common/uploader/pipes/file-type.pipe.js
import { NgModule, Pipe } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Directionality } from '@angular/cdk/bidi';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/bidi";
export class FileTypesPipe {
    constructor(_dir) {
        this._dir = _dir;
        this.isRtl = _dir.value === 'rtl';
    }
    transform(types, typesDivider) {
        const length = types.length;
        if (length === 1) {
            return types[0];
        }
        return !this.isRtl
            ? types.slice(0, length - 1).join(', ') + ` ${typesDivider} ` + types[length - 1]
            : types[length - 1] + ` ${typesDivider} ` + types.slice(0, length - 1).join(', ');
    }
}
FileTypesPipe.ɵfac = function FileTypesPipe_Factory(t) { return new (t || FileTypesPipe)(i0.ɵɵdirectiveInject(i1.Directionality, 16)); };
FileTypesPipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "fileTypes", type: FileTypesPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(FileTypesPipe, [{
        type: Pipe,
        args: [{
                name: 'fileTypes'
            }]
    }], function () { return [{ type: i1.Directionality }]; }, null); })();
export class AdaptFileTypesPipeModule {
}
AdaptFileTypesPipeModule.ɵfac = function AdaptFileTypesPipeModule_Factory(t) { return new (t || AdaptFileTypesPipeModule)(); };
AdaptFileTypesPipeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptFileTypesPipeModule });
AdaptFileTypesPipeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFileTypesPipeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [FileTypesPipe],
                exports: [FileTypesPipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptFileTypesPipeModule, { declarations: [FileTypesPipe], imports: [CommonModule], exports: [FileTypesPipe] }); })();
