// @bmc-ux/adapt-angular/esm2015/common/uploader/pipes/file-size.pipe.js
import { NgModule, Pipe } from '@angular/core';
import { CommonModule } from '@angular/common';
import { formatFileSize } from '../uploader';
import * as i0 from "@angular/core";
// TODO: move sizes to ???
export class FileSizePipe {
    transform(v) {
        return formatFileSize(v);
    }
}
FileSizePipe.ɵfac = function FileSizePipe_Factory(t) { return new (t || FileSizePipe)(); };
FileSizePipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "fileSize", type: FileSizePipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(FileSizePipe, [{
        type: Pipe,
        args: [{
                name: 'fileSize'
            }]
    }], null, null); })();
export class AdaptFileSizePipeModule {
}
AdaptFileSizePipeModule.ɵfac = function AdaptFileSizePipeModule_Factory(t) { return new (t || AdaptFileSizePipeModule)(); };
AdaptFileSizePipeModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptFileSizePipeModule });
AdaptFileSizePipeModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFileSizePipeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [FileSizePipe],
                exports: [FileSizePipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptFileSizePipeModule, { declarations: [FileSizePipe], imports: [CommonModule], exports: [FileSizePipe] }); })();
