// @bmc-ux/adapt-angular/esm2015/multi-card/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AdaptTabsControlsModule } from '../tabs-controls/index';
import { AdaptMultiCardSlideComponent, AdaptMultiCardComponent, AdaptMultiCardLineComponent } from './multi-card.component';
import { CardOrderByPipe } from './multi-card.pipe';
import * as i0 from "@angular/core";
export class AdaptMultiCardModule {
}
AdaptMultiCardModule.ɵfac = function AdaptMultiCardModule_Factory(t) { return new (t || AdaptMultiCardModule)(); };
AdaptMultiCardModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptMultiCardModule });
AdaptMultiCardModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptTabsControlsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMultiCardModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptTabsControlsModule
                ],
                exports: [AdaptMultiCardSlideComponent, AdaptMultiCardComponent, AdaptMultiCardLineComponent, CardOrderByPipe],
                declarations: [AdaptMultiCardSlideComponent, AdaptMultiCardComponent, AdaptMultiCardLineComponent, CardOrderByPipe]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptMultiCardModule, { declarations: [AdaptMultiCardSlideComponent, AdaptMultiCardComponent, AdaptMultiCardLineComponent, CardOrderByPipe], imports: [CommonModule,
        AdaptTabsControlsModule], exports: [AdaptMultiCardSlideComponent, AdaptMultiCardComponent, AdaptMultiCardLineComponent, CardOrderByPipe] }); })();
export * from './multi-card.component';
export * from './multi-card.pipe';
export * from './multi-card.model';
