// @bmc-ux/adapt-angular/esm2015/multi-card/multi-card.pipe.js
import { Pipe } from '@angular/core';
import { isUndefined } from '../common/utilities';
import * as i0 from "@angular/core";
export class CardOrderByPipe {
    // TODO: Refactor this if possible
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    transform(array, field) {
        if (!isUndefined(array)) {
            // TODO: Refactor this if possible
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
            array.sort((a, b) => {
                if (a[field] < b[field]) {
                    return -1;
                }
                else if (a[field] > b[field]) {
                    return 1;
                }
                else {
                    return 0;
                }
            });
            return array;
        }
    }
}
CardOrderByPipe.ɵfac = function CardOrderByPipe_Factory(t) { return new (t || CardOrderByPipe)(); };
CardOrderByPipe.ɵpipe = /*@__PURE__*/ i0.ɵɵdefinePipe({ name: "cardOrderBy", type: CardOrderByPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(CardOrderByPipe, [{
        type: Pipe,
        args: [{
                name: 'cardOrderBy'
            }]
    }], null, null); })();
