// @bmc-ux/adapt-angular/esm2015/rx/rx-list-builder/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptAlertModule } from '../../alert/index';
import { AdaptDragonInputHandlerModule } from '../../dragon-input-handler/index';
import { AdaptEmptyStateModule } from '../../empty-state/index';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxListItemModule } from '../rx-list-item/index';
import { AdaptRxListBuilderComponent } from './adapt-rx-list-builder.component';
import { AdaptCustomActionsModule } from '../../custom-actions/index';
import * as i0 from "@angular/core";
import * as i1 from "../form-control/rx-control-label/adapt-rx-control-label.component";
import * as i2 from "@angular/common";
import * as i3 from "../../dragon-input-handler/dragon-input-handler.directive";
import * as i4 from "../../custom-actions/adapt-custom-actions.component";
import * as i5 from "../rx-list-item/adapt-rx-list-item.component";
import * as i6 from "../../empty-state/empty-state.component";
import * as i7 from "../form-control/rx-feedback/adapt-rx-feedback.component";
export class AdaptRxListBuilderModule {
}
AdaptRxListBuilderModule.ɵfac = function AdaptRxListBuilderModule_Factory(t) { return new (t || AdaptRxListBuilderModule)(); };
AdaptRxListBuilderModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxListBuilderModule });
AdaptRxListBuilderModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            AdaptRxFormControlModule,
            AdaptEmptyStateModule,
            AdaptAlertModule,
            AdaptDragonInputHandlerModule,
            AdaptRxListItemModule,
            AdaptCustomActionsModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxListBuilderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    AdaptRxFormControlModule,
                    AdaptEmptyStateModule,
                    AdaptAlertModule,
                    AdaptDragonInputHandlerModule,
                    AdaptRxListItemModule,
                    AdaptCustomActionsModule,
                ],
                declarations: [AdaptRxListBuilderComponent],
                exports: [AdaptRxListBuilderComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxListBuilderModule, { declarations: [AdaptRxListBuilderComponent], imports: [CommonModule,
        AdaptRxFormControlModule,
        AdaptEmptyStateModule,
        AdaptAlertModule,
        AdaptDragonInputHandlerModule,
        AdaptRxListItemModule,
        AdaptCustomActionsModule], exports: [AdaptRxListBuilderComponent] }); })();
i0.ɵɵsetComponentScope(AdaptRxListBuilderComponent, [i1.AdaptRxControlLabelComponent, i2.NgIf, i3.AdaptDragonInputHandlerDirective, i2.NgStyle, i4.AdaptCustomActionsComponent, i2.NgForOf, i2.NgTemplateOutlet, i5.AdaptRxListItemComponent, i6.AdaptEmptyStateComponent, i7.AdaptRxFeedbackComponent], []);
export * from './adapt-rx-list-builder.component';
export * from './adapt-rx-list-builder.models';
