// @bmc-ux/adapt-angular/esm2015/rx/rx-select-editable/adapt-rx-select-editable-config.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptRxSelectEditableConfigService {
    constructor() {
        this.size = 'default';
        this.customOptionOnTop = false;
        this.menuMaxHeight = '200px';
        this.disabledStyleForReadonlyState = false;
        this.placement = 'auto';
        this.appendToBody = false;
        this.mobileView = {};
    }
}
AdaptRxSelectEditableConfigService.ɵfac = function AdaptRxSelectEditableConfigService_Factory(t) { return new (t || AdaptRxSelectEditableConfigService)(); };
AdaptRxSelectEditableConfigService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptRxSelectEditableConfigService, factory: AdaptRxSelectEditableConfigService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxSelectEditableConfigService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
