// @bmc-ux/adapt-angular/esm2015/rx/rx-select-editable/index.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptDropdownModule } from '../../dropdown/index';
import { AdaptRxSelectEditableComponent } from './adapt-rx-select-editable.component';
import * as i0 from "@angular/core";
export class AdaptRxSelectEditableModule {
}
AdaptRxSelectEditableModule.ɵfac = function AdaptRxSelectEditableModule_Factory(t) { return new (t || AdaptRxSelectEditableModule)(); };
AdaptRxSelectEditableModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxSelectEditableModule });
AdaptRxSelectEditableModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptDropdownModule,
            AdaptRxFormControlModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxSelectEditableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptDropdownModule,
                    AdaptRxFormControlModule,
                ],
                declarations: [
                    AdaptRxSelectEditableComponent
                ],
                exports: [
                    AdaptRxSelectEditableComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxSelectEditableModule, { declarations: [AdaptRxSelectEditableComponent], imports: [CommonModule,
        FormsModule,
        AdaptDropdownModule,
        AdaptRxFormControlModule], exports: [AdaptRxSelectEditableComponent] }); })();
export * from './adapt-rx-select-editable.component';
export * from './adapt-rx-select-editable-config.service';
