// @bmc-ux/adapt-angular/esm2015/rx/rx-textfield/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptBadgeModule } from '../../badge/index';
import { AdaptRxTextfieldComponent } from './adapt-rx-textfield.component';
import * as i0 from "@angular/core";
export class AdaptRxTextfieldModule {
}
AdaptRxTextfieldModule.ɵfac = function AdaptRxTextfieldModule_Factory(t) { return new (t || AdaptRxTextfieldModule)(); };
AdaptRxTextfieldModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxTextfieldModule });
AdaptRxTextfieldModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            AdaptRxFormControlModule,
            CommonModule,
            AdaptBadgeModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxTextfieldModule, [{
        type: NgModule,
        args: [{
                imports: [
                    AdaptRxFormControlModule,
                    CommonModule,
                    AdaptBadgeModule
                ],
                declarations: [
                    AdaptRxTextfieldComponent
                ],
                exports: [
                    AdaptRxTextfieldComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxTextfieldModule, { declarations: [AdaptRxTextfieldComponent], imports: [AdaptRxFormControlModule,
        CommonModule,
        AdaptBadgeModule], exports: [AdaptRxTextfieldComponent] }); })();
export * from './adapt-rx-textfield.component';
export * from './adapt-rx-textfield-config.service';
