// @bmc-ux/adapt-angular/esm2015/focus-tracker/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptFocusTrackerDirective } from './focus-tracker.directive';
import { AdaptFocusTrackerService } from './focus-tracker.service';
import * as i0 from "@angular/core";
export class AdaptFocusTrackerModule {
}
AdaptFocusTrackerModule.ɵfac = function AdaptFocusTrackerModule_Factory(t) { return new (t || AdaptFocusTrackerModule)(); };
AdaptFocusTrackerModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptFocusTrackerModule });
AdaptFocusTrackerModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [AdaptFocusTrackerService], imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFocusTrackerModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptFocusTrackerDirective],
                exports: [AdaptFocusTrackerDirective],
                providers: [AdaptFocusTrackerService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptFocusTrackerModule, { declarations: [AdaptFocusTrackerDirective], imports: [CommonModule], exports: [AdaptFocusTrackerDirective] }); })();
export * from './focus-tracker.directive';
export * from './focus-tracker.service';
