// @helix/platform/esm2015/view/components/record-grid/design/record-grid-design.constant.js
export const RX_RECORD_GRID_DESIGN = {
    keywords: {
        label: 'Keywords',
        children: [
            {
                label: 'Current user',
                icon: 'd-icon-dollar',
                expression: '$USER$'
            },
            {
                label: 'Current date',
                icon: 'd-icon-dollar',
                expression: '$DATE$'
            },
            {
                label: 'Current time',
                icon: 'd-icon-dollar',
                expression: '$TIME$'
            },
            {
                label: 'Current date and time',
                icon: 'd-icon-dollar',
                expression: '$TIMESTAMP$'
            },
            {
                label: 'Current week day',
                icon: 'd-icon-dollar',
                expression: '$WEEKDAY$'
            },
            {
                label: 'Current groups',
                icon: 'd-icon-dollar',
                expression: '$GROUPS$'
            },
            {
                label: 'Current group IDs',
                icon: 'd-icon-dollar',
                expression: '$GROUPIDS$'
            },
            {
                label: 'Current roles',
                icon: 'd-icon-dollar',
                expression: '$ROLES$'
            },
            {
                label: 'Current user locale',
                icon: 'd-icon-dollar',
                expression: '$LOCALE$'
            },
            {
                label: 'Current server URL',
                icon: 'd-icon-dollar',
                expression: '$SERVERURL$'
            },
            {
                label: 'NULL',
                icon: 'd-icon-dollar',
                expression: '$NULL$',
                hidden: true
            }
        ]
    }
};
