// @bmc-ux/adapt-angular/esm2015/rx/rx-select/adapt-rx-select-config.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptRxSelectConfigService {
    constructor() {
        this.size = 'default';
        this.placement = ['auto'];
        this.appendToBody = false;
        this.popupMaxHeight = 250;
        this.enableFilter = false;
        this.emptyOption = false;
        this.singleSelectStyle = 'line';
        this.inline = false;
        this.selectAllButton = false;
        this.deselectAllButton = false;
        this.closeOnSelect = true;
        this.focusFirst = true;
    }
}
AdaptRxSelectConfigService.ɵfac = function AdaptRxSelectConfigService_Factory(t) { return new (t || AdaptRxSelectConfigService)(); };
AdaptRxSelectConfigService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: AdaptRxSelectConfigService, factory: AdaptRxSelectConfigService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxSelectConfigService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
