// @bmc-ux/adapt-angular/esm2015/rx/rx-select/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ScrollingModule } from '@angular/cdk/scrolling';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptBusyModule } from '../../busy-loader/index';
import { AdaptDropdownModule } from '../../dropdown/index';
import { AdaptEmptyStateModule } from '../../empty-state/index';
import { AdaptRxSearchModule } from '../rx-search/index';
import { AdaptTooltipModule } from '../../tooltip/index';
import { AdaptCustomActionsModule } from '../../custom-actions/index';
import { AdaptRxSelectComponent } from './adapt-rx-select.component';
import { AdaptResizeModule } from '../../common/resize/index';
import { AdaptIsOptionSelectedPipe } from './adapt-rx-select.pipes';
import * as i0 from "@angular/core";
import * as i1 from "../form-control/rx-control-label/adapt-rx-control-label.component";
import * as i2 from "@angular/common";
import * as i3 from "../form-control/rx-feedback/adapt-rx-feedback.component";
import * as i4 from "../rx-search/adapt-rx-search.component";
import * as i5 from "@angular/forms";
import * as i6 from "../../custom-actions/adapt-custom-actions.component";
import * as i7 from "@angular/cdk/scrolling";
import * as i8 from "../../empty-state/empty-state.component";
import * as i9 from "../../dropdown/dropdown.component";
import * as i10 from "../../tooltip/tooltip.directive";
import * as i11 from "../../busy-loader/busy-inline-standalone.directive";
export class AdaptRxSelectModule {
}
AdaptRxSelectModule.ɵfac = function AdaptRxSelectModule_Factory(t) { return new (t || AdaptRxSelectModule)(); };
AdaptRxSelectModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxSelectModule });
AdaptRxSelectModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ providers: [
        AdaptIsOptionSelectedPipe
    ], imports: [[
            CommonModule,
            FormsModule,
            ScrollingModule,
            AdaptRxFormControlModule,
            AdaptBusyModule,
            AdaptDropdownModule,
            AdaptEmptyStateModule,
            AdaptRxSearchModule,
            AdaptTooltipModule,
            AdaptCustomActionsModule,
            AdaptResizeModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxSelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ScrollingModule,
                    AdaptRxFormControlModule,
                    AdaptBusyModule,
                    AdaptDropdownModule,
                    AdaptEmptyStateModule,
                    AdaptRxSearchModule,
                    AdaptTooltipModule,
                    AdaptCustomActionsModule,
                    AdaptResizeModule
                ],
                declarations: [
                    AdaptRxSelectComponent,
                    AdaptIsOptionSelectedPipe
                ],
                exports: [
                    AdaptRxSelectComponent
                ],
                providers: [
                    AdaptIsOptionSelectedPipe
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxSelectModule, { declarations: [AdaptRxSelectComponent,
        AdaptIsOptionSelectedPipe], imports: [CommonModule,
        FormsModule,
        ScrollingModule,
        AdaptRxFormControlModule,
        AdaptBusyModule,
        AdaptDropdownModule,
        AdaptEmptyStateModule,
        AdaptRxSearchModule,
        AdaptTooltipModule,
        AdaptCustomActionsModule,
        AdaptResizeModule], exports: [AdaptRxSelectComponent] }); })();
i0.ɵɵsetComponentScope(AdaptRxSelectComponent, [i1.AdaptRxControlLabelComponent, i2.NgIf, i3.AdaptRxFeedbackComponent, i2.NgClass, i2.NgTemplateOutlet, i2.NgStyle, i4.AdaptRxSearchComponent, i5.NgControlStatus, i5.NgModel, i6.AdaptCustomActionsComponent, i2.NgForOf, i7.CdkVirtualScrollViewport, i7.CdkFixedSizeVirtualScroll, i7.CdkVirtualForOf, i8.AdaptEmptyStateComponent, i9.AdaptDropdownDirective, i9.AdaptDropdownAnchorDirective, i10.AdaptTooltipDirective, i9.AdaptDropdownMenuTemplateDirective, i11.AdaptInlineStandaloneDirective], [AdaptIsOptionSelectedPipe, i2.SlicePipe]);
export * from './adapt-rx-select.component';
export * from './adapt-rx-select.types';
export * from './adapt-rx-select-config.service';
