// @bmc-ux/adapt-angular/esm2015/rx/rx-list-selector/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { AdaptEmptyStateModule } from '../../empty-state/index';
import { AdaptRxFormControlModule } from '../form-control/index';
import { AdaptRxListItemModule } from '../rx-list-item/index';
import { AdaptRxListSelectorComponent } from './adapt-rx-list-selector.component';
import { AdaptRxSelectModule } from '../rx-select/index';
import * as i0 from "@angular/core";
import * as i1 from "../form-control/rx-control-label/adapt-rx-control-label.component";
import * as i2 from "../rx-select/adapt-rx-select.component";
import * as i3 from "@angular/forms";
import * as i4 from "@angular/common";
import * as i5 from "../../empty-state/empty-state.component";
import * as i6 from "@angular/cdk/drag-drop";
import * as i7 from "../rx-list-item/adapt-rx-list-item.component";
import * as i8 from "../form-control/rx-feedback/adapt-rx-feedback.component";
export class AdaptRxListSelectorModule {
}
AdaptRxListSelectorModule.ɵfac = function AdaptRxListSelectorModule_Factory(t) { return new (t || AdaptRxListSelectorModule)(); };
AdaptRxListSelectorModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptRxListSelectorModule });
AdaptRxListSelectorModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            DragDropModule,
            AdaptRxFormControlModule,
            AdaptEmptyStateModule,
            AdaptRxListItemModule,
            AdaptRxSelectModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptRxListSelectorModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    DragDropModule,
                    AdaptRxFormControlModule,
                    AdaptEmptyStateModule,
                    AdaptRxListItemModule,
                    AdaptRxSelectModule
                ],
                declarations: [AdaptRxListSelectorComponent],
                exports: [AdaptRxListSelectorComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptRxListSelectorModule, { declarations: [AdaptRxListSelectorComponent], imports: [CommonModule,
        FormsModule,
        DragDropModule,
        AdaptRxFormControlModule,
        AdaptEmptyStateModule,
        AdaptRxListItemModule,
        AdaptRxSelectModule], exports: [AdaptRxListSelectorComponent] }); })();
i0.ɵɵsetComponentScope(AdaptRxListSelectorComponent, [i1.AdaptRxControlLabelComponent, i2.AdaptRxSelectComponent, i3.NgControlStatus, i3.NgModel, i4.NgIf, i5.AdaptEmptyStateComponent, i6.CdkDropList, i4.NgForOf, i6.CdkDrag, i6.CdkDragPlaceholder, i7.AdaptRxListItemComponent, i8.AdaptRxFeedbackComponent], []);
export * from './adapt-rx-list-selector.component';
export * from './adapt-rx-list-selector.models';
