// @bmc-ux/adapt-angular/esm2015/anchor-navigator/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptAnchorNavigatorComponent } from './anchor-navigator.component';
import { AdaptAnchorLinkComponent } from './anchor-link/anchor-link.component';
import * as i0 from "@angular/core";
export class AdaptAnchorNavigatorModule {
}
AdaptAnchorNavigatorModule.ɵfac = function AdaptAnchorNavigatorModule_Factory(t) { return new (t || AdaptAnchorNavigatorModule)(); };
AdaptAnchorNavigatorModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAnchorNavigatorModule });
AdaptAnchorNavigatorModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAnchorNavigatorModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptAnchorNavigatorComponent, AdaptAnchorLinkComponent],
                exports: [AdaptAnchorNavigatorComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAnchorNavigatorModule, { declarations: [AdaptAnchorNavigatorComponent, AdaptAnchorLinkComponent], imports: [CommonModule], exports: [AdaptAnchorNavigatorComponent] }); })();
export * from './anchor-navigator.component';
export * from './anchor-navigator.model';
