// @bmc-ux/adapt-angular/esm2015/tooltip/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptTooltipDirective } from './tooltip.directive';
import { AdaptTooltipComponent } from './tooltip.component';
import { OverlayModule } from '@angular/cdk/overlay';
import * as i0 from "@angular/core";
export class AdaptTooltipModule {
}
AdaptTooltipModule.ɵfac = function AdaptTooltipModule_Factory(t) { return new (t || AdaptTooltipModule)(); };
AdaptTooltipModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTooltipModule });
AdaptTooltipModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule, OverlayModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTooltipModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, OverlayModule],
                entryComponents: [AdaptTooltipComponent],
                declarations: [AdaptTooltipDirective, AdaptTooltipComponent],
                exports: [AdaptTooltipDirective, AdaptTooltipComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTooltipModule, { declarations: [AdaptTooltipDirective, AdaptTooltipComponent], imports: [CommonModule, OverlayModule], exports: [AdaptTooltipDirective, AdaptTooltipComponent] }); })();
export * from './tooltip.component';
export * from './tooltip.directive';
export * from './tooltip.service';
