// @helix/platform/esm2015/shared/components/form-controls/assignment-expression-list-form-control/assignment-expression-list-form-control.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AssignmentExpressionListFormControlComponent } from './assignment-expression-list-form-control.component';
import { TranslateModule } from '@ngx-translate/core';
import { ExpressionFormControlModule } from '../expression-form-control';
import { ReactiveFormsModule } from '@angular/forms';
import { TextFormControlModule } from '../text-form-control';
import { SelectFormControlModule } from '../select-form-control';
import * as i0 from "@angular/core";
export class AssignmentExpressionListFormControlModule {
}
AssignmentExpressionListFormControlModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AssignmentExpressionListFormControlModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AssignmentExpressionListFormControlModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AssignmentExpressionListFormControlModule, declarations: [AssignmentExpressionListFormControlComponent], imports: [CommonModule,
        TranslateModule,
        ExpressionFormControlModule,
        ReactiveFormsModule,
        TextFormControlModule,
        SelectFormControlModule], exports: [AssignmentExpressionListFormControlComponent] });
AssignmentExpressionListFormControlModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AssignmentExpressionListFormControlModule, imports: [[
            CommonModule,
            TranslateModule,
            ExpressionFormControlModule,
            ReactiveFormsModule,
            TextFormControlModule,
            SelectFormControlModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: AssignmentExpressionListFormControlModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [AssignmentExpressionListFormControlComponent],
                    entryComponents: [AssignmentExpressionListFormControlComponent],
                    imports: [
                        CommonModule,
                        TranslateModule,
                        ExpressionFormControlModule,
                        ReactiveFormsModule,
                        TextFormControlModule,
                        SelectFormControlModule
                    ],
                    exports: [AssignmentExpressionListFormControlComponent]
                }]
        }] });
