// @helix/platform/esm2015/ui-kit/breadcrumb-bar/breadcrumb-bar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RxBreadcrumbBarComponent } from './breadcrumb-bar.component';
import * as i0 from "@angular/core";
export class RxBreadcrumbBarModule {
}
RxBreadcrumbBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBreadcrumbBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxBreadcrumbBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBreadcrumbBarModule, declarations: [RxBreadcrumbBarComponent], imports: [CommonModule], exports: [RxBreadcrumbBarComponent] });
RxBreadcrumbBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBreadcrumbBarModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxBreadcrumbBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [RxBreadcrumbBarComponent],
                    declarations: [RxBreadcrumbBarComponent]
                }]
        }] });
