// @bmc-ux/obsolete/esm2015/lib/list-selector/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { ai_dots } from '@bmc-ux/dpl-iconfont';
import { AdaptListComponent } from './list-selector.component';
import { AdaptEmptyStateModule, AdaptIconModule, AdaptTooltipModule, AdaptPopoverModule, AdaptRxLabelModule, AdaptIconConfig } from '@bmc-ux/adapt-angular';
import { AdaptSelectModule } from '../select/index';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptListSelectorModule {
    constructor(iconService) {
        iconService.registerIcons([ai_dots]);
    }
    static forRoot() {
        return {
            ngModule: AdaptListSelectorModule
        };
    }
}
AdaptListSelectorModule.ɵfac = function AdaptListSelectorModule_Factory(t) { return new (t || AdaptListSelectorModule)(i0.ɵɵinject(i1.AdaptIconConfig)); };
AdaptListSelectorModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptListSelectorModule });
AdaptListSelectorModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            DragDropModule,
            AdaptSelectModule,
            AdaptEmptyStateModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptListSelectorModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    DragDropModule,
                    AdaptSelectModule,
                    AdaptEmptyStateModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule,
                    AdaptRxLabelModule
                ],
                declarations: [AdaptListComponent],
                exports: [AdaptListComponent]
            }]
    }], function () { return [{ type: i1.AdaptIconConfig }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptListSelectorModule, { declarations: [AdaptListComponent], imports: [CommonModule,
        FormsModule,
        DragDropModule,
        AdaptSelectModule,
        AdaptEmptyStateModule,
        AdaptIconModule,
        AdaptTooltipModule,
        AdaptPopoverModule,
        AdaptRxLabelModule], exports: [AdaptListComponent] }); })();
export * from './list-selector.component';
export * from './list-selector.model';
