// @bmc-ux/adapt-angular/esm2015/common/common.enums.js
export var ColorType;
(function (ColorType) {
    ColorType["RGB"] = "RGB";
    ColorType["HEX"] = "HEX";
})(ColorType || (ColorType = {}));
export var DismissReasons;
(function (DismissReasons) {
    DismissReasons[DismissReasons["CLOSE_BTN"] = 0] = "CLOSE_BTN";
    DismissReasons[DismissReasons["BACKDROP_CLICK"] = 1] = "BACKDROP_CLICK";
    DismissReasons[DismissReasons["ESC"] = 2] = "ESC";
    DismissReasons[DismissReasons["CLOSE_ALL"] = 3] = "CLOSE_ALL";
})(DismissReasons || (DismissReasons = {}));
export var AdaptDropdownEditableSize;
(function (AdaptDropdownEditableSize) {
    AdaptDropdownEditableSize["small"] = "btn-sm";
    AdaptDropdownEditableSize["medium"] = "medium";
    AdaptDropdownEditableSize["large"] = "btn-lg";
})(AdaptDropdownEditableSize || (AdaptDropdownEditableSize = {}));
export var MessageStatus;
(function (MessageStatus) {
    MessageStatus["Primary"] = "primary";
    MessageStatus["Info"] = "info";
    MessageStatus["Success"] = "success";
    MessageStatus["Warning"] = "warning";
    MessageStatus["Danger"] = "danger";
})(MessageStatus || (MessageStatus = {}));
export var CounterOperations;
(function (CounterOperations) {
    CounterOperations[CounterOperations["INCREASE"] = 0] = "INCREASE";
    CounterOperations[CounterOperations["DECREASE"] = 1] = "DECREASE";
})(CounterOperations || (CounterOperations = {}));
export var AdaptNumberType;
(function (AdaptNumberType) {
    AdaptNumberType[AdaptNumberType["Integer"] = 0] = "Integer";
    AdaptNumberType[AdaptNumberType["Float"] = 1] = "Float";
    AdaptNumberType[AdaptNumberType["Scientific"] = 2] = "Scientific";
})(AdaptNumberType || (AdaptNumberType = {}));
export var DateTimeKey;
(function (DateTimeKey) {
    DateTimeKey["Year"] = "year";
    DateTimeKey["Month"] = "month";
    DateTimeKey["Date"] = "date";
    DateTimeKey["Hours"] = "hours";
    DateTimeKey["Minutes"] = "minutes";
    DateTimeKey["Seconds"] = "seconds";
    DateTimeKey["Milliseconds"] = "milliseconds";
    DateTimeKey["TimeZone"] = "timeZone";
    DateTimeKey["Offset"] = "offset";
})(DateTimeKey || (DateTimeKey = {}));
