// @bmc-ux/adapt-angular/esm2015/common/focus-monitor.directive.js
import { Input, Directive, ElementRef, NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export class AdaptFocusMonitorDirective {
    constructor(_elementRef) {
        this._elementRef = _elementRef;
    }
    get disabled() {
        return this._elementRef.nativeElement.disabled;
    }
    // TODO: maybe remove this implementation
    getLabel() {
        return this._elementRef.nativeElement.innerText;
    }
    get nativeElement() {
        return this._elementRef.nativeElement;
    }
}
AdaptFocusMonitorDirective.ɵfac = function AdaptFocusMonitorDirective_Factory(t) { return new (t || AdaptFocusMonitorDirective)(i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptFocusMonitorDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: AdaptFocusMonitorDirective, selectors: [["", "adaptFocusMonitor", ""]], inputs: { optionId: ["adaptFocusMonitor", "optionId"] } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFocusMonitorDirective, [{
        type: Directive,
        args: [{
                selector: '[adaptFocusMonitor]'
            }]
    }], function () { return [{ type: i0.ElementRef }]; }, { optionId: [{
            type: Input,
            args: ['adaptFocusMonitor']
        }] }); })();
export class AdaptFocusMonitorDirectiveModule {
}
AdaptFocusMonitorDirectiveModule.ɵfac = function AdaptFocusMonitorDirectiveModule_Factory(t) { return new (t || AdaptFocusMonitorDirectiveModule)(); };
AdaptFocusMonitorDirectiveModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptFocusMonitorDirectiveModule });
AdaptFocusMonitorDirectiveModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptFocusMonitorDirectiveModule, [{
        type: NgModule,
        args: [{
                declarations: [AdaptFocusMonitorDirective],
                exports: [AdaptFocusMonitorDirective],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptFocusMonitorDirectiveModule, { declarations: [AdaptFocusMonitorDirective], exports: [AdaptFocusMonitorDirective] }); })();
