// @helix/com-bmc-arsys-rx-innovationstudio/esm2015/lib/components/bundle-action-wizard/select-approval-configurations-wizard-step/select-approval-configurations-wizard-step.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TranslateModule } from '@ngx-translate/core';
import { RecordGridModule } from '@helix/platform/view/components';
import { SelectApprovalConfigurationsWizardStepComponent } from './select-approval-configurations-wizard-step.component';
import * as i0 from "@angular/core";
export class SelectApprovalConfigurationsWizardStepModule {
}
/** @nocollapse */ SelectApprovalConfigurationsWizardStepModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectApprovalConfigurationsWizardStepModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ SelectApprovalConfigurationsWizardStepModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectApprovalConfigurationsWizardStepModule, declarations: [SelectApprovalConfigurationsWizardStepComponent], imports: [CommonModule, RecordGridModule, TranslateModule], exports: [SelectApprovalConfigurationsWizardStepComponent] });
/** @nocollapse */ SelectApprovalConfigurationsWizardStepModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectApprovalConfigurationsWizardStepModule, imports: [[CommonModule, RecordGridModule, TranslateModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: SelectApprovalConfigurationsWizardStepModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RecordGridModule, TranslateModule],
                    declarations: [SelectApprovalConfigurationsWizardStepComponent],
                    exports: [SelectApprovalConfigurationsWizardStepComponent]
                }]
        }] });
