// @bmc-ux/obsolete/esm2015/lib/textfield/index.js
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { AdaptTextFieldComponent } from './textfield.component';
import { NgModule } from '@angular/core';
import { AdaptTextFieldIconsModule } from './textfield.icons';
import { AdaptDragonInputHandlerModule, AdaptIconModule, AdaptPopoverModule, AdaptRxLabelModule, AdaptTooltipModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
import * as i1 from "@bmc-ux/adapt-angular";
export class AdaptTextFieldModule {
}
AdaptTextFieldModule.ɵfac = function AdaptTextFieldModule_Factory(t) { return new (t || AdaptTextFieldModule)(); };
AdaptTextFieldModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTextFieldModule });
AdaptTextFieldModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule,
            AdaptTextFieldIconsModule,
            AdaptIconModule,
            AdaptTooltipModule,
            AdaptPopoverModule.forRoot(),
            AdaptDragonInputHandlerModule,
            AdaptRxLabelModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTextFieldModule, [{
        type: NgModule,
        args: [{
                imports: [
                    FormsModule,
                    CommonModule,
                    AdaptTextFieldIconsModule,
                    AdaptIconModule,
                    AdaptTooltipModule,
                    AdaptPopoverModule.forRoot(),
                    AdaptDragonInputHandlerModule,
                    AdaptRxLabelModule
                ],
                exports: [AdaptTextFieldComponent],
                declarations: [AdaptTextFieldComponent],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTextFieldModule, { declarations: [AdaptTextFieldComponent], imports: [FormsModule,
        CommonModule,
        AdaptTextFieldIconsModule,
        AdaptIconModule,
        AdaptTooltipModule, i1.AdaptPopoverModule, AdaptDragonInputHandlerModule,
        AdaptRxLabelModule], exports: [AdaptTextFieldComponent] }); })();
export * from './textfield.component';
export * from './textfield.icons';
export * from './textfield.interface';
