// @bmc-ux/adapt-angular/esm2015/agreement/agreement-admin-content/agreement-admin-content.component.js
import { Component, ElementRef, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../../common/radar/adapt-radar';
import * as i0 from "@angular/core";
import * as i1 from "../../common/radar/adapt-radar";
import * as i2 from "@angular/common";
import * as i3 from "../../button/button.component";
import * as i4 from "../../collapse/collapse.directive";
import * as i5 from "../../rx/rx-switch/adapt-rx-switch.component";
import * as i6 from "@angular/forms";
import * as i7 from "../../alert/alert.component";
function AdaptAgreementAdminContentComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    const _r8 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "h4");
    i0.ɵɵtext(3, "Enable performance and quality improvement data collection:");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementStart(5, "div");
    i0.ɵɵtext(6, "\n      ");
    i0.ɵɵelementStart(7, "adapt-rx-switch", 9);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_7_listener($event) { i0.ɵɵrestoreView(_r8); const ctx_r7 = i0.ɵɵnextContext(); return ctx_r7.state.organizationPerformance = $event; })("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_7_listener() { i0.ɵɵrestoreView(_r8); const ctx_r9 = i0.ɵɵnextContext(); return ctx_r9.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n    ");
    i0.ɵɵelementStart(10, "div");
    i0.ɵɵtext(11, "\n      ");
    i0.ɵɵelementStart(12, "adapt-rx-switch", 10);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_12_listener($event) { i0.ɵɵrestoreView(_r8); const ctx_r10 = i0.ɵɵnextContext(); return ctx_r10.state.accountPerformance = $event; })("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_7_Template_adapt_rx_switch_ngModelChange_12_listener() { i0.ɵɵrestoreView(_r8); const ctx_r11 = i0.ɵɵnextContext(); return ctx_r11.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(7);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r0.state.organizationPerformance)("label", "For my organization");
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("disabled", !ctx_r0.state.organizationPerformance)("ngModel", ctx_r0.state.accountPerformance)("label", "For my account");
} }
function AdaptAgreementAdminContentComponent_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    const _r13 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "adapt-rx-switch", 9);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_9_Template_adapt_rx_switch_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r13); const ctx_r12 = i0.ɵɵnextContext(); return ctx_r12.state.accountPerformance = $event; })("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_9_Template_adapt_rx_switch_ngModelChange_2_listener() { i0.ɵɵrestoreView(_r13); const ctx_r14 = i0.ɵɵnextContext(); return ctx_r14.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r1.state.accountPerformance)("label", "Enable data collection for performance and product improvement");
} }
const _c0 = "Turning this feature on will enable data collection for all users of this product. Users may change these options at any time in their settings.";
const _c1 = function () { return { content: _c0, type: "inline", variant: "info" }; };
function AdaptAgreementAdminContentComponent_adapt_alert_13_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "adapt-alert", 11);
} if (rf & 2) {
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("config", i0.ɵɵpureFunction0(2, _c1));
} }
const _c2 = "Turning this feature on will enable data collection for all users of BMC\u2019s products and services. Users may change these options at any time in their settings.";
const _c3 = function () { return { content: _c2, type: "inline", variant: "info" }; };
function AdaptAgreementAdminContentComponent_adapt_alert_15_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "adapt-alert", 11);
} if (rf & 2) {
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("config", i0.ɵɵpureFunction0(2, _c3));
} }
function AdaptAgreementAdminContentComponent_span_29_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1, "this product.");
    i0.ɵɵelementEnd();
} }
function AdaptAgreementAdminContentComponent_span_31_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "span");
    i0.ɵɵtext(1, "BMC\u2019s products and services.");
    i0.ɵɵelementEnd();
} }
function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r20 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "h4");
    i0.ɵɵtext(3, "Enable marketing and outreach data collection:");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n    ");
    i0.ɵɵelementStart(5, "div");
    i0.ɵɵtext(6, "\n      ");
    i0.ɵɵelementStart(7, "adapt-rx-switch", 9);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_2_Template_adapt_rx_switch_ngModelChange_7_listener($event) { i0.ɵɵrestoreView(_r20); const ctx_r19 = i0.ɵɵnextContext(2); return ctx_r19.state.organizationMarketing = $event; })("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_2_Template_adapt_rx_switch_ngModelChange_7_listener() { i0.ɵɵrestoreView(_r20); const ctx_r21 = i0.ɵɵnextContext(2); return ctx_r21.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(8, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(9, "\n    ");
    i0.ɵɵelementStart(10, "div");
    i0.ɵɵtext(11, "\n      ");
    i0.ɵɵelementStart(12, "adapt-rx-switch", 10);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_2_Template_adapt_rx_switch_ngModelChange_12_listener($event) { i0.ɵɵrestoreView(_r20); const ctx_r22 = i0.ɵɵnextContext(2); return ctx_r22.state.accountMarketing = $event; })("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_2_Template_adapt_rx_switch_ngModelChange_12_listener() { i0.ɵɵrestoreView(_r20); const ctx_r23 = i0.ɵɵnextContext(2); return ctx_r23.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(14, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r15 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(7);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r15.state.organizationMarketing)("label", "For my organization");
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("disabled", !ctx_r15.state.organizationMarketing)("ngModel", ctx_r15.state.accountMarketing)("label", "For my account");
} }
function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    const _r25 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementStart(2, "adapt-rx-switch", 9);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_4_Template_adapt_rx_switch_ngModelChange_2_listener($event) { i0.ɵɵrestoreView(_r25); const ctx_r24 = i0.ɵɵnextContext(2); return ctx_r24.state.accountMarketing = $event; })("ngModelChange", function AdaptAgreementAdminContentComponent_ng_container_72_ng_container_4_Template_adapt_rx_switch_ngModelChange_2_listener() { i0.ɵɵrestoreView(_r25); const ctx_r26 = i0.ɵɵnextContext(2); return ctx_r26.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(3, "\n  ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r16 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r16.state.accountMarketing)("label", "Enable data collection for marketing and outreach");
} }
function AdaptAgreementAdminContentComponent_ng_container_72_adapt_alert_8_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "adapt-alert", 11);
} if (rf & 2) {
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("config", i0.ɵɵpureFunction0(2, _c1));
} }
function AdaptAgreementAdminContentComponent_ng_container_72_adapt_alert_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "adapt-alert", 11);
} if (rf & 2) {
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("config", i0.ɵɵpureFunction0(2, _c3));
} }
const _c4 = function (a0, a1) { return { "d-icon-angle_down": a0, "d-icon-angle_up": a1 }; };
const _c5 = function (a0, a1) { return { "mb-5": a0, "agreement__text-hidden": a1 }; };
function AdaptAgreementAdminContentComponent_ng_container_72_Template(rf, ctx) { if (rf & 1) {
    const _r28 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n  ");
    i0.ɵɵtemplate(2, AdaptAgreementAdminContentComponent_ng_container_72_ng_container_2_Template, 15, 7, "ng-container", 1);
    i0.ɵɵtext(3, "\n\n  ");
    i0.ɵɵtemplate(4, AdaptAgreementAdminContentComponent_ng_container_72_ng_container_4_Template, 4, 3, "ng-container", 1);
    i0.ɵɵtext(5, "\n\n  ");
    i0.ɵɵelementStart(6, "div", 2);
    i0.ɵɵtext(7, "\n\n    ");
    i0.ɵɵtemplate(8, AdaptAgreementAdminContentComponent_ng_container_72_adapt_alert_8_Template, 1, 3, "adapt-alert", 3);
    i0.ɵɵtext(9, "\n\n      ");
    i0.ɵɵtemplate(10, AdaptAgreementAdminContentComponent_ng_container_72_adapt_alert_10_Template, 1, 3, "adapt-alert", 3);
    i0.ɵɵtext(11, "\n\n    ");
    i0.ɵɵelementStart(12, "button", 4);
    i0.ɵɵlistener("click", function AdaptAgreementAdminContentComponent_ng_container_72_Template_button_click_12_listener() { i0.ɵɵrestoreView(_r28); const ctx_r27 = i0.ɵɵnextContext(); return ctx_r27.isCollapsedMarketingContent = !ctx_r27.isCollapsedMarketingContent; });
    i0.ɵɵtext(13, "\n      ");
    i0.ɵɵelementStart(14, "span");
    i0.ɵɵtext(15, "Learn more about what we collect");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(16, "\n      ");
    i0.ɵɵelement(17, "span", 5);
    i0.ɵɵtext(18, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(19, "\n\n    ");
    i0.ɵɵelementStart(20, "div", 6);
    i0.ɵɵtext(21, "\n      ");
    i0.ɵɵelementStart(22, "p", 12);
    i0.ɵɵtext(23, "Enabling this feature allows BMC to use the data we anonymously collect to\n        better\n        inform your company of relevant products and services, as well as to reach out to individuals at your\n        organization\n        for feedback and other requests. Enabling this option will improve the relevancy of BMC\u2019s communications, as\n        well\n        as provide your company with more targeted offerings to meet their needs.");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(24, "\n      ");
    i0.ɵɵelementStart(25, "ul", 8);
    i0.ɵɵtext(26, "\n        ");
    i0.ɵɵelementStart(27, "li");
    i0.ɵɵtext(28, "To better inform sales teams about the way your company leverages BMC products");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(29, "\n        ");
    i0.ɵɵelementStart(30, "li");
    i0.ɵɵtext(31, "To understand industry trends to better support your company");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(32, "\n        ");
    i0.ɵɵelementStart(33, "li");
    i0.ɵɵtext(34, "To identify communications topics relating to the industry");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(35, "\n        ");
    i0.ɵɵelementStart(36, "li");
    i0.ɵɵtext(37, "To better target messaging around new products and features");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(38, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(39, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(40, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(41, "\n\n  ");
    i0.ɵɵelement(42, "hr", 0);
    i0.ɵɵtext(43, "\n");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r6 = i0.ɵɵnextContext();
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r6.showOrganizationSettings);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", !ctx_r6.showOrganizationSettings);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", !ctx_r6.multiProductUsage);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r6.multiProductUsage);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true);
    i0.ɵɵadvance(5);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(9, _c4, ctx_r6.isCollapsedMarketingContent, !ctx_r6.isCollapsedMarketingContent));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("collapse", ctx_r6.isCollapsedMarketingContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(12, _c5, !ctx_r6.isCollapsedMarketingContent, ctx_r6.isCollapsedMarketingContent));
} }
export class AdaptAgreementAdminContentComponent extends AdaptRadarAngularGenericDirective {
    constructor(_adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptAgreementAdminContent);
        this.isCollapsedPerformanceContent = true;
        this.isCollapsedMarketingContent = true;
        this.multiProductUsage = false;
        this.stateChange = new EventEmitter();
    }
    handleStateChange() {
        if (this.showOrganizationSettings) {
            if (!this.state.organizationPerformance) {
                this.state.organizationMarketing = false;
                this.state.accountPerformance = false;
            }
            if (!this.state.organizationMarketing) {
                this.state.accountMarketing = false;
            }
        }
        else {
            if (!this.state.accountPerformance) {
                this.state.accountMarketing = false;
            }
        }
        this.stateChange.emit(this.state);
    }
}
AdaptAgreementAdminContentComponent.ɵfac = function AdaptAgreementAdminContentComponent_Factory(t) { return new (t || AdaptAgreementAdminContentComponent)(i0.ɵɵdirectiveInject(i1.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptAgreementAdminContentComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptAgreementAdminContentComponent, selectors: [["adapt-agreement-admin-content"]], inputs: { multiProductUsage: "multiProductUsage", state: "state", showOrganizationSettings: "showOrganizationSettings" }, outputs: { stateChange: "stateChange" }, features: [i0.ɵɵInheritDefinitionFeature], decls: 77, vars: 18, consts: [[1, "agreement__separator"], [4, "ngIf"], [1, "agreement__content", "mt-3"], [3, "adaptRadarDisableEventSending", "config", 4, "ngIf"], ["type", "button", "adapt-button", "", "btn-type", "tertiary", 1, "p-0", "mt-3", "mb-3", "agreement__collapse-btn", 3, "adaptRadarDisableEventSending", "click"], [1, "p-start-1", 3, "ngClass"], [3, "collapse", "adaptRadarDisableEventSending", "ngClass"], [1, "my-4"], [1, "m-0"], [1, "agreement__switcher", 3, "adaptRadarDisableEventSending", "ngModel", "label", "ngModelChange"], [1, "agreement__switcher", 3, "adaptRadarDisableEventSending", "disabled", "ngModel", "label", "ngModelChange"], [3, "adaptRadarDisableEventSending", "config"], [1, "agreement__content-text"]], template: function AdaptAgreementAdminContentComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "p");
        i0.ɵɵtext(1, "Change data collection settings");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(2, "\n\n");
        i0.ɵɵelement(3, "hr", 0);
        i0.ɵɵtext(4, "\n");
        i0.ɵɵelementContainerStart(5);
        i0.ɵɵtext(6, "\n  ");
        i0.ɵɵtemplate(7, AdaptAgreementAdminContentComponent_ng_container_7_Template, 15, 7, "ng-container", 1);
        i0.ɵɵtext(8, "\n\n  ");
        i0.ɵɵtemplate(9, AdaptAgreementAdminContentComponent_ng_container_9_Template, 4, 3, "ng-container", 1);
        i0.ɵɵtext(10, "\n\n  ");
        i0.ɵɵelementStart(11, "div", 2);
        i0.ɵɵtext(12, "\n\n    ");
        i0.ɵɵtemplate(13, AdaptAgreementAdminContentComponent_adapt_alert_13_Template, 1, 3, "adapt-alert", 3);
        i0.ɵɵtext(14, "\n\n      ");
        i0.ɵɵtemplate(15, AdaptAgreementAdminContentComponent_adapt_alert_15_Template, 1, 3, "adapt-alert", 3);
        i0.ɵɵtext(16, "\n\n    ");
        i0.ɵɵelementStart(17, "button", 4);
        i0.ɵɵlistener("click", function AdaptAgreementAdminContentComponent_Template_button_click_17_listener() { return ctx.isCollapsedPerformanceContent = !ctx.isCollapsedPerformanceContent; });
        i0.ɵɵtext(18, "\n      ");
        i0.ɵɵelementStart(19, "span");
        i0.ɵɵtext(20, "Learn more about what we collect");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(21, "\n      ");
        i0.ɵɵelement(22, "span", 5);
        i0.ɵɵtext(23, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(24, "\n\n    ");
        i0.ɵɵelementStart(25, "div", 6);
        i0.ɵɵtext(26, "\n      ");
        i0.ɵɵelementStart(27, "p", 7);
        i0.ɵɵtext(28, "Enabling this feature allows BMC to collect anonymous information about individual usage of\n        ");
        i0.ɵɵtemplate(29, AdaptAgreementAdminContentComponent_span_29_Template, 2, 0, "span", 1);
        i0.ɵɵtext(30, "\n        ");
        i0.ɵɵtemplate(31, AdaptAgreementAdminContentComponent_span_31_Template, 2, 0, "span", 1);
        i0.ɵɵtext(32, "\n        The data collected will be used to identify defects, optimize product performance, and aid in the\n        development of new product features. Enabling this option will ensure that your user\u2019s needs are taken\n        into account when planning for new product improvements.");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(33, "\n      ");
        i0.ɵɵelementStart(34, "ul", 8);
        i0.ɵɵtext(35, "\n        ");
        i0.ɵɵelementStart(36, "li");
        i0.ɵɵtext(37, "Time to complete certain tasks");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(38, "\n        ");
        i0.ɵɵelementStart(39, "li");
        i0.ɵɵtext(40, "Counts of clicks or taps on certain elements, like buttons");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(41, "\n        ");
        i0.ɵɵelementStart(42, "li");
        i0.ɵɵtext(43, "The number of errors experienced while using a BMC product, and their type");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(44, "\n        ");
        i0.ɵɵelementStart(45, "li");
        i0.ɵɵtext(46, "Time spent not actively using a BMC product");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(47, "\n        ");
        i0.ɵɵelementStart(48, "li");
        i0.ɵɵtext(49, "Responses to ratings and surveys provided by BMC products");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(50, "\n        ");
        i0.ɵɵelementStart(51, "li");
        i0.ɵɵtext(52, "How often a product is signed in to, or signed out of");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(53, "\n        ");
        i0.ɵɵelementStart(54, "li");
        i0.ɵɵtext(55, "Any movement from one BMC product to another");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(56, "\n        ");
        i0.ɵɵelementStart(57, "li");
        i0.ɵɵtext(58, "Information that allows us to identify a product, or instance of a product");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(59, "\n        ");
        i0.ɵɵelementStart(60, "li");
        i0.ɵɵtext(61, "Diagnostic information needed to technical troubleshooting");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(62, "\n        ");
        i0.ɵɵelementStart(63, "li");
        i0.ɵɵtext(64, "General information about your computer or web browser that could be used to identify technical problems\n        ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(65, "\n      ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(66, "\n    ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(67, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(68, "\n\n  ");
        i0.ɵɵelement(69, "hr", 0);
        i0.ɵɵtext(70, "\n");
        i0.ɵɵelementContainerEnd();
        i0.ɵɵtext(71, "\n\n");
        i0.ɵɵtemplate(72, AdaptAgreementAdminContentComponent_ng_container_72_Template, 44, 15, "ng-container", 1);
        i0.ɵɵtext(73, "\n\n");
        i0.ɵɵelementStart(74, "p");
        i0.ɵɵtext(75, "By enabling this feature you consent to BMC\u2019s collection, processing, and use of user analytics data\n  described, but not limited to, the examples outlined above. You can change your selection in your application\u2019s\n  settings at any time.");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(76, "\n");
    } if (rf & 2) {
        i0.ɵɵadvance(7);
        i0.ɵɵproperty("ngIf", ctx.showOrganizationSettings);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", !ctx.showOrganizationSettings);
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngIf", !ctx.multiProductUsage);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.multiProductUsage);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("adaptRadarDisableEventSending", true);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(12, _c4, ctx.isCollapsedPerformanceContent, !ctx.isCollapsedPerformanceContent));
        i0.ɵɵadvance(3);
        i0.ɵɵproperty("collapse", ctx.isCollapsedPerformanceContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(15, _c5, !ctx.isCollapsedPerformanceContent, ctx.isCollapsedPerformanceContent));
        i0.ɵɵadvance(4);
        i0.ɵɵproperty("ngIf", !ctx.multiProductUsage);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.multiProductUsage);
        i0.ɵɵadvance(41);
        i0.ɵɵproperty("ngIf", ctx.state.organizationPerformance || ctx.state.accountPerformance);
    } }, directives: [i2.NgIf, i3.AdaptButtonComponent, i2.NgClass, i4.AdaptCollapseDirective, i5.AdaptRxSwitchComponent, i6.NgControlStatus, i6.NgModel, i7.AdaptAlertComponent], styles: [".adapt-agreement.adapt-agreement_no-shadow{box-shadow:none}.agreement__main-title{margin:0 0 1rem}.agreement__separator{margin:1rem 0}.agreement__switcher{font-size:.9375rem}.agreement__content{margin-left:4.125em}.adapt-agreement__content_empty-state{font-size:.9375rem;text-align:center;width:100%}.agreement__content-text{margin:1rem 0}.agreement__badge .badge{min-width:70px}.agreement__collapse-btn{color:var(--text-primary)}adapt-alert .text-info{display:flex}.agreement__text-hidden .collapse-block{height:0;overflow:hidden;visibility:hidden}"], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAgreementAdminContentComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-agreement-admin-content',
                styleUrls: ['../agreement.scss'],
                templateUrl: './agreement-admin-content.component.html',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { multiProductUsage: [{
            type: Input
        }], state: [{
            type: Input
        }], showOrganizationSettings: [{
            type: Input
        }], stateChange: [{
            type: Output
        }] }); })();
