// @bmc-ux/obsolete/esm2015/lib/adapt-obsolete.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { AdaptCounterModule } from './counter/index';
import { AdaptDatetime2Module } from './datetime2/index';
import { AdaptDropdownEditableModule } from './dropdown-editable/index';
import { AdaptListBuilderModule } from './list-builder/index';
import { AdaptListSelectorModule } from './list-selector/index';
import { AdaptRatingModule } from './rating/index';
import { AdaptSearchModule } from './search/index';
import { AdaptTimeSlotModule } from './time-slot/index';
import { AdaptSwitcherModule } from './switcher/index';
import { AdaptCheckbox2Module } from './checkbox2/index';
import { AdaptSelectModule } from './select/index';
import { AdaptSelectInlineModule } from './select-inline/index';
import { AdaptRadioButtonModule } from './radiobutton/index';
import { AdaptUploadModule } from './uploader/index';
import { AdaptTextFieldModule } from './textfield/index';
import { AdaptTextFieldFormBuilderModule } from './textfield-formbuilder/index';
import { AdaptMentionModule } from './mention/index';
import { AdaptTypeaheadModule } from './typeahead/index';
import * as i0 from "@angular/core";
import * as i1 from "./list-builder/index";
import * as i2 from "./uploader/index";
import * as i3 from "./typeahead/index";
const ADAPT_MODULES = [
    AdaptCounterModule,
    AdaptDatetime2Module,
    AdaptDropdownEditableModule,
    AdaptListBuilderModule,
    AdaptListSelectorModule,
    AdaptRatingModule,
    AdaptSearchModule,
    AdaptTimeSlotModule,
    AdaptCheckbox2Module,
    AdaptSwitcherModule,
    AdaptSelectModule,
    AdaptSelectInlineModule,
    AdaptRadioButtonModule,
    AdaptUploadModule,
    AdaptTextFieldModule,
    AdaptTextFieldFormBuilderModule,
    AdaptMentionModule,
    AdaptTypeaheadModule
];
export class AdaptObsoleteRootModule {
}
AdaptObsoleteRootModule.ɵfac = function AdaptObsoleteRootModule_Factory(t) { return new (t || AdaptObsoleteRootModule)(); };
AdaptObsoleteRootModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptObsoleteRootModule });
AdaptObsoleteRootModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            FormsModule,
            AdaptCounterModule,
            AdaptDatetime2Module,
            AdaptDropdownEditableModule,
            AdaptListBuilderModule.forRoot(),
            AdaptListSelectorModule,
            AdaptRatingModule,
            AdaptSearchModule,
            AdaptTimeSlotModule,
            AdaptCheckbox2Module,
            AdaptSwitcherModule,
            AdaptSelectModule,
            AdaptSelectInlineModule,
            AdaptRadioButtonModule,
            AdaptUploadModule.forRoot(),
            AdaptTextFieldModule,
            AdaptTextFieldFormBuilderModule,
            AdaptMentionModule,
            AdaptTypeaheadModule.forRoot(),
        ], AdaptCounterModule,
        AdaptDatetime2Module,
        AdaptDropdownEditableModule,
        AdaptListBuilderModule,
        AdaptListSelectorModule,
        AdaptRatingModule,
        AdaptSearchModule,
        AdaptTimeSlotModule,
        AdaptCheckbox2Module,
        AdaptSwitcherModule,
        AdaptSelectModule,
        AdaptSelectInlineModule,
        AdaptRadioButtonModule,
        AdaptUploadModule,
        AdaptTextFieldModule,
        AdaptTextFieldFormBuilderModule,
        AdaptMentionModule,
        AdaptTypeaheadModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptObsoleteRootModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    AdaptCounterModule,
                    AdaptDatetime2Module,
                    AdaptDropdownEditableModule,
                    AdaptListBuilderModule.forRoot(),
                    AdaptListSelectorModule,
                    AdaptRatingModule,
                    AdaptSearchModule,
                    AdaptTimeSlotModule,
                    AdaptCheckbox2Module,
                    AdaptSwitcherModule,
                    AdaptSelectModule,
                    AdaptSelectInlineModule,
                    AdaptRadioButtonModule,
                    AdaptUploadModule.forRoot(),
                    AdaptTextFieldModule,
                    AdaptTextFieldFormBuilderModule,
                    AdaptMentionModule,
                    AdaptTypeaheadModule.forRoot(),
                ],
                exports: ADAPT_MODULES
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptObsoleteRootModule, { imports: [CommonModule,
        FormsModule,
        AdaptCounterModule,
        AdaptDatetime2Module,
        AdaptDropdownEditableModule, i1.AdaptListBuilderModule, AdaptListSelectorModule,
        AdaptRatingModule,
        AdaptSearchModule,
        AdaptTimeSlotModule,
        AdaptCheckbox2Module,
        AdaptSwitcherModule,
        AdaptSelectModule,
        AdaptSelectInlineModule,
        AdaptRadioButtonModule, i2.AdaptUploadModule, AdaptTextFieldModule,
        AdaptTextFieldFormBuilderModule,
        AdaptMentionModule, i3.AdaptTypeaheadModule], exports: [AdaptCounterModule,
        AdaptDatetime2Module,
        AdaptDropdownEditableModule,
        AdaptListBuilderModule,
        AdaptListSelectorModule,
        AdaptRatingModule,
        AdaptSearchModule,
        AdaptTimeSlotModule,
        AdaptCheckbox2Module,
        AdaptSwitcherModule,
        AdaptSelectModule,
        AdaptSelectInlineModule,
        AdaptRadioButtonModule,
        AdaptUploadModule,
        AdaptTextFieldModule,
        AdaptTextFieldFormBuilderModule,
        AdaptMentionModule,
        AdaptTypeaheadModule] }); })();
export class AdaptObsoleteModule {
    static forRoot() {
        return { ngModule: AdaptObsoleteRootModule };
    }
}
AdaptObsoleteModule.ɵfac = function AdaptObsoleteModule_Factory(t) { return new (t || AdaptObsoleteModule)(); };
AdaptObsoleteModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptObsoleteModule });
AdaptObsoleteModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [ADAPT_MODULES, AdaptCounterModule,
        AdaptDatetime2Module,
        AdaptDropdownEditableModule,
        AdaptListBuilderModule,
        AdaptListSelectorModule,
        AdaptRatingModule,
        AdaptSearchModule,
        AdaptTimeSlotModule,
        AdaptCheckbox2Module,
        AdaptSwitcherModule,
        AdaptSelectModule,
        AdaptSelectInlineModule,
        AdaptRadioButtonModule,
        AdaptUploadModule,
        AdaptTextFieldModule,
        AdaptTextFieldFormBuilderModule,
        AdaptMentionModule,
        AdaptTypeaheadModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptObsoleteModule, [{
        type: NgModule,
        args: [{
                imports: ADAPT_MODULES,
                exports: ADAPT_MODULES
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptObsoleteModule, { imports: [AdaptCounterModule,
        AdaptDatetime2Module,
        AdaptDropdownEditableModule,
        AdaptListBuilderModule,
        AdaptListSelectorModule,
        AdaptRatingModule,
        AdaptSearchModule,
        AdaptTimeSlotModule,
        AdaptCheckbox2Module,
        AdaptSwitcherModule,
        AdaptSelectModule,
        AdaptSelectInlineModule,
        AdaptRadioButtonModule,
        AdaptUploadModule,
        AdaptTextFieldModule,
        AdaptTextFieldFormBuilderModule,
        AdaptMentionModule,
        AdaptTypeaheadModule], exports: [AdaptCounterModule,
        AdaptDatetime2Module,
        AdaptDropdownEditableModule,
        AdaptListBuilderModule,
        AdaptListSelectorModule,
        AdaptRatingModule,
        AdaptSearchModule,
        AdaptTimeSlotModule,
        AdaptCheckbox2Module,
        AdaptSwitcherModule,
        AdaptSelectModule,
        AdaptSelectInlineModule,
        AdaptRadioButtonModule,
        AdaptUploadModule,
        AdaptTextFieldModule,
        AdaptTextFieldFormBuilderModule,
        AdaptMentionModule,
        AdaptTypeaheadModule] }); })();
