// @bmc-ux/adapt-angular/esm2015/tag/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptTagComponent } from './tag.component';
import * as i0 from "@angular/core";
export class AdaptTagModule {
}
AdaptTagModule.ɵfac = function AdaptTagModule_Factory(t) { return new (t || AdaptTagModule)(); };
AdaptTagModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptTagModule });
AdaptTagModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptTagModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [AdaptTagComponent],
                exports: [AdaptTagComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptTagModule, { declarations: [AdaptTagComponent], imports: [CommonModule], exports: [AdaptTagComponent] }); })();
export * from './tag.component';
