// @bmc-ux/adapt-angular/esm2015/agreement/agreement-user-modal/agreement-user-modal.component.js
import { Component, ElementRef, EventEmitter, Input, Output, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { AdaptModalService } from '../../modal/modal.service';
import { AdaptMessageService } from '../../toast/toast-message.service';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../../common/radar/adapt-radar';
import { AdaptTranslateService } from '../../common/i18n';
import { formatString } from '../../common/utilities';
import * as i0 from "@angular/core";
import * as i1 from "../../modal/modal.service";
import * as i2 from "../../toast/toast-message.service";
import * as i3 from "../../common/radar/adapt-radar";
import * as i4 from "../../common/i18n";
import * as i5 from "../../rx/rx-switch/adapt-rx-switch.component";
import * as i6 from "@angular/forms";
import * as i7 from "../../button/button.component";
import * as i8 from "@angular/common";
import * as i9 from "../../collapse/collapse.directive";
const _c0 = ["mTemplate"];
const _c1 = function (a0, a1) { return { "d-icon-angle_down": a0, "d-icon-angle_up": a1 }; };
const _c2 = function (a0, a1) { return { "mb-3": a0, "agreement__text-hidden": a1 }; };
function AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n        ");
    i0.ɵɵelement(2, "hr", 10);
    i0.ɵɵtext(3, "\n\n        ");
    i0.ɵɵelementStart(4, "adapt-rx-switch", 3);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_ng_container_2_Template_adapt_rx_switch_ngModelChange_4_listener($event) { i0.ɵɵrestoreView(_r6); const ctx_r5 = i0.ɵɵnextContext(3); return ctx_r5.state.accountMarketing = $event; })("ngModelChange", function AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_ng_container_2_Template_adapt_rx_switch_ngModelChange_4_listener() { i0.ɵɵrestoreView(_r6); const ctx_r7 = i0.ɵɵnextContext(3); return ctx_r7.handleStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n\n        ");
    i0.ɵɵelementStart(6, "div", 4);
    i0.ɵɵtext(7, "\n          ");
    i0.ɵɵelementStart(8, "p", 5);
    i0.ɵɵtext(9);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(10, "\n\n          ");
    i0.ɵɵelementStart(11, "button", 6);
    i0.ɵɵlistener("click", function AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_ng_container_2_Template_button_click_11_listener() { i0.ɵɵrestoreView(_r6); const ctx_r8 = i0.ɵɵnextContext(3); return ctx_r8.isCollapsedMarketingContent = !ctx_r8.isCollapsedMarketingContent; });
    i0.ɵɵtext(12, "\n            ");
    i0.ɵɵelementStart(13, "span");
    i0.ɵɵtext(14);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(15, "\n            ");
    i0.ɵɵelement(16, "span", 7);
    i0.ɵɵtext(17, "\n          ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n\n          ");
    i0.ɵɵelementStart(19, "ul", 8);
    i0.ɵɵtext(20, "\n            ");
    i0.ɵɵelementStart(21, "li");
    i0.ɵɵtext(22);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(23, "\n            ");
    i0.ɵɵelementStart(24, "li");
    i0.ɵɵtext(25);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(26, "\n            ");
    i0.ɵɵelementStart(27, "li");
    i0.ɵɵtext(28);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(29, "\n            ");
    i0.ɵɵelementStart(30, "li");
    i0.ɵɵtext(31);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(32, "\n          ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(33, "\n        ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(34, "\n      ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r4 = i0.ɵɵnextContext(3);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r4.state.accountMarketing)("label", ctx_r4.texts.marketingLabel);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r4.texts.agreementText);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.texts.loadMoreLabel2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(14, _c1, ctx_r4.isCollapsedMarketingContent, !ctx_r4.isCollapsedMarketingContent));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("collapse", ctx_r4.isCollapsedMarketingContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(17, _c2, !ctx_r4.isCollapsedMarketingContent, ctx_r4.isCollapsedMarketingContent));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.texts.learnMoreListItem11);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.texts.learnMoreListItem12);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.texts.learnMoreListItem13);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r4.texts.learnMoreListItem14);
} }
function AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementContainerStart(0);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵtemplate(2, AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_ng_container_2_Template, 35, 20, "ng-container", 9);
    i0.ɵɵtext(3, "\n    ");
    i0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r3 = i0.ɵɵnextContext(2);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngIf", ctx_r3.state.accountPerformance);
} }
function AdaptAgreementUserModalComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    const _r10 = i0.ɵɵgetCurrentView();
    i0.ɵɵtext(0, "\n  ");
    i0.ɵɵelementStart(1, "div", 1);
    i0.ɵɵtext(2, "\n    ");
    i0.ɵɵelementStart(3, "h5");
    i0.ɵɵtext(4);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(5, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(6, "\n  ");
    i0.ɵɵelementStart(7, "div", 2);
    i0.ɵɵtext(8, "\n    ");
    i0.ɵɵelementStart(9, "p");
    i0.ɵɵtext(10);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(11, "\n\n    ");
    i0.ɵɵelementStart(12, "adapt-rx-switch", 3);
    i0.ɵɵlistener("ngModelChange", function AdaptAgreementUserModalComponent_ng_template_0_Template_adapt_rx_switch_ngModelChange_12_listener($event) { i0.ɵɵrestoreView(_r10); const ctx_r9 = i0.ɵɵnextContext(); return ctx_r9.state.accountPerformance = $event; })("ngModelChange", function AdaptAgreementUserModalComponent_ng_template_0_Template_adapt_rx_switch_ngModelChange_12_listener() { i0.ɵɵrestoreView(_r10); const ctx_r11 = i0.ɵɵnextContext(); return ctx_r11.handleQualityStateChange(); });
    i0.ɵɵelementEnd();
    i0.ɵɵtext(13, "\n\n    ");
    i0.ɵɵelementStart(14, "div", 4);
    i0.ɵɵtext(15, "\n      ");
    i0.ɵɵelementStart(16, "p", 5);
    i0.ɵɵtext(17);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(18, "\n\n      ");
    i0.ɵɵelementStart(19, "button", 6);
    i0.ɵɵlistener("click", function AdaptAgreementUserModalComponent_ng_template_0_Template_button_click_19_listener() { i0.ɵɵrestoreView(_r10); const ctx_r12 = i0.ɵɵnextContext(); return ctx_r12.isCollapsedPerformanceContent = !ctx_r12.isCollapsedPerformanceContent; });
    i0.ɵɵtext(20, "\n        ");
    i0.ɵɵelementStart(21, "span");
    i0.ɵɵtext(22);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(23, "\n        ");
    i0.ɵɵelement(24, "span", 7);
    i0.ɵɵtext(25, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(26, "\n\n      ");
    i0.ɵɵelementStart(27, "ul", 8);
    i0.ɵɵtext(28, "\n        ");
    i0.ɵɵelementStart(29, "li");
    i0.ɵɵtext(30);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(31, "\n        ");
    i0.ɵɵelementStart(32, "li");
    i0.ɵɵtext(33);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(34, "\n        ");
    i0.ɵɵelementStart(35, "li");
    i0.ɵɵtext(36);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(37, "\n        ");
    i0.ɵɵelementStart(38, "li");
    i0.ɵɵtext(39);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(40, "\n        ");
    i0.ɵɵelementStart(41, "li");
    i0.ɵɵtext(42);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(43, "\n        ");
    i0.ɵɵelementStart(44, "li");
    i0.ɵɵtext(45);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(46, "\n        ");
    i0.ɵɵelementStart(47, "li");
    i0.ɵɵtext(48);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(49, "\n        ");
    i0.ɵɵelementStart(50, "li");
    i0.ɵɵtext(51);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(52, "\n        ");
    i0.ɵɵelementStart(53, "li");
    i0.ɵɵtext(54);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(55, "\n        ");
    i0.ɵɵelementStart(56, "li");
    i0.ɵɵtext(57);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(58, "\n      ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(59, "\n    ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(60, "\n\n    ");
    i0.ɵɵtemplate(61, AdaptAgreementUserModalComponent_ng_template_0_ng_container_61_Template, 4, 1, "ng-container", 9);
    i0.ɵɵtext(62, "\n    ");
    i0.ɵɵelement(63, "hr", 10);
    i0.ɵɵtext(64, "\n    ");
    i0.ɵɵelementStart(65, "p");
    i0.ɵɵtext(66);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(67, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(68, "\n  ");
    i0.ɵɵelementStart(69, "div", 11);
    i0.ɵɵtext(70, "\n    ");
    i0.ɵɵelementStart(71, "button", 12);
    i0.ɵɵlistener("click", function AdaptAgreementUserModalComponent_ng_template_0_Template_button_click_71_listener() { const restoredCtx = i0.ɵɵrestoreView(_r10); const close_r2 = restoredCtx.close; const ctx_r13 = i0.ɵɵnextContext(); return ctx_r13.handleSubmit(close_r2); });
    i0.ɵɵtext(72);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(73, "\n  ");
    i0.ɵɵelementEnd();
    i0.ɵɵtext(74, "\n");
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵadvance(3);
    i0.ɵɵclassMap("modal-title text-default");
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r1.texts.mainTitle);
    i0.ɵɵadvance(6);
    i0.ɵɵtextInterpolate(ctx_r1.texts.mainText);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("ngModel", ctx_r1.state.accountPerformance)("label", ctx_r1.texts.performanceLabel);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r1.learnMoreNoteText);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("adaptRadarDisableEventSending", true);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.loadMoreLabel);
    i0.ɵɵadvance(2);
    i0.ɵɵproperty("ngClass", i0.ɵɵpureFunction2(27, _c1, ctx_r1.isCollapsedPerformanceContent, !ctx_r1.isCollapsedPerformanceContent));
    i0.ɵɵadvance(3);
    i0.ɵɵproperty("collapse", ctx_r1.isCollapsedPerformanceContent)("adaptRadarDisableEventSending", true)("ngClass", i0.ɵɵpureFunction2(30, _c2, !ctx_r1.isCollapsedPerformanceContent, ctx_r1.isCollapsedPerformanceContent));
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem1);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem2);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem3);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem4);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem5);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem6);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem7);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem8);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem9);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r1.texts.learnMoreListItem10);
    i0.ɵɵadvance(4);
    i0.ɵɵproperty("ngIf", ctx_r1.showMarketingSection);
    i0.ɵɵadvance(5);
    i0.ɵɵtextInterpolate(ctx_r1.texts.agreementNote);
    i0.ɵɵadvance(6);
    i0.ɵɵtextInterpolate(ctx_r1.getSubmitButtonText());
} }
export class AdaptAgreementUserModalComponent extends AdaptRadarAngularGenericDirective {
    constructor(_modalService, _messageService, _adaptRadarService, _elem, _ts) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptAgreementUserModal);
        this._modalService = _modalService;
        this._messageService = _messageService;
        this._adaptRadarService = _adaptRadarService;
        this._ts = _ts;
        this.isCollapsedPerformanceContent = true;
        this.isCollapsedMarketingContent = true;
        this.texts = {};
        this.multiProductUsage = false;
        this.showMarketingSection = true;
        this.state = {
            accountPerformance: false,
            accountMarketing: false
        };
        this.openModal = false;
        this.showNotificationAfterSubmit = true;
        this.stateChange = new EventEmitter();
        this.submitButtonClick = new EventEmitter();
        this.modalOpen = new EventEmitter();
        this._currentLanguage = _ts.getCurrentLanguage();
        this._setDefaultTexts(this._currentLanguage);
    }
    ngOnChanges(changes) {
        var _a;
        if ((_a = changes['openModal']) === null || _a === void 0 ? void 0 : _a.currentValue) {
            this.open();
        }
    }
    handleQualityStateChange() {
        if (!this.state.accountPerformance) {
            this.state.accountMarketing = false;
        }
        this.handleStateChange();
    }
    handleStateChange() {
        this.stateChange.emit(this.state);
    }
    open() {
        const promise = this._modalService
            .open({
            title: null,
            content: this.mTemplate,
            blockKeyboard: true,
            testID: this.testID,
            beforeDismiss: () => false,
        });
        this.modalOpen.emit({ promise });
    }
    handleSubmit(close) {
        this.submitButtonClick.emit(this.state);
        close();
        if (this.showNotificationAfterSubmit) {
            this._messageService.add({
                severity: 'info',
                summary: this.texts.notificationSummaryText,
                detail: this.texts.notificationDetailsText,
                link: this.texts.notificationLinkText,
                linkHandler: this.notificationLinkHandler,
                life: 10000
            });
        }
    }
    getSubmitButtonText() {
        if (this.state.accountPerformance && this.state.accountMarketing) {
            return this.texts.submitButton1;
        }
        if (!this.state.accountPerformance) {
            return this.texts.submitButton2;
        }
        if (!this.state.accountMarketing) {
            return this.texts.submitButton3;
        }
    }
    get learnMoreNoteText() {
        return formatString(this._currentLanguage['adapt.agreement.common.learnMoreNote'], this.multiProductUsage ? this._currentLanguage['adapt.agreement.common.learnMoreNoteMultiProductUsage'] : this._currentLanguage['adapt.agreement.common.learnMoreNoteNotMultiProductUsage']);
    }
    _setDefaultTexts(translation) {
        this.texts.performanceLabel = translation['adapt.agreement.common.performanceLabel'];
        this.texts.marketingLabel = translation['adapt.agreement.common.marketingLabel'];
        this.texts.agreementNote = translation['adapt.agreement.common.marketingLabel'];
        this.texts.learnMoreListItem1 = translation['adapt.agreement.common.learnMoreListItem1'];
        this.texts.learnMoreListItem2 = translation['adapt.agreement.common.learnMoreListItem2'];
        this.texts.learnMoreListItem3 = translation['adapt.agreement.common.learnMoreListItem3'];
        this.texts.learnMoreListItem4 = translation['adapt.agreement.common.learnMoreListItem4'];
        this.texts.learnMoreListItem5 = translation['adapt.agreement.common.learnMoreListItem5'];
        this.texts.learnMoreListItem6 = translation['adapt.agreement.common.learnMoreListItem6'];
        this.texts.learnMoreListItem7 = translation['adapt.agreement.common.learnMoreListItem7'];
        this.texts.learnMoreListItem8 = translation['adapt.agreement.common.learnMoreListItem8'];
        this.texts.learnMoreListItem9 = translation['adapt.agreement.common.learnMoreListItem9'];
        this.texts.learnMoreListItem10 = translation['adapt.agreement.common.learnMoreListItem10'];
        this.texts.learnMoreListItem11 = translation['adapt.agreement.common.learnMoreListItem11'];
        this.texts.learnMoreListItem12 = translation['adapt.agreement.common.learnMoreListItem12'];
        this.texts.learnMoreListItem13 = translation['adapt.agreement.common.learnMoreListItem13'];
        this.texts.learnMoreListItem14 = translation['adapt.agreement.common.learnMoreListItem14'];
        this.texts.mainTitle = translation['adapt.agreement.userModal.mainTitle'];
        this.texts.mainText = translation['adapt.agreement.userModal.mainText'];
        this.texts.loadMoreLabel = translation['adapt.agreement.userModal.loadMoreLabel'];
        this.texts.loadMoreLabel2 = translation['adapt.agreement.userModal.loadMoreLabel2'];
        this.texts.agreementText = translation['adapt.agreement.userModal.agreementText'];
        this.texts.notificationSummaryText = translation['adapt.agreement.userModal.notificationSummaryText'];
        this.texts.notificationDetailsText = translation['adapt.agreement.userModal.notificationDetailsText'];
        this.texts.notificationLinkText = translation['adapt.agreement.userModal.notificationLinkText'];
        this.texts.submitButton1 = translation['adapt.agreement.common.submitButton1'];
        this.texts.submitButton2 = translation['adapt.agreement.common.submitButton2'];
        this.texts.submitButton3 = translation['adapt.agreement.common.submitButton3'];
    }
}
AdaptAgreementUserModalComponent.ɵfac = function AdaptAgreementUserModalComponent_Factory(t) { return new (t || AdaptAgreementUserModalComponent)(i0.ɵɵdirectiveInject(i1.AdaptModalService), i0.ɵɵdirectiveInject(i2.AdaptMessageService), i0.ɵɵdirectiveInject(i3.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i4.AdaptTranslateService)); };
AdaptAgreementUserModalComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptAgreementUserModalComponent, selectors: [["adapt-agreement-user-modal"]], viewQuery: function AdaptAgreementUserModalComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.mTemplate = _t.first);
    } }, inputs: { multiProductUsage: "multiProductUsage", showMarketingSection: "showMarketingSection", state: "state", openModal: "openModal", testID: "testID", showNotificationAfterSubmit: "showNotificationAfterSubmit", notificationLinkHandler: "notificationLinkHandler" }, outputs: { stateChange: "stateChange", submitButtonClick: "submitButtonClick", modalOpen: "modalOpen" }, features: [i0.ɵɵInheritDefinitionFeature, i0.ɵɵNgOnChangesFeature], decls: 3, vars: 0, consts: [["mTemplate", ""], [1, "modal-header"], [1, "modal-body"], [1, "agreement__switcher", 3, "adaptRadarDisableEventSending", "ngModel", "label", "ngModelChange"], [1, "agreement__content"], [1, "agreement__content-text"], ["type", "button", "adapt-button", "", "btn-type", "tertiary", 1, "p-0", "mt-3", "mb-3", "agreement__collapse-btn", 3, "adaptRadarDisableEventSending", "click"], [1, "p-start-1", 3, "ngClass"], [1, "m-0", 3, "collapse", "adaptRadarDisableEventSending", "ngClass"], [4, "ngIf"], [1, "agreement__separator"], [1, "modal-footer"], ["type", "button", 1, "btn", "btn-primary", 3, "click"]], template: function AdaptAgreementUserModalComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, AdaptAgreementUserModalComponent_ng_template_0_Template, 75, 33, "ng-template", null, 0, i0.ɵɵtemplateRefExtractor);
        i0.ɵɵtext(2, "\n");
    } }, directives: [i5.AdaptRxSwitchComponent, i6.NgControlStatus, i6.NgModel, i7.AdaptButtonComponent, i8.NgClass, i9.AdaptCollapseDirective, i8.NgIf], styles: [".adapt-agreement.adapt-agreement_no-shadow{box-shadow:none}.agreement__main-title{margin:0 0 1rem}.agreement__separator{margin:1rem 0}.agreement__switcher{font-size:.9375rem}.agreement__content{margin-left:4.125em}.adapt-agreement__content_empty-state{font-size:.9375rem;text-align:center;width:100%}.agreement__content-text{margin:1rem 0}.agreement__badge .badge{min-width:70px}.agreement__collapse-btn{color:var(--text-primary)}adapt-alert .text-info{display:flex}.agreement__text-hidden .collapse-block{height:0;overflow:hidden;visibility:hidden}"], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAgreementUserModalComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-agreement-user-modal',
                styleUrls: ['../agreement.scss'],
                templateUrl: './agreement-user-modal.component.html',
                encapsulation: ViewEncapsulation.None
            }]
    }], function () { return [{ type: i1.AdaptModalService }, { type: i2.AdaptMessageService }, { type: i3.AdaptRadarAngularService }, { type: i0.ElementRef }, { type: i4.AdaptTranslateService }]; }, { multiProductUsage: [{
            type: Input
        }], showMarketingSection: [{
            type: Input
        }], state: [{
            type: Input
        }], openModal: [{
            type: Input
        }], testID: [{
            type: Input
        }], showNotificationAfterSubmit: [{
            type: Input
        }], notificationLinkHandler: [{
            type: Input
        }], stateChange: [{
            type: Output
        }], submitButtonClick: [{
            type: Output
        }], modalOpen: [{
            type: Output
        }], mTemplate: [{
            type: ViewChild,
            args: ['mTemplate']
        }] }); })();
