// @helix/platform/esm2015/process/server-actions/record/get-record/components/get-record-input-map-inspector-widget/get-record-input-map-inspector-widget.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { ExpressionFormControlModule, RxDefinitionPickerModule } from '@helix/platform/shared/components';
import { RxGetRecordInputMapInspectorWidgetComponent } from './get-record-input-map-inspector-widget.component';
import * as i0 from "@angular/core";
export class RxGetRecordInputMapInspectorWidgetModule {
}
RxGetRecordInputMapInspectorWidgetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxGetRecordInputMapInspectorWidgetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxGetRecordInputMapInspectorWidgetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxGetRecordInputMapInspectorWidgetModule, declarations: [RxGetRecordInputMapInspectorWidgetComponent], imports: [CommonModule, RxDefinitionPickerModule, FormsModule, ExpressionFormControlModule], exports: [RxGetRecordInputMapInspectorWidgetComponent] });
RxGetRecordInputMapInspectorWidgetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxGetRecordInputMapInspectorWidgetModule, imports: [[CommonModule, RxDefinitionPickerModule, FormsModule, ExpressionFormControlModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxGetRecordInputMapInspectorWidgetModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RxGetRecordInputMapInspectorWidgetComponent],
                    imports: [CommonModule, RxDefinitionPickerModule, FormsModule, ExpressionFormControlModule],
                    exports: [RxGetRecordInputMapInspectorWidgetComponent]
                }]
        }] });
