// @bmc-ux/adapt-angular/esm2015/inline-editor/index.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AdaptCustomActionsModule } from '../custom-actions/index';
import { AdaptDragonInputHandlerModule } from '../dragon-input-handler/index';
import { AdaptRxTextfieldModule } from '../rx/rx-textfield/index';
import { AdaptInlineEditorComponent } from './inline-editor.component';
import * as i0 from "@angular/core";
export class AdaptInlineEditorModule {
}
AdaptInlineEditorModule.ɵfac = function AdaptInlineEditorModule_Factory(t) { return new (t || AdaptInlineEditorModule)(); };
AdaptInlineEditorModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptInlineEditorModule });
AdaptInlineEditorModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule,
            AdaptRxTextfieldModule,
            AdaptDragonInputHandlerModule,
            AdaptCustomActionsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptInlineEditorModule, [{
        type: NgModule,
        args: [{
                imports: [
                    FormsModule,
                    CommonModule,
                    AdaptRxTextfieldModule,
                    AdaptDragonInputHandlerModule,
                    AdaptCustomActionsModule
                ],
                declarations: [
                    AdaptInlineEditorComponent
                ],
                exports: [
                    AdaptInlineEditorComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptInlineEditorModule, { declarations: [AdaptInlineEditorComponent], imports: [FormsModule,
        CommonModule,
        AdaptRxTextfieldModule,
        AdaptDragonInputHandlerModule,
        AdaptCustomActionsModule], exports: [AdaptInlineEditorComponent] }); })();
export * from './inline-editor.component';
