// @helix/platform/esm2015/view/components/field-components/localized-character-field/design/localized-character-field-design.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { LocalizedCharacterFieldDesignComponent } from './localized-character-field-design.component';
import { AdaptRxTextfieldModule } from '@bmc-ux/adapt-angular';
import * as i0 from "@angular/core";
export class RxLocalizedCharacterFieldDesignModule {
}
RxLocalizedCharacterFieldDesignModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLocalizedCharacterFieldDesignModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RxLocalizedCharacterFieldDesignModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLocalizedCharacterFieldDesignModule, declarations: [LocalizedCharacterFieldDesignComponent], imports: [CommonModule, AdaptRxTextfieldModule, FormsModule] });
RxLocalizedCharacterFieldDesignModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLocalizedCharacterFieldDesignModule, imports: [[CommonModule, AdaptRxTextfieldModule, FormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0, type: RxLocalizedCharacterFieldDesignModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LocalizedCharacterFieldDesignComponent],
                    entryComponents: [LocalizedCharacterFieldDesignComponent],
                    imports: [CommonModule, AdaptRxTextfieldModule, FormsModule]
                }]
        }] });
