// @bmc-ux/adapt-angular/esm2015/accordion/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptAccordionComponent, AdaptAccordionTabComponent } from './accordion.component';
import * as i0 from "@angular/core";
export class AdaptAccordionModule {
}
AdaptAccordionModule.ɵfac = function AdaptAccordionModule_Factory(t) { return new (t || AdaptAccordionModule)(); };
AdaptAccordionModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptAccordionModule });
AdaptAccordionModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAccordionModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [AdaptAccordionComponent, AdaptAccordionTabComponent],
                exports: [AdaptAccordionComponent, AdaptAccordionTabComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptAccordionModule, { declarations: [AdaptAccordionComponent, AdaptAccordionTabComponent], imports: [CommonModule], exports: [AdaptAccordionComponent, AdaptAccordionTabComponent] }); })();
export * from './accordion.component';
export * from './accordion.model';
