// @bmc-ux/adapt-angular/esm2015/mention2/index.js
import { NgModule } from '@angular/core';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { AdaptMention2Directive } from './adapt-mention2.directive';
import { AdaptMention2ListComponent } from './adapt-mention2-list.component';
import * as i0 from "@angular/core";
export class AdaptMention2Module {
}
AdaptMention2Module.ɵfac = function AdaptMention2Module_Factory(t) { return new (t || AdaptMention2Module)(); };
AdaptMention2Module.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptMention2Module });
AdaptMention2Module.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            OverlayModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptMention2Module, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    OverlayModule
                ],
                declarations: [
                    AdaptMention2Directive,
                    AdaptMention2ListComponent
                ],
                exports: [
                    AdaptMention2Directive,
                    AdaptMention2ListComponent
                ],
                entryComponents: [
                    AdaptMention2ListComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptMention2Module, { declarations: [AdaptMention2Directive,
        AdaptMention2ListComponent], imports: [CommonModule,
        OverlayModule], exports: [AdaptMention2Directive,
        AdaptMention2ListComponent] }); })();
export * from './adapt-mention2.types';
export * from './adapt-mention2-utils';
export * from './adapt-mention2.directive';
export * from './adapt-mention2-list.component';
