// @bmc-ux/adapt-angular/esm2015/rx/rx-datetime/common/validate.utils.js
import { isDate, isNaN, isNil, isNumber } from 'lodash-es';
function isValidNumber(value) {
    return isNumber(value) && !isNaN(value);
}
/**
 * isInvalidDate() function
 *
 * Determines if the input date is an "Invalid Date" object
 */
export function isInvalidDate(date) {
    return isDate(date) && !isValidNumber(date.valueOf());
}
/**
 * isValidDate() function
 *
 * Determines if the input value is a Date but not an "Invalid Date"
 */
export function isValidDate(date) {
    return isDate(date) && isValidNumber(date.valueOf());
}
/**
 * isNilOrValidDate() function
 *
 * Determines if the input value is undefined, null or a valid Date
 */
export function isNilOrValidDate(date) {
    return isNil(date) || isValidDate(date);
}
/**
 * isInvalidDateStruct() function
 *
 * Determines if the input date struct corresponds to an "Invalid Date" object
 */
export function isInvalidDateStruct(date) {
    return Boolean(date) && Object.values(date).some(value => value !== value);
}
/**
 * getValidDate() function
 *
 * Helper method that converts Invalid Date objects to null
 */
export function getValidDate(date) {
    return isValidDate(date) ? date : null;
}
