// @bmc-ux/adapt-angular/esm2015/agreement/agreement-card/agreement-card.component.js
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { AdaptRadarAngularCustomEventName, AdaptRadarAngularGenericDirective, AdaptRadarAngularService } from '../../common/radar/adapt-radar';
import { isEmpty, isNull, isUndefined } from 'lodash-es';
import { AdaptTranslateService } from '../../common/i18n';
import * as i0 from "@angular/core";
import * as i1 from "../../common/i18n";
import * as i2 from "../../common/radar/adapt-radar";
import * as i3 from "@angular/common";
import * as i4 from "../../empty-state/empty-state.component";
import * as i5 from "../agreement-admin-content/agreement-admin-content.component";
import * as i6 from "../agreement-content/agreement-content.component";
function AdaptAgreementCardComponent_adapt_empty_state_10_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelementStart(0, "adapt-empty-state", 7);
    i0.ɵɵtext(1, "\n      ");
    i0.ɵɵelementStart(2, "p", 8);
    i0.ɵɵtext(3);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(4, "\n      ");
    i0.ɵɵelementStart(5, "p", 9);
    i0.ɵɵtext(6);
    i0.ɵɵelementEnd();
    i0.ɵɵtext(7, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("type", "config");
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r0.texts.cardEmptyStateMain);
    i0.ɵɵadvance(3);
    i0.ɵɵtextInterpolate(ctx_r0.texts.cardEmptyStateSecondary);
} }
function AdaptAgreementCardComponent_adapt_agreement_admin_content_12_Template(rf, ctx) { if (rf & 1) {
    const _r4 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "adapt-agreement-admin-content", 10);
    i0.ɵɵlistener("stateChange", function AdaptAgreementCardComponent_adapt_agreement_admin_content_12_Template_adapt_agreement_admin_content_stateChange_0_listener($event) { i0.ɵɵrestoreView(_r4); const ctx_r3 = i0.ɵɵnextContext(); return ctx_r3.handleStateChange($event); });
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = i0.ɵɵnextContext();
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("multiProductUsage", ctx_r1.multiProductUsage)("state", ctx_r1.state)("showOrganizationSettings", ctx_r1.showOrganizationSettings);
} }
function AdaptAgreementCardComponent_adapt_agreement_content_14_Template(rf, ctx) { if (rf & 1) {
    const _r6 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "adapt-agreement-content", 11);
    i0.ɵɵlistener("stateChange", function AdaptAgreementCardComponent_adapt_agreement_content_14_Template_adapt_agreement_content_stateChange_0_listener($event) { i0.ɵɵrestoreView(_r6); const ctx_r5 = i0.ɵɵnextContext(); return ctx_r5.handleStateChange($event); });
    i0.ɵɵtext(1, "\n    ");
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("adaptRadarDisableEventSending", true)("showMarketingSection", ctx_r2.showMarketingSection)("multiProductUsage", ctx_r2.multiProductUsage)("state", ctx_r2.state)("showOrganizationSettings", ctx_r2.showOrganizationSettings);
} }
export class AdaptAgreementCardComponent extends AdaptRadarAngularGenericDirective {
    constructor(_ts, _adaptRadarService, _elem) {
        super(_adaptRadarService, _elem, AdaptRadarAngularCustomEventName.AdaptAgreementCard);
        this._ts = _ts;
        this._accountDefaultState = {
            accountPerformance: false,
            accountMarketing: false,
        };
        this._organizationDefaultState = {
            organizationPerformance: false,
            organizationMarketing: false,
        };
        this.texts = {};
        this.multiProductUsage = false;
        this.state = {};
        this.showOrganizationSettings = false;
        this.showMarketingSection = true;
        this.showShadow = true;
        this.stateChange = new EventEmitter();
        this._setDefaultTexts(_ts.getCurrentLanguage());
    }
    ngOnInit() {
        const defaultOrganizationState = this.showOrganizationSettings ? Object.assign({}, this._organizationDefaultState) : {};
        const defaultState = Object.assign(Object.assign({}, this._accountDefaultState), defaultOrganizationState);
        if (!isNull(this.state)) {
            this.state = isUndefined(this.state) || isEmpty(this.state)
                ? defaultState
                : Object.assign(Object.assign({}, defaultState), this.state);
        }
    }
    handleStateChange(state) {
        this.stateChange.emit(state);
    }
    _setDefaultTexts(translation) {
        this.texts.cardMainTitle = translation['adapt.agreement.card.mainTitle'];
        this.texts.cardEmptyStateMain = translation['adapt.agreement.card.emptyStateMain'];
        this.texts.cardEmptyStateSecondary = translation['adapt.agreement.card.emptyStateSecondary'];
    }
}
AdaptAgreementCardComponent.ɵfac = function AdaptAgreementCardComponent_Factory(t) { return new (t || AdaptAgreementCardComponent)(i0.ɵɵdirectiveInject(i1.AdaptTranslateService), i0.ɵɵdirectiveInject(i2.AdaptRadarAngularService), i0.ɵɵdirectiveInject(i0.ElementRef)); };
AdaptAgreementCardComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: AdaptAgreementCardComponent, selectors: [["adapt-agreement-card"]], inputs: { multiProductUsage: "multiProductUsage", state: "state", showOrganizationSettings: "showOrganizationSettings", showMarketingSection: "showMarketingSection", showShadow: "showShadow" }, outputs: { stateChange: "stateChange" }, features: [i0.ɵɵInheritDefinitionFeature], decls: 18, vars: 8, consts: [[1, "adapt-agreement"], [1, "adapt-agreement__header", "p-3"], [1, "m-0"], [1, "adapt-agreement__content", "p-3"], [3, "adaptRadarDisableEventSending", "type", 4, "ngIf"], [3, "adaptRadarDisableEventSending", "multiProductUsage", "state", "showOrganizationSettings", "stateChange", 4, "ngIf"], [3, "adaptRadarDisableEventSending", "showMarketingSection", "multiProductUsage", "state", "showOrganizationSettings", "stateChange", 4, "ngIf"], [3, "adaptRadarDisableEventSending", "type"], [1, "adapt-agreement__content_empty-state"], [1, "text-center", "text-secondary"], [3, "adaptRadarDisableEventSending", "multiProductUsage", "state", "showOrganizationSettings", "stateChange"], [3, "adaptRadarDisableEventSending", "showMarketingSection", "multiProductUsage", "state", "showOrganizationSettings", "stateChange"]], template: function AdaptAgreementCardComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "div", 0);
        i0.ɵɵtext(1, "\n  ");
        i0.ɵɵelementStart(2, "div", 1);
        i0.ɵɵtext(3, "\n    ");
        i0.ɵɵelementStart(4, "h5", 2);
        i0.ɵɵtext(5);
        i0.ɵɵelementEnd();
        i0.ɵɵtext(6, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(7, "\n\n  ");
        i0.ɵɵelementStart(8, "div", 3);
        i0.ɵɵtext(9, "\n    ");
        i0.ɵɵtemplate(10, AdaptAgreementCardComponent_adapt_empty_state_10_Template, 8, 4, "adapt-empty-state", 4);
        i0.ɵɵtext(11, "\n\n    ");
        i0.ɵɵtemplate(12, AdaptAgreementCardComponent_adapt_agreement_admin_content_12_Template, 2, 4, "adapt-agreement-admin-content", 5);
        i0.ɵɵtext(13, "\n\n    ");
        i0.ɵɵtemplate(14, AdaptAgreementCardComponent_adapt_agreement_content_14_Template, 2, 5, "adapt-agreement-content", 6);
        i0.ɵɵtext(15, "\n  ");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(16, "\n");
        i0.ɵɵelementEnd();
        i0.ɵɵtext(17, "\n");
    } if (rf & 2) {
        i0.ɵɵclassProp("adapt-agreement_no-shadow", !ctx.showShadow)("adapt-agreement_has-shadow", ctx.showShadow);
        i0.ɵɵadvance(5);
        i0.ɵɵtextInterpolate(ctx.texts.cardMainTitle);
        i0.ɵɵadvance(5);
        i0.ɵɵproperty("ngIf", !ctx.state);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.state && ctx.showOrganizationSettings);
        i0.ɵɵadvance(2);
        i0.ɵɵproperty("ngIf", ctx.state && !ctx.showOrganizationSettings);
    } }, directives: [i3.NgIf, i4.AdaptEmptyStateComponent, i5.AdaptAgreementAdminContentComponent, i6.AdaptAgreementContentComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptAgreementCardComponent, [{
        type: Component,
        args: [{
                selector: 'adapt-agreement-card',
                templateUrl: './agreement-card.component.html'
            }]
    }], function () { return [{ type: i1.AdaptTranslateService }, { type: i2.AdaptRadarAngularService }, { type: i0.ElementRef }]; }, { multiProductUsage: [{
            type: Input
        }], state: [{
            type: Input
        }], showOrganizationSettings: [{
            type: Input
        }], showMarketingSection: [{
            type: Input
        }], showShadow: [{
            type: Input
        }], stateChange: [{
            type: Output
        }] }); })();
