// @bmc-ux/adapt-angular/esm2015/button/index.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AdaptButtonComponent } from './button.component';
import * as i0 from "@angular/core";
export class AdaptButtonModule {
}
AdaptButtonModule.ɵfac = function AdaptButtonModule_Factory(t) { return new (t || AdaptButtonModule)(); };
AdaptButtonModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: AdaptButtonModule });
AdaptButtonModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(AdaptButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [AdaptButtonComponent],
                declarations: [AdaptButtonComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(AdaptButtonModule, { declarations: [AdaptButtonComponent], imports: [CommonModule], exports: [AdaptButtonComponent] }); })();
export * from './button.component';
